{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "kernelspec": {
      "display_name": "R",
      "language": "R",
      "name": "ir"
    },
    "language_info": {
      "codemirror_mode": "r",
      "file_extension": ".r",
      "mimetype": "text/x-r-source",
      "name": "R",
      "pygments_lexer": "r",
      "version": "3.3.1"
    },
    "colab": {
      "name": "OSK_RNA_and_NOR_NFR_sf6ef.ipynb",
      "provenance": [],
      "collapsed_sections": [],
      "machine_shape": "hm",
      "toc_visible": true,
      "include_colab_link": true
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/gongx030/seatac_manuscript/blob/main/OSK_RNA_and_NOR_NFR_sf6ef.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# The notebook run Differential expression analysis between MEFs and D7 MEFs post OSK induction and lists genes with regions going from NFR to NOR and are significantly down at D7."
      ],
      "metadata": {
        "id": "Ab8RwYcN7Wfk"
      }
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "6OhqEEeZHroa"
      },
      "source": [
        "start_time0 <- Sys.time()\n",
        "options(max.print=1000)"
      ],
      "execution_count": 1,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "obcf6pAcw6TQ"
      },
      "source": [
        "### Install external packages"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "99vS8MZUxApi",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 156
        },
        "outputId": "d70e4332-742d-47a0-eebc-908ce46105ce"
      },
      "source": [
        "system('apt-get -q install libgsl-dev', intern = TRUE)"
      ],
      "execution_count": 2,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/html": [
              "<style>\n",
              ".list-inline {list-style: none; margin:0; padding: 0}\n",
              ".list-inline>li {display: inline-block}\n",
              ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
              "</style>\n",
              "<ol class=list-inline><li>'Reading package lists...'</li><li>'Building dependency tree...'</li><li>'Reading state information...'</li><li>'The following additional packages will be installed:'</li><li><span style=white-space:pre-wrap>'  libgsl23 libgslcblas0'</span></li><li>'Suggested packages:'</li><li><span style=white-space:pre-wrap>'  gsl-ref-psdoc | gsl-doc-pdf | gsl-doc-info | gsl-ref-html'</span></li><li>'The following NEW packages will be installed:'</li><li><span style=white-space:pre-wrap>'  libgsl-dev libgsl23 libgslcblas0'</span></li><li>'0 upgraded, 3 newly installed, 0 to remove and 39 not upgraded.'</li><li>'Need to get 1,926 kB of archives.'</li><li>'After this operation, 9,474 kB of additional disk space will be used.'</li><li>'Get:1 http://archive.ubuntu.com/ubuntu bionic/universe amd64 libgslcblas0 amd64 2.4+dfsg-6 [79.7 kB]'</li><li>'Get:2 http://archive.ubuntu.com/ubuntu bionic/universe amd64 libgsl23 amd64 2.4+dfsg-6 [823 kB]'</li><li>'Get:3 http://archive.ubuntu.com/ubuntu bionic/universe amd64 libgsl-dev amd64 2.4+dfsg-6 [1,023 kB]'</li><li>'Fetched 1,926 kB in 1s (2,388 kB/s)'</li><li>'Selecting previously unselected package libgslcblas0:amd64.\\r'</li><li>'(Reading database ... \\r(Reading database ... 5%\\r(Reading database ... 10%\\r(Reading database ... 15%\\r(Reading database ... 20%\\r(Reading database ... 25%\\r(Reading database ... 30%\\r(Reading database ... 35%\\r(Reading database ... 40%\\r(Reading database ... 45%\\r(Reading database ... 50%\\r(Reading database ... 55%\\r(Reading database ... 60%\\r(Reading database ... 65%\\r(Reading database ... 70%\\r(Reading database ... 75%\\r(Reading database ... 80%\\r(Reading database ... 85%\\r(Reading database ... 90%\\r(Reading database ... 95%\\r(Reading database ... 100%\\r(Reading database ... 155455 files and directories currently installed.)\\r'</li><li>'Preparing to unpack .../libgslcblas0_2.4+dfsg-6_amd64.deb ...\\r'</li><li>'Unpacking libgslcblas0:amd64 (2.4+dfsg-6) ...\\r'</li><li>'Selecting previously unselected package libgsl23:amd64.\\r'</li><li>'Preparing to unpack .../libgsl23_2.4+dfsg-6_amd64.deb ...\\r'</li><li>'Unpacking libgsl23:amd64 (2.4+dfsg-6) ...\\r'</li><li>'Selecting previously unselected package libgsl-dev.\\r'</li><li>'Preparing to unpack .../libgsl-dev_2.4+dfsg-6_amd64.deb ...\\r'</li><li>'Unpacking libgsl-dev (2.4+dfsg-6) ...\\r'</li><li>'Setting up libgslcblas0:amd64 (2.4+dfsg-6) ...\\r'</li><li>'Setting up libgsl23:amd64 (2.4+dfsg-6) ...\\r'</li><li>'Setting up libgsl-dev (2.4+dfsg-6) ...\\r'</li><li>'Processing triggers for man-db (2.8.3-2ubuntu0.1) ...\\r'</li><li>'Processing triggers for libc-bin (2.27-3ubuntu1.3) ...\\r'</li><li>'/sbin/ldconfig.real: /usr/local/lib/python3.7/dist-packages/ideep4py/lib/libmkldnn.so.0 is not a symbolic link\\r'</li><li>'\\r'</li></ol>\n"
            ],
            "text/markdown": "1. 'Reading package lists...'\n2. 'Building dependency tree...'\n3. 'Reading state information...'\n4. 'The following additional packages will be installed:'\n5. <span style=white-space:pre-wrap>'  libgsl23 libgslcblas0'</span>\n6. 'Suggested packages:'\n7. <span style=white-space:pre-wrap>'  gsl-ref-psdoc | gsl-doc-pdf | gsl-doc-info | gsl-ref-html'</span>\n8. 'The following NEW packages will be installed:'\n9. <span style=white-space:pre-wrap>'  libgsl-dev libgsl23 libgslcblas0'</span>\n10. '0 upgraded, 3 newly installed, 0 to remove and 39 not upgraded.'\n11. 'Need to get 1,926 kB of archives.'\n12. 'After this operation, 9,474 kB of additional disk space will be used.'\n13. 'Get:1 http://archive.ubuntu.com/ubuntu bionic/universe amd64 libgslcblas0 amd64 2.4+dfsg-6 [79.7 kB]'\n14. 'Get:2 http://archive.ubuntu.com/ubuntu bionic/universe amd64 libgsl23 amd64 2.4+dfsg-6 [823 kB]'\n15. 'Get:3 http://archive.ubuntu.com/ubuntu bionic/universe amd64 libgsl-dev amd64 2.4+dfsg-6 [1,023 kB]'\n16. 'Fetched 1,926 kB in 1s (2,388 kB/s)'\n17. 'Selecting previously unselected package libgslcblas0:amd64.\\r'\n18. '(Reading database ... \\r(Reading database ... 5%\\r(Reading database ... 10%\\r(Reading database ... 15%\\r(Reading database ... 20%\\r(Reading database ... 25%\\r(Reading database ... 30%\\r(Reading database ... 35%\\r(Reading database ... 40%\\r(Reading database ... 45%\\r(Reading database ... 50%\\r(Reading database ... 55%\\r(Reading database ... 60%\\r(Reading database ... 65%\\r(Reading database ... 70%\\r(Reading database ... 75%\\r(Reading database ... 80%\\r(Reading database ... 85%\\r(Reading database ... 90%\\r(Reading database ... 95%\\r(Reading database ... 100%\\r(Reading database ... 155455 files and directories currently installed.)\\r'\n19. 'Preparing to unpack .../libgslcblas0_2.4+dfsg-6_amd64.deb ...\\r'\n20. 'Unpacking libgslcblas0:amd64 (2.4+dfsg-6) ...\\r'\n21. 'Selecting previously unselected package libgsl23:amd64.\\r'\n22. 'Preparing to unpack .../libgsl23_2.4+dfsg-6_amd64.deb ...\\r'\n23. 'Unpacking libgsl23:amd64 (2.4+dfsg-6) ...\\r'\n24. 'Selecting previously unselected package libgsl-dev.\\r'\n25. 'Preparing to unpack .../libgsl-dev_2.4+dfsg-6_amd64.deb ...\\r'\n26. 'Unpacking libgsl-dev (2.4+dfsg-6) ...\\r'\n27. 'Setting up libgslcblas0:amd64 (2.4+dfsg-6) ...\\r'\n28. 'Setting up libgsl23:amd64 (2.4+dfsg-6) ...\\r'\n29. 'Setting up libgsl-dev (2.4+dfsg-6) ...\\r'\n30. 'Processing triggers for man-db (2.8.3-2ubuntu0.1) ...\\r'\n31. 'Processing triggers for libc-bin (2.27-3ubuntu1.3) ...\\r'\n32. '/sbin/ldconfig.real: /usr/local/lib/python3.7/dist-packages/ideep4py/lib/libmkldnn.so.0 is not a symbolic link\\r'\n33. '\\r'\n\n\n",
            "text/latex": "\\begin{enumerate*}\n\\item 'Reading package lists...'\n\\item 'Building dependency tree...'\n\\item 'Reading state information...'\n\\item 'The following additional packages will be installed:'\n\\item '  libgsl23 libgslcblas0'\n\\item 'Suggested packages:'\n\\item '  gsl-ref-psdoc \\textbar{} gsl-doc-pdf \\textbar{} gsl-doc-info \\textbar{} gsl-ref-html'\n\\item 'The following NEW packages will be installed:'\n\\item '  libgsl-dev libgsl23 libgslcblas0'\n\\item '0 upgraded, 3 newly installed, 0 to remove and 39 not upgraded.'\n\\item 'Need to get 1,926 kB of archives.'\n\\item 'After this operation, 9,474 kB of additional disk space will be used.'\n\\item 'Get:1 http://archive.ubuntu.com/ubuntu bionic/universe amd64 libgslcblas0 amd64 2.4+dfsg-6 {[}79.7 kB{]}'\n\\item 'Get:2 http://archive.ubuntu.com/ubuntu bionic/universe amd64 libgsl23 amd64 2.4+dfsg-6 {[}823 kB{]}'\n\\item 'Get:3 http://archive.ubuntu.com/ubuntu bionic/universe amd64 libgsl-dev amd64 2.4+dfsg-6 {[}1,023 kB{]}'\n\\item 'Fetched 1,926 kB in 1s (2,388 kB/s)'\n\\item 'Selecting previously unselected package libgslcblas0:amd64.\\textbackslash{}r'\n\\item '(Reading database ... \\textbackslash{}r(Reading database ... 5\\%\\textbackslash{}r(Reading database ... 10\\%\\textbackslash{}r(Reading database ... 15\\%\\textbackslash{}r(Reading database ... 20\\%\\textbackslash{}r(Reading database ... 25\\%\\textbackslash{}r(Reading database ... 30\\%\\textbackslash{}r(Reading database ... 35\\%\\textbackslash{}r(Reading database ... 40\\%\\textbackslash{}r(Reading database ... 45\\%\\textbackslash{}r(Reading database ... 50\\%\\textbackslash{}r(Reading database ... 55\\%\\textbackslash{}r(Reading database ... 60\\%\\textbackslash{}r(Reading database ... 65\\%\\textbackslash{}r(Reading database ... 70\\%\\textbackslash{}r(Reading database ... 75\\%\\textbackslash{}r(Reading database ... 80\\%\\textbackslash{}r(Reading database ... 85\\%\\textbackslash{}r(Reading database ... 90\\%\\textbackslash{}r(Reading database ... 95\\%\\textbackslash{}r(Reading database ... 100\\%\\textbackslash{}r(Reading database ... 155455 files and directories currently installed.)\\textbackslash{}r'\n\\item 'Preparing to unpack .../libgslcblas0\\_2.4+dfsg-6\\_amd64.deb ...\\textbackslash{}r'\n\\item 'Unpacking libgslcblas0:amd64 (2.4+dfsg-6) ...\\textbackslash{}r'\n\\item 'Selecting previously unselected package libgsl23:amd64.\\textbackslash{}r'\n\\item 'Preparing to unpack .../libgsl23\\_2.4+dfsg-6\\_amd64.deb ...\\textbackslash{}r'\n\\item 'Unpacking libgsl23:amd64 (2.4+dfsg-6) ...\\textbackslash{}r'\n\\item 'Selecting previously unselected package libgsl-dev.\\textbackslash{}r'\n\\item 'Preparing to unpack .../libgsl-dev\\_2.4+dfsg-6\\_amd64.deb ...\\textbackslash{}r'\n\\item 'Unpacking libgsl-dev (2.4+dfsg-6) ...\\textbackslash{}r'\n\\item 'Setting up libgslcblas0:amd64 (2.4+dfsg-6) ...\\textbackslash{}r'\n\\item 'Setting up libgsl23:amd64 (2.4+dfsg-6) ...\\textbackslash{}r'\n\\item 'Setting up libgsl-dev (2.4+dfsg-6) ...\\textbackslash{}r'\n\\item 'Processing triggers for man-db (2.8.3-2ubuntu0.1) ...\\textbackslash{}r'\n\\item 'Processing triggers for libc-bin (2.27-3ubuntu1.3) ...\\textbackslash{}r'\n\\item '/sbin/ldconfig.real: /usr/local/lib/python3.7/dist-packages/ideep4py/lib/libmkldnn.so.0 is not a symbolic link\\textbackslash{}r'\n\\item '\\textbackslash{}r'\n\\end{enumerate*}\n",
            "text/plain": [
              " [1] \"Reading package lists...\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      \n",
              " [2] \"Building dependency tree...\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   \n",
              " [3] \"Reading state information...\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  \n",
              " [4] \"The following additional packages will be installed:\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          \n",
              " [5] \"  libgsl23 libgslcblas0\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       \n",
              " [6] \"Suggested packages:\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           \n",
              " [7] \"  gsl-ref-psdoc | gsl-doc-pdf | gsl-doc-info | gsl-ref-html\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   \n",
              " [8] \"The following NEW packages will be installed:\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 \n",
              " [9] \"  libgsl-dev libgsl23 libgslcblas0\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            \n",
              "[10] \"0 upgraded, 3 newly installed, 0 to remove and 39 not upgraded.\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               \n",
              "[11] \"Need to get 1,926 kB of archives.\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             \n",
              "[12] \"After this operation, 9,474 kB of additional disk space will be used.\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         \n",
              "[13] \"Get:1 http://archive.ubuntu.com/ubuntu bionic/universe amd64 libgslcblas0 amd64 2.4+dfsg-6 [79.7 kB]\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          \n",
              "[14] \"Get:2 http://archive.ubuntu.com/ubuntu bionic/universe amd64 libgsl23 amd64 2.4+dfsg-6 [823 kB]\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               \n",
              "[15] \"Get:3 http://archive.ubuntu.com/ubuntu bionic/universe amd64 libgsl-dev amd64 2.4+dfsg-6 [1,023 kB]\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           \n",
              "[16] \"Fetched 1,926 kB in 1s (2,388 kB/s)\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           \n",
              "[17] \"Selecting previously unselected package libgslcblas0:amd64.\\r\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 \n",
              "[18] \"(Reading database ... \\r(Reading database ... 5%\\r(Reading database ... 10%\\r(Reading database ... 15%\\r(Reading database ... 20%\\r(Reading database ... 25%\\r(Reading database ... 30%\\r(Reading database ... 35%\\r(Reading database ... 40%\\r(Reading database ... 45%\\r(Reading database ... 50%\\r(Reading database ... 55%\\r(Reading database ... 60%\\r(Reading database ... 65%\\r(Reading database ... 70%\\r(Reading database ... 75%\\r(Reading database ... 80%\\r(Reading database ... 85%\\r(Reading database ... 90%\\r(Reading database ... 95%\\r(Reading database ... 100%\\r(Reading database ... 155455 files and directories currently installed.)\\r\"\n",
              "[19] \"Preparing to unpack .../libgslcblas0_2.4+dfsg-6_amd64.deb ...\\r\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               \n",
              "[20] \"Unpacking libgslcblas0:amd64 (2.4+dfsg-6) ...\\r\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               \n",
              "[21] \"Selecting previously unselected package libgsl23:amd64.\\r\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     \n",
              "[22] \"Preparing to unpack .../libgsl23_2.4+dfsg-6_amd64.deb ...\\r\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   \n",
              "[23] \"Unpacking libgsl23:amd64 (2.4+dfsg-6) ...\\r\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   \n",
              "[24] \"Selecting previously unselected package libgsl-dev.\\r\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         \n",
              "[25] \"Preparing to unpack .../libgsl-dev_2.4+dfsg-6_amd64.deb ...\\r\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 \n",
              "[26] \"Unpacking libgsl-dev (2.4+dfsg-6) ...\\r\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       \n",
              "[27] \"Setting up libgslcblas0:amd64 (2.4+dfsg-6) ...\\r\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              \n",
              "[28] \"Setting up libgsl23:amd64 (2.4+dfsg-6) ...\\r\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  \n",
              "[29] \"Setting up libgsl-dev (2.4+dfsg-6) ...\\r\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      \n",
              "[30] \"Processing triggers for man-db (2.8.3-2ubuntu0.1) ...\\r\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       \n",
              "[31] \"Processing triggers for libc-bin (2.27-3ubuntu1.3) ...\\r\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      \n",
              "[32] \"/sbin/ldconfig.real: /usr/local/lib/python3.7/dist-packages/ideep4py/lib/libmkldnn.so.0 is not a symbolic link\\r\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              \n",
              "[33] \"\\r\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            "
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "HgPXAFx1xZHE",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 208
        },
        "outputId": "ca8d2f5b-4ea2-4975-8e98-552989325b2c"
      },
      "source": [
        "system('apt-get install libcairo2-dev', intern = TRUE) # for Cairo package that is required by ComplexHeatmap package"
      ],
      "execution_count": 3,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/html": [
              "<style>\n",
              ".list-inline {list-style: none; margin:0; padding: 0}\n",
              ".list-inline>li {display: inline-block}\n",
              ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
              "</style>\n",
              "<ol class=list-inline><li>'Reading package lists...'</li><li>'Building dependency tree...'</li><li>'Reading state information...'</li><li>'The following additional packages will be installed:'</li><li><span style=white-space:pre-wrap>'  libcairo-script-interpreter2 libpixman-1-dev libxcb-shm0-dev'</span></li><li>'Suggested packages:'</li><li><span style=white-space:pre-wrap>'  libcairo2-doc'</span></li><li>'The following NEW packages will be installed:'</li><li><span style=white-space:pre-wrap>'  libcairo-script-interpreter2 libcairo2-dev libpixman-1-dev libxcb-shm0-dev'</span></li><li>'0 upgraded, 4 newly installed, 0 to remove and 39 not upgraded.'</li><li>'Need to get 930 kB of archives.'</li><li>'After this operation, 3,986 kB of additional disk space will be used.'</li><li>'Get:1 http://archive.ubuntu.com/ubuntu bionic-updates/main amd64 libcairo-script-interpreter2 amd64 1.15.10-2ubuntu0.1 [53.5 kB]'</li><li>'Get:2 http://archive.ubuntu.com/ubuntu bionic/main amd64 libpixman-1-dev amd64 0.34.0-2 [244 kB]'</li><li>'Get:3 http://archive.ubuntu.com/ubuntu bionic-updates/main amd64 libxcb-shm0-dev amd64 1.13-2~ubuntu18.04 [6,684 B]'</li><li>'Get:4 http://archive.ubuntu.com/ubuntu bionic-updates/main amd64 libcairo2-dev amd64 1.15.10-2ubuntu0.1 [626 kB]'</li><li>'Fetched 930 kB in 1s (1,309 kB/s)'</li><li>'Selecting previously unselected package libcairo-script-interpreter2:amd64.\\r'</li><li>'(Reading database ... \\r(Reading database ... 5%\\r(Reading database ... 10%\\r(Reading database ... 15%\\r(Reading database ... 20%\\r(Reading database ... 25%\\r(Reading database ... 30%\\r(Reading database ... 35%\\r(Reading database ... 40%\\r(Reading database ... 45%\\r(Reading database ... 50%\\r(Reading database ... 55%\\r(Reading database ... 60%\\r(Reading database ... 65%\\r(Reading database ... 70%\\r(Reading database ... 75%\\r(Reading database ... 80%\\r(Reading database ... 85%\\r(Reading database ... 90%\\r(Reading database ... 95%\\r(Reading database ... 100%\\r(Reading database ... 155727 files and directories currently installed.)\\r'</li><li>'Preparing to unpack .../libcairo-script-interpreter2_1.15.10-2ubuntu0.1_amd64.deb ...\\r'</li><li>'Unpacking libcairo-script-interpreter2:amd64 (1.15.10-2ubuntu0.1) ...\\r'</li><li>'Selecting previously unselected package libpixman-1-dev:amd64.\\r'</li><li>'Preparing to unpack .../libpixman-1-dev_0.34.0-2_amd64.deb ...\\r'</li><li>'Unpacking libpixman-1-dev:amd64 (0.34.0-2) ...\\r'</li><li>'Selecting previously unselected package libxcb-shm0-dev:amd64.\\r'</li><li>'Preparing to unpack .../libxcb-shm0-dev_1.13-2~ubuntu18.04_amd64.deb ...\\r'</li><li>'Unpacking libxcb-shm0-dev:amd64 (1.13-2~ubuntu18.04) ...\\r'</li><li>'Selecting previously unselected package libcairo2-dev:amd64.\\r'</li><li>'Preparing to unpack .../libcairo2-dev_1.15.10-2ubuntu0.1_amd64.deb ...\\r'</li><li>'Unpacking libcairo2-dev:amd64 (1.15.10-2ubuntu0.1) ...\\r'</li><li>'Setting up libcairo-script-interpreter2:amd64 (1.15.10-2ubuntu0.1) ...\\r'</li><li>'Setting up libxcb-shm0-dev:amd64 (1.13-2~ubuntu18.04) ...\\r'</li><li>'Setting up libpixman-1-dev:amd64 (0.34.0-2) ...\\r'</li><li>'Setting up libcairo2-dev:amd64 (1.15.10-2ubuntu0.1) ...\\r'</li><li>'Processing triggers for libc-bin (2.27-3ubuntu1.3) ...\\r'</li><li>'/sbin/ldconfig.real: /usr/local/lib/python3.7/dist-packages/ideep4py/lib/libmkldnn.so.0 is not a symbolic link\\r'</li><li>'\\r'</li></ol>\n"
            ],
            "text/markdown": "1. 'Reading package lists...'\n2. 'Building dependency tree...'\n3. 'Reading state information...'\n4. 'The following additional packages will be installed:'\n5. <span style=white-space:pre-wrap>'  libcairo-script-interpreter2 libpixman-1-dev libxcb-shm0-dev'</span>\n6. 'Suggested packages:'\n7. <span style=white-space:pre-wrap>'  libcairo2-doc'</span>\n8. 'The following NEW packages will be installed:'\n9. <span style=white-space:pre-wrap>'  libcairo-script-interpreter2 libcairo2-dev libpixman-1-dev libxcb-shm0-dev'</span>\n10. '0 upgraded, 4 newly installed, 0 to remove and 39 not upgraded.'\n11. 'Need to get 930 kB of archives.'\n12. 'After this operation, 3,986 kB of additional disk space will be used.'\n13. 'Get:1 http://archive.ubuntu.com/ubuntu bionic-updates/main amd64 libcairo-script-interpreter2 amd64 1.15.10-2ubuntu0.1 [53.5 kB]'\n14. 'Get:2 http://archive.ubuntu.com/ubuntu bionic/main amd64 libpixman-1-dev amd64 0.34.0-2 [244 kB]'\n15. 'Get:3 http://archive.ubuntu.com/ubuntu bionic-updates/main amd64 libxcb-shm0-dev amd64 1.13-2~ubuntu18.04 [6,684 B]'\n16. 'Get:4 http://archive.ubuntu.com/ubuntu bionic-updates/main amd64 libcairo2-dev amd64 1.15.10-2ubuntu0.1 [626 kB]'\n17. 'Fetched 930 kB in 1s (1,309 kB/s)'\n18. 'Selecting previously unselected package libcairo-script-interpreter2:amd64.\\r'\n19. '(Reading database ... \\r(Reading database ... 5%\\r(Reading database ... 10%\\r(Reading database ... 15%\\r(Reading database ... 20%\\r(Reading database ... 25%\\r(Reading database ... 30%\\r(Reading database ... 35%\\r(Reading database ... 40%\\r(Reading database ... 45%\\r(Reading database ... 50%\\r(Reading database ... 55%\\r(Reading database ... 60%\\r(Reading database ... 65%\\r(Reading database ... 70%\\r(Reading database ... 75%\\r(Reading database ... 80%\\r(Reading database ... 85%\\r(Reading database ... 90%\\r(Reading database ... 95%\\r(Reading database ... 100%\\r(Reading database ... 155727 files and directories currently installed.)\\r'\n20. 'Preparing to unpack .../libcairo-script-interpreter2_1.15.10-2ubuntu0.1_amd64.deb ...\\r'\n21. 'Unpacking libcairo-script-interpreter2:amd64 (1.15.10-2ubuntu0.1) ...\\r'\n22. 'Selecting previously unselected package libpixman-1-dev:amd64.\\r'\n23. 'Preparing to unpack .../libpixman-1-dev_0.34.0-2_amd64.deb ...\\r'\n24. 'Unpacking libpixman-1-dev:amd64 (0.34.0-2) ...\\r'\n25. 'Selecting previously unselected package libxcb-shm0-dev:amd64.\\r'\n26. 'Preparing to unpack .../libxcb-shm0-dev_1.13-2~ubuntu18.04_amd64.deb ...\\r'\n27. 'Unpacking libxcb-shm0-dev:amd64 (1.13-2~ubuntu18.04) ...\\r'\n28. 'Selecting previously unselected package libcairo2-dev:amd64.\\r'\n29. 'Preparing to unpack .../libcairo2-dev_1.15.10-2ubuntu0.1_amd64.deb ...\\r'\n30. 'Unpacking libcairo2-dev:amd64 (1.15.10-2ubuntu0.1) ...\\r'\n31. 'Setting up libcairo-script-interpreter2:amd64 (1.15.10-2ubuntu0.1) ...\\r'\n32. 'Setting up libxcb-shm0-dev:amd64 (1.13-2~ubuntu18.04) ...\\r'\n33. 'Setting up libpixman-1-dev:amd64 (0.34.0-2) ...\\r'\n34. 'Setting up libcairo2-dev:amd64 (1.15.10-2ubuntu0.1) ...\\r'\n35. 'Processing triggers for libc-bin (2.27-3ubuntu1.3) ...\\r'\n36. '/sbin/ldconfig.real: /usr/local/lib/python3.7/dist-packages/ideep4py/lib/libmkldnn.so.0 is not a symbolic link\\r'\n37. '\\r'\n\n\n",
            "text/latex": "\\begin{enumerate*}\n\\item 'Reading package lists...'\n\\item 'Building dependency tree...'\n\\item 'Reading state information...'\n\\item 'The following additional packages will be installed:'\n\\item '  libcairo-script-interpreter2 libpixman-1-dev libxcb-shm0-dev'\n\\item 'Suggested packages:'\n\\item '  libcairo2-doc'\n\\item 'The following NEW packages will be installed:'\n\\item '  libcairo-script-interpreter2 libcairo2-dev libpixman-1-dev libxcb-shm0-dev'\n\\item '0 upgraded, 4 newly installed, 0 to remove and 39 not upgraded.'\n\\item 'Need to get 930 kB of archives.'\n\\item 'After this operation, 3,986 kB of additional disk space will be used.'\n\\item 'Get:1 http://archive.ubuntu.com/ubuntu bionic-updates/main amd64 libcairo-script-interpreter2 amd64 1.15.10-2ubuntu0.1 {[}53.5 kB{]}'\n\\item 'Get:2 http://archive.ubuntu.com/ubuntu bionic/main amd64 libpixman-1-dev amd64 0.34.0-2 {[}244 kB{]}'\n\\item 'Get:3 http://archive.ubuntu.com/ubuntu bionic-updates/main amd64 libxcb-shm0-dev amd64 1.13-2\\textasciitilde{}ubuntu18.04 {[}6,684 B{]}'\n\\item 'Get:4 http://archive.ubuntu.com/ubuntu bionic-updates/main amd64 libcairo2-dev amd64 1.15.10-2ubuntu0.1 {[}626 kB{]}'\n\\item 'Fetched 930 kB in 1s (1,309 kB/s)'\n\\item 'Selecting previously unselected package libcairo-script-interpreter2:amd64.\\textbackslash{}r'\n\\item '(Reading database ... \\textbackslash{}r(Reading database ... 5\\%\\textbackslash{}r(Reading database ... 10\\%\\textbackslash{}r(Reading database ... 15\\%\\textbackslash{}r(Reading database ... 20\\%\\textbackslash{}r(Reading database ... 25\\%\\textbackslash{}r(Reading database ... 30\\%\\textbackslash{}r(Reading database ... 35\\%\\textbackslash{}r(Reading database ... 40\\%\\textbackslash{}r(Reading database ... 45\\%\\textbackslash{}r(Reading database ... 50\\%\\textbackslash{}r(Reading database ... 55\\%\\textbackslash{}r(Reading database ... 60\\%\\textbackslash{}r(Reading database ... 65\\%\\textbackslash{}r(Reading database ... 70\\%\\textbackslash{}r(Reading database ... 75\\%\\textbackslash{}r(Reading database ... 80\\%\\textbackslash{}r(Reading database ... 85\\%\\textbackslash{}r(Reading database ... 90\\%\\textbackslash{}r(Reading database ... 95\\%\\textbackslash{}r(Reading database ... 100\\%\\textbackslash{}r(Reading database ... 155727 files and directories currently installed.)\\textbackslash{}r'\n\\item 'Preparing to unpack .../libcairo-script-interpreter2\\_1.15.10-2ubuntu0.1\\_amd64.deb ...\\textbackslash{}r'\n\\item 'Unpacking libcairo-script-interpreter2:amd64 (1.15.10-2ubuntu0.1) ...\\textbackslash{}r'\n\\item 'Selecting previously unselected package libpixman-1-dev:amd64.\\textbackslash{}r'\n\\item 'Preparing to unpack .../libpixman-1-dev\\_0.34.0-2\\_amd64.deb ...\\textbackslash{}r'\n\\item 'Unpacking libpixman-1-dev:amd64 (0.34.0-2) ...\\textbackslash{}r'\n\\item 'Selecting previously unselected package libxcb-shm0-dev:amd64.\\textbackslash{}r'\n\\item 'Preparing to unpack .../libxcb-shm0-dev\\_1.13-2\\textasciitilde{}ubuntu18.04\\_amd64.deb ...\\textbackslash{}r'\n\\item 'Unpacking libxcb-shm0-dev:amd64 (1.13-2\\textasciitilde{}ubuntu18.04) ...\\textbackslash{}r'\n\\item 'Selecting previously unselected package libcairo2-dev:amd64.\\textbackslash{}r'\n\\item 'Preparing to unpack .../libcairo2-dev\\_1.15.10-2ubuntu0.1\\_amd64.deb ...\\textbackslash{}r'\n\\item 'Unpacking libcairo2-dev:amd64 (1.15.10-2ubuntu0.1) ...\\textbackslash{}r'\n\\item 'Setting up libcairo-script-interpreter2:amd64 (1.15.10-2ubuntu0.1) ...\\textbackslash{}r'\n\\item 'Setting up libxcb-shm0-dev:amd64 (1.13-2\\textasciitilde{}ubuntu18.04) ...\\textbackslash{}r'\n\\item 'Setting up libpixman-1-dev:amd64 (0.34.0-2) ...\\textbackslash{}r'\n\\item 'Setting up libcairo2-dev:amd64 (1.15.10-2ubuntu0.1) ...\\textbackslash{}r'\n\\item 'Processing triggers for libc-bin (2.27-3ubuntu1.3) ...\\textbackslash{}r'\n\\item '/sbin/ldconfig.real: /usr/local/lib/python3.7/dist-packages/ideep4py/lib/libmkldnn.so.0 is not a symbolic link\\textbackslash{}r'\n\\item '\\textbackslash{}r'\n\\end{enumerate*}\n",
            "text/plain": [
              " [1] \"Reading package lists...\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      \n",
              " [2] \"Building dependency tree...\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   \n",
              " [3] \"Reading state information...\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  \n",
              " [4] \"The following additional packages will be installed:\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          \n",
              " [5] \"  libcairo-script-interpreter2 libpixman-1-dev libxcb-shm0-dev\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                \n",
              " [6] \"Suggested packages:\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           \n",
              " [7] \"  libcairo2-doc\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               \n",
              " [8] \"The following NEW packages will be installed:\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 \n",
              " [9] \"  libcairo-script-interpreter2 libcairo2-dev libpixman-1-dev libxcb-shm0-dev\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  \n",
              "[10] \"0 upgraded, 4 newly installed, 0 to remove and 39 not upgraded.\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               \n",
              "[11] \"Need to get 930 kB of archives.\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               \n",
              "[12] \"After this operation, 3,986 kB of additional disk space will be used.\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         \n",
              "[13] \"Get:1 http://archive.ubuntu.com/ubuntu bionic-updates/main amd64 libcairo-script-interpreter2 amd64 1.15.10-2ubuntu0.1 [53.5 kB]\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              \n",
              "[14] \"Get:2 http://archive.ubuntu.com/ubuntu bionic/main amd64 libpixman-1-dev amd64 0.34.0-2 [244 kB]\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              \n",
              "[15] \"Get:3 http://archive.ubuntu.com/ubuntu bionic-updates/main amd64 libxcb-shm0-dev amd64 1.13-2~ubuntu18.04 [6,684 B]\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           \n",
              "[16] \"Get:4 http://archive.ubuntu.com/ubuntu bionic-updates/main amd64 libcairo2-dev amd64 1.15.10-2ubuntu0.1 [626 kB]\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              \n",
              "[17] \"Fetched 930 kB in 1s (1,309 kB/s)\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             \n",
              "[18] \"Selecting previously unselected package libcairo-script-interpreter2:amd64.\\r\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 \n",
              "[19] \"(Reading database ... \\r(Reading database ... 5%\\r(Reading database ... 10%\\r(Reading database ... 15%\\r(Reading database ... 20%\\r(Reading database ... 25%\\r(Reading database ... 30%\\r(Reading database ... 35%\\r(Reading database ... 40%\\r(Reading database ... 45%\\r(Reading database ... 50%\\r(Reading database ... 55%\\r(Reading database ... 60%\\r(Reading database ... 65%\\r(Reading database ... 70%\\r(Reading database ... 75%\\r(Reading database ... 80%\\r(Reading database ... 85%\\r(Reading database ... 90%\\r(Reading database ... 95%\\r(Reading database ... 100%\\r(Reading database ... 155727 files and directories currently installed.)\\r\"\n",
              "[20] \"Preparing to unpack .../libcairo-script-interpreter2_1.15.10-2ubuntu0.1_amd64.deb ...\\r\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       \n",
              "[21] \"Unpacking libcairo-script-interpreter2:amd64 (1.15.10-2ubuntu0.1) ...\\r\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       \n",
              "[22] \"Selecting previously unselected package libpixman-1-dev:amd64.\\r\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              \n",
              "[23] \"Preparing to unpack .../libpixman-1-dev_0.34.0-2_amd64.deb ...\\r\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              \n",
              "[24] \"Unpacking libpixman-1-dev:amd64 (0.34.0-2) ...\\r\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              \n",
              "[25] \"Selecting previously unselected package libxcb-shm0-dev:amd64.\\r\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              \n",
              "[26] \"Preparing to unpack .../libxcb-shm0-dev_1.13-2~ubuntu18.04_amd64.deb ...\\r\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    \n",
              "[27] \"Unpacking libxcb-shm0-dev:amd64 (1.13-2~ubuntu18.04) ...\\r\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    \n",
              "[28] \"Selecting previously unselected package libcairo2-dev:amd64.\\r\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                \n",
              "[29] \"Preparing to unpack .../libcairo2-dev_1.15.10-2ubuntu0.1_amd64.deb ...\\r\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      \n",
              "[30] \"Unpacking libcairo2-dev:amd64 (1.15.10-2ubuntu0.1) ...\\r\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      \n",
              "[31] \"Setting up libcairo-script-interpreter2:amd64 (1.15.10-2ubuntu0.1) ...\\r\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      \n",
              "[32] \"Setting up libxcb-shm0-dev:amd64 (1.13-2~ubuntu18.04) ...\\r\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   \n",
              "[33] \"Setting up libpixman-1-dev:amd64 (0.34.0-2) ...\\r\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             \n",
              "[34] \"Setting up libcairo2-dev:amd64 (1.15.10-2ubuntu0.1) ...\\r\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     \n",
              "[35] \"Processing triggers for libc-bin (2.27-3ubuntu1.3) ...\\r\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      \n",
              "[36] \"/sbin/ldconfig.real: /usr/local/lib/python3.7/dist-packages/ideep4py/lib/libmkldnn.so.0 is not a symbolic link\\r\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              \n",
              "[37] \"\\r\"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            "
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "wFuTF2MUEiAk"
      },
      "source": [
        "## Install R packages"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "0c02HPWEh2nU",
        "outputId": "f6667eca-33ea-41b4-d090-b07c46d2d89d"
      },
      "source": [
        "install.packages('BiocManager')"
      ],
      "execution_count": 4,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Installing package into ‘/usr/local/lib/R/site-library’\n",
            "(as ‘lib’ is unspecified)\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "8EZ8Zm7zagc0"
      },
      "source": [
        "required_packages <- c(\n",
        "  'devtools',\n",
        "  'DESeq2',\n",
        "  'org.Mm.eg.db',\n",
        "  'clusterProfiler',\n",
        "  'TxDb.Mmusculus.UCSC.mm10.knownGene',\n",
        "  'motifmatchr',\n",
        "  'BSgenome.Mmusculus.UCSC.mm10'\n",
        ")"
      ],
      "execution_count": 5,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "icve41ciuJim",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "cbc39ab8-0b4a-47e3-fbb0-46e8974a0abd"
      },
      "source": [
        "missing_packages <- required_packages[!required_packages %in% rownames(installed.packages())]\n",
        "if (length(missing_packages) > 0){\n",
        "    BiocManager::install(missing_packages)\n",
        "}"
      ],
      "execution_count": 6,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
            "'?repositories' for details\n",
            "\n",
            "replacement repositories:\n",
            "    CRAN: https://cran.rstudio.com\n",
            "\n",
            "\n",
            "Bioconductor version 3.14 (BiocManager 1.30.16), R 4.1.3 (2022-03-10)\n",
            "\n",
            "Installing package(s) 'BiocVersion', 'DESeq2', 'org.Mm.eg.db',\n",
            "  'clusterProfiler', 'TxDb.Mmusculus.UCSC.mm10.knownGene', 'motifmatchr',\n",
            "  'BSgenome.Mmusculus.UCSC.mm10'\n",
            "\n",
            "also installing the dependencies ‘formatR’, ‘gridGraphics’, ‘tweenr’, ‘polyclip’, ‘RcppEigen’, ‘lazyeval’, ‘rjson’, ‘filelock’, ‘pracma’, ‘R.oo’, ‘R.methodsS3’, ‘lambda.r’, ‘futile.options’, ‘xtable’, ‘plogr’, ‘png’, ‘gridExtra’, ‘fastmatch’, ‘ggfun’, ‘ggplotify’, ‘patchwork’, ‘ggforce’, ‘ggrepel’, ‘viridis’, ‘tidygraph’, ‘graphlayouts’, ‘ape’, ‘tidytree’, ‘treeio’, ‘GenomicAlignments’, ‘restfulr’, ‘BiocFileCache’, ‘poweRlaw’, ‘R.utils’, ‘XVector’, ‘MatrixGenerics’, ‘DelayedArray’, ‘futile.logger’, ‘snow’, ‘BH’, ‘annotate’, ‘RSQLite’, ‘KEGGREST’, ‘DO.db’, ‘fgsea’, ‘reshape2’, ‘aplot’, ‘ggraph’, ‘igraph’, ‘scatterpie’, ‘shadowtext’, ‘ggtree’, ‘RCurl’, ‘BiocIO’, ‘rtracklayer’, ‘biomaRt’, ‘caTools’, ‘CNEr’, ‘DirichletMultinomial’, ‘gtools’, ‘seqLogo’, ‘TFMPvalue’, ‘XML’, ‘matrixStats’, ‘zlibbioc’, ‘bitops’, ‘Rhtslib’, ‘GenomeInfoDbData’, ‘S4Vectors’, ‘IRanges’, ‘GenomicRanges’, ‘SummarizedExperiment’, ‘BiocGenerics’, ‘Biobase’, ‘BiocParallel’, ‘genefilter’, ‘locfit’, ‘geneplotter’, ‘Rcpp’, ‘RcppArmadillo’, ‘AnnotationDbi’, ‘downloader’, ‘DOSE’, ‘enrichplot’, ‘GO.db’, ‘GOSemSim’, ‘plyr’, ‘qvalue’, ‘yulab.utils’, ‘GenomicFeatures’, ‘TFBSTools’, ‘Biostrings’, ‘BSgenome’, ‘Rsamtools’, ‘GenomeInfoDb’\n",
            "\n",
            "\n",
            "Old packages: 'blob', 'broom', 'haven', 'RColorBrewer', 'scales', 'vctrs'\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Download packages from Github"
      ],
      "metadata": {
        "id": "0LXZzGqK7_mK"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "devtools::install_github(\"GreenleafLab/chromVARmotifs\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "6k1WsNXQVM-L",
        "outputId": "3aeb66e0-e816-4483-834b-de55e088edff"
      },
      "execution_count": 7,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Downloading GitHub repo GreenleafLab/chromVARmotifs@HEAD\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "vctrs        (0.4.0 -> 0.4.1) [CRAN]\n",
            "blob         (1.2.2 -> 1.2.3) [CRAN]\n",
            "RColorBrewer (1.1-2 -> 1.1-3) [CRAN]\n",
            "scales       (1.1.1 -> 1.2.0) [CRAN]\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Installing 4 packages: vctrs, blob, RColorBrewer, scales\n",
            "\n",
            "Installing packages into ‘/usr/local/lib/R/site-library’\n",
            "(as ‘lib’ is unspecified)\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "* checking for file ‘/tmp/RtmpB0zzEJ/remotes565176349d/GreenleafLab-chromVARmotifs-38bed55/DESCRIPTION’ ... OK\n",
            "* preparing ‘chromVARmotifs’:\n",
            "* checking DESCRIPTION meta-information ... OK\n",
            "* checking for LF line-endings in source and make files and shell scripts\n",
            "* checking for empty or unneeded directories\n",
            "* building ‘chromVARmotifs_0.2.0.tar.gz’\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Installing package into ‘/usr/local/lib/R/site-library’\n",
            "(as ‘lib’ is unspecified)\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "devtools::install_github(\"GuangchuangYu/ChIPseeker\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Tal1Qux-ZqHI",
        "outputId": "ec5a3810-d9e8-4bc4-e5d5-eecc140b85b1"
      },
      "execution_count": 8,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Downloading GitHub repo GuangchuangYu/ChIPseeker@HEAD\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "enrichplot   (1.14.2 -> d11b60bb6...) [GitHub]\n",
            "TxDb.Hsap... (NA     -> 3.2.2       ) [CRAN]\n",
            "plotrix      (NA     -> 3.8-2       ) [CRAN]\n",
            "gplots       (NA     -> 3.1.1       ) [CRAN]\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Installing 3 packages: TxDb.Hsapiens.UCSC.hg19.knownGene, plotrix, gplots\n",
            "\n",
            "Installing packages into ‘/usr/local/lib/R/site-library’\n",
            "(as ‘lib’ is unspecified)\n",
            "\n",
            "Downloading GitHub repo GuangchuangYu/enrichplot@HEAD\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "* checking for file ‘/tmp/RtmpB0zzEJ/remotes563a87cf36/GuangchuangYu-enrichplot-d11b60b/DESCRIPTION’ ... OK\n",
            "* preparing ‘enrichplot’:\n",
            "* checking DESCRIPTION meta-information ... OK\n",
            "* checking for LF line-endings in source and make files and shell scripts\n",
            "* checking for empty or unneeded directories\n",
            "* building ‘enrichplot_1.13.1.992.tar.gz’\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Installing package into ‘/usr/local/lib/R/site-library’\n",
            "(as ‘lib’ is unspecified)\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "* checking for file ‘/tmp/RtmpB0zzEJ/remotes5610c4fa2f/YuLab-SMU-ChIPseeker-5945269/DESCRIPTION’ ... OK\n",
            "* preparing ‘ChIPseeker’:\n",
            "* checking DESCRIPTION meta-information ... OK\n",
            "* checking for LF line-endings in source and make files and shell scripts\n",
            "* checking for empty or unneeded directories\n",
            "* looking to see if a ‘data/datalist’ file should be added\n",
            "* building ‘ChIPseeker_1.31.4.tar.gz’\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Installing package into ‘/usr/local/lib/R/site-library’\n",
            "(as ‘lib’ is unspecified)\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "-r47NH7jiD99"
      },
      "source": [
        "## Load R packages"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "nQFQk04TzjFj",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "589d4412-54d6-4e8e-9a9e-cc7ca876ca9a"
      },
      "source": [
        "library(tidyr)\n",
        "library(dplyr)\n",
        "library(ggplot2) # colorRamp2\n",
        "library(parallel)\n",
        "library(DESeq2)\n",
        "library(ChIPseeker)\n",
        "library(clusterProfiler)\n",
        "library(TxDb.Mmusculus.UCSC.mm10.knownGene)\n",
        "library(org.Mm.eg.db)\n",
        "library(motifmatchr)\n",
        "library(chromVARmotifs)\n",
        "library(BSgenome.Mmusculus.UCSC.mm10)\n",
        "library(data.table)"
      ],
      "execution_count": 9,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "\n",
            "Attaching package: ‘dplyr’\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:stats’:\n",
            "\n",
            "    filter, lag\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:base’:\n",
            "\n",
            "    intersect, setdiff, setequal, union\n",
            "\n",
            "\n",
            "Loading required package: S4Vectors\n",
            "\n",
            "Loading required package: stats4\n",
            "\n",
            "Loading required package: BiocGenerics\n",
            "\n",
            "\n",
            "Attaching package: ‘BiocGenerics’\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:dplyr’:\n",
            "\n",
            "    combine, intersect, setdiff, union\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:stats’:\n",
            "\n",
            "    IQR, mad, sd, var, xtabs\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:base’:\n",
            "\n",
            "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
            "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
            "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
            "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
            "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
            "    union, unique, unsplit, which.max, which.min\n",
            "\n",
            "\n",
            "\n",
            "Attaching package: ‘S4Vectors’\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:dplyr’:\n",
            "\n",
            "    first, rename\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:tidyr’:\n",
            "\n",
            "    expand\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:base’:\n",
            "\n",
            "    expand.grid, I, unname\n",
            "\n",
            "\n",
            "Loading required package: IRanges\n",
            "\n",
            "\n",
            "Attaching package: ‘IRanges’\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:dplyr’:\n",
            "\n",
            "    collapse, desc, slice\n",
            "\n",
            "\n",
            "Loading required package: GenomicRanges\n",
            "\n",
            "Loading required package: GenomeInfoDb\n",
            "\n",
            "Loading required package: SummarizedExperiment\n",
            "\n",
            "Loading required package: MatrixGenerics\n",
            "\n",
            "Loading required package: matrixStats\n",
            "\n",
            "\n",
            "Attaching package: ‘matrixStats’\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:dplyr’:\n",
            "\n",
            "    count\n",
            "\n",
            "\n",
            "\n",
            "Attaching package: ‘MatrixGenerics’\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:matrixStats’:\n",
            "\n",
            "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
            "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
            "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
            "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
            "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
            "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
            "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
            "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
            "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
            "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
            "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
            "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
            "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
            "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
            "    rowWeightedSds, rowWeightedVars\n",
            "\n",
            "\n",
            "Loading required package: Biobase\n",
            "\n",
            "Welcome to Bioconductor\n",
            "\n",
            "    Vignettes contain introductory material; view with\n",
            "    'browseVignettes()'. To cite Bioconductor, see\n",
            "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
            "\n",
            "\n",
            "\n",
            "Attaching package: ‘Biobase’\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:MatrixGenerics’:\n",
            "\n",
            "    rowMedians\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:matrixStats’:\n",
            "\n",
            "    anyMissing, rowMedians\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "Registered S3 method overwritten by 'ggtree':\n",
            "  method      from \n",
            "  identify.gg ggfun\n",
            "\n",
            "ChIPseeker v1.31.4  For help: https://guangchuangyu.github.io/software/ChIPseeker\n",
            "\n",
            "If you use ChIPseeker in published research, please cite:\n",
            "Guangchuang Yu, Li-Gen Wang, Qing-Yu He. ChIPseeker: an R/Bioconductor package for ChIP peak annotation, comparison and visualization. Bioinformatics 2015, 31(14):2382-2383\n",
            "\n",
            "clusterProfiler v4.2.2  For help: https://yulab-smu.top/biomedical-knowledge-mining-book/\n",
            "\n",
            "If you use clusterProfiler in published research, please cite:\n",
            "T Wu, E Hu, S Xu, M Chen, P Guo, Z Dai, T Feng, L Zhou, W Tang, L Zhan, X Fu, S Liu, X Bo, and G Yu. clusterProfiler 4.0: A universal enrichment tool for interpreting omics data. The Innovation. 2021, 2(3):100141\n",
            "\n",
            "\n",
            "Attaching package: ‘clusterProfiler’\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:IRanges’:\n",
            "\n",
            "    slice\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:S4Vectors’:\n",
            "\n",
            "    rename\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:stats’:\n",
            "\n",
            "    filter\n",
            "\n",
            "\n",
            "Loading required package: GenomicFeatures\n",
            "\n",
            "Loading required package: AnnotationDbi\n",
            "\n",
            "\n",
            "Attaching package: ‘AnnotationDbi’\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:clusterProfiler’:\n",
            "\n",
            "    select\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:dplyr’:\n",
            "\n",
            "    select\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "Loading required package: BSgenome\n",
            "\n",
            "Loading required package: Biostrings\n",
            "\n",
            "Loading required package: XVector\n",
            "\n",
            "\n",
            "Attaching package: ‘Biostrings’\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:base’:\n",
            "\n",
            "    strsplit\n",
            "\n",
            "\n",
            "Loading required package: rtracklayer\n",
            "\n",
            "\n",
            "Attaching package: ‘data.table’\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:SummarizedExperiment’:\n",
            "\n",
            "    shift\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:GenomicRanges’:\n",
            "\n",
            "    shift\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:IRanges’:\n",
            "\n",
            "    shift\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:S4Vectors’:\n",
            "\n",
            "    first, second\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:dplyr’:\n",
            "\n",
            "    between, first, last\n",
            "\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Read regions from SeATAC for the putative OSK motifs"
      ],
      "metadata": {
        "id": "zXcfEy-Q7tW0"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "tab <- read.table('https://s3.msi.umn.edu/gongx030/projects/seatac_project/config/seatac_OSKM/results.tsv', sep = '\\t', header = TRUE)"
      ],
      "metadata": {
        "id": "B-CSShxCE-FG"
      },
      "execution_count": 10,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "GmhFQtDRRXcw"
      },
      "source": [
        "mef <- read.table(tab[tab$vae_latent == 5,]$remote_output_file, header = FALSE, sep = '\\t')"
      ],
      "execution_count": 11,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "colnames(mef) <- c('seqnames','start','end','width','strand','pvalue_z','padj','nucleosome_treatment','nucleosome_control','log_ratio')"
      ],
      "metadata": {
        "id": "idD6YFdQpqvt"
      },
      "execution_count": 12,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "mef <- GRanges(mef)"
      ],
      "metadata": {
        "id": "rDfqsIpe5XSK"
      },
      "execution_count": 13,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "QijS2vz2fJAI",
        "outputId": "27aaf7b0-568a-4b7f-e7d0-cf1fa05edc43"
      },
      "source": [
        "sprintf('number of motif-centric regions: %s', length(mef)) %>% message()"
      ],
      "execution_count": 14,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "number of motif-centric regions: 282789\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "is_signif <- mef$padj < 0.05\n",
        "sprintf('# of significantly differential V-plots called by SeATAC: %d', sum(is_signif)) %>% message()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "aKcjBPE6E9Rb",
        "outputId": "8e97efba-d4cb-43d4-c793-a2213b845601"
      },
      "execution_count": 15,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "# of significantly differential V-plots called by SeATAC: 19603\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "#OSK individual motifs"
      ],
      "metadata": {
        "id": "W8DMfzjBYgto"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "osk_dat <- read.table('https://s3.msi.umn.edu/dsouz072/datasets/dataset=seatac_file_version=20200209a/Seatac_dataset.tsv', header = TRUE, sep = '\\t')\n",
        "head(osk_dat)\n",
        "osk_bam <- osk_dat[grepl('https://s3.msi.umn.edu/dsouz072/datasets/dataset=iPSC_reprogramming_version=20200208a/', osk_dat$BAM),]\n",
        "osk <- osk_bam[grepl('OSK-D0-rep|OSK-D1-rep|OSK-D7-rep|_mES-OSK-|MEF-rep1|_ESCs-rep1|_iPS-rep1', osk_bam$BAM),]\n",
        "\n",
        "ARGS <- list(\n",
        "  bin_size = 5L,\n",
        "  fragment_size_range_max = 640L,\n",
        "  fragment_size_interval = 10L,\n",
        "  window_size = 1000L,\n",
        "  heatmap_window_size = 2000L,\n",
        "  block_size = 640L,\n",
        "  sliding_step = 20L,\n",
        "  min_reads = 5L,\n",
        "  step_size = 200L,\n",
        "  \n",
        "  vae_num_peaks = 5000L,\n",
        "  vae_latent = c(10L),\n",
        "  vae_filters0 = c(128L),\n",
        "  vae_filters = c(32L),\n",
        "  vae_downsample_layers = 4L,\n",
        "  vae_upsample_layers = 4L,\n",
        "  vae_base_learning_rate = c(1e-2),\n",
        "  vae_epochs = 200L,\n",
        "  vae_beta_schedule_method = c('constant'),\n",
        "  vae_beta0 = 1,\n",
        "  vae_batch_size = c(128L),\n",
        "  vae_warmup_ratio = 0.1,\n",
        "  \n",
        "  target_width = 100L,\n",
        "  min_test_reads = 1L,\n",
        "  max_test_reads = 10L,\n",
        "  center_width = 100L # the center width\n",
        ")\n",
        "\n",
        "gr_all <- lapply(as.character(osk$BED), function(bed_file){\n",
        "  peaks <- read.table(bed_file, header = FALSE, sep = '\\t')\n",
        "  peaks <- GRanges(seqnames = peaks[, 1], range = IRanges(start = peaks[, 2], end = peaks[, 3]))\n",
        "  peaks <- resize(peaks, fix = 'center', width = ARGS[['step_size']])\n",
        "  peaks\n",
        "}) %>%\n",
        "  GRangesList() %>%\n",
        "  coverage() %>%\n",
        "  IRanges::slice(lower = 1, rangesOnly = TRUE) %>%\n",
        "  GRanges() %>%\n",
        "  resize(fix = 'center', width = ARGS[['window_size']])\n",
        "gr_all <- gr_all[!seqnames(gr_all) %in% c('chrM', 'chrY')]\n",
        "\n",
        "\n",
        "\n",
        "data('homer_pwms')\n",
        "\n",
        "\n",
        "\n",
        "gr_oct4 <- matchMotifs(homer_pwms[\"Oct4(POU,Homeobox)/mES-Oct4-ChIP-Seq(GSE11431)/Homer\"], gr_all, genome = BSgenome.Mmusculus.UCSC.mm10, out = 'positions')[[1]] %>%\n",
        "  resize(width = ARGS[['block_size']], fix = 'center')\n",
        "gr_sox2 <- matchMotifs(homer_pwms[\"Sox2(HMG)/mES-Sox2-ChIP-Seq(GSE11431)/Homer\"], gr_all, genome = BSgenome.Mmusculus.UCSC.mm10, out = 'positions')[[1]] %>%\n",
        "  resize(width = ARGS[['block_size']], fix = 'center')\n",
        "gr_klf4 <- matchMotifs(homer_pwms[\"Klf4(Zf)/mES-Klf4-ChIP-Seq(GSE11431)/Homer\"], gr_all, genome = BSgenome.Mmusculus.UCSC.mm10, out = 'positions')[[1]] %>%\n",
        "  resize(width = ARGS[['block_size']], fix = 'center')"
      ],
      "metadata": {
        "id": "Sntz3vKLYmaZ",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 390
        },
        "outputId": "d3b4ec18-12da-4167-eb4e-9c7ec5d4aadf"
      },
      "execution_count": 16,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/html": [
              "<table class=\"dataframe\">\n",
              "<caption>A data.frame: 6 × 4</caption>\n",
              "<thead>\n",
              "\t<tr><th></th><th scope=col>BAM</th><th scope=col>BED</th><th scope=col>Nucleoatac</th><th scope=col>Genome</th></tr>\n",
              "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
              "</thead>\n",
              "<tbody>\n",
              "\t<tr><th scope=row>1</th><td>https://s3.msi.umn.edu/dsouz072/datasets/dataset=Baf_version=20200129a/Baf_rep1.dedup.bam          </td><td>https://s3.msi.umn.edu/dsouz072/datasets/dataset=Baf_version=20200129a/Baf_rep1_summits.bed          </td><td>https://s3.msi.umn.edu/dsouz072/datasets/dataset=Baf_version=20200129a/Baf_rep1.nucleoatac_signal.smooth.bw </td><td>hg19</td></tr>\n",
              "\t<tr><th scope=row>2</th><td>https://s3.msi.umn.edu/dsouz072/datasets/dataset=Baf_version=20200129a/Baf_rep2.dedup.bam          </td><td>https://s3.msi.umn.edu/dsouz072/datasets/dataset=Baf_version=20200129a/Baf_rep2_summits.bed          </td><td>https://s3.msi.umn.edu/dsouz072/datasets/dataset=Baf_version=20200129a/Baf_rep2.nucleoatac_signal.smooth.bw </td><td>hg19</td></tr>\n",
              "\t<tr><th scope=row>3</th><td>https://s3.msi.umn.edu/dsouz072/datasets/dataset=Baf_version=20200129a/Ctrl_rep1.dedup.bam         </td><td>https://s3.msi.umn.edu/dsouz072/datasets/dataset=Baf_version=20200129a/Ctrl_rep1_summits.bed         </td><td>https://s3.msi.umn.edu/dsouz072/datasets/dataset=Baf_version=20200129a/Ctrl_rep1.nucleoatac_signal.smooth.bw</td><td>hg19</td></tr>\n",
              "\t<tr><th scope=row>4</th><td>https://s3.msi.umn.edu/dsouz072/datasets/dataset=Baf_version=20200129a/Ctrl_rep2.dedup.bam         </td><td>https://s3.msi.umn.edu/dsouz072/datasets/dataset=Baf_version=20200129a/Ctrl_rep2_summits.bed         </td><td>https://s3.msi.umn.edu/dsouz072/datasets/dataset=Baf_version=20200129a/Ctrl_rep2.nucleoatac_signal.smooth.bw</td><td>hg19</td></tr>\n",
              "\t<tr><th scope=row>5</th><td>https://s3.msi.umn.edu/dsouz072/datasets/dataset=Ascl1_version=20200128a/MEF.dedup.bam             </td><td>https://s3.msi.umn.edu/dsouz072/datasets/dataset=Ascl1_version=20200128a/MEF_summits.bed             </td><td>NA                                                                                                          </td><td>mm10</td></tr>\n",
              "\t<tr><th scope=row>6</th><td>https://s3.msi.umn.edu/dsouz072/datasets/dataset=Ascl1_version=20200128a/MEF_rtTA_B1_rep1.dedup.bam</td><td>https://s3.msi.umn.edu/dsouz072/datasets/dataset=Ascl1_version=20200128a/MEF_rtTA_B1_rep1_summits.bed</td><td>NA                                                                                                          </td><td>mm10</td></tr>\n",
              "</tbody>\n",
              "</table>\n"
            ],
            "text/markdown": "\nA data.frame: 6 × 4\n\n| <!--/--> | BAM &lt;chr&gt; | BED &lt;chr&gt; | Nucleoatac &lt;chr&gt; | Genome &lt;chr&gt; |\n|---|---|---|---|---|\n| 1 | https://s3.msi.umn.edu/dsouz072/datasets/dataset=Baf_version=20200129a/Baf_rep1.dedup.bam           | https://s3.msi.umn.edu/dsouz072/datasets/dataset=Baf_version=20200129a/Baf_rep1_summits.bed           | https://s3.msi.umn.edu/dsouz072/datasets/dataset=Baf_version=20200129a/Baf_rep1.nucleoatac_signal.smooth.bw  | hg19 |\n| 2 | https://s3.msi.umn.edu/dsouz072/datasets/dataset=Baf_version=20200129a/Baf_rep2.dedup.bam           | https://s3.msi.umn.edu/dsouz072/datasets/dataset=Baf_version=20200129a/Baf_rep2_summits.bed           | https://s3.msi.umn.edu/dsouz072/datasets/dataset=Baf_version=20200129a/Baf_rep2.nucleoatac_signal.smooth.bw  | hg19 |\n| 3 | https://s3.msi.umn.edu/dsouz072/datasets/dataset=Baf_version=20200129a/Ctrl_rep1.dedup.bam          | https://s3.msi.umn.edu/dsouz072/datasets/dataset=Baf_version=20200129a/Ctrl_rep1_summits.bed          | https://s3.msi.umn.edu/dsouz072/datasets/dataset=Baf_version=20200129a/Ctrl_rep1.nucleoatac_signal.smooth.bw | hg19 |\n| 4 | https://s3.msi.umn.edu/dsouz072/datasets/dataset=Baf_version=20200129a/Ctrl_rep2.dedup.bam          | https://s3.msi.umn.edu/dsouz072/datasets/dataset=Baf_version=20200129a/Ctrl_rep2_summits.bed          | https://s3.msi.umn.edu/dsouz072/datasets/dataset=Baf_version=20200129a/Ctrl_rep2.nucleoatac_signal.smooth.bw | hg19 |\n| 5 | https://s3.msi.umn.edu/dsouz072/datasets/dataset=Ascl1_version=20200128a/MEF.dedup.bam              | https://s3.msi.umn.edu/dsouz072/datasets/dataset=Ascl1_version=20200128a/MEF_summits.bed              | NA                                                                                                           | mm10 |\n| 6 | https://s3.msi.umn.edu/dsouz072/datasets/dataset=Ascl1_version=20200128a/MEF_rtTA_B1_rep1.dedup.bam | https://s3.msi.umn.edu/dsouz072/datasets/dataset=Ascl1_version=20200128a/MEF_rtTA_B1_rep1_summits.bed | NA                                                                                                           | mm10 |\n\n",
            "text/latex": "A data.frame: 6 × 4\n\\begin{tabular}{r|llll}\n  & BAM & BED & Nucleoatac & Genome\\\\\n  & <chr> & <chr> & <chr> & <chr>\\\\\n\\hline\n\t1 & https://s3.msi.umn.edu/dsouz072/datasets/dataset=Baf\\_version=20200129a/Baf\\_rep1.dedup.bam           & https://s3.msi.umn.edu/dsouz072/datasets/dataset=Baf\\_version=20200129a/Baf\\_rep1\\_summits.bed           & https://s3.msi.umn.edu/dsouz072/datasets/dataset=Baf\\_version=20200129a/Baf\\_rep1.nucleoatac\\_signal.smooth.bw  & hg19\\\\\n\t2 & https://s3.msi.umn.edu/dsouz072/datasets/dataset=Baf\\_version=20200129a/Baf\\_rep2.dedup.bam           & https://s3.msi.umn.edu/dsouz072/datasets/dataset=Baf\\_version=20200129a/Baf\\_rep2\\_summits.bed           & https://s3.msi.umn.edu/dsouz072/datasets/dataset=Baf\\_version=20200129a/Baf\\_rep2.nucleoatac\\_signal.smooth.bw  & hg19\\\\\n\t3 & https://s3.msi.umn.edu/dsouz072/datasets/dataset=Baf\\_version=20200129a/Ctrl\\_rep1.dedup.bam          & https://s3.msi.umn.edu/dsouz072/datasets/dataset=Baf\\_version=20200129a/Ctrl\\_rep1\\_summits.bed          & https://s3.msi.umn.edu/dsouz072/datasets/dataset=Baf\\_version=20200129a/Ctrl\\_rep1.nucleoatac\\_signal.smooth.bw & hg19\\\\\n\t4 & https://s3.msi.umn.edu/dsouz072/datasets/dataset=Baf\\_version=20200129a/Ctrl\\_rep2.dedup.bam          & https://s3.msi.umn.edu/dsouz072/datasets/dataset=Baf\\_version=20200129a/Ctrl\\_rep2\\_summits.bed          & https://s3.msi.umn.edu/dsouz072/datasets/dataset=Baf\\_version=20200129a/Ctrl\\_rep2.nucleoatac\\_signal.smooth.bw & hg19\\\\\n\t5 & https://s3.msi.umn.edu/dsouz072/datasets/dataset=Ascl1\\_version=20200128a/MEF.dedup.bam              & https://s3.msi.umn.edu/dsouz072/datasets/dataset=Ascl1\\_version=20200128a/MEF\\_summits.bed              & NA                                                                                                           & mm10\\\\\n\t6 & https://s3.msi.umn.edu/dsouz072/datasets/dataset=Ascl1\\_version=20200128a/MEF\\_rtTA\\_B1\\_rep1.dedup.bam & https://s3.msi.umn.edu/dsouz072/datasets/dataset=Ascl1\\_version=20200128a/MEF\\_rtTA\\_B1\\_rep1\\_summits.bed & NA                                                                                                           & mm10\\\\\n\\end{tabular}\n",
            "text/plain": [
              "  BAM                                                                                                \n",
              "1 https://s3.msi.umn.edu/dsouz072/datasets/dataset=Baf_version=20200129a/Baf_rep1.dedup.bam          \n",
              "2 https://s3.msi.umn.edu/dsouz072/datasets/dataset=Baf_version=20200129a/Baf_rep2.dedup.bam          \n",
              "3 https://s3.msi.umn.edu/dsouz072/datasets/dataset=Baf_version=20200129a/Ctrl_rep1.dedup.bam         \n",
              "4 https://s3.msi.umn.edu/dsouz072/datasets/dataset=Baf_version=20200129a/Ctrl_rep2.dedup.bam         \n",
              "5 https://s3.msi.umn.edu/dsouz072/datasets/dataset=Ascl1_version=20200128a/MEF.dedup.bam             \n",
              "6 https://s3.msi.umn.edu/dsouz072/datasets/dataset=Ascl1_version=20200128a/MEF_rtTA_B1_rep1.dedup.bam\n",
              "  BED                                                                                                  \n",
              "1 https://s3.msi.umn.edu/dsouz072/datasets/dataset=Baf_version=20200129a/Baf_rep1_summits.bed          \n",
              "2 https://s3.msi.umn.edu/dsouz072/datasets/dataset=Baf_version=20200129a/Baf_rep2_summits.bed          \n",
              "3 https://s3.msi.umn.edu/dsouz072/datasets/dataset=Baf_version=20200129a/Ctrl_rep1_summits.bed         \n",
              "4 https://s3.msi.umn.edu/dsouz072/datasets/dataset=Baf_version=20200129a/Ctrl_rep2_summits.bed         \n",
              "5 https://s3.msi.umn.edu/dsouz072/datasets/dataset=Ascl1_version=20200128a/MEF_summits.bed             \n",
              "6 https://s3.msi.umn.edu/dsouz072/datasets/dataset=Ascl1_version=20200128a/MEF_rtTA_B1_rep1_summits.bed\n",
              "  Nucleoatac                                                                                                  \n",
              "1 https://s3.msi.umn.edu/dsouz072/datasets/dataset=Baf_version=20200129a/Baf_rep1.nucleoatac_signal.smooth.bw \n",
              "2 https://s3.msi.umn.edu/dsouz072/datasets/dataset=Baf_version=20200129a/Baf_rep2.nucleoatac_signal.smooth.bw \n",
              "3 https://s3.msi.umn.edu/dsouz072/datasets/dataset=Baf_version=20200129a/Ctrl_rep1.nucleoatac_signal.smooth.bw\n",
              "4 https://s3.msi.umn.edu/dsouz072/datasets/dataset=Baf_version=20200129a/Ctrl_rep2.nucleoatac_signal.smooth.bw\n",
              "5 NA                                                                                                          \n",
              "6 NA                                                                                                          \n",
              "  Genome\n",
              "1 hg19  \n",
              "2 hg19  \n",
              "3 hg19  \n",
              "4 hg19  \n",
              "5 mm10  \n",
              "6 mm10  "
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "gr_oct4 <- matchMotifs(homer_pwms[\"Oct4(POU,Homeobox)/mES-Oct4-ChIP-Seq(GSE11431)/Homer\"], mef, genome = BSgenome.Mmusculus.UCSC.mm10, out = 'positions')[[1]] %>%\n",
        "  resize(width = ARGS[['block_size']], fix = 'center')\n",
        "gr_sox2 <- matchMotifs(homer_pwms[\"Sox2(HMG)/mES-Sox2-ChIP-Seq(GSE11431)/Homer\"], mef, genome = BSgenome.Mmusculus.UCSC.mm10, out = 'positions')[[1]] %>%\n",
        "  resize(width = ARGS[['block_size']], fix = 'center')\n",
        "gr_klf4 <- matchMotifs(homer_pwms[\"Klf4(Zf)/mES-Klf4-ChIP-Seq(GSE11431)/Homer\"], mef, genome = BSgenome.Mmusculus.UCSC.mm10, out = 'positions')[[1]] %>%\n",
        "  resize(width = ARGS[['block_size']], fix = 'center')"
      ],
      "metadata": {
        "id": "ZWU2HC4wbVOf"
      },
      "execution_count": 17,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "mef_oct4 <- subsetByOverlaps(mef, gr_oct4)\n",
        "mef_sox2 <- subsetByOverlaps(mef, gr_sox2)\n",
        "mef_klf4 <- subsetByOverlaps(mef, gr_klf4)"
      ],
      "metadata": {
        "id": "DsOJ3BHibySX"
      },
      "execution_count": 18,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "osk_gr <- list(oct4 = mef_oct4, sox2 = mef_sox2, klf4 = mef_klf4)"
      ],
      "metadata": {
        "id": "XAHMlru9cLhW"
      },
      "execution_count": 19,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "suGsvJ4wtKZD"
      },
      "source": [
        "## Read bulk RNA-seq of MEF reprogramming"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "remote_res <- c(rna = 'https://s3.msi.umn.edu/dsouz072/datasets/dataset=iPSC_reprogramming_version=20200208a/GSE93027_RNAseq_tag_count.tsv.gz')\n",
        "for (f in remote_res){\n",
        "  system(sprintf('wget -cx %s', f))\n",
        "}\n",
        "rna <- gsub('https://', './', remote_res)"
      ],
      "metadata": {
        "id": "pKqGl75rDJkA"
      },
      "execution_count": 20,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "frgpTS3xhprU"
      },
      "source": [
        "bulk <- read.table(gzfile(rna[['rna']]), header = TRUE, sep = '\\t')"
      ],
      "execution_count": 21,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Assigning gene symbol to Ensembl gene ID"
      ],
      "metadata": {
        "id": "EhxyUDnf8pXk"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "t2g_dat <- read.table('https://s3.msi.umn.edu/dsouz072/datasets/dataset=Foxk1_version=20191130a/transcripts_to_genes.txt', sep = '\\t')"
      ],
      "metadata": {
        "id": "sAvwPBTAFPOH"
      },
      "execution_count": 22,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "colnames(t2g_dat) <- c('ensembl_transcript', 'ensembl_gene', 'gene_name')\n",
        "head(t2g_dat)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 286
        },
        "id": "QhAkb5TsFRQg",
        "outputId": "8e12537b-8822-4c91-e7b9-6139b34f360a"
      },
      "execution_count": 23,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/html": [
              "<table class=\"dataframe\">\n",
              "<caption>A data.frame: 6 × 3</caption>\n",
              "<thead>\n",
              "\t<tr><th></th><th scope=col>ensembl_transcript</th><th scope=col>ensembl_gene</th><th scope=col>gene_name</th></tr>\n",
              "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
              "</thead>\n",
              "<tbody>\n",
              "\t<tr><th scope=row>1</th><td>ENSMUST00000193812.1</td><td>ENSMUSG00000102693.1</td><td>4933401J01Rik</td></tr>\n",
              "\t<tr><th scope=row>2</th><td>ENSMUST00000082908.1</td><td>ENSMUSG00000064842.1</td><td>Gm26206      </td></tr>\n",
              "\t<tr><th scope=row>3</th><td>ENSMUST00000162897.1</td><td>ENSMUSG00000051951.5</td><td>Xkr4         </td></tr>\n",
              "\t<tr><th scope=row>4</th><td>ENSMUST00000159265.1</td><td>ENSMUSG00000051951.5</td><td>Xkr4         </td></tr>\n",
              "\t<tr><th scope=row>5</th><td>ENSMUST00000070533.4</td><td>ENSMUSG00000051951.5</td><td>Xkr4         </td></tr>\n",
              "\t<tr><th scope=row>6</th><td>ENSMUST00000192857.1</td><td>ENSMUSG00000102851.1</td><td>Gm18956      </td></tr>\n",
              "</tbody>\n",
              "</table>\n"
            ],
            "text/markdown": "\nA data.frame: 6 × 3\n\n| <!--/--> | ensembl_transcript &lt;chr&gt; | ensembl_gene &lt;chr&gt; | gene_name &lt;chr&gt; |\n|---|---|---|---|\n| 1 | ENSMUST00000193812.1 | ENSMUSG00000102693.1 | 4933401J01Rik |\n| 2 | ENSMUST00000082908.1 | ENSMUSG00000064842.1 | Gm26206       |\n| 3 | ENSMUST00000162897.1 | ENSMUSG00000051951.5 | Xkr4          |\n| 4 | ENSMUST00000159265.1 | ENSMUSG00000051951.5 | Xkr4          |\n| 5 | ENSMUST00000070533.4 | ENSMUSG00000051951.5 | Xkr4          |\n| 6 | ENSMUST00000192857.1 | ENSMUSG00000102851.1 | Gm18956       |\n\n",
            "text/latex": "A data.frame: 6 × 3\n\\begin{tabular}{r|lll}\n  & ensembl\\_transcript & ensembl\\_gene & gene\\_name\\\\\n  & <chr> & <chr> & <chr>\\\\\n\\hline\n\t1 & ENSMUST00000193812.1 & ENSMUSG00000102693.1 & 4933401J01Rik\\\\\n\t2 & ENSMUST00000082908.1 & ENSMUSG00000064842.1 & Gm26206      \\\\\n\t3 & ENSMUST00000162897.1 & ENSMUSG00000051951.5 & Xkr4         \\\\\n\t4 & ENSMUST00000159265.1 & ENSMUSG00000051951.5 & Xkr4         \\\\\n\t5 & ENSMUST00000070533.4 & ENSMUSG00000051951.5 & Xkr4         \\\\\n\t6 & ENSMUST00000192857.1 & ENSMUSG00000102851.1 & Gm18956      \\\\\n\\end{tabular}\n",
            "text/plain": [
              "  ensembl_transcript   ensembl_gene         gene_name    \n",
              "1 ENSMUST00000193812.1 ENSMUSG00000102693.1 4933401J01Rik\n",
              "2 ENSMUST00000082908.1 ENSMUSG00000064842.1 Gm26206      \n",
              "3 ENSMUST00000162897.1 ENSMUSG00000051951.5 Xkr4         \n",
              "4 ENSMUST00000159265.1 ENSMUSG00000051951.5 Xkr4         \n",
              "5 ENSMUST00000070533.4 ENSMUSG00000051951.5 Xkr4         \n",
              "6 ENSMUST00000192857.1 ENSMUSG00000102851.1 Gm18956      "
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "t2g_dat$ensg <- gsub('\\\\.[0-9]+','', t2g_dat$ensembl_gene)"
      ],
      "metadata": {
        "id": "m85WTiEKF1Bl"
      },
      "execution_count": 24,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "gene_dat <- data.frame(ensg = bulk$X)"
      ],
      "metadata": {
        "id": "Z00mC-TTGZli"
      },
      "execution_count": 25,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "gene_dat$gene_name <-t2g_dat$gene_name[match(gene_dat$ensg, t2g_dat$ensg)]"
      ],
      "metadata": {
        "id": "b4QtKGidGm77"
      },
      "execution_count": 26,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "bulk$X <- gene_dat$gene_name[match(bulk$X, gene_dat$ensg)]"
      ],
      "metadata": {
        "id": "4yfiF_RbG2ZY"
      },
      "execution_count": 27,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Summarized experiment file with count data"
      ],
      "metadata": {
        "id": "dZWs6W_I8u1I"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "se_bulk <- SummarizedExperiment(list(counts = as.matrix(bulk[!colnames(bulk) %in% c('X')])), \n",
        "  rowData = DataFrame(SYMBOL = bulk$X),\n",
        "  colData = DataFrame(id = colnames(bulk[!colnames(bulk) %in% c('X')])))"
      ],
      "metadata": {
        "id": "ixRmqDDf-Hxl"
      },
      "execution_count": 28,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "se_bulk$group <- gsub('\\\\.','_', se_bulk$id)\n",
        "se_bulk$group <- gsub('_rep[0-9]','', se_bulk$group)"
      ],
      "metadata": {
        "id": "ih8XywFw_Lbf"
      },
      "execution_count": 29,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "se_bulk"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 181
        },
        "id": "MydLwiRCCnuR",
        "outputId": "d2bd3a95-aea8-4f66-b1bf-f7299bdc67e6"
      },
      "execution_count": 30,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "class: SummarizedExperiment \n",
              "dim: 18198 26 \n",
              "metadata(0):\n",
              "assays(1): counts\n",
              "rownames: NULL\n",
              "rowData names(1): SYMBOL\n",
              "colnames(26): ESCs.rep1 ESCs.rep2 ... OSK.shSap30.D7.rep1\n",
              "  OSK.shSap30.D7.rep2\n",
              "colData names(2): id group"
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "#Deseq2"
      ],
      "metadata": {
        "id": "_8zGNjolN7dH"
      }
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "l4iZYk_Ihqzu"
      },
      "source": [
        "## Find  DE genes between MEF and D7 cells post induction"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "BomFTf6KsnEp",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "54670db1-0060-406e-9d56-78590d992a5e"
      },
      "source": [
        "se_bulk <- DESeqDataSet(se_bulk, design = ~ group)\n",
        "se_bulk <- estimateSizeFactors(se_bulk)\n",
        "se_bulk <- DESeq(se_bulk)\n",
        "assays(se_bulk)$normalized_counts <- log2(counts(se_bulk, normalized = TRUE) + 1)\n",
        "res <- results(se_bulk, contrast = c('group', 'OKS_D7', 'MEF'))"
      ],
      "execution_count": 31,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message in DESeqDataSet(se_bulk, design = ~group):\n",
            "“some variables in design formula are characters, converting to factors”\n",
            "using pre-existing size factors\n",
            "\n",
            "estimating dispersions\n",
            "\n",
            "gene-wise dispersion estimates\n",
            "\n",
            "mean-dispersion relationship\n",
            "\n",
            "-- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
            "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
            "   specify fitType='local' or 'mean' to avoid this message next time.\n",
            "\n",
            "final dispersion estimates\n",
            "\n",
            "fitting model and testing\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "res$SYMBOL <- rowData(se_bulk)$SYMBOL"
      ],
      "metadata": {
        "id": "oDj4t6dQDy8n"
      },
      "execution_count": 32,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "summary(res)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "htqAln5jD2jG",
        "outputId": "1f693192-56dd-460a-be5c-e705364f3043"
      },
      "execution_count": 33,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "out of 17950 with nonzero total read count\n",
            "adjusted p-value < 0.1\n",
            "LFC > 0 (up)       : 3548, 20%\n",
            "LFC < 0 (down)     : 3248, 18%\n",
            "outliers [1]       : 0, 0%\n",
            "low counts [2]     : 1391, 7.7%\n",
            "(mean count < 2)\n",
            "[1] see 'cooksCutoff' argument of ?results\n",
            "[2] see 'independentFiltering' argument of ?results\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# NFR -> NOR \n",
        "(MEF to D7 MEFs post OSK induction)"
      ],
      "metadata": {
        "id": "cv544KQl52bW"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "down <- lapply(osk_gr, function(o){\n",
        "  o[o$log_ratio > 0.1 & o$padj < 0.05]\n",
        "})"
      ],
      "metadata": {
        "id": "Bc-STbROeDdV"
      },
      "execution_count": 34,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "anno <- lapply(down, function(d){\n",
        "  annotatePeak(d, tssRegion = c(-5000, 1000),TxDb = TxDb.Mmusculus.UCSC.mm10.knownGene,\n",
        "        annoDb = 'org.Mm.eg.db')\n",
        "})"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "5p6xKiv6eYAR",
        "outputId": "6fda3312-d697-48eb-acfa-b427fdba9384"
      },
      "execution_count": 35,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            ">> preparing features information...\t\t 2022-04-15 09:18:38 PM \n",
            ">> identifying nearest features...\t\t 2022-04-15 09:18:39 PM \n",
            ">> calculating distance from peak to TSS...\t 2022-04-15 09:18:39 PM \n",
            ">> assigning genomic annotation...\t\t 2022-04-15 09:18:39 PM \n",
            ">> adding gene annotation...\t\t\t 2022-04-15 09:18:56 PM \n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "'select()' returned 1:many mapping between keys and columns\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            ">> assigning chromosome lengths\t\t\t 2022-04-15 09:18:56 PM \n",
            ">> done...\t\t\t\t\t 2022-04-15 09:18:56 PM \n",
            ">> preparing features information...\t\t 2022-04-15 09:18:56 PM \n",
            ">> identifying nearest features...\t\t 2022-04-15 09:18:56 PM \n",
            ">> calculating distance from peak to TSS...\t 2022-04-15 09:18:57 PM \n",
            ">> assigning genomic annotation...\t\t 2022-04-15 09:18:57 PM \n",
            ">> adding gene annotation...\t\t\t 2022-04-15 09:18:59 PM \n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "'select()' returned 1:many mapping between keys and columns\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            ">> assigning chromosome lengths\t\t\t 2022-04-15 09:18:59 PM \n",
            ">> done...\t\t\t\t\t 2022-04-15 09:18:59 PM \n",
            ">> preparing features information...\t\t 2022-04-15 09:18:59 PM \n",
            ">> identifying nearest features...\t\t 2022-04-15 09:18:59 PM \n",
            ">> calculating distance from peak to TSS...\t 2022-04-15 09:19:00 PM \n",
            ">> assigning genomic annotation...\t\t 2022-04-15 09:19:00 PM \n",
            ">> adding gene annotation...\t\t\t 2022-04-15 09:19:03 PM \n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "'select()' returned 1:many mapping between keys and columns\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            ">> assigning chromosome lengths\t\t\t 2022-04-15 09:19:03 PM \n",
            ">> done...\t\t\t\t\t 2022-04-15 09:19:03 PM \n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "dat <- lapply(anno, function(a){\n",
        "  data.frame(a)\n",
        "})"
      ],
      "metadata": {
        "id": "ssNgLbLAewXw"
      },
      "execution_count": 36,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "for (i in 1:length(dat)){\n",
        "dat[[i]]$log2FC <- res$log2FoldChange[match(dat[[i]]$SYMBOL, res$SYMBOL)]\n",
        "}"
      ],
      "metadata": {
        "id": "Rx98qnL8fBgR"
      },
      "execution_count": 37,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "for (i in 1:length(dat)){\n",
        "dat[[i]]$padj_RNA <- res$padj[match(dat[[i]]$SYMBOL, res$SYMBOL)]\n",
        "}"
      ],
      "metadata": {
        "id": "gLhWBGvXfgwE"
      },
      "execution_count": 38,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "for (i in 1:length(dat)){\n",
        "  dat[[i]]$group <- 'NFR_NOR'\n",
        "}"
      ],
      "metadata": {
        "id": "NysXw9QbjEwl"
      },
      "execution_count": 39,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "#NOR -> NFR\n",
        "(MEF to D7 MEFs post OSK induction)"
      ],
      "metadata": {
        "id": "70kJauYtlAtu"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "up <- lapply(osk_gr, function(o){\n",
        "  o[o$log_ratio < -0.1 & o$padj < 0.05]\n",
        "  })"
      ],
      "metadata": {
        "id": "Zkl1qJf7KZ0s"
      },
      "execution_count": 40,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "anno_up <- lapply(up, function(u){\n",
        "  annotatePeak(u, tssRegion = c(-5000, 1000),TxDb = TxDb.Mmusculus.UCSC.mm10.knownGene,\n",
        "        annoDb = 'org.Mm.eg.db')\n",
        "})"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "92a6e3e5-e771-4fa1-8b80-f8f2bf373781",
        "id": "aKnhx2VTKZ0w"
      },
      "execution_count": 41,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            ">> preparing features information...\t\t 2022-04-15 09:19:03 PM \n",
            ">> identifying nearest features...\t\t 2022-04-15 09:19:03 PM \n",
            ">> calculating distance from peak to TSS...\t 2022-04-15 09:19:03 PM \n",
            ">> assigning genomic annotation...\t\t 2022-04-15 09:19:03 PM \n",
            ">> adding gene annotation...\t\t\t 2022-04-15 09:19:06 PM \n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "'select()' returned 1:1 mapping between keys and columns\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            ">> assigning chromosome lengths\t\t\t 2022-04-15 09:19:06 PM \n",
            ">> done...\t\t\t\t\t 2022-04-15 09:19:06 PM \n",
            ">> preparing features information...\t\t 2022-04-15 09:19:06 PM \n",
            ">> identifying nearest features...\t\t 2022-04-15 09:19:06 PM \n",
            ">> calculating distance from peak to TSS...\t 2022-04-15 09:19:07 PM \n",
            ">> assigning genomic annotation...\t\t 2022-04-15 09:19:07 PM \n",
            ">> adding gene annotation...\t\t\t 2022-04-15 09:19:10 PM \n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "'select()' returned 1:many mapping between keys and columns\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            ">> assigning chromosome lengths\t\t\t 2022-04-15 09:19:10 PM \n",
            ">> done...\t\t\t\t\t 2022-04-15 09:19:10 PM \n",
            ">> preparing features information...\t\t 2022-04-15 09:19:10 PM \n",
            ">> identifying nearest features...\t\t 2022-04-15 09:19:10 PM \n",
            ">> calculating distance from peak to TSS...\t 2022-04-15 09:19:10 PM \n",
            ">> assigning genomic annotation...\t\t 2022-04-15 09:19:10 PM \n",
            ">> adding gene annotation...\t\t\t 2022-04-15 09:19:13 PM \n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "'select()' returned 1:many mapping between keys and columns\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            ">> assigning chromosome lengths\t\t\t 2022-04-15 09:19:13 PM \n",
            ">> done...\t\t\t\t\t 2022-04-15 09:19:13 PM \n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "dat_up <- lapply(anno_up, function(a){\n",
        "  data.frame(a)\n",
        "})"
      ],
      "metadata": {
        "id": "1iTTbfQDKZ0x"
      },
      "execution_count": 42,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "for (i in 1:length(dat_up)){\n",
        "  dat_up[[i]]$log2FC <- res$log2FoldChange[match(dat_up[[i]]$SYMBOL, res$SYMBOL)]\n",
        "}"
      ],
      "metadata": {
        "id": "EfTDe7DiKZ0z"
      },
      "execution_count": 43,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "for (i in 1:length(dat_up)){\n",
        "dat_up[[i]]$padj_RNA <- res$padj[match(dat_up[[i]]$SYMBOL, res$SYMBOL)]\n",
        "}"
      ],
      "metadata": {
        "id": "qCeRMq7JKZ00"
      },
      "execution_count": 44,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "for (i in 1:length(dat_up)){\n",
        "  dat_up[[i]]$group <- 'NOR_NFR'\n",
        "}\n"
      ],
      "metadata": {
        "id": "wH95XeKXjbTu"
      },
      "execution_count": 45,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "#Combined bar plots for Down, Up and NoChange genes"
      ],
      "metadata": {
        "id": "4tx-FoEF_Qu_"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "df_oct4 <- rbind(dat$oct4, dat_up$oct4)\n",
        "df_oct4$TF <- 'Oct4'\n",
        "df_sox2 <- rbind(dat$sox2, dat_up$sox2)\n",
        "df_sox2$TF <- 'Sox2'\n",
        "df_klf4 <- rbind(dat$klf4, dat_up$klf4)\n",
        "df_klf4$TF <- 'Klf4'"
      ],
      "metadata": {
        "id": "4o6WFpNTjkr0"
      },
      "execution_count": 46,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "df_promoter_oct4 <- df_oct4[grepl('Promoter',df_oct4$annotation),]"
      ],
      "metadata": {
        "id": "wyZ44MXRnN10"
      },
      "execution_count": 47,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "o <- df_promoter_oct4 %>%\n",
        "  filter(!is.na(padj_RNA)) %>%\n",
        "  mutate(  Direction = case_when(\n",
        "    padj_RNA < 0.05 & log2FC < 0 ~ 'Down',\n",
        "    padj_RNA < 0.05 & log2FC > 0 ~ 'Up',\n",
        "    padj_RNA > 0.05 ~ 'NC'\n",
        "    )\n",
        "  ) %>% \n",
        "  group_by(group) %>%\n",
        "  summarize(Total = n(),\n",
        "  across()\n",
        "  ) %>%\n",
        "  group_by(group, Direction, Total) %>%\n",
        "  summarise(Count = n()) %>%\n",
        "  mutate(Percent = 100*Count/Total) %>%\n",
        "  ggplot(aes(x = Direction, y = Percent, fill = group)) +\n",
        "  geom_bar(stat = 'identity', position = 'dodge') +\n",
        "  scale_fill_manual(values = c('indianred1','black')) +\n",
        "  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),\n",
        "  panel.background = element_blank(), axis.line = element_line(colour = \"black\"))"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "EuP3CyPJowIa",
        "outputId": "2ef07a4d-5b8f-4453-e456-5040164af008"
      },
      "execution_count": 48,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "\u001b[1m\u001b[22m`summarise()` has grouped output by 'group'. You can override using the\n",
            "`.groups` argument.\n",
            "\u001b[1m\u001b[22m`summarise()` has grouped output by 'group', 'Direction'. You can override\n",
            "using the `.groups` argument.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "df_promoter_sox2 <- df_sox2[grepl('Promoter',df_sox2$annotation),]"
      ],
      "metadata": {
        "id": "61Bw3wR2gqy4"
      },
      "execution_count": 49,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "s <- df_promoter_sox2 %>%\n",
        "  filter(!is.na(padj_RNA)) %>%\n",
        "  mutate(  Direction = case_when(\n",
        "    padj_RNA < 0.05 & log2FC < 0 ~ 'Down',\n",
        "    padj_RNA < 0.05 & log2FC > 0 ~ 'Up',\n",
        "    padj_RNA > 0.05 ~ 'NC'\n",
        "    )\n",
        "  ) %>% \n",
        "  group_by(group) %>%\n",
        "  summarize(Total = n(),\n",
        "  across()\n",
        "  ) %>%\n",
        "  group_by(group, Direction, Total) %>%\n",
        "  summarise(Count = n()) %>%\n",
        "  mutate(Percent = 100*Count/Total) %>%\n",
        "  ggplot(aes(x = Direction, y = Percent, fill = group)) +\n",
        "  geom_bar(stat = 'identity', position = 'dodge') +\n",
        "  scale_fill_manual(values = c('indianred1','black')) +\n",
        "  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),\n",
        "  panel.background = element_blank(), axis.line = element_line(colour = \"black\"))"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "pOJgM9g7gxzn",
        "outputId": "d1a09147-b7d3-4025-9b66-66239ea8d754"
      },
      "execution_count": 50,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "\u001b[1m\u001b[22m`summarise()` has grouped output by 'group'. You can override using the\n",
            "`.groups` argument.\n",
            "\u001b[1m\u001b[22m`summarise()` has grouped output by 'group', 'Direction'. You can override\n",
            "using the `.groups` argument.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "df_promoter_klf4 <- df_klf4[grepl('Promoter',df_klf4$annotation),]"
      ],
      "metadata": {
        "id": "sBDsVCFxg-71"
      },
      "execution_count": 51,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "k <- df_promoter_klf4 %>%\n",
        "  filter(!is.na(padj_RNA)) %>%\n",
        "  mutate(  Direction = case_when(\n",
        "    padj_RNA < 0.05 & log2FC < 0 ~ 'Down',\n",
        "    padj_RNA < 0.05 & log2FC > 0 ~ 'Up',\n",
        "    padj_RNA > 0.05 ~ 'NC'\n",
        "    )\n",
        "  ) %>% \n",
        "  group_by(group) %>%\n",
        "  summarize(Total = n(),\n",
        "  across()\n",
        "  ) %>%\n",
        "  group_by(group, Direction, Total) %>%\n",
        "  summarise(Count = n()) %>%\n",
        "  mutate(Percent = 100*Count/Total) %>%\n",
        "  ggplot(aes(x = Direction, y = Percent, fill = group)) +\n",
        "  geom_bar(stat = 'identity', position = 'dodge') +\n",
        "  scale_fill_manual(values = c('indianred1','black')) +\n",
        "  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),\n",
        "  panel.background = element_blank(), axis.line = element_line(colour = \"black\"))"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "6dbea945-9f09-4dea-fba9-1e19a7a46e9e",
        "id": "cxuVLdFmg-72"
      },
      "execution_count": 52,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "\u001b[1m\u001b[22m`summarise()` has grouped output by 'group'. You can override using the\n",
            "`.groups` argument.\n",
            "\u001b[1m\u001b[22m`summarise()` has grouped output by 'group', 'Direction'. You can override\n",
            "using the `.groups` argument.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "options(repr.plot.width = 12, repr.plot.height = 8)\n",
        "o+s+k"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 497
        },
        "id": "cnozbzu-iH4s",
        "outputId": "45240217-c4d4-48e0-f691-b75bc49073a3"
      },
      "execution_count": 53,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "plot without title"
            ],
            "image/png": "iVBORw0KGgoAAAANSUhEUgAABaAAAAPACAMAAAD0Wi6aAAAC91BMVEUAAAABAQECAgIDAwME\nBAQFBQUGBgYHBwcICAgJCQkKCgoLCwsMDAwNDQ0ODg4PDw8QEBARERESEhITExMUFBQVFRUW\nFhYXFxcYGBgZGRkaGhobGxscHBwdHR0eHh4fHx8gICAhISEiIiIkJCQlJSUmJiYnJycoKCgp\nKSkqKiorKyssLCwtLS0uLi4vLy8wMDAxMTEyMjIzMzM0NDQ1NTU2NjY3Nzc4ODg5OTk6Ojo7\nOzs8PDw9PT0+Pj4/Pz9BQUFCQkJDQ0NERERFRUVGRkZHR0dISEhJSUlMTExNTU1OTk5PT09Q\nUFBRUVFSUlJTU1NUVFRVVVVWVlZXV1dYWFhZWVlaWlpbW1tcXFxdXV1eXl5fX19gYGBhYWFi\nYmJjY2NkZGRlZWVmZmZnZ2doaGhpaWlqampra2tsbGxtbW1ubm5vb29wcHBxcXFycnJzc3N0\ndHR1dXV2dnZ3d3d4eHh5eXl6enp7e3t8fHx9fX1+fn5/f3+AgICBgYGCgoKDg4OEhISFhYWG\nhoaHh4eIiIiJiYmKioqLi4uMjIyNjY2Ojo6Pj4+QkJCRkZGSkpKTk5OUlJSVlZWWlpaXl5eY\nmJiZmZmampqbm5ucnJydnZ2enp6fn5+goKChoaGioqKjo6OkpKSlpaWmpqanp6eoqKipqamq\nqqqrq6usrKytra2urq6vr6+wsLCxsbGysrKzs7O0tLS1tbW2tra3t7e4uLi5ubm6urq7u7u8\nvLy9vb2+vr6/v7/AwMDBwcHCwsLDw8PExMTFxcXGxsbHx8fIyMjJycnKysrLy8vMzMzNzc3O\nzs7Pz8/Q0NDR0dHS0tLT09PU1NTV1dXW1tbX19fY2NjZ2dna2trb29vc3Nzd3d3e3t7f39/g\n4ODh4eHi4uLj4+Pk5OTl5eXm5ubn5+fo6Ojp6enq6urr6+vs7Ozt7e3u7u7v7+/w8PDx8fHy\n8vLz8/P09PT19fX29vb39/f4+Pj5+fn6+vr7+/v8/Pz9/f3+/v7/amr///+Yw9gkAAAACXBI\nWXMAABJ0AAASdAHeZh94AAAgAElEQVR4nO3de4DdZX3n8QMYCITLhottXasWVm2FuDgWRcQG\nrLo7k2AENIRcQBAILZBqUbASGhu6YAziPRYFdXUpgiKXhgQvbZGESw2LIIWYLHJtiIkkWQwQ\nk33+2HPmIrl8D+f5cZ5Jfp8P79cfk5njOY/f+X3PeROGTKaRAAC11NjRAwAAYgQaAGqKQANA\nTRFoAKgpAg0ANUWgAaCmCDQA1BSBBoCaItAAUFMEGgBqikADQE0RaACoKQINADVFoAGgpgg0\nANQUgQaAmiLQAFBT3Qb6infeVWQO1M2l7/zFjh4Bw+KCd67a0SMgV7eB/kTj5iJzoG5OayzZ\n0SNgWBzTeHxHj4BcBBoxAu2KQAsh0IgRaFcEWgiBRoxAuyLQQgg0YgTaFYEWQqARI9CuCLQQ\nAo0YgXZFoIUQaMQItCsCLYRAI0agXRFoIQQaMQLtikALIdCIEWhXBFoIgUaMQLsi0EIINGIE\n2hWBFkKgESPQrgi0EAKNGIF2RaCFEGjECLQrAi2EQCNGoF0RaCEEGjEC7YpACyHQiBFoVwRa\nCIFGjEC7ItBCCDRiBNoVgRZCoBEj0K4ItBACjRiBdkWghRBoxAi0KwIthEAjRqBdEWghBBox\nAu2KQAsh0IgRaFcEWgiBRoxAuyLQQgg0YgTaFYEWQqARI9CuCLQQAo0YgXZFoIUQaMQItCsC\nLYRAI0agXRFoIQQaMQLtikALIdCIEWhXBFoIgUaMQLsi0EIINGIE2hWBFkKgESPQrgi0EAKN\nGIF2RaCFmAS6kWNHD6ll2AP9/7IM7wwvScMeaDZbDoFGjEC7ItBCCDRiBNoVgRZCoBEj0K4I\ntBACjRiBdkWghRBoxAi0KwIthEAjRqBdEWghBBoxAu2KQAsh0IgRaFcEWgiBRoxAuyLQQgg0\nYgTaFYEWQqARI9CuCLQQAo0YgXZFoIXkV+svepuOTWndnCkTL1wxdCuBdkWgXRFoIfnVmnb9\nypUrV6U069zlj10yfePgrQTaFYF2RaCF5FfrfXf2/7Kyb1nzd9Hj7x68lUC7ItCuCLSQ7Go9\n13vZWSfNfjTdNmFT86Mzrxq8mUC7ItCuCLSQ7Go9deKnH3hg5on/d/7U1kfnz2vddM0110wd\nSaA9EWhXBFpItWr95tgF86e13ukP9AM9TfsTaE8E2hWBFlKxWmd8a/HAlziuTvwO2huBdkWg\nhWRX66HPbkhp/bE/XNW3NKU14+4dvJmvQbsi0K4ItJDsaq2dOPeJR2dPeyZddPbyR2fO2DR4\nM4F2RaBdEWgh+dVa9vHjJ836j5Senjv5hNmrh24l0K4ItCsCLYRv9UaMQLsi0EIINGIE2hWB\nFkKgESPQrgi0EAKNGIF2RaCFEGjECLQrAi2EQCNGoF0RaCEEGjEC7YpACyHQiBFoVwRaCIFG\njEC7ItBCCDRiBNoVgRZCoBEj0K4ItBACjRiBdkWghRBoxAi0KwIthEAjRqBdEWghBBoxAu2K\nQAsh0IgRaFcEWgiBRoxAuyLQQgg0YgTaFYEWQqARI9CuCLQQAo0YgXZFoIUQaMQItCsCLaT2\ngc5bNoEujkC7ItBCCDRiBNoVgRZCoBEj0K4ItBACjRiBdkWghRBoxAi0KwIthEAjRqBdEWgh\nBBoxAu2KQAsh0IgRaFcEWgiBRoxAuyLQQgg0YgTaFYEWQqARI9CuCLQQAo0YgXZFoIUQaMQI\ntCsCLYRAI1aTQLPZ4gi0EAKNGIF2RaCFEGjECLQrAi2EQCNGoF0RaCEEGjEC7YpACyHQiBFo\nVwRaCIFGjEC7qkmg2WwOAo0YgXZFoIUQaMQItCsCLYRAI0agXRFoIQQaMQLtikALIdCIEWhX\nBFoIgUaMQLsi0EIINGIE2hWBFkKgESPQrgi0EAKNGIF2RaCFEGjECLQrAi2EQCNGoF0RaCEE\nGjEC7YpACyHQiBFoVwRaCIFGjEC7ItBCCDRiBNoVgRZCoBEj0K4ItBACjRiBdkWghRBoxAi0\nKwIthEAjRqBdEWghBBoxAu2KQAsh0IgRaFcEWgiBRoxAuyLQQgg0YgTaFYEWQqARI9CuCLQQ\nAo0YgXZFoIUQaMQItCsCLYRAI0agXRFoIQQaMQLtikALIdCIEWhXBFoIgUaMQLsi0EIINGIE\n2hWBFkKgESPQrgi0EAKNGIF2RaCFEGjECLQrAi2EQCNGoF0RaCEEGjEC7YpACyHQiBFoVwRa\nCIFGjEC7ItBCCDRiBNoVgRZCoBEj0K4ItBACjRiBdkWghRBoxAi0KwIthEAjRqBdEWghBBox\nAu2KQAsh0IgRaFcEWgiBRoxAuyLQQgg0YgTaFYEWQqARI9CuCLQQAo0YgXZFoIUQaMQItCsC\nLYRAI0agXRFoIQQaMQLtikALIdCIEWhXBFoIgUaMQLsi0EIINGIE2hWBFkKgESPQrgi0EAKN\nGIF2RaCFEGjECLQrAi2EQCNGoF0RaCEEGjEC7YpACyHQiBFoVwRaCIFGjEC7ItBCCDRiBNoV\ngRZCoBEj0K4ItBACjRiBdkWghRBoxAi0KwIthEAjRqBdEWghBBoxAu2KQAsh0IgRaFcEWgiB\nRoxAuyLQQgg0YgTaFYEWQqARI9CuCLQQAo0YgXZFoIUQaMQItCsCLYRAI0agXRFoIQQaMQLt\nikALIdCIEWhXBFoIgUaMQLsi0EIINGIE2hWBFkKgESPQrgi0EAKNGIF2RaCFEGjECLQrAi2E\nQCNGoF0RaCEEGjEC7YpACyHQiBFoVwRaCIFGjEC7ItBCCDRiBNoVgRZCoBEj0K4ItBACjRiB\ndkWghRBoxAi0KwIthEAjRqBdEWghBBoxAu2KQAsh0IgRaFcEWgiBRoxAuyLQQgg0YgTaFYEW\nQqARI9CuCLQQAo0YgXZFoIUQaMQItCsCLYRAI0agXRFoIQQaMQLtikALIdCIEWhXBFoIgUaM\nQLsi0EIINGIE2hWBFkKgESPQrgi0EAKNGIF2RaCFEGjECLQrAi2k0hW4pXdRSuvmTJl44Yqh\nmwi0KwLtikALqXIFfn3ihGagZ527/LFLpm8cvI1AuyLQrgi0kCpX4KLLT1yUVvYta/4uevzd\ng7cRaFcE2hWBFlLhCtz2wfXNQN82YVPz/TOvGryRQLsi0K4ItJD8K7Bu8pLUDPT8qa0Pzp/X\nfPOLsWPHvnVfAu2JQLsi0ELyr8Cll6b+QE9rfdAf6OV9fX1HjibQngi0KwItJPsKLJm8tj/Q\niwe+xHH14M18icMVgXZFoIVkX4GLJ0ycOLHvuNmr+pamtGbcvYM3E2hXBNoVgRaSfQXWrmya\ntGBNuujs5Y/OnLFp8GYC7YpAuyLQQqpdgRMXpfT03MknzF49dAuBdkWgXRFoIXyrN2IE2hWB\nFkKgESPQrgi0EAKNGIF2RaCFEGjECLQrAi2EQCNGoF0RaCEEGjEC7YpACyHQiBFoVwRaCIFG\njEC7ItBCCDRiBNoVgRZCoBEj0K4ItBACjRiBdkWghRBoxAi0KwIthEAjhT+vnUC7ItBCCDTi\nn9dOoF0RaCEEGvHPayfQrgi0EAKN+Oe1E2hXBFoIgca2P699SV9f39v2ItCeCLQQAo1tf177\nXWPHjn3L3gTaE4EWQqBf8tr8vHa+xOGKQAsh0C95bX5eO4F2RaCFEOiXvDY/r51AuyLQQgg0\nWrb9ee0E2hWBFkKgESPQrgi0EAKNGIF2RaCFEGjECLQrAi2EQCNGoF0RaCEEGjEC7YpACyHQ\niBFoVwRaCIFGjEC7ItBCCDRiBNoVgRZCoBEj0K4ItBACjRiBdkWghRBoxAi0KwIthEAjRqBd\nEWghBBoxAu2KQAsh0IgRaFcEWgiBRoxAuyLQQgg0YgTaFYEWQqARI9AWbuldlNK6OVMmXrhi\n6CYCLYRAI0agHfz6xAnNQM86d/ljl0zfOHgbgRZCoBEj0A4uuvzERWll37Lm76LH3z14G4EW\nQqARI9AGbvvg+magb5vQ+knAZ141eCOBFkKgESPQ+tZNXtL6ccDzp7Y+OH9e882NPU2jCLQM\nAo0YgdZ36aX9P699/rTWB/2B/smkSZMOH0mgZRBoxAi0vCWT1/YHevHAlziuHryZL3EIIdCI\nEWh5F0+YOHFi33GzV/UtTWnNuHsHbybQQgg0YgRa3tqVTZMWrEkXnb380ZkzNg3eTKCFEGjE\nCLSHExel9PTcySfMXj10C4EWQqARI9CuCLQQAo0YgXZFoIUQaMQItCsCLYRAI0agXRFoIQQa\nMQLtikALIdCIEWhXBFoIgUaMQLsi0EIINGIE2hWBFkKgESPQrgi0EAKNGIF2RaCFEGjECLQr\nAi2EQCNGoF0RaCEEGjEC7YpACyHQiBFoVwRaCIFGjEC7ItBCCDRiBNoVgRZCoBEj0K4ItBAC\njRiBdkWghRBoxAi0KwIthEAjRqBdEWghBBoxAu2KQAsh0IgRaFcEWgiBRoxAuyLQQgg0YgTa\nFYEWQqARI9CuCLQQAo0YgXZFoIUQaMQItCsCLYRAI0agXRFoIQQaMQLtikALIdCIEWhXBFoI\ngUaMQLsi0EIINGIE2hWBFkKgESPQrgi0EAKNGIF2RaCFEGjECLQrAi2EQCNGoF0RaCEEGjEC\n7YpACyHQiBFoVwRaCIFGjEC7ItBCCDRiBNoVgRZCoBEj0K4ItBACjRiBdkWghRBoxAi0KwIt\nhEAjRqBdEWghBBoxAu2KQAsh0IgRaFcEWgiBRoxAuyLQQgg0YgTaFYEWQqARI9CuCLQQAo0Y\ngXZFoIUQaMQItCsCLYRAI0agXRFoIQQaMQLtikALIdCIEWhXBFoIgUaMQLsi0EIINGIE2hWB\nFkKgESPQrgi0EAKNGIF2RaCFEGjECLQrAi2EQCNGoF0RaCEEGjEC7YpACyHQiBFoVwRaCIFG\njEC7ItBCCDRiBNoVgRZCoBEj0K4ItBACjRiBdkWghRBoxAi0KwIthEAjRqBdEWghBBoxAu2K\nQAsh0IgRaFcEWgiBRoxAuyLQQgg0YgTaFYEWQqARI9CuCLQQAo0YgXZFoIUQaMQItCsCLYRA\nI0agXRFoIQQaMQLtikALIdCIEWhXBFoIgUaMQLsi0EIINGIE2hWBFkKgESPQrgi0EAKNGIF2\nRaCFEGjECLQrAi2EQCNGoF0RaCEEGjEC7YpACyHQiBFoVwRaCIFGjEC7ItBCCDRiBNoVgRZC\noBEj0K4ItBACjRiBdkWghRBoxAi0KwIthEAjRqBdEWghBBoxAu2KQAsh0IgRaFcEWgiBbifr\nSOMnEIF2RaCFEOh2so40fgIRaFcEWgiBbifrSOMnEIF2RaCFEOh2so40fgIRaFcEWkj+FXh4\n1sQPnHd/SuvmTJl44YqhWwm0KwLtikALyb4CG6Z8+tHH577/N2nWucsfu2T6xsGbCbQrAu2K\nQAvJvgJPXfublB7tXbayb1nzd9Hj7x68mUC7ItCuCLSQSldg7RdOf+62CZua75151eBNBNoV\ngXZFoIVUuAIb39v7sV+l+VNb758/r/nmqWuuuWbqSALtiUC7ItBCqlyBR+656NR186e13u0P\n9AM9TfsTaE8E2hWBFlLtCmw8/obFA1/iuLr5Zu3ChQtP2Y1AeyLQrgi0kOwr8NNTnklp0wk3\nrOpbmtKacfcO3szXoF0RaFcEWkj2FVh34t8//MS8CU+ki85e/ujMGZsGbybQrgi0KwItJP8K\nPHTBscd9+O6Unp47+YTZq4duJdCuCLQrAi2Eb/VuJ+tI4ycQgXZFoIUQ6HayjjR+AhFoVwRa\nCIFuJ+tI4ycQgXZFoIUQ6HayjjR+AhFoVwRaCIFuJ+tI4ycQgXZFoIUQ6HayjjR+AhFoVwRa\nCIFuJ+tI4ycQgXZFoIUQ6HayjjR+AhFoVwRaCIFuJ+tI4ycQgXZFoIUQ6HayjjR+AhFoVwRa\nCIFuJ+tI4ycQgXZFoIUQ6HayjjR+AhFoVwRaCIFuJ+tI4ycQgXZFoIUQ6HayjjR+AhFoVwRa\nCIFuJ+tI4ycQgXZFoIUQ6HayjjR+AhFoVwRaCIFuJ+tI4ycQgXZFoIUQ6HayjjR+AhFoVwRa\nCIFuJ+tI4ycQgXZFoIUQ6HayjjR+AhFoVwRaCIFuJ+tI4ycQgXZFoIUQ6HayjjR+AhFoVwRa\nCIFuJ+tI4ycQgXZFoIUQ6HayjjR+AhFoVwRaCIFuJ+tI4ycQgXZFoIUQ6HayjjR+AhFoVwRa\nCIFuJ+tI4ycQgXZFoIUQ6HayjjR+AhFoVwRaCIFuJ+vI/DM3XvDK3d604MwRKR1x5PWvPDyl\nm47cc+Qb5mxK6Y1vbP3v4/ZL6U1v/cGf7j562lPVBh0mBDqP3GIJdKZabJZAt5N1ZP6Zf9c4\n7uZ/+IPDRqV01JjXf/6G9N2d3vO9W2Y0PrL5tg8/4M0/WfmNEcdUG3SYEOg8cosl0JlqsVkC\n3U7Wkdlnbvq9g5v/5F3caG776Ma1zRte/6pnm2/Hj/jVZts+ovEvzfdObjxcbdLhQaCz6C2W\nQOepx2YJdDtZR2af+XjjnNYvB7e2vetzKT3WOK318eWNGzbf9qjmMyJd2bip2qTDg0Bn0Vss\ngc5Tj80S6Hayjsw+c0nj4tYvE1rbfkXznTsas1of39SYt/m2D2y9d0Pja9UmHR4EOoveYgl0\nnnpslkC3k3Vk9pmLG59q/fK+1rZf3XznzsaFrY9vbPzDNtv+XuPKapMODwKdRW+xBDpPPTYb\nX4Genw/8+p0/7vR4Ap3nwdZ/W0jpkKFtP9E4tfXxvMb8dOjBrffe0tr27r9tvvfl5m3DJn+z\nBDpLXRZbYbMEOks9Nhtfgcad/b9suHDXTo8n0Hk27NPa6R2NoW2ng1+xvvn2PXusSUftvyml\nFbv3/xeH1peyxu+2utqkVeRvlkBnqctiK2yWQGepx2ajK7DZ9XlTp8cT6EwzGlNv/vJrjvjd\ntm/c+V3X/dPpjYtSurRx0X/8dOwbWtv+w9d+ceGHG5OrDfriPqeOmyXQeWqx2EqbJdB5arHZ\n6Arc/ZnGuJNbPviJRzo9nkBneuYv9h915O0T9xzadlrw9lG7HfrV5jvPzvjPu73x+ul7Nbf9\n+rvescfoU9ZVG7SCKpsl0HlqsdhKmyXQeWqx2fgKvPvB3McT6EqO/oMX+l+PeF3F46rL3yyB\nrmKHL7bCZgl0FTt4s/wpjnayjsw/c+57N6T0633+2wvdZ3u8jrMR6DxyiyXQmWqx2fgKrJjy\nip3zLhCBzvT1xn+/7qrDd7rlhe6zHV7H+Zsl0HlqstgKmyXQeWqx2fgKHPuyo6f0f0Xr5E6P\nJ9C5vn7oqD3eduML3mU7vI7zN0ugM9VjsRU2S6Az1WGz8RXY73u5jyfQWvI3S6C15G+WQAuJ\nr8AeT+Y+XjLQ5Y6UewLlb5ZAa8nfLIEWEl+BI3+U+3gCrSV/swRaS/5mCbSQ+ArcddhtmY8n\n0FryN0ugteRvlkALia/AEX/Y2OPV/To9nkBryd8sgdaSv1kCLaTNlziOHtLp8QRaS/5mCbSW\n/M0SaCEvyW9UKXek8ROIQLsi0ELaXYH1d1y7Mm3o/HgCnWftCxiW69ZW7mYJdJYXWmxNN0ug\ns9Rjs22uwKf2ajQWpfOmdlw3gc6ztv3/0fZ9GWdvlkBnWau3WQKdpR6bja/AvEbfl5rLvvJl\nF3d6PIHOU4tlpyqb9Q100c3W42WcqmyWQGepx2bjKzDmtLS+uez0sdd2ejyBzlOLZacqmyXQ\nWerxMk5VNkugs9Rjs/EVGLlwYNk3j+j0eAKdpxbLTlU2S6Cz1ONlnKpslkBnqcdm4yvw8usH\nlv2Pe3d6PIHOU4tlpyqbJdBZ6vEyTlU2S6Cz1GOz8RV455/9prXsVQe/q9PjCXSeWiw7Vdks\ngc5Sj5dxqrJZAp2lHpuNr8CPdjnorMZJU/YecWunxxPoPLVYdqqyWQKdpR4v41RlswQ6Sz02\n2+YK3HJo6+oc9uOOjyfQeWqx7JbszRLoLPV4Gbdkb5ZAZ6nHZttegRVLluT8KHECnSdv2T0j\n+3+w3Bu+2Hx34PyvDbyz/7sXp/Bu6Zenv2bX0X9+fUrRHUOZmyXQWTJfxnXaLIHOUo/NtrkC\nT1zWfPPkhSs6fhYEOk/msvcb2/qlf9mTlrasTT1TH3nkkTuO23N5eLd7Ro+59r5bP7TT36bo\njoHszRLoLLkv4xptlkBnqcdm4yvw77+/a/PtQ43fX9bpsyDQeTKX/cl9L0+Dy54+dFv/Oxv2\n/Fx4t8N6nm3dMmene6I7bit/swQ6S+7LuEabJdBZ6rHZ+AqMP+iO1i8/P+i9nT4LAp0nc9mf\n++roFeGyN+7z6ehuyxrf7b9lw/7nRXfcVv5mCXSW3JdxjTZLoLPUY7PxFTjgqwO/fnmvTp8F\ngc6TuezPpqPeHy177Yd3XR7dbX5j6cBN7zg2uuO28jdLoLPkvoxrtFkCnaUem42vwO7fHPj1\nf+7R6bMg0Hmyl7105I0Dy95pl5a7Us+IUaNGNcZsfpmfv9vNjfsHbjri+OiO28rfLIHOkv8y\nrs1mCXSWemw2vgJve/dv+0f80yM6fRYEOk/2stPsV607uLXsD/ysZX3/f3n4t9Gf3/yw5+/2\ncOPb/bc8t9/M6I7byt8sgc6S/zKuzWYJdJZ6bDa+AvN3OnD6zE9MO2Dn+Z0+CwKdJ3/ZGw45\n69Bt/nXpG7vfl8K7HXnIM61bPrvLg9Edt5W/WQKdpcLLuC6bJdBZ6rHZNldgQf8f0BtzY8fP\ngkDnyV92un3EAdv+B4f3jnkmvNv9+435/gN3zth5TnjHQPZmCXSWCi/jumyWQGepx2bbXoFf\n3XNfzrfLEOg8FZad/rKx7bKffPlZ4d3SL0951YjR71kQ3zGUuVkCnaXKy7gmmyXQWeqx2fgK\nHN75H8ODCHSeWnzbaKqyWQKdpR7fEJyqbJZAZ6nHZuMr8Mo5uY8n0HlqsexUZbMEOks9Xsap\nymYJdJZ6bDa+Atf98Xefy3s8gc5TYNl37jPouqrXcDP5myXQWUq8jLfzZgl0lnpsNr4CRx7S\n2PUVr27p9HgCnacW/zROVTZLoLPU4/dZqcpmCXSWemw2vgJHHHX0oE6PJ9B5arHsVGWzBDpL\nPV7GqcpmCXSWemy22ytAoPPUYtmVEOgs9XgZV0Kgs9Rjs+2uwPo7rl2ZNnR+PIHOU4tl98vd\nLIHOUo+Xcb/czRLoLPXYbJsr8Km9Go1F6bypHddNoPPUYtkt2Zsl0Fnq8TJuyd4sgc5Sj83G\nV2Beo+9LzWVf+bKLOz2eQOdZ+wKG5bq1kb9ZAp3lhRZb080S6Cz12Gx8BcacNvAj3D/22k6P\nJ9B5arHsVGWzBDpLPV7GqcpmCXSWemw2vgIjFw4s++YRnR5PoPPUYtmpymYJdJZ6vIxTlc0S\n6Cz12Gx8BV5+/cCy/3HvTo/fMtBZi6k24HAsu9yR+YFuf8R2fRnnb5ZAZ1mrt1kCnaUem42v\nwDv/7DetZa86+F2dHk+g89Ri2anKZgl0lnq8jFOVzRLoLPXYbHwFfrTLQWc1Tpqy94hbOz2e\nQOepxbJTlc0S6Cz1eBmnKpsl0Fnqsdk2V+CWQ1tjHPbjjo8n0HlqseyW7M0S6Cz1eBm3ZG+W\nQGepx2bbXoEVS5aszng8gc5Ti2UPyNwsgc5Sj5fxgMzNEugs9dhseAXW337rU5mPJ9B5arHs\nSpsl0Fnq8TKutFkCnaUem42uwKV7NRojznjhn68zhEDnqcWyK22WQGepx8u40mYJdJZ6bDa4\nAtc0XjPj429vnJn1eAKdpxbLrrRZAp2lHi/jSpv1DXTRI+ux2WDcI1+zrvn2pBFrch5PoPPk\nLbtn5IOtX97Q/4PLTn/NrqP//PrWra177f/uxSm8Y8/AKV+L7raVSpsl0FkyX8Z12iyBzlKP\nzQbj7vk3rbd3NDr+cZ0WAp0nc9n7jW390trhPaPHXHvfrR/a6W+bt0595JFH7jhuz+XhHXsm\nLW1ZG91tK5U2S6Cz5L6Ma7RZAp2lHpsNxm18ufX28cb8nM+CQOfJXPYn9708DezwsJ5nW7fM\n2emewR/9u2HPz4V33OpHCW9xt61U2iyBzpL7Mq7RZgl0lppsNrjpK623TzT+KeezINB5Mpf9\nua+OXtG/w2WN7/bfsmH/8wa3uHGfT0d33HrZW9xtK5U2S6Cz5L6Ma7RZAp2lJpsNbiLQ2UeW\nDvRn01Hv79/h/MbSgZvecezAFtd+eNfl0R23WvaWd9sKge5XdLO5L+MabZZAZ6nJZoObPrao\n6cbG3NYvnT4LAp0ne9lLR97Y2uHNjfsHbjri+NQzYtSoUY0xm1/o5++YenbapeWu6G5bqbRZ\nAp0l/2Vcm80S6Cw12WyHz7LTZ0Gg82QvO81+1bqDv5gebny7/5bn9pvZ/58U/m305zc/7vk7\npp4P/KxlfXS3rVSanEBnyX8Z12azBDpLTTa77U0XbK7TZ0Gg8+Qve8MhZx3a3OGRh/R/18Fn\nd3lw4N+DvrH7fSm845b/urTl3bZSabMEOkuFl3FdNkugs9Rjs91+qw6BzpO/7HT7iAOaO7x/\nvzHff+DOGTvPGVrne8c8E95xq//gsMXdukGgs1R4GddlswQ6Sz02S6C7OnI4Ap3+stH/h95P\nedWI0e9ZkIa2+OTLzwrvuNWyt7hbNwh0liov45pslkBnqcdmCXRXR2p9q3clBDpLPb4huBIC\nnaUemyXQXR1JoF+84dhslqKbrcfLuBICnaUemyXQXR25fQN95z6Drqt2DV8UAp2lzMt4u26W\nQGepx2YJdFdH8jvoF284Npul6Gbr8fusSgh0lnpslkB3dSSBfvGGY7NZim62Hi/jSgh0lnps\nlkB3deRLJtAim81SdLP1eBlXQqCz1GOzBLqrIwl0vTabpehm6/EyroRAZ6nHZgl0V0cS6Hpt\nNkvRzdbjZVwJgc5Sj80S6K6OrLDt9qpdkO2FQGd5ocXWdLMEOks9Nkuguzqy28tXYwTaFYEW\nQqC7OlJs23jjpd4AABrCSURBVFUQaFcEWgiB7upIsW1XQaBdEWghBLqrI8W2XQWBdkWghRDo\nro4U23YVBNoVgRZCoLs6UmzbVRBoVwRaCIHu6kixbVdBoF0RaCEEuqsjxbZdxYsIdLVLNhyb\nzfJS3yyBFkKguzpSbNtVEGhXBFoIge7qSLFtV0GgXRFoIQS6qyPFtl0FgXZFoIUQ6K6OFNt2\nFQTaFYEWQqC7OlJs21UQaFcEWgiB7upIsW1XIRloNpuBQAsh0F0dKbbtKgi0KwIthEB3daTY\ntqsg0K4ItBAC3dWRYtuugkC7ItBCCHRXR4ptuwoC7YpACyHQXR0ptu0qCLSrLQOddcmq/R8M\nx2azGC6WQHd1pNi2qyDQrgi0EALd1ZFi266CQLsi0EIIdFdHim27CgLtikALIdBdHSm27SoI\ntCsCLYRAd3Wk2LarINCuCLQQAt3VkWLbroJAuyLQQgh0V0eKbbsKAu2KQAsh0F0dKbbtKgi0\nKwIthEB3daTYtqsg0K4ItBAC3dWRYtuugkC7ItBCCHRXR4ptuwoC7YpACyHQXR0ptu0qCLQr\nAi2EQHd1pNi2qyDQrl5EoHf8ZrMYLpZAd3Wk2LarINCuCLQQAt3VkWLbroJAuyLQQgh0V0eK\nbbsKAu2KQAsh0F0dKbbtKgi0KwIthEB3daTYtqsg0K4ItBAC3dWRYtuugkC7ItBCCHRXR4pt\nuwoC7YpACyHQXR0ptu0qCLQrAi2EQHd1pNi2qyDQrgi0kPxxV10y6biPPpDSujlTJl64YuhW\nAu2KQLsi0ELyxz3n3GWPf+qE9WnWucsfu2T6xsFbCbQrAu2KQAvJHnft7IdTerL3wZV9y5q/\nix5/9+DNBNoVgXZFoIVUG/f+catvm7Cp+c6ZVw3eQqBdEWhXBFpIpXHXnvG1NH9q673z5zXf\n/GLs2LFv3ZdAeyLQrgi0kCrjPnLqFzal+dNa7/YHenlfX9+Rowm0JwLtikALqTDu3ROvb75d\nPPAljqsHb+RLHK4ItCsCLSR/3Ps+cFfrl1V9S1NaM+7ewVsJtCsC7YpAC8ke99lTvr2yaX26\n6Ozlj86csWnwZgLtikC7ItBCsse9u7ffDenpuZNPmL166GYC7YpAuyLQQvhW766OFNt2FQTa\nFYEWQqC7OlJs27H4m/gJtCsCLYRAd3Wk2LZj8TfxE2hXBFoIge7qSLFth9p8Ez+BdkWghRDo\nro4U23Z7234TP4F2pRjoYThSA4Hu6kixbbe11Tfx3zV27Ni37E2gPRFoIQS6qyPFtt3O1t/E\nv6Svr+9texFoTwRaCIHu6kixbbcRfhM/X+JwRaCFEOiujhTbdiz+Jn4C7YpACyHQXR0ptu1Q\nm2/iJ9CuCLQQAt3VkWLbDrX5Jn4C7YpACyHQXR0ptu0qCLQrAi2EQHd1pNi2qyDQrgi0EALd\n1ZFi266CQLsi0EIIdFdHim27CgKtL/5rsAi0EALd1ZFi266CQOuL/xosAi2EQHd1pNi2qyDQ\n8tr8NVgEWgiB7upIsW1XQaA9bPvXYBFoIQS6qyPFtl0Fgbaw1V+DdWNP0ygCLYNAd3Wk2Lar\nINAOtv5rsH4yadKkw0cSaBkEuqsjxbZdBYE2EP41WHyJQwiB7upIsW1XQaD1xX8NFoEWQqC7\nOlJs21UQaHlt/hosAi1kBwR6hy+bl3EOAi2vzV+DRaCFEOgyY/oh0K4ItBACXWZMPwTaFYEW\nQqDLjOmHQLsi0EIIdJkx/RBoVwRaCIEuM6YfAu2KQAsh0GXG9EOgXRFoIQS6zJh+CLQrAi2E\nQJcZ0w+BdkWghRDoMmP6IdCuCLQQAl1mTD8E2hWBFkKgy4zph0C7ItBCCHSZMf0QaFcEWgiB\nLjOmHwLtikALIdBlxvRDoF0RaCEEusyYfgi0KwIthECXGdMPgXZFoIUQ6DJj+iHQrgi0EAJd\nZkw/BNoVgRZCoMuM6YdAuyLQQgh0mTH9EGhXBFoIgS4zph8C7YpACyHQZcb0Q6BdEWghBLrM\nmH4ItCsCLYRAlxnTD4F2RaCFEOgyY/oh0K4ItBACXWZMPwTaFYEWQqDLjOmHQLsi0EIIdJkx\n/RBoVwRaCIEuM6YfAu2KQAsh0GXG9EOgXRFoIQS6zJh+CLQrAi2EQJcZ0w+BdkWghRDoMmP6\nIdCuCLQQAl1mTD8E2hWBFkKgy4zph0C7ItBCCHSZMf0QaFcEWgiBLjOmHwLtikALIdBlxvRD\noF0RaCEEusyYfgi0KwIthECXGdMPgXZFoIUQ6DJj+iHQrgi0EAJdZkw/BNoVgRZCoMuM6YdA\nuyLQQgh0mTH9EGhXBFoIgS4zph8C7YpACyHQZcb0Q6BdEWghBLrMmH4ItCsCLYRAlxnTD4F2\nRaCFEOgyY/oh0K4ItBACXWZMPwTaFYEWQqDLjOmHQLsi0EIIdJkx/RBoVwRaCIEuM6YfAu2K\nQAsh0GXG9EOgXRFoIQS6zJh+CLQrAi2EQJcZ0w+BdkWghRDoMmP6IdCuCLQQAl1mTD8E2hWB\nFkKgy4zph0C7ItBCCHSZMf0QaFcEWgiBLjOmHwLtikALIdBlxvRDoF0RaCEEusyYfgi0KwIt\nhECXGdMPgXZFoIUQ6DJj+iHQrgi0EAJdZkw/BNoVgRZCoMuM6YdAuyLQQgh0mTH9EGhXBFoI\ngS4zph8C7YpACyHQZcb0Q6BdEWghBLrMmH4ItCsCLYRAlxnTD4F2RaCFEOgyY/oh0K4ItBAC\nXWZMPwTaFYEWQqDLjOmHQLsi0EIIdJkx/RBoVwRaCIEuM6YfAu2KQAsh0GXG9EOgXRFoIQS6\nzJh+CLQrAi2EQJcZ0w+BdkWghRDoMmP6IdCuCLQQAl1mTD8E2hWBFkKgy4zph0C7ItBCCHSZ\nMf0QaFcEWgiBLjOmHwLtikALIdBlxvRDoF0RaCEEusyYfgi0KwIthECXGdMPgXZFoIUQ6DJj\n+iHQrgi0EAJdZkw/BNoVgRZCoMuM6YdAuyLQQgh0mTH9EGhXBFoIgS4zph8C7YpACyHQZcb0\nQ6BdEWghBLrMmH4ItCsCLYRAlxnTD4F2RaCFEOgyY/oh0K4ItBACXWZMPwTaFYEWQqDLjOmH\nQLsi0EIIdJkx/RBoVwRaCIEuM6YfAu2KQAsh0GXG9EOgXRFoIQS6zJh+CLQrAi2EQJcZ0w+B\ndkWghRDoMmP6IdCuCLQQAl1mTD8E2hWBFkKgy4zph0C7ItBCCHSZMf0QaFcEWgiBLjOmHwLt\nikALIdBlxvRDoF0RaCEEusyYfgi0KwIthECXGdMPgXZFoIUQ6DJj+iHQrgi0EAJdZkw/BNoV\ngRZCoMuM6YdAuyLQQgh0mTH9EGhXBFoIgS4zph8C7YpACyHQZcb0Q6BdEWghBLrMmH4ItCsC\nLYRAlxnTD4F2RaCFEOgyY/oh0K4ItBACXWZMPwTaFYEWUmHcR/9qXOuXdXOmTLxwxdCNBNoV\ngXZFoIXkj/svk+f2B3rWucsfu2T6xsFbCbQrAu2KQAvJH/cHTy5qBXpl37Lm76LH3z14K4F2\nRaBdEWghVcbtD/RtEzY135551eBtBNoVgXZFoIVUDvT8qa13z5/XfLPqiiuuOGF3Au2JQLsi\n0EKqB3pa693+QD/Q07Q/gfZEoF0RaCGVA7144EscVzffrF24cOEpuxFoTwTaFYEWUjnQq/qW\nprRm3L2Dt/E1aFcE2hWBFpI/7uqVC8atXLk+XXT28kdnztg0eCuBdkWgXRFoIfnjntzbcl16\neu7kE2avHrqVQLsi0K4ItBC+1bvMmH4ItCsCLYRAlxnTD4F2RaCFEOgyY/oh0K4ItBACXWZM\nPwTaFYEWQqDLjOmHQLsi0EIIdJkx/RBoVwRaCIEuM6YfAu2KQAsh0GXG9EOgXRFoIQS6zJh+\nCLQrAi2EQJcZ0w+BdkWghRDoMmP6IdCuCLQQAl1mTD8E2hWBFkKgy4zph0C7ItBCCHSZMf0Q\naFcEWgiBLjOmHwLtikALIdBlxvRDoF0RaCEEusyYfgi0KwIthECXGdMPgXZFoIUQ6DJj+iHQ\nrgi0EAJdZkw/BNoVgRZCoMuM6YdAuyLQQgh0mTH9EGhXBFoIgS4zph8C7YpACyHQZcb0Q6Bd\nEWghBLrMmH4ItCsCLYRAlxnTD4F2RaCFEOgyY/oh0K4ItBACXWZMPwTaFYEWQqDLjOmHQLsi\n0EIIdJkx/RBoVwRaCIEuM6YfAu2KQAsh0GXG9EOgXRFoIQS6zJh+CLQrAi2EQJcZ0w+BdkWg\nhRDoMmP6IdCuCLQQAl1mTD8E2hWBFkKgy4zph0C7ItBCCHSZMf0QaFcEWgiBLjOmHwLtikAL\nIdBlxvRDoF0RaCEEusyYfgi0KwIthECXGdMPgXZFoIUQ6DJj+iHQrgi0EAJdZkw/BNoVgRZC\noMuM6YdAuyLQQgh0mTH9EGhXBFoIgS4zph8C7YpACyHQZcb0Q6BdEWghBLrMmH4ItCsCLYRA\nlxnTD4F2RaCFEOgyY/oh0K4ItBACXWZMPwTaFYEWQqDLjOmHQLsi0EIIdJkx/RBoVwRaCIEu\nM6YfAu2KQAsh0GXG9EOgXRFoIQS6zJh+CLQrAi2EQJcZ0w+BdkWghRDoMmP6IdCuCLQQAl1m\nTD8E2hWBFkKgy4zph0C7ItBCCHSZMf0QaFcEWgiBLjOmHwLtikALIdBlxvRDoF0RaCEEusyY\nfgi0KwIthECXGdMPgXZFoIUQ6DJj+iHQrgi0EAJdZkw/BNoVgRZCoMuM6YdAuyLQQgh0mTH9\nEGhXBFoIgS4zph8C7YpACyHQZcb0Q6BdEWghBLrMmH4ItCsCLYRAlxnTD4F2RaCFEOgyY/oh\n0K4ItBACXWZMPwTaFYEWQqDLjOmHQLsi0EIIdJkx/RBoVwRaCIEuM6YfAu2KQAsh0GXG9EOg\nXRFoIQS6zJh+CLQrAi2EQJcZ0w+BdkWghRDoMmP6IdCuCLQQAl1mTD8E2hWBFkKgy4zph0C7\nItBCCHSZMf0QaFcEWgiBLjOmHwLtikALIdBlxvRDoF0RaCEEusyYfgi0KwIthECXGdMPgXZF\noIUQ6DJj+iHQrgi0EAJdZkw/BNoVgRZCoMuM6YdAuyLQQgh0mTH9EGhXBFoIgS4zph8C7YpA\nCyHQZcb0Q6BdEWghBLrMmH4ItCsCLYRAlxnTD4F2RaCFEOgyY/oh0K4ItBACXWZMPwTaFYEW\nQqDLjOmHQLsi0EIIdJkx/RBoVwRaCIEuM6YfAu2KQAsh0GXG9EOgXRFoIQS6zJh+CLQrAi2E\nQJcZ0w+BdkWghRDoMmP6IdCuCLQQAl1mTD8E2hWBFkKgy4zph0C7ItBCCHSZMf0QaFcEWgiB\nLjOmHwLtikALIdBlxvRDoF0RaCEEusyYfgi0KwIthECXGdMPgXZFoIUQ6DJj+iHQrgi0EAJd\nZkw/BNoVgRZCoMuM6YdAuyLQQgh0mTH9EGhXBFoIgS4zph8C7YpACyHQZcb0Q6BdEWghBLrM\nmH4ItCsCLYRAlxnTD4F2RaCFEOgyY/oh0K4ItBACXWZMPwTaFYEWQqDLjOmHQLsi0EIIdJkx\n/RBoVwRaCIEuM6YfAu2KQAsh0GXG9EOgXRFoIQS6zJh+CLQrAi2EQJcZ0w+BdkWghRDoMmP6\nIdCuCLQQAl1mTD8E2hWBFkKgy4zph0C7ItBCCHSZMf0QaFcEWgiBLjOmHwLtikALIdBlxvRD\noF0RaCEEusyYfgi0KwIthECXGdMPgXZFoIUQ6DJj+iHQrgi0EAJdZkw/BNoVgRZCoMuM6YdA\nuyLQQgh0mTH9EGhXBFoIgS4zph8C7YpACyHQZcb0Q6BdEWghBLrMmH4ItCsCLYRAlxnTD4F2\nRaCFEOgyY/oh0K4ItJDq466bM2XihSuGPiLQPrbcLIH2seVmCbSQ6uPOOnf5Y5dM3zj4EYH2\nseVmCbSPLTdLoIVUHndl37LmP5HH3z34IYG2sdVmCbSNrTZLoIVUHve2CZuab8+8avBDAm1j\nq80SaBtbbZZAC6k87vyprbfnz2u+eaCnaX8CbWKzzS5ubXYfAm1is83e2NrsKAIto3qgp7Xe\n9i/7l5MmTTrqP93c4QEQsdlmf9bc7Dv2XNLhARCx2WZ/0tzs4SMf7/AA1EblQC8e+Nelqwc/\n3PJLHBC21Wa3/BIHhG212S2/xIFaqxzoVX1LU1oz7t7BDwm0ja02S6BtbLVZAi2k+ldkLjp7\n+aMzZ2wa/IhA+9hyswTax5abJdBCqgf66bmTT5i9eugjAu1jy80SaB9bbpZACyn7rd7wQaBd\nEWghBBoxAu2KQAsh0IgRaFcEWgiBRoxAuyLQQgg0YgTaFYEWQqARI9CuCLQQAo0YgXZFoIUQ\naMQItCsCLYRAI0agXRFoIQQaMQLtikALIdCIEWhXBFoIgUaMQLsi0EIINGIE2hWBFkKgESPQ\nrgi0EAKNGIF2RaCFEGjECLQrAi2EQCNGoF0RaCEEGjEC7YpACyHQiBFoVwRaCIFGjEC7ItBC\nCDRiBNoVgRZCoBEj0K4ItBACjRiBdkWghRBoxAi0KwIthEAjRqBdEWghBBoxAu2KQAsh0IgR\naFcEWgiBRoxAuyLQQgg0YgTaFYEWQqARI9CuCLQQAo0YgXZFoIUQaMQItCsCLYRAI0agXRFo\nIQQaMQLtikAL6TbQs3fba3Q1++63b8VHdDYcR+5X+SGPFNlIXczYde+Kn/8wbHZYnizVN/vM\njl5GUZNetk/Fz3/f6pes85F12Oyrd/QuOuo20OncAyv6k57XVX1IJwf1HFL6yAPf2FP5IY+V\nWEh9fLDq539wz3+pfM06eG3PG0ofeeChh1Z+yLM7ehdlja/6+R/Sc1Dla9bB63r+pPSRB77p\nv1Z9xBt39Co66jrQlX255yelj3y4529KH5mmvLn4ke7+R8/PSx95d8/c0kem3ncVP9LdX/es\nKH3kD3uuKH1ketsHih+5wxHoNgh0ZQTaFYHeYQh0GwS6MgLtikDvMNs/0Ldf9lDpI5+6rPwf\nJbnqs8WPdPfjy4q/jB+/7F9LH5muuLz4ke5uumxd6SOXXfZvpY9MX/hW8SN3uO0faABAFgIN\nADVFoAGgpoY90Of09vZN/PgPNw3L2RP6/+jx9Juab1Z+4eRjTpx1b/eHnjzwhaypV3d/lDc2\n60pts86LHf5AX7pyxb1XHf/3w7Htcyae1/qltexHTph+28P3zBnX/Z8Qcd52UWzWldpmnRc7\n/IH+Uuvt0nE/SunXF0+e8JGfp5N+kNLXe1ek9NGrNvX+8yfOaH38Is++auLCNLDs887o/26v\nK7v/77jPb/vZ3vkfPfmkxV2faIrNulLbrPNit1Og06yZKf3VBU89c/nxay67LKUPn/HD9Ox7\nl6Zx5zyVFrxv/Ys9+4aFH3iqf9lP9d5SauDnt/3b3rN+nRaOf6rUyWbYrCu1zTovdnsF+hsf\nSst6H07p2WN/+C8fSuuPve6y9L9P2JTGfT+lFb2/fLFn35DOv7h/2Q/0Li018Obbvi6ljccv\nKHWyGTbrSm2zzovdXoG+4vR0a1/ra1pn/K+1favvPPehU9M3L0lp3G0prX7Ri2ou+/EJdw4s\n+99LDbz5tm9vvnO64Z9+L4LNulLbrPNit1egPzp7cNmnfzOdc+vl39w0cdW5P2gue1GXy05X\nn7T+zJvSur6b+m/Z2P1/1/jQV1pvN73v+81tN5+K6dRvd32kJzbrSm2zzovdToFe1LsoLW/9\na9H6CT9IV37pnLvTrIXHrC6x7N+eOe+s5qIvOOnp1i1fP7/rgT95TusJc0/vfc1tf6f573cT\nin0N1AybdaW2WefFbo8/Zrfy/q+P/0zz3Y/MXLv+i5OeTvecNuHZdO2Zf5FKLDs9MH5Sc9mP\nTfrgPz/8s7kTftr1wA8fO+ffH75l8pzm86h3+kPPfuuYNV0f6YnNulLbrPNit8c3qvS+78P9\n/0h78pOTTrjw0ZQ2HHtuc0e9X01Flp3m9fb/offPTTtm8t8vKzDx/5k9+X3Tv/tca9s3/fWE\nk+8scKQlNutKbrPGi+Vbvdv7bW/5v3ALdcBmTfktlkC357dtDGCzpvwWS6Db89s2BrBZU36L\nJdAAUFMEGgBqikADQE0RaACoKQINADWlEOgLGk17v+mvl7c+eMvrOt298z1QE2zWFZstRCPQ\nH/vKvIuPGzmy9X1Mc2e/wD2XNDrdA3XCZl2x2UI0Ar2o9csjb955fod7Xqbw6eB32KwrNluI\nwtUZXHZ6fI8xA/8ydMSR17/y8JR+/M69dj/08tb/suAde/7esUvTu5v/WtUz8K9LNx2558g3\nzNmU0pFv/+lRex3w/hU78hNAG2zWFZstRCnQaXLjF/2rPGrM6z9/Q7pll3dcv+C0xqeau97p\nXd+8/MA/eOLBcY07f95/j+/u9J7v3TKj8ZGUjv7DP1244ju7TNmxnwJCbNYVmy1EKtCXNW7q\nX+XRjWubHx16UOuvk+3ba3168x9tSOn2XT+TTm59Oq17vP5VrZ9HOX7Er5r3vbX53tGv2GHT\noz0264rNFiIV6K82rhpY9q7PpbSicdb6pi817vhV44zBew4t+7HGaa0PL2/ckI7eo/XelJ13\nxODogM26YrOFSAX64sbCgWW3/tm6pDHo2p81Zg7ec2jZdzRmtT68qTEvHf3q3/0PqBs264rN\nFqJwEX637L6dVg4su7XAJY2TFvVbeV/jE4P3HFr2nY0LWx/e2PgHll1nbNYVmy1E4SIMLfv+\nlx2dnl/2qsbQf0RY25jW+uWhJ3+37Ccap7ZumdeYz7LrjM26YrOFKFyEwWU/9McjFm+27HTY\nPr9uvr3y/A3pkAPWNp8KzX9p+mBjw8A9Dn7F+ub/9p491rDsOmOzrthsIQoXofVdSV/5zLQ9\ndvtm2nzZPx4x5sqbPz5iako37PyWb8177cufSJ9oXPid/nvcuPO7rvun0xsXJZZdZ2zWFZst\nROEi9H9f/65/dOoDrQ+eX3b61z/fa8RrL27+8zfd+NY9Xn7Mgyk9cuiI1w38ofcFbx+126Gt\nbzNl2TXGZl2x2UK4CABQUwQaAGqKQANATRFoAKgpAg0ANUWgAaCmCDQA1BSBBoCaItAAUFME\nGgBqikADQE0RaACoKQINADVFoAGgpgg0ANQUgQaAmiLQAFBTBBoAaopAA0BNEWgAqCkCDQA1\nRaABoKb+P4xLg0NMZ9MPAAAAAElFTkSuQmCC"
          },
          "metadata": {
            "image/png": {
              "width": 720,
              "height": 480
            }
          }
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "#Box plots with replicates to show the normalized counts calculated by DESeq2"
      ],
      "metadata": {
        "id": "eXYLDS0l1i_q"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "mat <- counts(se_bulk,normalized=TRUE)"
      ],
      "metadata": {
        "id": "Y8cGoOQf6GFv"
      },
      "execution_count": 54,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "colnames(mat) <- se_bulk$group"
      ],
      "metadata": {
        "id": "6P3ZTBCn2Acf"
      },
      "execution_count": 55,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "rownames(mat) <- rowData(se_bulk)$SYMBOL"
      ],
      "metadata": {
        "id": "MZ3hw3gjAOaF"
      },
      "execution_count": 56,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "genes_selected <- data.frame(t(mat[rownames(mat) %in% c('Maf','Smad3'),]))"
      ],
      "metadata": {
        "id": "-l8DKg-F2IHK"
      },
      "execution_count": 57,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "genes_selected$group <- se_bulk$group"
      ],
      "metadata": {
        "id": "8cljpSeG2nhg"
      },
      "execution_count": 58,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "unique(genes_selected$group)"
      ],
      "metadata": {
        "id": "OJ8vpQDFAhsd",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 53
        },
        "outputId": "8fee0301-a042-4a6d-a325-da06bc9d95a0"
      },
      "execution_count": 59,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/html": [
              "<style>\n",
              ".list-inline {list-style: none; margin:0; padding: 0}\n",
              ".list-inline>li {display: inline-block}\n",
              ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
              "</style>\n",
              "<ol class=list-inline><li>ESCs</li><li>MEF</li><li>OKS_D0</li><li>OKS_D1</li><li>OKS_D3</li><li>OKS_D5</li><li>OKS_D7</li><li>OKS_cJun_D0</li><li>OKS_cJun_D1</li><li>OKS_cJun_D3</li><li>OKS_cJun_D7</li><li>iPS</li><li>OSK_shSap30_D7</li></ol>\n",
              "\n",
              "<details>\n",
              "\t<summary style=display:list-item;cursor:pointer>\n",
              "\t\t<strong>Levels</strong>:\n",
              "\t</summary>\n",
              "\t<style>\n",
              "\t.list-inline {list-style: none; margin:0; padding: 0}\n",
              "\t.list-inline>li {display: inline-block}\n",
              "\t.list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
              "\t</style>\n",
              "\t<ol class=list-inline><li>'ESCs'</li><li>'iPS'</li><li>'MEF'</li><li>'OKS_cJun_D0'</li><li>'OKS_cJun_D1'</li><li>'OKS_cJun_D3'</li><li>'OKS_cJun_D7'</li><li>'OKS_D0'</li><li>'OKS_D1'</li><li>'OKS_D3'</li><li>'OKS_D5'</li><li>'OKS_D7'</li><li>'OSK_shSap30_D7'</li></ol>\n",
              "</details>"
            ],
            "text/markdown": "1. ESCs\n2. MEF\n3. OKS_D0\n4. OKS_D1\n5. OKS_D3\n6. OKS_D5\n7. OKS_D7\n8. OKS_cJun_D0\n9. OKS_cJun_D1\n10. OKS_cJun_D3\n11. OKS_cJun_D7\n12. iPS\n13. OSK_shSap30_D7\n\n\n\n**Levels**: 1. 'ESCs'\n2. 'iPS'\n3. 'MEF'\n4. 'OKS_cJun_D0'\n5. 'OKS_cJun_D1'\n6. 'OKS_cJun_D3'\n7. 'OKS_cJun_D7'\n8. 'OKS_D0'\n9. 'OKS_D1'\n10. 'OKS_D3'\n11. 'OKS_D5'\n12. 'OKS_D7'\n13. 'OSK_shSap30_D7'\n\n\n",
            "text/latex": "\\begin{enumerate*}\n\\item ESCs\n\\item MEF\n\\item OKS\\_D0\n\\item OKS\\_D1\n\\item OKS\\_D3\n\\item OKS\\_D5\n\\item OKS\\_D7\n\\item OKS\\_cJun\\_D0\n\\item OKS\\_cJun\\_D1\n\\item OKS\\_cJun\\_D3\n\\item OKS\\_cJun\\_D7\n\\item iPS\n\\item OSK\\_shSap30\\_D7\n\\end{enumerate*}\n\n\\emph{Levels}: \\begin{enumerate*}\n\\item 'ESCs'\n\\item 'iPS'\n\\item 'MEF'\n\\item 'OKS\\_cJun\\_D0'\n\\item 'OKS\\_cJun\\_D1'\n\\item 'OKS\\_cJun\\_D3'\n\\item 'OKS\\_cJun\\_D7'\n\\item 'OKS\\_D0'\n\\item 'OKS\\_D1'\n\\item 'OKS\\_D3'\n\\item 'OKS\\_D5'\n\\item 'OKS\\_D7'\n\\item 'OSK\\_shSap30\\_D7'\n\\end{enumerate*}\n",
            "text/plain": [
              " [1] ESCs           MEF            OKS_D0         OKS_D1         OKS_D3        \n",
              " [6] OKS_D5         OKS_D7         OKS_cJun_D0    OKS_cJun_D1    OKS_cJun_D3   \n",
              "[11] OKS_cJun_D7    iPS            OSK_shSap30_D7\n",
              "13 Levels: ESCs iPS MEF OKS_cJun_D0 OKS_cJun_D1 OKS_cJun_D3 ... OSK_shSap30_D7"
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "genes_selected <- genes_selected[genes_selected$group %in% c('MEF','OKS_D7'),]"
      ],
      "metadata": {
        "id": "0ddBdHKX2rkr"
      },
      "execution_count": 60,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "genes_selected"
      ],
      "metadata": {
        "id": "pV7xCcl36WcV",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 223
        },
        "outputId": "0401baa5-3c48-4a1f-8f66-880382da16b5"
      },
      "execution_count": 61,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/html": [
              "<table class=\"dataframe\">\n",
              "<caption>A data.frame: 4 × 3</caption>\n",
              "<thead>\n",
              "\t<tr><th></th><th scope=col>Maf</th><th scope=col>Smad3</th><th scope=col>group</th></tr>\n",
              "\t<tr><th></th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;fct&gt;</th></tr>\n",
              "</thead>\n",
              "<tbody>\n",
              "\t<tr><th scope=row>MEF</th><td> 60.471000</td><td>443.1236</td><td>MEF   </td></tr>\n",
              "\t<tr><th scope=row>MEF.1</th><td>174.629643</td><td>339.3932</td><td>MEF   </td></tr>\n",
              "\t<tr><th scope=row>OKS_D7</th><td>  8.924387</td><td>227.0761</td><td>OKS_D7</td></tr>\n",
              "\t<tr><th scope=row>OKS_D7.1</th><td>  5.915184</td><td>207.0314</td><td>OKS_D7</td></tr>\n",
              "</tbody>\n",
              "</table>\n"
            ],
            "text/markdown": "\nA data.frame: 4 × 3\n\n| <!--/--> | Maf &lt;dbl&gt; | Smad3 &lt;dbl&gt; | group &lt;fct&gt; |\n|---|---|---|---|\n| MEF |  60.471000 | 443.1236 | MEF    |\n| MEF.1 | 174.629643 | 339.3932 | MEF    |\n| OKS_D7 |   8.924387 | 227.0761 | OKS_D7 |\n| OKS_D7.1 |   5.915184 | 207.0314 | OKS_D7 |\n\n",
            "text/latex": "A data.frame: 4 × 3\n\\begin{tabular}{r|lll}\n  & Maf & Smad3 & group\\\\\n  & <dbl> & <dbl> & <fct>\\\\\n\\hline\n\tMEF &  60.471000 & 443.1236 & MEF   \\\\\n\tMEF.1 & 174.629643 & 339.3932 & MEF   \\\\\n\tOKS\\_D7 &   8.924387 & 227.0761 & OKS\\_D7\\\\\n\tOKS\\_D7.1 &   5.915184 & 207.0314 & OKS\\_D7\\\\\n\\end{tabular}\n",
            "text/plain": [
              "         Maf        Smad3    group \n",
              "MEF       60.471000 443.1236 MEF   \n",
              "MEF.1    174.629643 339.3932 MEF   \n",
              "OKS_D7     8.924387 227.0761 OKS_D7\n",
              "OKS_D7.1   5.915184 207.0314 OKS_D7"
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "df_genes <- gather(genes_selected, \"genes\", \"normalized_counts\", -group)"
      ],
      "metadata": {
        "id": "oVPEC_6N4-NY"
      },
      "execution_count": 62,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Barplot for normalized read counts showing a decrease at D7 MEFs post OSK induction"
      ],
      "metadata": {
        "id": "stC16Pck9ctX"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "options(repr.plot.width = 8, repr.plot.height = 10)\n",
        "df_genes %>%\n",
        "  ggplot(aes(group, normalized_counts, fill = group)) +\n",
        "  geom_bar(stat = 'identity', position = 'dodge') +\n",
        "  geom_jitter(size = 2) + \n",
        "  scale_fill_manual(values = c('lightgrey','lightgrey')) +\n",
        "  facet_grid(genes ~ ., scales = 'free') + \n",
        "  theme_linedraw()\n"
      ],
      "metadata": {
        "id": "KmBuUr2J5Ol2",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 617
        },
        "outputId": "1767e3ea-30b8-46be-952e-fda9d332169b"
      },
      "execution_count": 63,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "plot without title"
            ],
            "image/png": "iVBORw0KGgoAAAANSUhEUgAAA8AAAASwCAMAAAADjdPmAAAC/VBMVEUAAAABAQECAgIDAwME\nBAQFBQUGBgYHBwcICAgJCQkKCgoLCwsMDAwNDQ0ODg4PDw8QEBARERESEhITExMUFBQVFRUW\nFhYXFxcYGBgZGRkaGhobGxscHBwdHR0eHh4fHx8gICAhISEiIiIjIyMkJCQlJSUmJiYnJyco\nKCgpKSkqKiorKyssLCwtLS0uLi4vLy8wMDAxMTEyMjIzMzM0NDQ1NTU2NjY3Nzc4ODg5OTk6\nOjo7Ozs8PDw9PT0+Pj4/Pz9AQEBBQUFCQkJDQ0NERERFRUVGRkZHR0dISEhJSUlKSkpLS0tM\nTExNTU1OTk5PT09QUFBRUVFSUlJTU1NUVFRVVVVWVlZXV1dYWFhZWVlaWlpbW1tcXFxdXV1e\nXl5fX19gYGBhYWFiYmJjY2NkZGRlZWVmZmZnZ2doaGhpaWlqampra2tsbGxtbW1ubm5vb29w\ncHBxcXFycnJzc3N0dHR1dXV2dnZ3d3d4eHh5eXl6enp7e3t9fX1+fn5/f3+AgICBgYGCgoKD\ng4OEhISFhYWGhoaHh4eIiIiJiYmKioqLi4uMjIyNjY2Ojo6Pj4+QkJCRkZGSkpKTk5OUlJSV\nlZWWlpaXl5eYmJiZmZmampqbm5ucnJydnZ2enp6fn5+goKChoaGioqKjo6OkpKSlpaWmpqan\np6eoqKipqamqqqqrq6usrKytra2urq6vr6+wsLCxsbGysrKzs7O0tLS1tbW2tra3t7e4uLi5\nubm6urq7u7u8vLy9vb2+vr6/v7/AwMDBwcHCwsLDw8PExMTFxcXGxsbHx8fIyMjJycnKysrL\ny8vMzMzNzc3Ozs7Pz8/Q0NDR0dHS0tLT09PU1NTV1dXW1tbX19fY2NjZ2dna2trb29vc3Nzd\n3d3e3t7f39/g4ODh4eHi4uLj4+Pk5OTl5eXm5ubn5+fo6Ojp6enq6urr6+vs7Ozt7e3u7u7v\n7+/w8PDx8fHy8vLz8/P09PT19fX29vb39/f4+Pj5+fn6+vr7+/v8/Pz9/f3+/v7///9eq/7C\nAAAACXBIWXMAABJ0AAASdAHeZh94AAAgAElEQVR4nO3deZxddXn48TOZ7AFpSAATwlJAhJKE\nJSxBCGBQSopFw6JSZHFfsKC4Vi0uCIrys5XaGn9YCtQfKi0JEJAdEWRLNBGqbKYIaYgICZuW\nLeS8fvfM3DszNyQn9zvP9zzP8z3zef8xd7abOTzP98OdSSaZLAeQrMz6AgAMHgEDCSNgIGEE\nDCSMgIGEETCQMAIGEkbAQMKkAR+3Qx1sY30BtVGTSW46YsOeihJeLNKAD3koymUY+5X1BdRG\nPSZ54PBsw1ZZX10bAi7U49h5UI9JEnBi6nHsPKjHJAk4MfU4dh7UY5IEnJh6HDsP6jFJAk5M\nPY6dB/WYJAEnph7HzoN6TJKAE1OPY+dBPSZJwImpx7HzoB6TJODE1OPYeVCPSRJwYupx7Dyo\nxyQJODH1OHYe1GOSBJyYehw7D+oxSQJOTD2OnQf1mCQBJ6Yex86DekySgBNTj2PnQT0mScCJ\nqcex86AekyTgxNTj2HlQj0kScGLqcew8qMckCTgx9Th2HtRjkgScmHocOw/qMUkCHry1t513\n1hX/G/kX3Zh6HDsP6jFJAh60lQcWM5r007i/6sbU49h5UI9JDqGA979pWUwP7d47pLE/jfrL\nbsxVqh+tzuoxyb27h0zAkR+Br2pN6SNRf9mNqcfjhgf1mOQQegQOD/jeMh9oTen1pe82JAg3\nY4OAlTkL+MTWlLZTqsQx4WZsELAyZwF/oTWlg5UqcUy4GRsErMxZwLeMa07p/yhV4phwMzYI\nWJmzgO89d2TPkI7VacQ14WZsELAybwHf+x9H7jT54HNVCnFOuBkbBKzMXcBoEW7GBgErI2C3\nhJuxQcDKCNgt4WZsELAyAnZLuBkbBKyMgN0SbsYGASsjYLeEm7FBwMoI2C3hZmwQsDICdku4\nGRsErIyA3RJuxgYBKyNgt4SbsUHAygjYLeFmbBCwMgJ2S7gZGwSsjIDdEm7GBgErI2C3hJux\nQcDKCNgt4WZsELAyAnZLuBkbBKyMgN0SbsYGASsjYLeEm7FBwMoI2C3hZmwQsDICdku4GRsE\nrIyA3RJuxgYBKyNgt4SbsUHAygjYLeFmbBCwMgJ2S7gZGwSsjIDdEm7GBgErI2C3hJuxQcDK\nCNgt4WZsELAyAnZLuBkbBKyMgN0SbsYGASsjYLeEm7FBwMoI2C3hZmwQsDICdku4GRsErIyA\n3RJuxgYBKyNgt4SbsUHAygjYLeFmbBCwMgJ2S7gZGwSsjIDdEm7GBgErI2C3hJuxQcDKCNgt\n4WZsELAyAnZLuBkbBKyMgN0SbsYGASsjYLeEm7FBwMoI2C3hZmwQsDICdku4GRsErIyA3RJu\nxgYBKyNgt4SbsUHAygjYLeFmbBCwMgJ2S7gZGwSsjIDdEm7GBgErI2C3hJuxQcDKCNgt4WZs\nELAyAnZLuBkbBKyMgN0SbsYGASsjYLeEm7FBwMoI2C3hZmwQsDICdku4GRsErIyA3RJuxgYB\nKyNgt4SbsUHAygjYLeFmbBCwMgJ2S7gZGwSsjIDdEm7GBgEr23jA9+/XXdxML659XJ6vPn7y\n5kc83PdWAq5M6GBdIGBlGw34h5NO7Al4yreXL1++Is+PPHDpg8fttqb1ZgKuTOhgXSBgZRsN\n+MJH5vcEPPaqnhcf7VrSeBQefn3rzQRcmdDBukDAyjr4Grgn4Bey9+657VEP5JeNWdt4adqZ\nrTcScGVCB+sCASvrNOA/bHXCXXfO2eqpeVsXr5p9auPJU4sbZhFwVUIH6wIBK+s04B7Pjjt/\n3pTimdmnNZ7c9qaG6TctC2TdRTJCB+vCVdYXEMXe3fUMON/1jAWji0+hp57Veg2fQlcmdLAu\n8AisrNOA733fi3n+3LiLVnQtyvMnum9pvZGAKxM6WBcIWNlGA165/Pzu5cufe3LCicvuP2qb\nP+VH77X0gTn7rG29mYArEzpYFwhY2UYD3q7nor+VLzl0sy2O/O88f+akSRPnPtb3ZgKuTOhg\nXSBgZXwrpVvCzdggYGUE7JZwMzYIWBkBuyXcjA0CVkbAbgk3Y4OAlRGwW8LN2CBgZQTslnAz\nNghYGQG7JdyMDQJWRsBuCTdjg4CVEbBbws3YIGBlBOyWcDM2CFgZAbsl3IwNAlZGwG4JN2OD\ngJURsFvCzdggYGUE7JZwMzYIWBkBuyXcjA0CVkbAbgk3Y4OAlRGwW8LN2CBgZQTslnAzNghY\nGQG7JdyMDQJWRsBuCTdjg4CVEbBbws3YIGBlBOyWcDM2CFgZAbsl3IwNAlZGwG4JN2ODgJUR\nsFvCzdggYGUE7JZwMzYIWBkBuyXcjA0CVkbAbgk3Y4OAlRGwW8LN2CBgZQTslnAzNghYGQG7\nJdyMDQJWRsBuCTdjg4CVEbBbws3YIGBlBOyWcDM2CFgZAbsl3IwNAlZGwG4JN2ODgJURsFvC\nzdggYGUE7JZwMzYIWBkBuyXcjA0CVkbAbgk3Y4OAlRGwW8LN2CBgZQTslnAzNghYGQG7JdyM\nDQJWRsBuCTdjg4CVEbBbws3YIGBlBOyWcDM2CFgZAbsl3IwNAlZGwG4JN2ODgJURsFvCzdgg\nYGUE7JZwMzYIWBkBuyXcjA0CVkbAbgk3Y4OAlRGwW8LN2CBgZQTslnAzNghYGQG7JdyMDQJW\nRsBuCTdjg4CVEbBbws3YIGBlBOyWcDM2CFgZAbsl3IwNAlZGwG4JN2ODgJURsFvCzdggYGUE\n7JZwMzYIWBkBuyXcjA0CVkbAbgk3Y4OAlRGwW8LN2CBgZQTslnAzNghYGQG7JdyMDQJWRsBu\nCTdjg4CVSQOetXh1IOsukhE6WBdutb6AKGYOnYB5BK6McDM2eARWRsBuCTdjg4CVEbBbws3Y\nIGBlBOyWcDM2CFgZAbsl3IwNAlZGwG4JN2ODgJURsFvCzdggYGUE7JZwMzYIWBkBuyXcjA0C\nVkbAbgk3Y4OAlRGwW8LN2CBgZQTslnAzNghYGQG7JdyMDQJWRsBuCTdjg4CVEbBbws3YIGBl\nBOyWcDM2CFgZAbsl3IwNAlZGwG4JN2ODgJURsFvCzdggYGUE7JZwMzYIWBkBuyXcjA0CVkbA\nbgk3Y4OAlRGwW8LN2CBgZQTslnAzNghYGQG7JdyMDQJWRsBuCTdjg4CVEbBbws3YIGBlBOyW\ncDM2CFgZAbsl3IwNAlZGwG4JN2ODgJURsFvCzdggYGUE7JZwMzYIWBkBuyXcjA0CVkbAbgk3\nY4OAlRGwW8LN2CBgZQTslnAzNghYGQG7JdyMDQJWRsBuCTdjg4CVEbBbws3YIGBlBOyWcDM2\nCFgZAbsl3IwNAlZGwG4JN2ODgJURsFvCzdggYGUE7JZwMzYIWBkBuyXcjA0CVkbAbgk3Y4OA\nlRGwW8LN2CBgZQTslnAzNghYGQG7JdyMDQJWRsBuCTdjg4CVEbBbws3YIGBlBOyWcDM2CFgZ\nAbsl3IwNAlZGwG4JN2ODgJURsFvCzdggYGUE7JZwMzYIWBkBuyXcjA0CVkbAbgk3Y4OAlRGw\nW8LN2CBgZQTslnAzNghYGQG7JdyMDQJWRsBuCTdjg4CVEbBbws3YIGBlBOyWcDM2CFgZAbsl\n3IwNAlZGwG4JN2ODgJURsFvCzdggYGUE7JZwMzYIWNnGA75/v+7iZvXxkzc/4uH+2yYCrkzo\nYF0gYGUbDfiHk07sCfjIA5c+eNxua/pumwi4MqGDdYGAlW004AsfmV8E/GjXksaj7/DrW7et\nNxNwZUIH6wIBK+vga+CegC8bs7bxdNqZrdvWGwm4MqGDdYGAlXUa8Lyti2dnn9q6bTx5aXXD\nwQRcldDBukDAyjoOeErx7OzTWreNJzfNaNht/uJA1l0kI3SwLlxifQFR7NFdu4AXjC4+dZ56\nVuu29UY+ha5M6GBd4BFYWacBr+halOdPdN/Sum29kYArEzpYFwhY2UYDXrn8/O7ly5/Lj95r\n6QNz9lnbd9tEwJUJHawLBKxsowFv13PR38qfOWnSxLmP5X23TQRcmdDBukDAyvhWSreEm7FB\nwMoI2C3hZmwQsDICdku4GRsErIyA3RJuxgYBKyNgt4SbsUHAygjYLeFmbBCwMgJ2S7gZGwSs\njIDdEm7GBgErI2C3hJuxQcDKCNgt4WZsELAyAnZLuBkbBKyMgN0SbsYGASsjYLeEm7FBwMoI\n2C3hZmwQsDICdku4GRsErIyA3RJuxgYBKyNgt4SbsUHAygjYLeFmbBCwMgJ2S7gZGwSsjIDd\nEm7GBgErI2C3hJuxQcDKCNgt4WZsELAyAnZLuBkbdQu4/8etEHCTdRfJEG7GRt0C/mivCx/P\nCbjJuotkCDdjo24BFyb/3UNP/uMeBNxk3UUyhJuxUbuAh7/1ihcWHj2Sr4H7WHeRDOFmbNQt\n4K+v/NWnJ/GbWANZd5EM4WZs1C3gX/zuS6/nd6HbWHeRDOFmbNQt4Gz3bz95x4fGE3A/6y6S\nIdyMjdoFnGUj337Nn370V90E3GTdRTKEm7FRt4Bn9jj6kidXEnCTdRfJEG7GRt0C7n8dATdZ\nd5EM4WZs1C3g4X0IuMm6i2QIN2OjbgFn+75zR34Xuo11F8kQbsZG3QI+O1/z/BvH3/8XBNzH\nuotkCDdjo24Brz54xOdvHPfv1xBwH+sukiHcjI26BbxqVDZmZTZuFQH3se4iGcLN2KhbwN9/\nR5b9Ptv0aQLuY91FMoSbsVG3gDe7+ctvfeqoS68l4D7WXSRDuBkbdQv4ruLFl27fmYD7WHeR\nDOFmbNQt4Ke3Hz58eBd/jDSAdRfJEG7GRt0C3r735mcE3Me6i2QIN2OjbgEv7LXmyotmEHAv\n6y6SIdyMjboF/N1eL3/3tocIuJd1F8kQbsZG3QJu+mM2/gUC7mXdRTKEm7FR04C3zIZtQcC9\nrLtIhnAzNmoaML8L3c+6i2QIN2ODgJURsFvCzdggYGUE7JZwMzYIWBkBuyXcjA0CVkbAbgk3\nY4OAlRGwW8LN2CBgZQTslnAzNghYGQG7JdyMDQJWRsBuCTdjg4CVEbBbws3YIGBlBOyWcDM2\nCFgZAbsl3IwNAlZGwG4JN2ODgJURsFvCzdggYGUE7JZwMzYIWBkBuyXcjA0CVkbAbgk3Y4OA\nlRGwW8LN2CBgZQTslnAzNghYGQG7JdyMDQJWRsBuCTdjg4CVSQOetXh1IOsukhE6WBdutb6A\nKGYOnYB5BK6McDM2eARWRsBuCTdjg4CVEbBbws3YIGBlBOyWcDM2CFgZAbsl3IwNAlZGwG4J\nN2ODgJURsFvCzdggYGUE7JZwMzYIWBkBuyXcjA0CVkbAbgk3Y4OAlRGwW8LN2CBgZQTslnAz\nNghYGQG7JdyMDQJWRsBuCTdjg4CVEbBbws3YIGBlBOyWcDM2CFgZAbsl3IwNAlZGwG4JN2OD\ngJURsFvCzdggYGUE7JZwMzYIWBkBuyXcjA0CVkbAbgk3Y4OAlRGwW8LN2CBgZQTslnAzNghY\nGQG7JdyMDQJWRsBuCTdjg4CVEbBbws3YIGBlBOyWcDM2CFgZAbsl3IwNAlZGwG4JN2ODgJUR\nsFvCzdggYGUE7JZwMzYIWBkBuyXcjA0CVkbAbgk3Y4OAlRGwW8LN2CBgZQTslnAzNghYGQG7\nJdyMDQJWRsBuCTdjg4CVEbBbws3YIGBlBOyWcDM2CFgZAVfnjou+c43g7sLN2CBgZQRclSUn\nj2xse/rlg/4FhJuxQcDKCLgqx/Sue/MbB/sLCDdjg4CVEXBFFrT2fdxgfwXhZmwQsDICrshn\nW/vebrC/gnAzNghYGQFX5COtfY8f7K8g3IwNAlZGwBX5Umvfuw72VxBuxgYBKyPgitw8qrnv\n0wb7Kwg3Y4OAlRFwVT7fu+6piwf7Cwg3Y4OAlRFwZb49bUS25cl3Dfr+ws3YIGBlBFyhJbdL\n7i3cjA0CVkbAbgk3Y4OAlRGwW8LN2CBgZQTslnAzNghYGQG7JdyMDQJWRsBuCTdjg4CVEbBb\nws3YIGBlBOyWcDM2CFgZAbsl3IwNAlZGwG4JN2ODgJURsFvCzdggYGUE7JZwMzYIWBkBuyXc\njA0CVkbAbgk3Y4OAlRGwW8LN2CBgZQTslnAzNghYGQG7JdyMDQJWRsBuCTdjg4CVdRzw9OLa\nx+X56uMnb37Ew32vJuDKhA7WBQJW1nHAU769fPnyFXl+5IFLHzxutzWtVxNwZUIH6wIBK+s4\n4LFX9dw82rWk8Sg8/PrWqwm4MqGDdYGAlXUa8AvZe/fc9qgH8svGrG28NO3M1usJuDKhg3WB\ngJV1GvAftjrhrjvnbPXUvK2Ll2af2no9AVcmdLAuELCyoN+Ffnbc+fOmFM/MPq3x5NodGv5i\n/uJA1l0kI3SwLlxifQFR7NFdz4DzXc9YMLr4FHrqWa3X8AhcmdDBusAjsLJOA773fS/m+XPj\nLlrRtSjPn+i+pfV6Aq5M6GBdIGBlnQb85IQTl91/1DZ/yo/ea+kDc/ZZ23o9AVcmdLAuELCy\njj+FXnLoZlsc+d95/sxJkybOfazv1QRcmdDBukDAyvhWSreEm7FBwMoI2C3hZmwQsDICdku4\nGRsErIyA3RJuxgYBKyNgt4SbsUHAygjYLeFmbBCwMgJ2S7gZGwSsjIDdEm7GBgErI2C3hJux\nQcDKCNgt4WZsELAyAnZLuBkbBKyMgN0SbsYGASsjYLeEm7FBwMoI2C3hZmwQsDICdku4GRsE\nrIyA3RJuxgYBKyNgt4SbsUHAygjYLeFmbBCwMgJ2S7gZGwSsjIDdEm7GBgErI2C3hJuxQcDK\nCNgt4WZsELAyAnZLuBkbBKyMgN0SbsYGASsjYLeEm7FBwMoI2C3hZmwQsDICdku4GRsErIyA\n3RJuxgYBKyNgt4SbsUHAygjYLeFmbBCwMgJ2S7gZGwSsjIDdEm7GBgErI2C3hJuxQcDKCNgt\n4WZsELAyAnZLuBkbBKyMgN0SbsYGASsjYLeEm7FBwMoI2C3hZmwQsDICdku4GRsErIyA3RJu\nxgYBKyNgt4SbsUHAygjYLeFmbBCwMgJ2S7gZGwSsjIDdEm7GBgErI2C3hJuxQcDKCNgt4WZs\nELAyAnZLuBkbBKyMgN0SbsYGASsjYLeEm7FBwMoI2C3hZmwQsDICdku4GRsErIyA3RJuxgYB\nKyNgt4SbsUHAygjYLeFmbBCwMgJ2S7gZGwSsjIDdEm7GBgErI2C3hJuxQcDKCNgt4WZsELAy\nAnZLuBkbBKyMgN0SbsYGASsjYLeEm7FBwMoI2C3hZmwQsDICdku4GRsErIyA3RJuxgYBKyNg\nt4SbsUHAygjYLeFmbBCwMgJ2S7gZGwSsjIDdEm7GBgErkwY8a/HqQNZdJCN0sC7can0BUcwc\nOgHzCFwZ4WZs8AisjIDdEm7GBgErI2C3hJuxQcDKCNgt4WZsELAyAnZLuBkbBKyMgN0SbsYG\nASsjYLeEm7FBwMoI2C3hZmwQsDICdku4GRsErIyA3RJuxgYBKyNgt4SbsUHAygjYLeFmbBCw\nMgJ2S7gZGwSsjIDdEm7GBgErI2C3hJuxQcDKCNgt4WZsELAyAnZLuBkbBKyMgN0SbsYGASsj\nYLeEm7FBwMoI2C3hZmwQsDICdku4GRsErIyA3RJuxgYBKyNgt4SbsUHAygjYLeFmbBCwMgJ2\nS7gZGwSsjIDdEm7GBgErI2C3hJuxQcDKCNgt4WZsELCy9oD/9Fie/+8F31zW+f0JuDKhg3WB\ngJW1BXzflmfnL++dZZv9suP7E3BlQgfrAgErawv4qGm/zS/O/vm3bzim4/sTcGVCB+sCAStr\nC3jLH+T53Kl5/oNtOr4/AVcmdLAuELCytoBH3pSvGf/pPL9uZMf3J+DKhA7WBQJW1hbwNufn\n12U35fn3J3V8fwKuTOhgXSBgZW0Bv/e1n91uxzX549P5GtiB0MG6QMDK2gJ+bGY28Y48f8dm\nna+BgCsTOlgXCFjZOt/I8cxLjSeLft/5/Qm4MqGDdYGAlbUFPOM3vbf/sWvH9yfgyoQO1gUC\nVtYWcLao5+blL/G70A6EDtYFAlY2IOAB17hXx/cn4MqEDtYFAlY2IOCl/5i99b2F9/398o7v\nT8CVCR2sCwSsrO1T6L98MPj+BFyZ4F14QMDK+OuEbgk3Y4OAlbUF/PhJk4f1XmTH9yfgyoQO\n1oVKA150xt984j9ervIjNKUa8LHDDz2p56vg93Z8fwKuTOhgXagy4E/19DPzyQo/RFOqAU9Y\nEHx/Aq5M8C48qDDg/9sM6IjqPkRLqgGP/UPw/Qm4MsG78KDCgHdrFXR/dR+jKdWAZ90cfH8C\nrgyTbPOLrlZBXw++b+gZTTXgxfveHnp/Aq4Mk2yzuK+grwXfN/SMphrwAdtkY7fr0fH9Cbgy\naUzyzh/835t0PtJ2rYLmB9819IymGvCsQ1s6vj8BVyaFSf7ihJGNI73/tRof67PNgPYJv2vo\nGU014EEg4MqkMMm/6j3Tk3+u8LF+NbfnY+14Y/hdQ88oAZeIv9maSmCSP2gd6g+qfLh5x73h\nyDMWD+KOoWc01YAntGza8f0JuDIJTPKjrUM9zfpKyoWe0VQDfmuPfcdMPaXj+xNwZRKY5Amt\nQ72N9ZWUCz2jqQbctPKghR3fn4Ark8AkP9E61HtbX0m50DOaeMD5ohkd35+AK5PAJBeOaB7q\nv7O+knKhZzT1gFeO6fj+BFyZFCZ5Su+Z3uuX1hdSLvSMJh7w2q9O6fj+BFyZJCb51R2HZRPf\ns8j6MjYi9IymGvDuPaZOzD7Z8f0JuDKJTHLRz6yvYONCz2jaAe85+x9f7Pj+BFwZJhlN6BlN\nNeBBIODKMMloQs9ougE/ufB751/zbMD9CbgyTDKa0DOaasCvfKLnjwXGndP5/Qm4MkwymtAz\nmmrA52Rzv3/1wnl/mV3Y8f0JuDJMMprQM5pqwLue3nv7AX4ygwNMMprQM5pqwKNu7L29im/k\ncIBJRhN6RlMNeNyVvbcLNun4/gRcGSYZTegZHXTAr5wxZdRe1310RJ4fMOvKKfvn+dWzNhm9\n27lriz+iLd7+1gl5vtfMG/cZM/7dT4de1Pq1BXzgG3v+APj5ww7p+P4EXBkmGU3oGR10wF/N\n3n7t+ZP2HZfns6fv8p2F+fyuwxfccHr2qYEB77/F3j9/4uIRc0Mvav3aAr6qa9sPfeXL7588\n7PoN32H18ZM3P+LhvhcJuDJMMprQMzrYgNduNbXxYHtn1gj40Oyyxit22bZ4SHzbiCcHBHxA\n9rPGc+/NHg29qvVq/3Pg+bsUVzjtqpI7HHng0geP221N60UCrgyTjCb0jA424Meyjxc3U4uA\nR76U5yuyDxUvfz9bODDgcY3I8wuzq0Ovar3W/U6sFXcv+n3Z+z/ataTxKDy87yGagCvDJKMJ\nPaODDXhJ1vMtFEcXAU9uPHN39pXi5auz7w0MeIfiuYXZBaFXtV7tAa/8duPJH770+Ibf/7Ix\nxf8+pp3ZepmAK8Mkowk9o4MN+M7sm8XNMUXAxT/NvCj7UvHyVdn5rwp4QcA3W5RpC/j+145s\nPP1d9tplG3z/eVsXT2ef2niycHzD1JuWBbJeZjKYZDShZ3Tv7sEF/GDx21WNx7dWwCuzDxQv\nfy+7Jt9zavHcfkXAY4ovQOc1XhdDW8Bv2+nu4uY3Ox21wfef1/NXhWef1no5/BHYo3r8VFsP\n6jHJwT4Cv7xZkendWSvgfOrk5xtPDx/7TD57YuNT18fH9PwmVvHV79tGrY5yqW0Bb/Gvvbfz\nNvyvUi4YXXwKPfWs1ssEjIHqMclB/zHS6dnJ187b/oC+gK8adtjlP/lwdnae/0N29u9/+cbd\nioC32flfrv9kdmKcS20LeMy/997+YOwG339F16I8f6L7ltbLBIyB6jHJQQf8wt9OHDfrrr/Z\npBVwft2B40btWTwuvnj61qN2v/KUxkPjAbssPmjs+Pc/F+dS2wJ+w1/2/PHQs/scsOE7HL3X\n0gfm7LO29SIBY6B6TFL4rZSHTip76wGvj3ad+ToBX9O1wylf/Pt3bzGs5OvrZ06aNHHuY30v\nEjAGqsckBx3wt456Oc+f2mxO2S9eYcD5dTOKK5xe9o0c6yBgDFSPSQ464Iuyt1z+o/27bij7\nxasMOM+fvOfXzX+Q4186+l0yAsZA9Zjk4D+FvmjPcWPfUP4AWG3AA97S0R9/EzAGqsckU/3r\nhO1vIWAEq8ckCTgx9Th2HtRjkgScmHocOw/qMUkCTkw9jp0H9ZgkASemHsfOg3pMkoATU49j\n50E9JknAianHsfOgHpMk4MTU49h5UI9JDj7gDf+d5IoulYAL9Th2HtRjkrUI+Ccd/X0nAsZA\n9ZhkkgGPG2Bkx/cnYAxUj0kmGfA7Gl4/Yv+j37ZH14yPdnx/AsZA9ZhkkgE3XDq15y/63r/L\nFR3fn4AxUD0mmWrAU3/ce/svu3d8fwLGQPWYZKoBj2z+ReRLR3V8fwLGQPWYZKoBTz6+52bt\nO0r/TZ82BIyB6jHJGAH/6sK//9rl6w94RrakuFmzVfZyPqP3F90xbz13QdiltgV8Rjbt1K98\n5ZRds892fP9D5hxbA4dZX0Bt1GOSE+UB/3i74n0Pvm29AW/Z8++qL5xQBPyuhwqP5K3nng3J\nd52A1359UvFRJ35hzYbe/VUeWFwHP7e+gNqoySR/UfK2l8t6aOV642a9te99z/oCPnFi8UML\njz22CPiUvtee8upfbuPW+UaOtY/cdeeyVwbzCwHoC/ik1uP1d9cX8Hk7/meerx57afyAn7/7\nsify0v/FANigVq5TWwG/b70Bf/Utef7Ph99RBDyi51unvhMn4G9ummV35J87mYSBwWjlukMr\n4OPWG/CKUSvz/X58R//XwE83XtvVXVgc9gHbAv5eduR3GwFfOPycCP8pwNDTyvWgVsCfXm/A\n+VvOuW/Ci3e0fwrd2/rzYR+wLeDpH8qfbwSc/93Oov8IYKhq5fr1Zr+jblh/wPP3/Pyp+R2x\nvwYefX1vwNeOGOTlA7gGcWIAACAASURBVENbK9d75vQG/MX1/jHSefnLr91uaT7gU+iHXooR\n8JZX9gb849eI/iOAoaqv13vO2GOTLQ6+YP3fyHFenn9qr7w34OZD9X0xAn7Twf9bBLxq6mGi\n/whgqLp3gyr6gG0B39y902nZe056zYjbOr7/GdbfNBPFEdYXUBs1meQuUzas9B+6sA04v2HP\n4tF83592fv9D5qt9d0yFLrG+gNqoxyT3SPQvMzQ8vmRJRz+VsIW/zICB6jHJVP820r1P994u\nO6Pj+xMwBqrHJFMNOJtyV8/t9Rv+p+7WRcAYqB6TTDbgiaPOL24JGINUj0kmG/CV784+8CIB\nY9DqMclkA74+/+eRM/+HgDFY9Zhkqj9apRFwfvvkrW4hYAxSPSaZcsD5ygOHv52AMTj1mGTS\nAecvfTQjYAxOPSaZ6tfAi5r/oNYPT+v4/gSMgeoxyVQDHgQCxkD1mGSSAb/+rPz1fTq+PwFj\noHpMMsmA9/tWvl+fju9PwBioHpNMMuDBIWAMVI9JEnBi6nHsPKjHJJMM+PUDdXx/AsZA9Zhk\nkgEfMFDH9ydgDFSPScYI+MbPHnPi2YvXH/AjH95+5Pg3X5k3/ynKByednudrzp6+ycidzxrw\nY1FmZFnX5odetDbPb+79yOet5wOu91Po5x7s+D911uLVNXCr9QXURj0mOVMe8DdHF++7zeXr\nC/ie8dMv+/VtH+z6cm/A/z3lY41XfnzywpWP/eA1XxgQ8MnLH77lzNcc80r+wvKGWzf5dacB\n37B5xwHzCIyB6jFJ+SPwf47sfedtF68n4H1nFD/bLD+3654i4Ee3/9vipak9PxL02p/0v1vv\nP1O5uPvi3hffdMb6PmB7wAuPn9X4/HnmphM7/k+tYcBP/MN7TvjGCqtLSRwB95rbeu9zXh3w\nsmx+z+3LEz/XiPSx132k56V37bTuD1Vp/juzR87publk+xfX9wHbAr4kGz4lmzw6e+NVHf+n\n1i/gWyYUc99kgdnFJI2Ae+3Seu+TXh3wNVkzmoOOzWe8fdfxvf+Q1ap3dv/5u+Y9PuDXagb8\n+dcVT9fs/P31fsC2gGcc/mzefe/L3z6k8x8yXLuAnxjfO/jRy+wuJ2EE3Gun1nuf8OqAr83u\n633mgHfkM7rPnHFY82cJrrrsk7uOuqj//ZoBf2aX4uklW6//Jw62BbzpwjzvvifPP9b5vxFf\nu4DPbU3+k3aXkzAC7jWn9d5ffnXAj2aX9Ny+NOGL+YwP5ismv6//bR8b199pM+CD5hZP/+rU\n9X/A9p+N1PgK+jU/y/NbJ3f8n1q7gE9oTf5Qu8tJGAH3uqD5zpvfsZ7fxJo17YXi5rzuB3si\nXTTm7Dx/5NhHitddOuyPfe/WG/D8nq+Ynxpx/fo/YFvAex7zYr7b5/P88nEd/6fWLuB3tfY0\n2+5yEkbATaf3vO+m/7a+P0a6b8L0Kx5YdPqwc5uR/mjYj/I106df8fAjl+94eP+7zTh5+fLb\nPzfiPcXzN2S/W/8HbAv44sbDzhe63/+lrd/Q8X9qeMD3+nb6q754sRI6WBcIuOWHx+93yEdu\nXP86H3n/tiPGH35d3nqU/eLo2/NVp79u7MidPzPgt5+Kn3s2duYFPc9f3LXe34Ne94+RLjk7\n/9Obs2ybRR3/p9Yu4J9u0rumkVdaX0noYF0gYOV1rucbOR76zUud3792Ad87r6fgUeds/D0r\nFjpYFwhYeZ3rBvzsUz06vn/9Ar73xtMOe9OHf2J9FQRsyDLgRZs1Xd7Ru7cFvOyIcc2L7PSj\n1TFgL0IH6wIBK6+zLdVDNjv+k5/p0fH9CbgyoYN1gYCV19kW8LifB9+fgCsTvAsPCFh5nW0B\nbxn+LfwEXJngXXhAwMrrbAv4E18Jvj8BVyZ4Fx4QsPI62wJ+8U0HfPLsHh3fn4ArEzpYF4Z6\nwOraAj677yI7vj8BVyZ0sC4QsLK2VCcdfdtvH+7R8f0JuDKhg3WBgJW1BTyK38RyJHgXHhCw\nsva/jbQ0+P4EXJngXXhAwMraAr5ldvD8CbgyoYN1gYCVtQV8wJRsk+16dHx/Aq5M6GBdIGBl\nbQHPOrSl4/sTcGVCB+sCASvT/9lI1l0kQ7gZGwSsrC3g/Tv/52RbCLgywbvwgICVtQU85dzg\n+xNwZYJ34QEBK2sL+PJd5wf8Yxw9CLgyoYN1gYCVtf8m1rRs5GR+F9qJ0MG6QMDK2v8YaTa/\nC+1H6GBdIGBl/C60W8LN2CBgZesE/OTC751/Tec/GYmAKxQ6WBcIWFlbwK98YkRxhePO6fz+\nBFyZ0MG6QMDK2gI+J5v7/asXzvvL7MKO70/AlQkdrAsErKwt4F1P7739wF4d35+AKxM6WBcI\nWFn73we+sff2qjEd35+AKxM6WBcIWFn7Pyt7Ze/tgk06vj8BVyZ0sC4QsLK2gA98Y8+PQHv+\nsEM6vj8BVyZ0sC4QsLK2gK/q2vZDX/ny+ycP28APE14PAq5M6GBdIGBl7X8OPH+X4gqnBfyl\nJAKuTOhgXSBgZet+J9aKuxf9PuT+BFyZ0MG6QMDK+FZKt4SbsUHAytoCfvykycN6L7Lj+xNw\nZUIH6wIBK2tL9djhh5703h4d35+AKxM6WBcIWFlbwBMWBN+fgCsTvAsPCFhZW8Bj/xB8fwKu\nTPAuPCBgZe3/IsfNwfcn4MoE78IDAlbWFvDifW8PvT8BVyZ0sC4QsLL2f1Jnm2wsP5nBi9DB\nukDAyvjJDG6FDtYFAlbGN3K4JdyMDQJWRsBuCTdjg4CVdRLwBdn8PF99/OTNj3i4/7aJgCsT\nOlgXCFhZBwH/fqsxjYCPPHDpg8fttqbvtomAKxM6WBcIWFkHAR99+lbz80e7ljQefYdf37pt\nvZGAKxM6WBcIWNnGA/7PP/9jI+DLxqxtPD/tzNZt660EXJnQwbpAwMo2GvDqSdfljYDnbV28\nMPvU1m3jycPzGvZbvDqQdRfJCB2sC7daX0AUM2sU8Mkn5z0BTylemH1a67bx5L8+07DnTcsC\nWXeRjNDBunCV9QVEsXd3bQK+7rWregJeMLr41HnqWa3b1tv5FLoyoYN1gU+hlW0s4HeOnjBh\nQtemR63oWpTnT3Tf0rptvZ2AKxO8Sw8IWNnGAl61vGGL85/Ij95r6QNz9lnbd9tEwJUJ3qUH\nBKyso+/EanwKnT9z0qSJcx/rv20i4MqEDtYFAlbGt1K6JdyMDQJWRsBuCTdjg4CVEbBbws3Y\nIGBlBOyWcDM2CFgZAbsl3IwNAlZGwG4JN2ODgJURsFvCzdggYGUE7JZwMzYIWBkBuyXcjA0C\nVkbAbgk3Y4OAlRGwW8LN2CBgZQTslnAzNghYGQG7JdyMDQJWRsBuCTdjg4CVEbBbws3YIGBl\nBOyWcDM2CFgZAbsl3IwNAlZGwG4JN2ODgJURsFvCzdggYGUE7JZwMzYIWBkBuyXcjA0CVkbA\nbgk3Y4OAlRGwW8LN2CBgZQTslnAzNghYGQG7JdyMDQJWRsBuCTdjg4CVEbBbws3YIGBlBOyW\ncDM2CFgZAbsl3IwNAlZGwG4JN2ODgJURsFvCzdggYGUE7JZwMzYIWBkBuyXcjA0CVkbAbgk3\nY4OAlRGwW8LN2CBgZQTslnAzNghYGQG7JdyMDQJWRsBuCTdjg4CVEbBbws3YIGBlBOyWcDM2\nCFgZAbsl3IwNAlZGwG4JN2ODgJURsFvCzdggYGUE7JZwMzYIWBkBuyXcjA0CVkbAbgk3Y4OA\nlRGwW8LN2CBgZQTslnAzNghYGQG7JdyMDQJWRsBuCTdjg4CVEbBbws3YIGBlBOyWcDM2CFiZ\nNOBZi1cHsu4iGaGDdeFW6wuIYubQCZhH4MoIN2ODR2BlBOyWcDM2CFgZAbsl3IwNAlZGwG4J\nN2ODgJURsFvCzdggYGUE7JZwMzYIWJnvgG/6+ofPuLyqQLwTbsYGAStzHfDnRhYDm7ukskZc\nE27GBgEr8xzwuc2JHV9dJJ4JN2ODgJV5Dnjn5sS6f1ZdJY4JN2ODgJU5DvjuvpF9p8JM/BJu\nxgYBK3Mc8M/6RnZuhZn4JdyMDQJW5jjge/6sNbLLKszEL+FmbBCwMscB33tSc2JTq4vEM+Fm\nbBCwMs8B37VHz8AmXFFdJJ4JN2ODgJV5DvjeX35q5mt3f/ctlSXim3AzNghYmeuAhzbhZmwQ\nsDICdku4GRsErIyA3RJuxgYBKyNgt4SbsUHAygjYLeFmbBCwMgJ2S7gZGwSsjIDdEm7GBgEr\nI2C3hJuxQcDKCNgt4WZsELAyAnZLuBkbBKyMgN0SbsYGASsjYLeEm7FBwMoI2C3hZmwQsDIC\ndku4GRsErIyA3RJuxgYBKyNgt4SbsUHAygjYLeFmbBCwMgJ2S7gZGwSsjIDdEm7GBgErI2C3\nhJuxQcDKCNgt4WZsELAyAnZLuBkbBKyMgN0SbsYGASsjYLeEm7FBwMoI2C3hZmwQsDICdku4\nGRsErIyA3RJuxgYBKyNgt4SbsUHAygjYLeFmbBCwMgJ2S7gZGwSsjIDdEm7GBgErI2C3hJux\nQcDKCNgt4WZsELAyAnZLuBkbBKyMgN0SbsYGASsjYLeEm7FBwMoI2C3hZmwQsLKNBvybv958\n/Btvz/PVx0/e/IiH+2+bCLgyoYN1gYCVbSzgFyefcP9DJ/7Zs/mRBy598Ljd1vTdNhFwZYJ3\n6QEBK9tYwH/4xrN5fn+25NGuJY1H3+HXt25bbyfgygTv0gMCVtbJ18CrPrzLC5eNWdt4btqZ\nrdvW2wi4MqGDdYGAlW084DWjsoP/J5+3dfH87FNbt40nt72pYer8xYGsu0hG6GBduMT6AqLY\no7tGAef33Xz0TqvnTSmenX1a67bx5KniP3UWj8BVCR2sCzwCK+voj5HWbHbegtHFp85Tz2rd\ntt7Ep9CVCR2sCwSsbGMBX7vjn/J87cR/WtG1KM+f6L6lddt6OwFXJniXHhCwso0FvHqrY3+9\n7LQxv82P3mvpA3P2Wdt320TAlQnepQcErGyjn0Lfe/i4TWfekOfPnDRp4tzH+m+bCLgyoYN1\ngYCV8a2Ubgk3Y4OAlRGwW8LN2CBgZQTslnAzNghYGQG7JdyMDQJWRsBuCTdjg4CVEbBbws3Y\nIGBlBOyWcDM2CFgZAbsl3IwNAlZGwG4JN2ODgJURsFvCzdggYGUE7JZwMzYIWBkBuyXcjA0C\nVkbAbgk3Y4OAlRGwW8LN2CBgZQTslnAzNghYGQG7JdyMDQJWRsBuCTdjg4CVEbBbws3YIGBl\nBOyWcDM2CFgZAbsl3IwNAlZGwG4JN2ODgJURsFvCzdggYGUE7JZwMzYIWBkBuyXcjA0CVkbA\nbgk3Y4OAlRGwW8LN2CBgZQTslnAzNghYGQG7JdyMDQJWRsBuCTdjg4CVEbBbws3YIGBlBOyW\ncDM2CFgZAbsl3IwNAlZGwG4JN2ODgJURsFvCzdggYGUE7JZwMzYIWBkBuyXcjA0CVkbAbgk3\nY4OAlRGwW8LN2CBgZQTslnAzNghYGQG7JdyMDQJWRsBuCTdjg4CVEbBbws3YIGBlBOyWcDM2\nCFgZAbsl3IwNAlZGwG4JN2ODgJURsFvCzdggYGUE7JZwMzYIWBkBuyXcjA0CVkbAbgk3Y4OA\nlRGwW8LN2CBgZQTslnAzNghYGQG7JdyMDQJWJg141uLVgay7SEboYF241foCopg5dALmEbgy\nws3Y4BFYGQG7JdyMDQJWRsBuCTdjg4CVEbBbws3YIGBlBOyWcDM2CFgZAbsl3IwNAlZGwG4J\nN2ODgJURsFvCzdggYGUE7JZwMzYIWBkBuyXcjI3aBdy9LQG3s+4iGcLN2KhbwH+25MdZtvnF\n935jGAE3WXeRDOFmbNQt4G/c03gE/n/LvvbYZwm4ybqLZAg3Y6NuAT94ZOMBeM3e2Vt+RcBN\n1l0kQ7gZG3UL+E/bZ9nch7Pstc8RcJN1F8kQbsZG3QJ+sfEZ9DcuzrJJLxJwk3UXyRBuxkbd\nAn7gr7Ps1x/Isnc+QMBN1l0kQ7gZG3UL+OsPHfWNF7bMDnv67wi4ybqLZAg3Y6NuAW963do/\nnpBlc77MHyO1WHeRDOFmbNQt4CzbZN3vyiJgdES4GRv1C/hVCBgdEW7GRt0CXtPrFQLuY91F\nMoSbsVG3gN/S8Ncfuu3dBNzHuotkCDdjwyjgO+dut9kb/mlNrF9u3U+hN72JgPtYd5EM4WZs\n2AT8w1E9aR0Zq+B1Ax72OwLuY91FMoSbsWES8JOvabb13Ui/YH/AXyycdfNvCLiPdRfJEG7G\nhknAF7TaOjDSL9gf8J0Ndy1/ak8C7mPdRTKYZKc+2Gpriw28Q+gZXfdT6KO+QcB9VFebMibZ\nqY+32tpuA+8QekZf9efASwi4j+pqU8YkO/WDVltHb+AdQs9of8DDC2MPW0nAfVRXmzIm2bGD\netMad/UG3h56RvsD7n15zWcJuI/qZlPGJDt2x5uLsiZftKG3h57R/oAPKczcKiPgPpqLTRqT\nDHDF1z73r0s2+NbQMzrga+B937njOl8PEzA6wiSjCT2j/QGfna95/o3j7/8LAu5jvcxkMMlo\nQs9of8CrDx7x+RvH/fs1BNzHepnJYJLRhJ7R/oBXjcrGrMzGrSLgPtbLTAaTjCb0jPYH/P13\nZNnvs02fJuA+1stMBpOMJvSM9ge82c1ffutTR116LQH3sV5mMphkNKFntD/gu4oXX7p9ZwLu\nY73MZDDJaELPaH/AT28/fPjwLv4YaQDrZSaDSUYTekb7A95+1CEnvH3mMALuZ73MZDDJaELP\naH/A+/whX7s2f3h3Au5jvcxkMMloQs9of8C/uHCn4d07XngXAfexXmYymGQ0oWe0P+DnxxdP\nN/kTAfexXmYyfE1y4Xn/drvqB4wp9Iz2B3zP64qn295DwH2sl5kMT5OcX3wz8IgTf6n4IWMK\nPaP9AR/202Om7frXN71z++23J+Be1stMhqNJ3jC+9+wepfchowo9o+v+feAeBNzLepnJcDTJ\nd7YO73y9jxlT6BntD3jPqX0IuJf1MpPhaJLbtQ7vp/U+ZkyhZ3TA3wcef1jxdOJIvgZusV5m\nMhxNcnzr8H5Q72PGFHpG+wOe+vvfFjefXLELATdZLzMZjia5a+vwflHvY8YUekb7A/7JpZtk\nWXe26cWXE3CT9TKT4WiSH2ue3dE3633MmELPaH/Af/yLLNvlhSx7/SoCbrJeZjIcTXLxtN6z\n+wW9DxlV6BntD/jZHbLs4HzzbKenUg14xXFbbnrQXXm++vjJmx/xcP9tEwFXxtMk7373VtmI\n6f+k+BGjCj2jA360yv/pGnbpT/9p53+7KdWA9z7wlw/9zcQ/5kceuPTB43Zb03fbRMCVcTbJ\nO5bqfryYQs9of8BvfuEPTy3b+oH8mTckGvCqo36T549kdz/ataTx6Dv8+tZt6+0EXBkmGU3o\nGR3wx0g7n3rqhGzUHpsl/cdIt3evvGzM2sYz085s3bbeRMCVYZLRhJ7RV/1spIHSC3jVrp/O\n521dPDf71NZt48lLqxsOJuCqMMloQs9ovQK+b6cPr83nTSmenX1a67bx5NodGqbdtCyQ9TKT\nwSSjCT2je3fXKOAbJny78XTB6OJT56lntW5bb+VT6MowyWhCz2idHoFvHX91cbOia1GeP9F9\nS+u29WYCrgyTjCb0jNYo4P/d4UvLG/6YH73X0gfm7LO277aJgCvDJKMJPaM1CviG3os+L3/m\npEkT5z6W9902EXBlmGQ0oWe0RgFvDAFXhklGE3pGCbiE9TKTwSSjCT2jBFzCepnJYJLRhJ5R\nAi5hvcxkMMloQs8oAZewXmYymGQ0oWeUgEtYLzMZTDKa0DNKwCWsl5kMJhlN6Bkl4BLWy0wG\nk4wm9IwScAnrZSaDSUYTekYJuIT1MpPBJKMJPaMEXMJ6mclgktGEnlECLmG9zGQwyWhCzygB\nl7BeZjKYZDShZ5SAS1gvMxlMMprQM0rAJayXmQwmGU3oGSXgEtbLTAaTjCb0jBJwCetlJoNJ\nRhN6Rgm4hPUyk8Ekowk9owRcwnqZyWCS0YSeUQIuYb3MZDDJaELPKAGXsF5mMphkNKFnlIBL\nWC8zGUwymtAzSsAlrJeZDCYZTegZJeAS1stMBpOMJvSMEnAJ62Umg0lGE3pGCbiE9TKTwSSj\nCT2jBFzCepnJYJLRhJ5RAi5hvcxkMMloQs8oAZewXmYymGQ0oWeUgEtYLzMZTDKa0DNKwCWs\nl5kMJhlN6Bkl4BLWy0wGk4wm9IwScAnrZSaDSUYTekYJuIT1MpPBJKMJPaMEXMJ6mclgktGE\nnlECLmG9zGQwyWhCzygBl7BeZjKYZDShZ5SAS1gvMxlMMprQM0rAJayXmQwmGU3oGSXgEtbL\nTAaTjCb0jBJwCetlJoNJRhN6Rgm4hPUyk8Ekowk9owRcwnqZyWCS0YSeUQIuYb3MZDDJaELP\nKAGXsF5mMphkNKFnlIBLWC8zGUwymtAzSsAlrJeZDCYZTegZJeAS1stMBpOMJvSMEnAJ62Um\ng0lGE3pGCbiE9TKTwSSjCT2jBFzCepnJYJLRhJ5RAi5hvcxkMMloQs8oAZewXmYymGQ0oWd0\nCAU8a/HqQNbLTAaTjCb0jM4cOgHzCFwZJhlN6BkdQo/ABFwZJhlN6Bkl4BLWy0wGk4wm9IwS\ncAnrZSaDSUYTekYJuIT1MpPBJKMJPaMEXMJ6mclgktGEnlECLmG9zGQwyWhCzygBl7BeZjKY\nZDShZ5SAS1gvMxlMMprQM0rAJayXmQwmGU3oGSXgEtbLTAaTjCb0jBJwCetlJoNJRhN6Rgm4\nhPUyk8Ekowk9owRcwnqZyWCS0YSeUQIuYb3MZDDJaELPKAGXsF5mMphkNKFnlIBLWC8zGUwy\nmtAzSsAlrJeZDCYZTegZJeAS1stMBpOMJvSMEnAJ62Umg0lGE3pGCbiE9TKTwSSjCT2jBFzC\nepnJYJLRhJ5RAi5hvcxkMMloQs8oAZewXmYymGQ0oWeUgEtYLzMZTDKa0DNKwCWsl5kMJhlN\n6Bkl4BLWy0wGk4wm9IwScAnrZSaDSUYTekYJuIT1MpPBJKMJPaMEXMJ6mclgktGEnlECLmG9\nzGQwyWhCzygBl7BeZjKYZDShZ5SAS1gvMxlMMprQM0rAJayXmQwmGU3oGSXgEtbLTAaTjCb0\njBJwCetlJoNJRhN6Rgm4hPUyk8Ekowk9owRcwnqZyWCS0YSeUQIuYb3MZDDJaELPKAGXsF5m\nMphkNKFnlIBLWC8zGUwymtAzSsAlrJeZDCYZTegZJeAS1stMBpOMJvSMEnAJ62Umg0lGE3pG\nCbiE9TKTwSSjCT2jBFzCepnJYJLRhJ5RAi5hvcxkMMloQs8oAZewXmYymGQ0oWeUgEtYLzMZ\nTDKa0DNKwCWsl5kMJhlN6Bkl4BLWy0wGk4wm9IwScAnrZSaDSUYTekZrFfD9+3UXN6uPn7z5\nEQ/33zYRcGWYZDShZ7ROAf9w0ok9AR954NIHj9ttTd9tEwFXhklGE3pG6xTwhY/MLwJ+tGtJ\n49F3+PWt29abCbgyTDKa0DNap4DzvCfgy8asbTyddmbrtvVGAq4Mk4wm9IzWMOB5WxfPzj61\nddt48vC8hv0JuCpMMprQM1rHgKcUz84+rXXbePJfn2nY86ZlgayXmQwmGU3oGd27u3YBLxhd\nfOo89azWbeuNfApdGSYZTegZreEj8IquRXn+RPctrdvWGwm4MkwymtAzWqeAVy4/v3v58ufy\no/da+sCcfdb23TYRcGWYZDShZ7ROAW/Xc9Hfyp85adLEuY/lfbdNBFwZJhlN6BmtU8AbQcCV\nYZLRhJ5RAi5hvcxkMMloQs8oAZewXmYymGQ0oWeUgEtYLzMZTDKa0DNKwCWsl5kMJhlN6Bkl\n4BLWy0wGk4wm9IwScAnrZSaDSUYTekYJuIT1MpPBJKMJPaMEXMJ6mclgktGEnlECLmG9zGQw\nyWhCzygBl7BeZjKYZDShZ5SAS1gvMxlMMprQM0rAJayXmQwmGU3oGSXgEtbLTAaTjCb0jBJw\nCetlJoNJRhN6Rgm4hPUyk8Ekowk9owRcwnqZyWCS0YSeUQIuYb3MZDDJaELPKAGXsF5mMphk\nNKFnlIBLWC8zGUwymtAzSsAlrJeZDCYZTegZJeAS1stMBpOMJvSMEnAJ62Umg0lGE3pGCbiE\n9TKTwSSjCT2jBFzCepnJYJLRhJ5RAi5hvcxkMMloQs8oAZewXmYymGQ0oWeUgEtYLzMZTDKa\n0DNKwCWsl5kMJhlN6Bkl4BLWy0wGk4wm9IwScAnrZSaDSUYTekYJuIT1MpPBJKMJPaMEXMJ6\nmclgktGEnlECLmG9zGQwyWhCzygBl7BeZjKYZDShZ5SAS1gvMxlMMprQM0rAJayXmQwmGU3o\nGSXgEtbLTAaTjCb0jBJwCetlJoNJRhN6Rgm4hPUyk8Ekowk9owRcwnqZyWCS0YSeUQIuYb3M\nZDDJaELPKAGXsF5mMphkNKFnlIBLWC8zGUwymtAzSsAlrJeZDCYZTegZJeAS1stMBpOMJvSM\nEnAJ62Umg0lGE3pGCbiE9TKTwSSjCT2jBFzCepnJYJLRhJ5RAi5hvcxkMMloQs8oAZewXmYy\nmGQ0oWeUgEtYLzMZTDKa0DM6hAKetXh1IOtlJoNJRhN6RmcOnYB5BK4Mk4wm9IwOoUdgAq4M\nk4wm9IwScAnrZSaDSUYTekYJuIT1MpPBJKMJPaMEXMJ6mclgktGEnlECLmG9zGQwyWhCzygB\nl7BeZjKYZDShZ5SAS1gvMxlMMprQM0rAJayXmQwmGU3oGSXgEtbLTAaTjCb0jBJwCetlJoNJ\nRhN6Rgm4hPUyk8Ekowk9owRcwnqZyWCS0YSeUQIuYb3MZDDJaELPKAGXsF5mMphkNKFnlIBL\nWC8zGUwymtAzSsAlrJeZDCYZTegZJeAS1stMBpOMJvSMEnAJ62Umg0lGE3pGCbiE9TKTwSSj\nCT2jBFzCepnJF8BtUAAACEVJREFUYJLRhJ5RAi5hvcxkMMloQs8oAZewXmYymGQ0oWeUgEtY\nLzMZTDKa0DNKwCWsl5kMJhlN6Bkl4BLWy0wGk4wm9IwScAnrZSaDSUYTekYJuIT1MpPBJKMJ\nPaMEXMJ6mclgktGEnlECLmG9zGQwyWhCzygBl7BeZjKYZDShZ5SAS1gvMxlMMprQM0rAJayX\nmQwmGU3oGSXgEtbLTAaTjCb0jBJwCetlJoNJRhN6Rgm4hPUyk8Ekowk9owRcwnqZyWCS0YSe\nUQIuYb3MZDDJaELPKAGXsF5mMphkNKFnlIBLWC8zGUwymtAzSsAlrJeZDCYZTegZJeAS1stM\nBpOMJvSMEnAJ62Umg0lGE3pGCbiE9TKTwSSjCT2jBFzCepnJYJLRhJ5RAi5hvcxkMMloQs8o\nAZewXmYymGQ0oWeUgEtYLzMZTDKa0DNKwCWsl5kMJhlN6Bkl4BLWy0wGk4wm9IwScAnrZSaD\nSUYTekYJuIT1MpPBJKMJPaMEXMJ6mclgktGEntE6B7z6+MmbH/Fw34sEXBkmGU3oGa1zwEce\nuPTB43Zb03qRgCvDJKMJPaM1DvjRriWNR+Hh17deJuDKMMloQs9ojQO+bMzaxtNpZ7ZeJuDK\nMMloQs9ojQOet3XxdPapjSe3valh+k3LAlkvMxlMMprQM7p3d30DnlI8nX1a48lTixtmBT8C\ne/Qr6wuojXpMssaPwAtGF59CTz2r9XL4p9Ae1ePYeVCPSdY44BVdi/L8ie5bWi8TMAaqxyRr\nHHB+9F5LH5izz9rWiwSMgeoxyToH/MxJkybOfazvRQLGQPWYZJ0DXgcBY6B6TJKAE1OPY+dB\nPSZJwImpx7HzoB6TJODE1OPYeVCPSRJwYupx7DyoxyQJODH1OHYe1GOSBJyYehw7D+oxSQJO\nTD2OnQf1mCQBJ6Yex86DekySgBNTj2PnQT0mScCJqcex86AekyTgxNTj2HlQj0kScGLqcew8\nqMckCTgx9Th2HtRjkgScmHocOw/qMUkCTkw9jp0H9ZgkASemHsfOg3pMkoATU49j50E9JknA\nianHsfOgHpMcSgF/6ms1cJr1BdRGPSa5/dAJ+Ip5dTDb+gJq46TPWV9BFP9a8rYXooQXizTg\nephifQG18fEfW1/BEEPABQKOhYCVEXCBgGMhYGUEXCDgWAhYGQEDCSNgIGEEDCSMgIGEDdWA\nZ2RLips1W2UvN57vsWPeeu4C44tz75EPbz9y/JuvbDw345TGkwcnnd4Y5dnTNxm581mv9L9X\nY5pdmx960do8v7l3rudZXW+NDdmAtzytuFk4oQj4XQ8VHslbzz1rfXXO3TN++mW/vu2DXV/u\nDfi/p3ys8cqPT1648rEfvOYL/e824+TlD99y5muOeSV/YXnDrZv82uyK62vIBnzixBcbN8ce\nWwR8St9rTym5C1r2nVHMLj+3655iYo9u/7fFS1M/Wzy99if979Y7zcXdF/e++KYzVC9yiBiy\nAZ+343/m+eqxlxJwsGXZ/J7blyd+rjGxx173kZ6X3rXT4nXerznNI+f03Fyy/YtqVziEDN2A\nv/qWPP/nw+8oAh4xrvAdAu7MNVnz75AedGw+4+27jn+654VV7+z+83fNe3zA+zWn+fnXFU/X\n7Px91YscKoZuwCtGrcz3+/Ed/V8DN87hjK7uwroPJWhzbXZf7zMHvCOf0X3mjMNe7n1x1WWf\n3HXURf3v1wz4M7sUTy/Z+mXVixwqhm7A+VvOuW/Ci3e0fwp93L2F5y2vzL9Hs0t6bl+a8MV8\nxgfzFZPf1/+2j43r77Q514PmFk//6lTFKxxChnDA8/f8/Kn5HXwNHG7WtJ6/E3te94M9E1s0\n5uw8f+TYR4rXXTrsj33v1jvN+T1fMT814nqLK62/IRzwy6/dbmk+4FPoh14i4M7cN2H6FQ8s\nOn3Yuc1IfzTsR/ma6dOvePiRy3c8vP/dZpy8fPntnxvxnuL5G7LfGV1szQ3hgPNP7ZX3Btz8\nt1LuI+AOPfL+bUeMP/y6vPUo+8XRt+erTn/d2JE7f2bAn6EXcx0784Ke5y/u4vegKzFUAwZq\ngYCBhBEw4lq0WdPl1lcyJBAwkDACBhJGwEDCCBhIGAEDCSNgIGEEDCSMgIGEEbBHr5wxZdRe\n1310RJ4fMOvKKfvn+dWzNhm927lr83z33Yu3v3VCnu8188Z9xox/99PW1wpTBOzRV7O3X3v+\npH3H5fns6bt8Z2E+v+vwBTecnn1qYMD7b7H3z5+4eMRc62uFKQJ2aO1WUxsPtndmjYAPzS5r\nvGKXbYu/y/O2EU8OCPiA7GeN596bPWp6qTBGwA49ln28uJlaBDzypTxfkX2oePn72cKBAY9r\nRJ5fmF1teKEwR8AOLcnOKW6OLgKe3Hjm7uwrxctXZ98bGPAOxXML+VfohzYCdujO7JvFzTFF\nwNs1nlmUfal4+ars/FcFvCC70Owy4QABO/Rg8dtVeT6tFfDK7APFy9/Lrsn3nFo8t18R8Jg1\njefmNV6HIYyAHXp5syLTu7NWwPnUycU/lHn42Gfy2RMbX/k+PqbnN7GKr37fNmq16aXCGAF7\ndHp28rXztj+gL+Crhh12+U8+nJ2d5/+Qnf37X75xtyLgbXb+l+s/mZ1ofKmwRcAevfC3E8fN\nuutvNmkFnF934LhRe/5r45kXT9961O5XnrJpI+BdFh80dvz7n7O9UhgjYL8OnVT21gNer3Ud\ncIyAPfrWUS/n+VObzSl7HwJGTsA+XZS95fIf7d91Q9n7EDByAnbqoj3HjX3DVaXvQsDICRhI\nGgEDCSNgIGEEDCSMgIGEETCQMAIGEkbAQML+P25dvXVZjF2fAAAAAElFTkSuQmCC"
          },
          "metadata": {
            "image/png": {
              "width": 480,
              "height": 600
            }
          }
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "res[res$SYMBOL %in% c('Maf','Smad3'),]"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 217
        },
        "id": "OEVgMsWPGyg3",
        "outputId": "a606ed7d-1cdd-4887-b473-1d47b9d3db5f"
      },
      "execution_count": 66,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "log2 fold change (MLE): group OKS_D7 vs MEF \n",
              "Wald test p-value: group OKS_D7 vs MEF \n",
              "DataFrame with 2 rows and 7 columns\n",
              "   baseMean log2FoldChange     lfcSE      stat      pvalue        padj\n",
              "  <numeric>      <numeric> <numeric> <numeric>   <numeric>   <numeric>\n",
              "1   34.2625      -3.986266  0.662245  -6.01932 1.75154e-09 2.27124e-08\n",
              "2  279.9890      -0.850028  0.152551  -5.57210 2.51692e-08 2.75121e-07\n",
              "       SYMBOL\n",
              "  <character>\n",
              "1         Maf\n",
              "2       Smad3"
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "DLYh9wPhxed2",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "outputId": "8ab81aff-4239-4347-ab7b-4fba92329455"
      },
      "source": [
        "sessionInfo()"
      ],
      "execution_count": 64,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "R version 4.1.3 (2022-03-10)\n",
              "Platform: x86_64-pc-linux-gnu (64-bit)\n",
              "Running under: Ubuntu 18.04.5 LTS\n",
              "\n",
              "Matrix products: default\n",
              "BLAS:   /usr/lib/x86_64-linux-gnu/openblas/libblas.so.3\n",
              "LAPACK: /usr/lib/x86_64-linux-gnu/libopenblasp-r0.2.20.so\n",
              "\n",
              "locale:\n",
              " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
              " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
              " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
              " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
              " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
              "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
              "\n",
              "attached base packages:\n",
              "[1] stats4    parallel  stats     graphics  grDevices utils     datasets \n",
              "[8] methods   base     \n",
              "\n",
              "other attached packages:\n",
              " [1] data.table_1.14.2                        \n",
              " [2] BSgenome.Mmusculus.UCSC.mm10_1.4.3       \n",
              " [3] BSgenome_1.62.0                          \n",
              " [4] rtracklayer_1.54.0                       \n",
              " [5] Biostrings_2.62.0                        \n",
              " [6] XVector_0.34.0                           \n",
              " [7] chromVARmotifs_0.2.0                     \n",
              " [8] motifmatchr_1.16.0                       \n",
              " [9] org.Mm.eg.db_3.14.0                      \n",
              "[10] TxDb.Mmusculus.UCSC.mm10.knownGene_3.10.0\n",
              "[11] GenomicFeatures_1.46.5                   \n",
              "[12] AnnotationDbi_1.56.2                     \n",
              "[13] clusterProfiler_4.2.2                    \n",
              "[14] ChIPseeker_1.31.4                        \n",
              "[15] DESeq2_1.34.0                            \n",
              "[16] SummarizedExperiment_1.24.0              \n",
              "[17] Biobase_2.54.0                           \n",
              "[18] MatrixGenerics_1.6.0                     \n",
              "[19] matrixStats_0.61.0                       \n",
              "[20] GenomicRanges_1.46.1                     \n",
              "[21] GenomeInfoDb_1.30.1                      \n",
              "[22] IRanges_2.28.0                           \n",
              "[23] S4Vectors_0.32.4                         \n",
              "[24] BiocGenerics_0.40.0                      \n",
              "[25] ggplot2_3.3.5                            \n",
              "[26] dplyr_1.0.8                              \n",
              "[27] tidyr_1.2.0                              \n",
              "\n",
              "loaded via a namespace (and not attached):\n",
              "  [1] utf8_1.2.2                             \n",
              "  [2] R.utils_2.11.0                         \n",
              "  [3] tidyselect_1.1.2                       \n",
              "  [4] poweRlaw_0.70.6                        \n",
              "  [5] RSQLite_2.2.12                         \n",
              "  [6] grid_4.1.3                             \n",
              "  [7] BiocParallel_1.28.3                    \n",
              "  [8] devtools_2.4.3                         \n",
              "  [9] scatterpie_0.1.7                       \n",
              " [10] munsell_0.5.0                          \n",
              " [11] pbdZMQ_0.3-7                           \n",
              " [12] withr_2.5.0                            \n",
              " [13] colorspace_2.0-3                       \n",
              " [14] GOSemSim_2.20.0                        \n",
              " [15] filelock_1.0.2                         \n",
              " [16] uuid_1.0-4                             \n",
              " [17] DOSE_3.20.1                            \n",
              " [18] labeling_0.4.2                         \n",
              " [19] repr_1.1.4                             \n",
              " [20] GenomeInfoDbData_1.2.7                 \n",
              " [21] polyclip_1.10-0                        \n",
              " [22] bit64_4.0.5                            \n",
              " [23] farver_2.1.0                           \n",
              " [24] rprojroot_2.0.3                        \n",
              " [25] downloader_0.4                         \n",
              " [26] vctrs_0.4.0                            \n",
              " [27] treeio_1.18.1                          \n",
              " [28] generics_0.1.2                         \n",
              " [29] BiocFileCache_2.2.1                    \n",
              " [30] R6_2.5.1                               \n",
              " [31] graphlayouts_0.8.0                     \n",
              " [32] locfit_1.5-9.5                         \n",
              " [33] bitops_1.0-7                           \n",
              " [34] cachem_1.0.6                           \n",
              " [35] fgsea_1.20.0                           \n",
              " [36] gridGraphics_0.5-1                     \n",
              " [37] DelayedArray_0.20.0                    \n",
              " [38] assertthat_0.2.1                       \n",
              " [39] BiocIO_1.4.0                           \n",
              " [40] scales_1.2.0                           \n",
              " [41] ggraph_2.0.5                           \n",
              " [42] enrichplot_1.13.1.992                  \n",
              " [43] gtable_0.3.0                           \n",
              " [44] processx_3.5.3                         \n",
              " [45] tidygraph_1.2.1                        \n",
              " [46] seqLogo_1.60.0                         \n",
              " [47] rlang_1.0.2                            \n",
              " [48] genefilter_1.76.0                      \n",
              " [49] splines_4.1.3                          \n",
              " [50] lazyeval_0.2.2                         \n",
              " [51] BiocManager_1.30.16                    \n",
              " [52] yaml_2.3.5                             \n",
              " [53] reshape2_1.4.4                         \n",
              " [54] qvalue_2.26.0                          \n",
              " [55] tools_4.1.3                            \n",
              " [56] usethis_2.1.5                          \n",
              " [57] ggplotify_0.1.0                        \n",
              " [58] ellipsis_0.3.2                         \n",
              " [59] gplots_3.1.1                           \n",
              " [60] RColorBrewer_1.1-3                     \n",
              " [61] sessioninfo_1.2.2                      \n",
              " [62] Rcpp_1.0.8.3                           \n",
              " [63] plyr_1.8.7                             \n",
              " [64] base64enc_0.1-3                        \n",
              " [65] progress_1.2.2                         \n",
              " [66] zlibbioc_1.40.0                        \n",
              " [67] purrr_0.3.4                            \n",
              " [68] RCurl_1.98-1.6                         \n",
              " [69] ps_1.6.0                               \n",
              " [70] prettyunits_1.1.1                      \n",
              " [71] viridis_0.6.2                          \n",
              " [72] ggrepel_0.9.1                          \n",
              " [73] fs_1.5.2                               \n",
              " [74] magrittr_2.0.3                         \n",
              " [75] DO.db_2.9                              \n",
              " [76] pkgload_1.2.4                          \n",
              " [77] hms_1.1.1                              \n",
              " [78] patchwork_1.1.1                        \n",
              " [79] evaluate_0.15                          \n",
              " [80] xtable_1.8-4                           \n",
              " [81] XML_3.99-0.9                           \n",
              " [82] gridExtra_2.3                          \n",
              " [83] testthat_3.1.3                         \n",
              " [84] compiler_4.1.3                         \n",
              " [85] biomaRt_2.50.3                         \n",
              " [86] tibble_3.1.6                           \n",
              " [87] KernSmooth_2.23-20                     \n",
              " [88] crayon_1.5.1                           \n",
              " [89] shadowtext_0.1.1                       \n",
              " [90] R.oo_1.24.0                            \n",
              " [91] htmltools_0.5.2                        \n",
              " [92] tzdb_0.3.0                             \n",
              " [93] ggfun_0.0.6                            \n",
              " [94] TFBSTools_1.32.0                       \n",
              " [95] geneplotter_1.72.0                     \n",
              " [96] aplot_0.1.3                            \n",
              " [97] DBI_1.1.2                              \n",
              " [98] tweenr_1.0.2                           \n",
              " [99] dbplyr_2.1.1                           \n",
              "[100] MASS_7.3-56                            \n",
              "[101] rappdirs_0.3.3                         \n",
              "[102] boot_1.3-28                            \n",
              "[103] readr_2.1.2                            \n",
              "[104] Matrix_1.4-1                           \n",
              "[105] brio_1.1.3                             \n",
              "[106] cli_3.2.0                              \n",
              "[107] R.methodsS3_1.8.1                      \n",
              "[108] igraph_1.3.0                           \n",
              "[109] pkgconfig_2.0.3                        \n",
              "[110] TxDb.Hsapiens.UCSC.hg19.knownGene_3.2.2\n",
              "[111] TFMPvalue_0.0.8                        \n",
              "[112] GenomicAlignments_1.30.0               \n",
              "[113] IRdisplay_1.1                          \n",
              "[114] xml2_1.3.3                             \n",
              "[115] ggtree_3.2.1                           \n",
              "[116] annotate_1.72.0                        \n",
              "[117] DirichletMultinomial_1.36.0            \n",
              "[118] yulab.utils_0.0.4                      \n",
              "[119] stringr_1.4.0                          \n",
              "[120] callr_3.7.0                            \n",
              "[121] digest_0.6.29                          \n",
              "[122] pracma_2.3.8                           \n",
              "[123] CNEr_1.30.0                            \n",
              "[124] fastmatch_1.1-3                        \n",
              "[125] tidytree_0.3.9                         \n",
              "[126] restfulr_0.0.13                        \n",
              "[127] curl_4.3.2                             \n",
              "[128] Rsamtools_2.10.0                       \n",
              "[129] gtools_3.9.2                           \n",
              "[130] rjson_0.2.21                           \n",
              "[131] lifecycle_1.0.1                        \n",
              "[132] nlme_3.1-157                           \n",
              "[133] jsonlite_1.8.0                         \n",
              "[134] desc_1.4.1                             \n",
              "[135] viridisLite_0.4.0                      \n",
              "[136] fansi_1.0.3                            \n",
              "[137] pillar_1.7.0                           \n",
              "[138] lattice_0.20-45                        \n",
              "[139] KEGGREST_1.34.0                        \n",
              "[140] fastmap_1.1.0                          \n",
              "[141] httr_1.4.2                             \n",
              "[142] plotrix_3.8-2                          \n",
              "[143] pkgbuild_1.3.1                         \n",
              "[144] survival_3.3-1                         \n",
              "[145] GO.db_3.14.0                           \n",
              "[146] glue_1.6.2                             \n",
              "[147] remotes_2.4.2                          \n",
              "[148] png_0.1-7                              \n",
              "[149] bit_4.0.4                              \n",
              "[150] ggforce_0.3.3                          \n",
              "[151] stringi_1.7.6                          \n",
              "[152] blob_1.2.3                             \n",
              "[153] caTools_1.18.2                         \n",
              "[154] memoise_2.0.1                          \n",
              "[155] IRkernel_1.3                           \n",
              "[156] ape_5.6-2                              "
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "t8JLxMUeybHs",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 35
        },
        "outputId": "1f47d503-f94d-4274-c3d2-a3235788241e"
      },
      "source": [
        "Sys.time() - start_time0"
      ],
      "execution_count": 65,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "Time difference of 36.85204 mins"
            ]
          },
          "metadata": {}
        }
      ]
    }
  ]
}