{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "kernelspec": {
      "display_name": "R",
      "language": "R",
      "name": "ir"
    },
    "language_info": {
      "codemirror_mode": "r",
      "file_extension": ".r",
      "mimetype": "text/x-r-source",
      "name": "R",
      "pygments_lexer": "r",
      "version": "3.3.1"
    },
    "colab": {
      "provenance": [],
      "machine_shape": "hm",
      "toc_visible": true,
      "collapsed_sections": [
        "wFuTF2MUEiAk"
      ],
      "include_colab_link": true
    },
    "accelerator": "GPU"
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/gongx030/seatac_manuscript/blob/main/notebooks/pred_histone_bs640_k10.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "4TLq2MbDD3zq"
      },
      "source": [
        "# Generate V-plot representations of GM12878, K562, and Etv2 reprogramming data"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Background"
      ],
      "metadata": {
        "id": "Vjj4t4zwi0Ok"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Reviewer #1's Issue #2: Further, as SeATAC appears to be unique in documenting decreased chromatin accessibility at promoters of regulated genes, this aspect would be particularly interesting to biologists and browser views would be very helpful to convince the readers. To be really helpful, these browser views should include other typical marks of promoters and enhancers such as H3K4me3, H3K4me1, H3K27ac, etc."
      ],
      "metadata": {
        "id": "hDfuFM8T0GWl"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Tasks"
      ],
      "metadata": {
        "id": "tO_DYRmfXDkx"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "* Train SeATAC model on GM12878, K562 and Etv2 reprogramming ATAC-seq\n",
        "* Get V-plot representations\n",
        "* Get PCA representations of MACS2 and NucleoATAC signals\n",
        "* Train a simple MLP model to predict histone codes (H3K4me1, H3K4me3, H3K27ac, H3K27me3 and etc.) from V-plots representations.\n",
        "* Evaluate the prediction performance.\n",
        "* Verify the prediction of H3K27ac on Etv2 reprogramming data.\n",
        "* Predict the missing histone codes for the Etv2 binding sites in Etv2 reprogramming data."
      ],
      "metadata": {
        "id": "_6nErFC40hiJ"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Setup"
      ],
      "metadata": {
        "id": "mKVG3W-LjCh-"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "### MSI Parition"
      ],
      "metadata": {
        "id": "8NUcGWdphu6E"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "```salloc --nodes=1 --ntasks-per-node=16 --mem=64gb -t 24:00:00 -p interactive```"
      ],
      "metadata": {
        "id": "CRZgXi9Th2Yn"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Conda"
      ],
      "metadata": {
        "id": "BBq9MjHUh4OG"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "```r421_py37_tf26```"
      ],
      "metadata": {
        "id": "24xsV1WuiTsg"
      }
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "wFuTF2MUEiAk"
      },
      "source": [
        "### Install R packages"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "hv8nq9sjEzhx"
      },
      "source": [
        "required_packages <- c(\n",
        "  'BiocManager', 'devtools', 'Matrix',\n",
        "  'dplyr', 'tidyr', 'stringr',\n",
        "  'pROC', 'irlba',\n",
        "  'chromVAR', 'chromVARmotifs', 'motifmatchr'\n",
        ")"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "FDu62M5JFJZt"
      },
      "source": [
        "missing_packages <- required_packages[!required_packages %in% rownames(installed.packages())]\n",
        "if (length(missing_packages) > 0){\n",
        "    BiocManager::install(missing_packages)\n",
        "}"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "for (p in required_packages){\n",
        "  library(p, character.only  = TRUE)\n",
        "}"
      ],
      "metadata": {
        "id": "J3iIivkyr2ZR",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "5226327b-fa94-404c-f75d-626ca73652ba"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Bioconductor version '3.15' is out-of-date; the current release version '3.16'\n",
            "  is available with R version '4.2'; see https://bioconductor.org/install\n",
            "\n",
            "Loading required package: usethis\n",
            "\n",
            "\n",
            "Attaching package: ‘devtools’\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:BiocManager’:\n",
            "\n",
            "    install\n",
            "\n",
            "\n",
            "Warning message:\n",
            "“package ‘Matrix’ was built under R version 4.2.2”\n",
            "\n",
            "Attaching package: ‘dplyr’\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:stats’:\n",
            "\n",
            "    filter, lag\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:base’:\n",
            "\n",
            "    intersect, setdiff, setequal, union\n",
            "\n",
            "\n",
            "\n",
            "Attaching package: ‘tidyr’\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:Matrix’:\n",
            "\n",
            "    expand, pack, unpack\n",
            "\n",
            "\n",
            "Warning message:\n",
            "“package ‘stringr’ was built under R version 4.2.2”\n",
            "Type 'citation(\"pROC\")' for a citation.\n",
            "\n",
            "\n",
            "Attaching package: ‘pROC’\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:stats’:\n",
            "\n",
            "    cov, smooth, var\n",
            "\n",
            "\n",
            "\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "JV2N0ehhUtZm"
      },
      "source": [
        "### Install SeATAC from github"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "p9iI0Qk5Usy7"
      },
      "source": [
        "#devtools::install('../seatac')\n",
        "#install_github('gongx030/seatac@v0.4.0', force = TRUE, quiet = FALSE)"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#reload(pkgload::inst(\"seatac\"))\n",
        "#library(seatac)"
      ],
      "metadata": {
        "id": "2jPtosE8LdWd"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "devtools::load_all('../seatac')"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "gA94tv2IDp4V",
        "outputId": "fb66633a-6b73-42fb-cf71-6cb76ec5d98d"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Loading \u001b[34mseatac\u001b[39m\n",
            "Loading required package: tensorflow\n",
            "\n",
            "Warning message:\n",
            "“package ‘tensorflow’ was built under R version 4.2.2”\n",
            "Loading required package: tfprobability\n",
            "\n",
            "Loading required package: tfdatasets\n",
            "\n",
            "Loading required package: GenomicRanges\n",
            "\n",
            "Loading required package: stats4\n",
            "\n",
            "Loading required package: BiocGenerics\n",
            "\n",
            "\n",
            "Attaching package: ‘BiocGenerics’\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:pROC’:\n",
            "\n",
            "    var\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:dplyr’:\n",
            "\n",
            "    combine, intersect, setdiff, union\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:stats’:\n",
            "\n",
            "    IQR, mad, sd, var, xtabs\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:base’:\n",
            "\n",
            "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
            "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
            "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
            "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
            "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
            "    union, unique, unsplit, which.max, which.min\n",
            "\n",
            "\n",
            "Loading required package: S4Vectors\n",
            "\n",
            "\n",
            "Attaching package: ‘S4Vectors’\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:tidyr’:\n",
            "\n",
            "    expand\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:dplyr’:\n",
            "\n",
            "    first, rename\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:Matrix’:\n",
            "\n",
            "    expand, unname\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:base’:\n",
            "\n",
            "    expand.grid, I, unname\n",
            "\n",
            "\n",
            "Loading required package: IRanges\n",
            "\n",
            "\n",
            "Attaching package: ‘IRanges’\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:dplyr’:\n",
            "\n",
            "    collapse, desc, slice\n",
            "\n",
            "\n",
            "Loading required package: GenomeInfoDb\n",
            "\n",
            "Loading required package: SummarizedExperiment\n",
            "\n",
            "Loading required package: MatrixGenerics\n",
            "\n",
            "Loading required package: matrixStats\n",
            "\n",
            "\n",
            "Attaching package: ‘matrixStats’\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:dplyr’:\n",
            "\n",
            "    count\n",
            "\n",
            "\n",
            "\n",
            "Attaching package: ‘MatrixGenerics’\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:matrixStats’:\n",
            "\n",
            "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
            "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
            "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
            "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
            "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
            "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
            "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
            "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
            "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
            "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
            "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
            "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
            "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
            "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
            "    rowWeightedSds, rowWeightedVars\n",
            "\n",
            "\n",
            "Loading required package: Biobase\n",
            "\n",
            "Welcome to Bioconductor\n",
            "\n",
            "    Vignettes contain introductory material; view with\n",
            "    'browseVignettes()'. To cite Bioconductor, see\n",
            "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
            "\n",
            "\n",
            "\n",
            "Attaching package: ‘Biobase’\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:MatrixGenerics’:\n",
            "\n",
            "    rowMedians\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:matrixStats’:\n",
            "\n",
            "    anyMissing, rowMedians\n",
            "\n",
            "\n",
            "Creating a new generic function for ‘fit’ in package ‘seatac’\n",
            "\n",
            "Creating a new generic function for ‘predict’ in package ‘seatac’\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Load `tfa`"
      ],
      "metadata": {
        "id": "09ioiaCiDY13"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Required for `tfa.image.gaussian_filter2d`"
      ],
      "metadata": {
        "id": "LPu7C1_iDaog"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "tfa <- reticulate::import('tensorflow_addons')"
      ],
      "metadata": {
        "id": "WctoNewHAHrg"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Load bpt"
      ],
      "metadata": {
        "id": "QAFMnWr_EGBW"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "devtools::load_all('../bpt')"
      ],
      "metadata": {
        "id": "lMZPDXvpEInR",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "c6a621b3-7a75-431c-fd24-bdad66e7f570"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Loading \u001b[34mbpt\u001b[39m\n",
            "Warning message:\n",
            "“replacing previous import ‘GenomicRanges::union’ by ‘dplyr::union’ when loading ‘bpt’”\n",
            "Warning message:\n",
            "“replacing previous import ‘GenomicRanges::intersect’ by ‘dplyr::intersect’ when loading ‘bpt’”\n",
            "Warning message:\n",
            "“replacing previous import ‘GenomicRanges::setdiff’ by ‘dplyr::setdiff’ when loading ‘bpt’”\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "hl2lXOAv9ljx"
      },
      "source": [
        "## Parameters"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "set.seed(123)"
      ],
      "metadata": {
        "id": "5IlR-BldsR6i"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "KMfU5_KD9nN9"
      },
      "source": [
        "ARGS <- list(\n",
        "\n",
        "  remote_bam_files = c(\n",
        "    'GM12878' = 'https://s3.msi.umn.edu/gongx030/datasets/dataset=Buenrostro_version=20170721a/GM12878_50k_cells.bam',\n",
        "    'K562' = 'https://s3.msi.umn.edu/gongx030/datasets/dataset=K562_version=20230104a/K562.bam',\n",
        "    'MEF_NoDox' = 'https://s3.msi.umn.edu/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/MEF_NoDox.bam',\n",
        "    'MEF_Dox_D7_Flk1pos' = 'https://s3.msi.umn.edu/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/MEF_Dox_D7_Flk1pos.bam',\n",
        "    'EB_NoDox_D25' = 'https://s3.msi.umn.edu/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/EB_NoDox_D25.bam',\n",
        "    'EB_Dox_D25_Flk1pos' = 'https://s3.msi.umn.edu/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/EB_Dox_D25_Flk1pos.bam'\n",
        "  ), \n",
        "  genome_versions = c('hg19', 'hg19', 'mm10', 'mm10', 'mm10', 'mm10'),\n",
        "  remote_bw_files = list(\n",
        "    'GM12878' = c(\n",
        "      'H3K27ac' = 'https://www.encodeproject.org/files/ENCFF440GZA/@@download/ENCFF440GZA.bigWig',\n",
        "      'H3K4me1' = 'https://www.encodeproject.org/files/ENCFF372EXQ/@@download/ENCFF372EXQ.bigWig',\n",
        "      'H3K4me3' = 'https://www.encodeproject.org/files/ENCFF154XCY/@@download/ENCFF154XCY.bigWig'\n",
        "    ),\n",
        "    'K562' = c(\n",
        "      'H3K27ac' = 'https://www.encodeproject.org/files/ENCFF488FYZ/@@download/ENCFF488FYZ.bigWig',\n",
        "      'H3K4me1' = 'https://www.encodeproject.org/files/ENCFF700ZWZ/@@download/ENCFF700ZWZ.bigWig',\n",
        "      'H3K4me3' = 'https://www.encodeproject.org/files/ENCFF285EUA/@@download/ENCFF285EUA.bigWig'\n",
        "    ),\n",
        "    'MEF_NoDox' = c(\n",
        "      'H3K27ac' = 'https://s3.msi.umn.edu/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/MEF_NoDox_d0_H3K27ac_FE.bw'\n",
        "    ),\n",
        "    'MEF_Dox_D7_Flk1pos' = c(\n",
        "      'H3K27ac' = 'https://s3.msi.umn.edu/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/MEF_Dox_d7_H3K27ac_FE.bw'\n",
        "    ),\n",
        "    'EB_NoDox_D25' = c(\n",
        "      'H3K27ac' = 'https://s3.msi.umn.edu/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/EB_NoDox_12h_H3K27ac_FE.bw'\n",
        "    ),\n",
        "    'EB_Dox_D25_Flk1pos' = c(\n",
        "      'H3K27ac' = 'https://s3.msi.umn.edu/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/EB_Dox_12h_H3K27ac_FE.bw'\n",
        "    )\n",
        "  ),\n",
        "  block_size = 640L,\n",
        "  latent_dim = 10L,\n",
        "  min_reads_train_seatac = 5L,  # minimum reads in blocks to train SeATAC\n",
        "  min_reads_train_histone = 5L,  # minimum reads in blocks to train histone model\n",
        "  min_reads_test_histone = 5L, # minimum reads in blocks to test histone model\n",
        "  step_size = 200L,\n",
        "  fragment_size_range_max = 640L,\n",
        "  bin_size = 5L,\n",
        "  fragment_size_interval = 10L,\n",
        "  window_size = 2000L,\n",
        "  sliding_step = 10L,\n",
        "  max_peaks_train_seatac = 2000L, # maximum peaks per sample for training SeATAC\n",
        "  max_peaks_count_reads = 10000L,  # maximum peaks for counting reads\n",
        "  max_peaks_train_histone = 50000L, # maximum peaks for training histone model\n",
        "  max_peaks_test_histone = 20000L, # maximum peaks for testing histone model\n",
        "  batch_size = 256L,\n",
        "  epochs = 100L,\n",
        "  validation_split = 0.2,\n",
        "  dataset_train_histone = c('GM12878'),\n",
        "  min_histone_signal_train = 100,\n",
        "  min_histone_signal_test = 10,\n",
        "\n",
        "  remote_representation_file = 'https://s3.msi.umn.edu/gongx030/projects/seatac_project/etv2_20230108a.RData'\n",
        "\n",
        ")"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "ARGS$local_bam_files <- gsub('https://', './', ARGS$remote_bam_files)"
      ],
      "metadata": {
        "id": "DrihFrVyJBnb"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "ARGS$remote_summit_files <- gsub('.bam', '_summits.bed', ARGS$remote_bam_files)\n",
        "ARGS$local_summit_files <- gsub('https://', './', ARGS$remote_summit_files)"
      ],
      "metadata": {
        "id": "j1ARnDoIJCHe"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "ARGS$local_bw_files <- list()\n",
        "for (i in names(ARGS$remote_bw_files)){\n",
        "  for (j in names(ARGS$remote_bw_files[[i]])){\n",
        "    ARGS$local_bw_files[[i]][j] <- gsub('https://', './', ARGS$remote_bw_files[[i]][j])\n",
        "  }\n",
        "  stopifnot(all(file.exists(ARGS$local_bw_files[[i]])))\n",
        "}"
      ],
      "metadata": {
        "id": "tE2dZNCQJHkr"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "ARGS$remote_nucleoatac_files <- gsub('.bam', '.nucleoatac_signal.smooth.bw', ARGS$remote_bam_files)\n",
        "ARGS$local_nucleoatac_files <- gsub('https://', './', ARGS$remote_nucleoatac_files)"
      ],
      "metadata": {
        "id": "qEcKLfKaK3nM"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "ARGS$remote_macs2_files <- gsub('.bam', '_treat_pileup.bw', ARGS$remote_bam_files)\n",
        "ARGS$local_macs2_files <- gsub('https://', './', ARGS$remote_macs2_files)"
      ],
      "metadata": {
        "id": "ykp7e5duLBbw"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "ARGS$local_representation_file <- gsub('https://', './', ARGS$remote_representation_file)\n",
        "ARGS$s3_representation_file <- gsub('https://s3.msi.umn.edu/', 's3://', ARGS$remote_representation_file)"
      ],
      "metadata": {
        "id": "9UVwTcZ0xGT7"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Download files"
      ],
      "metadata": {
        "id": "RKSmgudb_WsU"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "### BAM files"
      ],
      "metadata": {
        "id": "8HEq54UY_n_a"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "download_remote_file(ARGS$remote_bam_files)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "ymnU3dMZ_Vi3",
        "outputId": "766f3caf-8d72-4703-b923-52d4ede8cb38"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "wget -cx https://s3.msi.umn.edu/gongx030/datasets/dataset=Buenrostro_version=20170721a/GM12878_50k_cells.bam\n",
            "\n",
            "wget -cx https://s3.msi.umn.edu/gongx030/datasets/dataset=K562_version=20230104a/K562.bam\n",
            "\n",
            "wget -cx https://s3.msi.umn.edu/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/MEF_NoDox.bam\n",
            "\n",
            "wget -cx https://s3.msi.umn.edu/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/MEF_Dox_D7_Flk1pos.bam\n",
            "\n",
            "wget -cx https://s3.msi.umn.edu/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/EB_NoDox_D25.bam\n",
            "\n",
            "wget -cx https://s3.msi.umn.edu/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/EB_Dox_D25_Flk1pos.bam\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "stopifnot(all(file.exists(ARGS$local_bam_files)))"
      ],
      "metadata": {
        "id": "nKTE-WNX_jrl"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Summit files"
      ],
      "metadata": {
        "id": "MU7a_aUb_qSl"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "download_remote_file(ARGS$remote_summit_files)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Zf8WMgNE_GJK",
        "outputId": "eb08bc2a-d058-4271-8dca-78be14307e63"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "wget -cx https://s3.msi.umn.edu/gongx030/datasets/dataset=Buenrostro_version=20170721a/GM12878_50k_cells_summits.bed\n",
            "\n",
            "wget -cx https://s3.msi.umn.edu/gongx030/datasets/dataset=K562_version=20230104a/K562_summits.bed\n",
            "\n",
            "wget -cx https://s3.msi.umn.edu/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/MEF_NoDox_summits.bed\n",
            "\n",
            "wget -cx https://s3.msi.umn.edu/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/MEF_Dox_D7_Flk1pos_summits.bed\n",
            "\n",
            "wget -cx https://s3.msi.umn.edu/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/EB_NoDox_D25_summits.bed\n",
            "\n",
            "wget -cx https://s3.msi.umn.edu/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/EB_Dox_D25_Flk1pos_summits.bed\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "stopifnot(file.exists(ARGS$local_summit_files))"
      ],
      "metadata": {
        "id": "uXJpb9YE-Leb"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### bigwig files"
      ],
      "metadata": {
        "id": "w1uDNQzwDT3b"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "for (i in names(ARGS$remote_bw_files)){\n",
        "  download_remote_file(ARGS$remote_bw_files[[i]])\n",
        "  stopifnot(all(file.exists(ARGS$local_bw_files[[i]])))\n",
        "}"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "f0OqvDYFDhC-",
        "outputId": "9650771c-6f7f-4a0d-809f-7c9a9247f453"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "wget -cx https://www.encodeproject.org/files/ENCFF440GZA/@@download/ENCFF440GZA.bigWig\n",
            "\n",
            "wget -cx https://www.encodeproject.org/files/ENCFF372EXQ/@@download/ENCFF372EXQ.bigWig\n",
            "\n",
            "wget -cx https://www.encodeproject.org/files/ENCFF154XCY/@@download/ENCFF154XCY.bigWig\n",
            "\n",
            "wget -cx https://www.encodeproject.org/files/ENCFF488FYZ/@@download/ENCFF488FYZ.bigWig\n",
            "\n",
            "wget -cx https://www.encodeproject.org/files/ENCFF700ZWZ/@@download/ENCFF700ZWZ.bigWig\n",
            "\n",
            "wget -cx https://www.encodeproject.org/files/ENCFF285EUA/@@download/ENCFF285EUA.bigWig\n",
            "\n",
            "wget -cx https://s3.msi.umn.edu/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/MEF_NoDox_d0_H3K27ac_FE.bw\n",
            "\n",
            "wget -cx https://s3.msi.umn.edu/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/MEF_Dox_d7_H3K27ac_FE.bw\n",
            "\n",
            "wget -cx https://s3.msi.umn.edu/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/EB_NoDox_12h_H3K27ac_FE.bw\n",
            "\n",
            "wget -cx https://s3.msi.umn.edu/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/EB_Dox_12h_H3K27ac_FE.bw\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### NucleoATAC files"
      ],
      "metadata": {
        "id": "h3MfAa5TLUig"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "download_remote_file(ARGS$remote_nucleoatac_files)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "86a219c5-4d30-4db8-fa65-81b965d5131a",
        "id": "g6d_IXmqLUig"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "wget -cx https://s3.msi.umn.edu/gongx030/datasets/dataset=Buenrostro_version=20170721a/GM12878_50k_cells.nucleoatac_signal.smooth.bw\n",
            "\n",
            "wget -cx https://s3.msi.umn.edu/gongx030/datasets/dataset=K562_version=20230104a/K562.nucleoatac_signal.smooth.bw\n",
            "\n",
            "wget -cx https://s3.msi.umn.edu/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/MEF_NoDox.nucleoatac_signal.smooth.bw\n",
            "\n",
            "wget -cx https://s3.msi.umn.edu/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/MEF_Dox_D7_Flk1pos.nucleoatac_signal.smooth.bw\n",
            "\n",
            "wget -cx https://s3.msi.umn.edu/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/EB_NoDox_D25.nucleoatac_signal.smooth.bw\n",
            "\n",
            "wget -cx https://s3.msi.umn.edu/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/EB_Dox_D25_Flk1pos.nucleoatac_signal.smooth.bw\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "stopifnot(all(file.exists(ARGS$local_nucleoatac_files)))"
      ],
      "metadata": {
        "id": "zNGUdPnLLcw9"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### MACS2 pileup files"
      ],
      "metadata": {
        "id": "lbWxvbsvLpNu"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "download_remote_file(ARGS$remote_macs2_files)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "236a2cff-28db-47fd-8c25-dc18f2df4f52",
        "id": "93X2KoWiLpNu"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "wget -cx https://s3.msi.umn.edu/gongx030/datasets/dataset=Buenrostro_version=20170721a/GM12878_50k_cells_treat_pileup.bw\n",
            "\n",
            "wget -cx https://s3.msi.umn.edu/gongx030/datasets/dataset=K562_version=20230104a/K562_treat_pileup.bw\n",
            "\n",
            "wget -cx https://s3.msi.umn.edu/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/MEF_NoDox_treat_pileup.bw\n",
            "\n",
            "wget -cx https://s3.msi.umn.edu/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/MEF_Dox_D7_Flk1pos_treat_pileup.bw\n",
            "\n",
            "wget -cx https://s3.msi.umn.edu/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/EB_NoDox_D25_treat_pileup.bw\n",
            "\n",
            "wget -cx https://s3.msi.umn.edu/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/EB_Dox_D25_Flk1pos_treat_pileup.bw\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "stopifnot(all(file.exists(ARGS$local_macs2_files)))"
      ],
      "metadata": {
        "id": "K324zP2-LpNu"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Read summits for training SeATAC"
      ],
      "metadata": {
        "id": "mPYL5SnL-Dan"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "x <- list()\n",
        "for (i in 1:length(ARGS$local_bam_files)){\n",
        "  sprintf('reading %s', ARGS$local_summit_files[i]) %>% message()\n",
        "  gr0 <- read.table(ARGS$local_summit_files[i], header = FALSE, sep = '\\t')\n",
        "  gr0 <- GRanges(seqnames = gr0[, 1], range = IRanges(start = gr0[, 2], end = gr0[, 3]))\n",
        "  gr0 <- gr0[seqnames(gr0) %in% sprintf('chr%s', c(1:22, 'X'))]\n",
        "  gr0 <- resize(gr0, fix = 'center', width = ARGS$window_size)\n",
        "\n",
        "  gr <- slidingWindows(gr0, width = ARGS$block_size, step = ARGS$bin_size) %>% unlist()\n",
        "  sprintf('total blocks: %d', length(gr)) %>% message()\n",
        "  gr <- sample(gr, min(length(gr), ARGS$max_peaks_count_reads))\n",
        "  sprintf('total blocks for training: %d', length(gr)) %>% message()\n",
        "\n",
        "  counts <- count_reads(gr, ARGS$local_bam_files[i], genome = get_bsgenome(ARGS$genome_versions[i]),  fragment_size_range = c(0, ARGS$fragment_size_range_max))\n",
        "\n",
        "  valid <- counts >= ARGS$min_reads_train_seatac\n",
        "  sprintf('number of valid peaks for VaeModel(>=%d reads): %d', ARGS$min_reads_train_seatac, sum(valid)) %>% message()\n",
        "\n",
        "  gr <- gr[valid]\n",
        "  gr <- sample(gr, min(length(gr), ARGS$max_peaks_train_seatac))\n",
        "  sprintf('sampling %d valid peaks for training VaeModel', length(gr)) %>% message()\n",
        "\n",
        "  x[[i]] <- read_vplot(\n",
        "    gr,\n",
        "    filenames = ARGS$local_bam_files[i], \n",
        "    genome = get_bsgenome(ARGS$genome_versions[i]),\n",
        "    bin_size = ARGS$bin_size,\n",
        "    fragment_size_range = c(0, ARGS$fragment_size_range_max),\n",
        "    fragment_size_interval = ARGS$fragment_size_interval \n",
        "  )\n",
        "}"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "kxl2tCmOQiND",
        "outputId": "edd5e109-11f5-48d7-e250-4dfc9e454be7"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "reading ./s3.msi.umn.edu/gongx030/datasets/dataset=Buenrostro_version=20170721a/GM12878_50k_cells_summits.bed\n",
            "\n",
            "total blocks: 28502565\n",
            "\n",
            "total blocks for training: 10000\n",
            "\n",
            "Loading required package: BSgenome.Mmusculus.UCSC.mm10\n",
            "\n",
            "Loading required package: BSgenome\n",
            "\n",
            "Loading required package: Biostrings\n",
            "\n",
            "Loading required package: XVector\n",
            "\n",
            "\n",
            "Attaching package: ‘Biostrings’\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:base’:\n",
            "\n",
            "    strsplit\n",
            "\n",
            "\n",
            "Loading required package: rtracklayer\n",
            "\n",
            "Loading required package: BSgenome.Hsapiens.UCSC.hg19\n",
            "\n",
            "read_bam | reading ./s3.msi.umn.edu/gongx030/datasets/dataset=Buenrostro_version=20170721a/GM12878_50k_cells.bam\n",
            "\n",
            "number of valid peaks for VaeModel(>=5 reads): 9978\n",
            "\n",
            "Warning message in NSBS(i, x, exact = exact, strict.upper.bound = !allow.append, :\n",
            "“subscript is an array, passing it thru as.vector() first”\n",
            "sampling 2000 valid peaks for training VaeModel\n",
            "\n",
            "read_bam | reading ./s3.msi.umn.edu/gongx030/datasets/dataset=Buenrostro_version=20170721a/GM12878_50k_cells.bam\n",
            "\n",
            "as(<ngCMatrix>, \"dgCMatrix\") is deprecated since Matrix 1.5-0; do as(., \"dMatrix\") instead\n",
            "\n",
            "reading ./s3.msi.umn.edu/gongx030/datasets/dataset=K562_version=20230104a/K562_summits.bed\n",
            "\n",
            "total blocks: 54366312\n",
            "\n",
            "total blocks for training: 10000\n",
            "\n",
            "read_bam | reading ./s3.msi.umn.edu/gongx030/datasets/dataset=K562_version=20230104a/K562.bam\n",
            "\n",
            "number of valid peaks for VaeModel(>=5 reads): 9975\n",
            "\n",
            "Warning message in NSBS(i, x, exact = exact, strict.upper.bound = !allow.append, :\n",
            "“subscript is an array, passing it thru as.vector() first”\n",
            "sampling 2000 valid peaks for training VaeModel\n",
            "\n",
            "read_bam | reading ./s3.msi.umn.edu/gongx030/datasets/dataset=K562_version=20230104a/K562.bam\n",
            "\n",
            "reading ./s3.msi.umn.edu/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/MEF_NoDox_summits.bed\n",
            "\n",
            "total blocks: 16933644\n",
            "\n",
            "total blocks for training: 10000\n",
            "\n",
            "read_bam | reading ./s3.msi.umn.edu/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/MEF_NoDox.bam\n",
            "\n",
            "number of valid peaks for VaeModel(>=5 reads): 9910\n",
            "\n",
            "Warning message in NSBS(i, x, exact = exact, strict.upper.bound = !allow.append, :\n",
            "“subscript is an array, passing it thru as.vector() first”\n",
            "sampling 2000 valid peaks for training VaeModel\n",
            "\n",
            "read_bam | reading ./s3.msi.umn.edu/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/MEF_NoDox.bam\n",
            "\n",
            "reading ./s3.msi.umn.edu/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/MEF_Dox_D7_Flk1pos_summits.bed\n",
            "\n",
            "total blocks: 8058960\n",
            "\n",
            "total blocks for training: 10000\n",
            "\n",
            "read_bam | reading ./s3.msi.umn.edu/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/MEF_Dox_D7_Flk1pos.bam\n",
            "\n",
            "number of valid peaks for VaeModel(>=5 reads): 9877\n",
            "\n",
            "Warning message in NSBS(i, x, exact = exact, strict.upper.bound = !allow.append, :\n",
            "“subscript is an array, passing it thru as.vector() first”\n",
            "sampling 2000 valid peaks for training VaeModel\n",
            "\n",
            "read_bam | reading ./s3.msi.umn.edu/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/MEF_Dox_D7_Flk1pos.bam\n",
            "\n",
            "reading ./s3.msi.umn.edu/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/EB_NoDox_D25_summits.bed\n",
            "\n",
            "total blocks: 23340954\n",
            "\n",
            "total blocks for training: 10000\n",
            "\n",
            "read_bam | reading ./s3.msi.umn.edu/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/EB_NoDox_D25.bam\n",
            "\n",
            "number of valid peaks for VaeModel(>=5 reads): 9913\n",
            "\n",
            "Warning message in NSBS(i, x, exact = exact, strict.upper.bound = !allow.append, :\n",
            "“subscript is an array, passing it thru as.vector() first”\n",
            "sampling 2000 valid peaks for training VaeModel\n",
            "\n",
            "read_bam | reading ./s3.msi.umn.edu/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/EB_NoDox_D25.bam\n",
            "\n",
            "reading ./s3.msi.umn.edu/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/EB_Dox_D25_Flk1pos_summits.bed\n",
            "\n",
            "total blocks: 15835092\n",
            "\n",
            "total blocks for training: 10000\n",
            "\n",
            "read_bam | reading ./s3.msi.umn.edu/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/EB_Dox_D25_Flk1pos.bam\n",
            "\n",
            "number of valid peaks for VaeModel(>=5 reads): 9712\n",
            "\n",
            "Warning message in NSBS(i, x, exact = exact, strict.upper.bound = !allow.append, :\n",
            "“subscript is an array, passing it thru as.vector() first”\n",
            "sampling 2000 valid peaks for training VaeModel\n",
            "\n",
            "read_bam | reading ./s3.msi.umn.edu/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/EB_Dox_D25_Flk1pos.bam\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "names(x) <- names(ARGS$local_bam_files)\n",
        "x <- VplotsList(x)"
      ],
      "metadata": {
        "id": "iaTy1u8-Lh4_"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Training SeATAC"
      ],
      "metadata": {
        "id": "fKlh1WaFHi41"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "model <- VaeModel(x, latent_dim = ARGS$latent_dim)"
      ],
      "metadata": {
        "id": "n6hLf5vMM0tp"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "model <- model %>% fit(x, batch_size = ARGS$batch_size, epochs = ARGS$epochs, compile = TRUE)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "WpmMawIsM0wQ",
        "outputId": "d1e57224-7818-4cff-de82-be1b0a90d621"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "epoch=     1/   100 | beta=1.000e+00 | loss= 6128.3309820 | loss_reconstruction= 6125.6838015 | loss_kl=    2.6471532\n",
            "\n",
            "epoch=     2/   100 | beta=1.000e+00 | loss= 4062.7965166 | loss_reconstruction= 4061.0384444 | loss_kl=    1.7580885\n",
            "\n",
            "epoch=     3/   100 | beta=1.000e+00 | loss= 1676.7342711 | loss_reconstruction= 1676.4609245 | loss_kl=    0.2733526\n",
            "\n",
            "epoch=     4/   100 | beta=1.000e+00 | loss=  549.2022004 | loss_reconstruction=  549.0192540 | loss_kl=    0.1829454\n",
            "\n",
            "epoch=     5/   100 | beta=1.000e+00 | loss=  375.0864005 | loss_reconstruction=  373.8919106 | loss_kl=    1.1944921\n",
            "\n",
            "epoch=     6/   100 | beta=1.000e+00 | loss=  337.0319805 | loss_reconstruction=  334.8865837 | loss_kl=    2.1453968\n",
            "\n",
            "epoch=     7/   100 | beta=1.000e+00 | loss=  325.4783419 | loss_reconstruction=  322.0617731 | loss_kl=    3.4165678\n",
            "\n",
            "epoch=     8/   100 | beta=1.000e+00 | loss=  321.1480596 | loss_reconstruction=  317.4270020 | loss_kl=    3.7210572\n",
            "\n",
            "epoch=     9/   100 | beta=1.000e+00 | loss=  316.3684209 | loss_reconstruction=  312.1369301 | loss_kl=    4.2314911\n",
            "\n",
            "epoch=    10/   100 | beta=1.000e+00 | loss=  314.2934304 | loss_reconstruction=  309.3346636 | loss_kl=    4.9587679\n",
            "\n",
            "epoch=    11/   100 | beta=1.000e+00 | loss=  311.9223315 | loss_reconstruction=  305.6621808 | loss_kl=    6.2601504\n",
            "\n",
            "epoch=    12/   100 | beta=1.000e+00 | loss=  308.3014617 | loss_reconstruction=  301.6694271 | loss_kl=    6.6320346\n",
            "\n",
            "epoch=    13/   100 | beta=1.000e+00 | loss=  305.8138386 | loss_reconstruction=  298.8980041 | loss_kl=    6.9158329\n",
            "\n",
            "epoch=    14/   100 | beta=1.000e+00 | loss=  303.7753543 | loss_reconstruction=  296.6873711 | loss_kl=    7.0879833\n",
            "\n",
            "epoch=    15/   100 | beta=1.000e+00 | loss=  302.5262675 | loss_reconstruction=  295.1139692 | loss_kl=    7.4122976\n",
            "\n",
            "epoch=    16/   100 | beta=1.000e+00 | loss=  301.7984827 | loss_reconstruction=  293.9155848 | loss_kl=    7.8828987\n",
            "\n",
            "epoch=    17/   100 | beta=1.000e+00 | loss=  302.1595245 | loss_reconstruction=  294.1140497 | loss_kl=    8.0454746\n",
            "\n",
            "epoch=    18/   100 | beta=1.000e+00 | loss=  301.7289796 | loss_reconstruction=  293.6040357 | loss_kl=    8.1249442\n",
            "\n",
            "epoch=    19/   100 | beta=1.000e+00 | loss=  300.5349933 | loss_reconstruction=  292.5009107 | loss_kl=    8.0340816\n",
            "\n",
            "epoch=    20/   100 | beta=1.000e+00 | loss=  298.5338622 | loss_reconstruction=  290.3020373 | loss_kl=    8.2318251\n",
            "\n",
            "epoch=    21/   100 | beta=1.000e+00 | loss=  298.5018996 | loss_reconstruction=  290.2817428 | loss_kl=    8.2201562\n",
            "\n",
            "epoch=    22/   100 | beta=1.000e+00 | loss=  298.8872957 | loss_reconstruction=  290.7304129 | loss_kl=    8.1568824\n",
            "\n",
            "epoch=    23/   100 | beta=1.000e+00 | loss=  298.1546345 | loss_reconstruction=  289.8921006 | loss_kl=    8.2625341\n",
            "\n",
            "epoch=    24/   100 | beta=1.000e+00 | loss=  298.9325526 | loss_reconstruction=  290.6906819 | loss_kl=    8.2418700\n",
            "\n",
            "epoch=    25/   100 | beta=1.000e+00 | loss=  298.3492126 | loss_reconstruction=  290.0529116 | loss_kl=    8.2963009\n",
            "\n",
            "epoch=    26/   100 | beta=1.000e+00 | loss=  297.3158894 | loss_reconstruction=  289.0347400 | loss_kl=    8.2811499\n",
            "\n",
            "epoch=    27/   100 | beta=1.000e+00 | loss=  299.8180708 | loss_reconstruction=  291.4314530 | loss_kl=    8.3866166\n",
            "\n",
            "epoch=    28/   100 | beta=1.000e+00 | loss=  297.4151618 | loss_reconstruction=  289.1341160 | loss_kl=    8.2810461\n",
            "\n",
            "epoch=    29/   100 | beta=1.000e+00 | loss=  297.8026448 | loss_reconstruction=  289.4216805 | loss_kl=    8.3809633\n",
            "\n",
            "epoch=    30/   100 | beta=1.000e+00 | loss=  297.7627826 | loss_reconstruction=  289.4500934 | loss_kl=    8.3126892\n",
            "\n",
            "epoch=    31/   100 | beta=1.000e+00 | loss=  296.8073347 | loss_reconstruction=  288.4857817 | loss_kl=    8.3215528\n",
            "\n",
            "epoch=    32/   100 | beta=1.000e+00 | loss=  296.6586930 | loss_reconstruction=  288.3812350 | loss_kl=    8.2774594\n",
            "\n",
            "epoch=    33/   100 | beta=1.000e+00 | loss=  296.8430608 | loss_reconstruction=  288.4695288 | loss_kl=    8.3735309\n",
            "\n",
            "epoch=    34/   100 | beta=1.000e+00 | loss=  296.2273430 | loss_reconstruction=  287.8728024 | loss_kl=    8.3545388\n",
            "\n",
            "epoch=    35/   100 | beta=1.000e+00 | loss=  296.4337168 | loss_reconstruction=  288.0780257 | loss_kl=    8.3556900\n",
            "\n",
            "epoch=    36/   100 | beta=1.000e+00 | loss=  296.2318862 | loss_reconstruction=  287.7896381 | loss_kl=    8.4422492\n",
            "\n",
            "epoch=    37/   100 | beta=1.000e+00 | loss=  297.4825806 | loss_reconstruction=  289.0895908 | loss_kl=    8.3929881\n",
            "\n",
            "epoch=    38/   100 | beta=1.000e+00 | loss=  297.2415966 | loss_reconstruction=  288.7895388 | loss_kl=    8.4520580\n",
            "\n",
            "epoch=    39/   100 | beta=1.000e+00 | loss=  296.0601703 | loss_reconstruction=  287.6950138 | loss_kl=    8.3651559\n",
            "\n",
            "epoch=    40/   100 | beta=1.000e+00 | loss=  295.8499642 | loss_reconstruction=  287.4122477 | loss_kl=    8.4377163\n",
            "\n",
            "epoch=    41/   100 | beta=1.000e+00 | loss=  296.0570708 | loss_reconstruction=  287.6181176 | loss_kl=    8.4389524\n",
            "\n",
            "epoch=    42/   100 | beta=1.000e+00 | loss=  295.4797220 | loss_reconstruction=  287.0725117 | loss_kl=    8.4072103\n",
            "\n",
            "epoch=    43/   100 | beta=1.000e+00 | loss=  295.5425733 | loss_reconstruction=  287.0181885 | loss_kl=    8.5243865\n",
            "\n",
            "epoch=    44/   100 | beta=1.000e+00 | loss=  295.3477390 | loss_reconstruction=  286.8887144 | loss_kl=    8.4590265\n",
            "\n",
            "epoch=    45/   100 | beta=1.000e+00 | loss=  295.0007974 | loss_reconstruction=  286.5296975 | loss_kl=    8.4710990\n",
            "\n",
            "epoch=    46/   100 | beta=1.000e+00 | loss=  295.3218906 | loss_reconstruction=  286.9235957 | loss_kl=    8.3982950\n",
            "\n",
            "epoch=    47/   100 | beta=1.000e+00 | loss=  298.2076380 | loss_reconstruction=  289.5903317 | loss_kl=    8.6173065\n",
            "\n",
            "epoch=    48/   100 | beta=1.000e+00 | loss=  295.6444764 | loss_reconstruction=  287.1953862 | loss_kl=    8.4490924\n",
            "\n",
            "epoch=    49/   100 | beta=1.000e+00 | loss=  295.0846928 | loss_reconstruction=  286.5520289 | loss_kl=    8.5326645\n",
            "\n",
            "epoch=    50/   100 | beta=1.000e+00 | loss=  294.9569007 | loss_reconstruction=  286.4620342 | loss_kl=    8.4948649\n",
            "\n",
            "epoch=    51/   100 | beta=1.000e+00 | loss=  296.2002112 | loss_reconstruction=  287.7108138 | loss_kl=    8.4893970\n",
            "\n",
            "epoch=    52/   100 | beta=1.000e+00 | loss=  294.9740721 | loss_reconstruction=  286.3764713 | loss_kl=    8.5975981\n",
            "\n",
            "epoch=    53/   100 | beta=1.000e+00 | loss=  294.5983345 | loss_reconstruction=  286.0606293 | loss_kl=    8.5377078\n",
            "\n",
            "epoch=    54/   100 | beta=1.000e+00 | loss=  294.3866830 | loss_reconstruction=  285.8291402 | loss_kl=    8.5575422\n",
            "\n",
            "epoch=    55/   100 | beta=1.000e+00 | loss=  294.4893471 | loss_reconstruction=  285.8456642 | loss_kl=    8.6436828\n",
            "\n",
            "epoch=    56/   100 | beta=1.000e+00 | loss=  294.4435831 | loss_reconstruction=  285.8446688 | loss_kl=    8.5989141\n",
            "\n",
            "epoch=    57/   100 | beta=1.000e+00 | loss=  294.4420588 | loss_reconstruction=  285.8518738 | loss_kl=    8.5901862\n",
            "\n",
            "epoch=    58/   100 | beta=1.000e+00 | loss=  294.5927504 | loss_reconstruction=  285.8806143 | loss_kl=    8.7121351\n",
            "\n",
            "epoch=    59/   100 | beta=1.000e+00 | loss=  294.8298298 | loss_reconstruction=  286.1473298 | loss_kl=    8.6825001\n",
            "\n",
            "epoch=    60/   100 | beta=1.000e+00 | loss=  294.1966686 | loss_reconstruction=  285.5343427 | loss_kl=    8.6623266\n",
            "\n",
            "epoch=    61/   100 | beta=1.000e+00 | loss=  294.0220506 | loss_reconstruction=  285.3599526 | loss_kl=    8.6620980\n",
            "\n",
            "epoch=    62/   100 | beta=1.000e+00 | loss=  293.8774411 | loss_reconstruction=  285.1681564 | loss_kl=    8.7092834\n",
            "\n",
            "epoch=    63/   100 | beta=1.000e+00 | loss=  293.7481530 | loss_reconstruction=  284.9986988 | loss_kl=    8.7494542\n",
            "\n",
            "epoch=    64/   100 | beta=1.000e+00 | loss=  293.7230043 | loss_reconstruction=  285.0135011 | loss_kl=    8.7095044\n",
            "\n",
            "epoch=    65/   100 | beta=1.000e+00 | loss=  293.7162677 | loss_reconstruction=  284.9731689 | loss_kl=    8.7430993\n",
            "\n",
            "epoch=    66/   100 | beta=1.000e+00 | loss=  293.5301494 | loss_reconstruction=  284.7226478 | loss_kl=    8.8075007\n",
            "\n",
            "epoch=    67/   100 | beta=1.000e+00 | loss=  293.5638554 | loss_reconstruction=  284.7742313 | loss_kl=    8.7896232\n",
            "\n",
            "epoch=    68/   100 | beta=1.000e+00 | loss=  293.3756198 | loss_reconstruction=  284.5592668 | loss_kl=    8.8163531\n",
            "\n",
            "epoch=    69/   100 | beta=1.000e+00 | loss=  293.3139950 | loss_reconstruction=  284.4853210 | loss_kl=    8.8286730\n",
            "\n",
            "epoch=    70/   100 | beta=1.000e+00 | loss=  293.3082279 | loss_reconstruction=  284.5064509 | loss_kl=    8.8017780\n",
            "\n",
            "epoch=    71/   100 | beta=1.000e+00 | loss=  293.2914341 | loss_reconstruction=  284.4375724 | loss_kl=    8.8538615\n",
            "\n",
            "epoch=    72/   100 | beta=1.000e+00 | loss=  293.0906203 | loss_reconstruction=  284.2260349 | loss_kl=    8.8645844\n",
            "\n",
            "epoch=    73/   100 | beta=1.000e+00 | loss=  292.9612557 | loss_reconstruction=  284.0707255 | loss_kl=    8.8905311\n",
            "\n",
            "epoch=    74/   100 | beta=1.000e+00 | loss=  292.8743189 | loss_reconstruction=  284.0061094 | loss_kl=    8.8682105\n",
            "\n",
            "epoch=    75/   100 | beta=1.000e+00 | loss=  292.7925814 | loss_reconstruction=  283.8461901 | loss_kl=    8.9463900\n",
            "\n",
            "epoch=    76/   100 | beta=1.000e+00 | loss=  292.7500010 | loss_reconstruction=  283.8282708 | loss_kl=    8.9217304\n",
            "\n",
            "epoch=    77/   100 | beta=1.000e+00 | loss=  292.5401319 | loss_reconstruction=  283.6380771 | loss_kl=    8.9020556\n",
            "\n",
            "epoch=    78/   100 | beta=1.000e+00 | loss=  292.5070031 | loss_reconstruction=  283.5225613 | loss_kl=    8.9844405\n",
            "\n",
            "epoch=    79/   100 | beta=1.000e+00 | loss=  292.4896036 | loss_reconstruction=  283.4982624 | loss_kl=    8.9913411\n",
            "\n",
            "epoch=    80/   100 | beta=1.000e+00 | loss=  292.3921635 | loss_reconstruction=  283.4358741 | loss_kl=    8.9562902\n",
            "\n",
            "epoch=    81/   100 | beta=1.000e+00 | loss=  292.3177510 | loss_reconstruction=  283.2431592 | loss_kl=    9.0745903\n",
            "\n",
            "epoch=    82/   100 | beta=1.000e+00 | loss=  292.2365898 | loss_reconstruction=  283.2400983 | loss_kl=    8.9964926\n",
            "\n",
            "epoch=    83/   100 | beta=1.000e+00 | loss=  292.1996889 | loss_reconstruction=  283.1369652 | loss_kl=    9.0627250\n",
            "\n",
            "epoch=    84/   100 | beta=1.000e+00 | loss=  292.0746236 | loss_reconstruction=  283.0182304 | loss_kl=    9.0563944\n",
            "\n",
            "epoch=    85/   100 | beta=1.000e+00 | loss=  292.0743457 | loss_reconstruction=  283.0274752 | loss_kl=    9.0468689\n",
            "\n",
            "epoch=    86/   100 | beta=1.000e+00 | loss=  291.9578646 | loss_reconstruction=  282.8730157 | loss_kl=    9.0848484\n",
            "\n",
            "epoch=    87/   100 | beta=1.000e+00 | loss=  291.9268302 | loss_reconstruction=  282.8270906 | loss_kl=    9.0997391\n",
            "\n",
            "epoch=    88/   100 | beta=1.000e+00 | loss=  291.8140853 | loss_reconstruction=  282.7307580 | loss_kl=    9.0833276\n",
            "\n",
            "epoch=    89/   100 | beta=1.000e+00 | loss=  291.8173354 | loss_reconstruction=  282.6757689 | loss_kl=    9.1415660\n",
            "\n",
            "epoch=    90/   100 | beta=1.000e+00 | loss=  291.7772691 | loss_reconstruction=  282.6422635 | loss_kl=    9.1350064\n",
            "\n",
            "epoch=    91/   100 | beta=1.000e+00 | loss=  291.6686755 | loss_reconstruction=  282.5625883 | loss_kl=    9.1060865\n",
            "\n",
            "epoch=    92/   100 | beta=1.000e+00 | loss=  291.6776104 | loss_reconstruction=  282.5026680 | loss_kl=    9.1749432\n",
            "\n",
            "epoch=    93/   100 | beta=1.000e+00 | loss=  291.6538281 | loss_reconstruction=  282.5063477 | loss_kl=    9.1474812\n",
            "\n",
            "epoch=    94/   100 | beta=1.000e+00 | loss=  291.5867394 | loss_reconstruction=  282.4340798 | loss_kl=    9.1526589\n",
            "\n",
            "epoch=    95/   100 | beta=1.000e+00 | loss=  291.5936279 | loss_reconstruction=  282.4863470 | loss_kl=    9.1072799\n",
            "\n",
            "epoch=    96/   100 | beta=1.000e+00 | loss=  291.5540736 | loss_reconstruction=  282.3726674 | loss_kl=    9.1814060\n",
            "\n",
            "epoch=    97/   100 | beta=1.000e+00 | loss=  291.5382207 | loss_reconstruction=  282.3604830 | loss_kl=    9.1777369\n",
            "\n",
            "epoch=    98/   100 | beta=1.000e+00 | loss=  291.5310477 | loss_reconstruction=  282.3957763 | loss_kl=    9.1352724\n",
            "\n",
            "epoch=    99/   100 | beta=1.000e+00 | loss=  291.5212539 | loss_reconstruction=  282.4013584 | loss_kl=    9.1198976\n",
            "\n",
            "epoch=   100/   100 | beta=1.000e+00 | loss=  291.5308913 | loss_reconstruction=  282.3913227 | loss_kl=    9.1395681\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Etv2 motifs in Etv2 pioneer data"
      ],
      "metadata": {
        "id": "rMtJw9_E_6y1"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "gr0 <- lapply(ARGS$local_summit_files[c('MEF_NoDox','MEF_Dox_D7_Flk1pos','EB_NoDox_D25','EB_Dox_D25_Flk1pos')], function(bed_file){\n",
        "  peaks <- read.table(bed_file, header = FALSE, sep = '\\t')\n",
        "  peaks <- GRanges(seqnames = peaks[, 1], range = IRanges(start = peaks[, 2], end = peaks[, 3]))\n",
        "  peaks <- resize(peaks, fix = 'center', width = ARGS[['step_size']])\n",
        "  peaks\n",
        "}) %>%\n",
        "  GRangesList() %>%\n",
        "  coverage() %>%\n",
        "  IRanges::slice(lower = 1, rangesOnly = TRUE) %>%\n",
        "  GRanges() %>%\n",
        "  resize(fix = 'center', width = 800L)\n",
        "gr0 <- gr0[seqnames(gr0) %in% sprintf('chr%s', c(1:22, 'X'))]   "
      ],
      "metadata": {
        "id": "VfOtAtty_6y1"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "data('homer_pwms')"
      ],
      "metadata": {
        "id": "Q2nSKfk0_6y1"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "j <- 82 # Etv2\n",
        "homer_pwms[j]"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 53
        },
        "outputId": "e2cf39f3-1eb8-4e88-fde2-f8d1f0a2898f",
        "id": "jq7qEVo-_6y1"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "PWMatrixList of length 1\n",
              "names(1): Etv2(ETS)/ES-ER71-ChIP-Seq(GSE59402)/Homer(0.967)"
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "gr_etv2 <- matchMotifs(homer_pwms[j], gr0, genome = BSgenome.Mmusculus.UCSC.mm10, out = 'positions')[[1]] %>%\n",
        "  resize(width = ARGS$block_size, fix = 'center')"
      ],
      "metadata": {
        "id": "uAShIOl5_6y2"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "sprintf('number of motif-centric regions: %s', length(gr_etv2)) %>% message()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "428ed0a0-ca56-471e-907d-e467ab739481",
        "id": "oQ18FEeA_6y2"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "number of motif-centric regions: 32222\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Read summits for histone model\n",
        "\n"
      ],
      "metadata": {
        "id": "eYUtdQ55s7cx"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "x <- list()"
      ],
      "metadata": {
        "id": "wEHB-SsHpQje"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "for (i in 1:length(ARGS$local_bam_files)){\n",
        "  sprintf('reading %s', ARGS$local_summit_files[i]) %>% message()\n",
        "\n",
        "  if (names(ARGS$local_bam_files)[i] %in% c('MEF_NoDox','MEF_Dox_D7_Flk1pos','EB_NoDox_D25','EB_Dox_D25_Flk1pos')){\n",
        "    gr <- gr_etv2\n",
        "  }else{\n",
        "    gr0 <- read.table(ARGS$local_summit_files[i], header = FALSE, sep = '\\t')\n",
        "    gr0 <- GRanges(seqnames = gr0[, 1], range = IRanges(start = gr0[, 2], end = gr0[, 3]))\n",
        "    gr0 <- gr0[seqnames(gr0) %in% sprintf('chr%s', c(1:22, 'X'))]\n",
        "    gr0 <- resize(gr0, fix = 'center', width = ARGS$window_size)\n",
        "\n",
        "    gr <- slidingWindows(gr0, width = ARGS$block_size, step = ARGS$bin_size) %>% unlist()\n",
        "  \n",
        "    if (names(ARGS$local_bam_files)[i] %in% c(ARGS$dataset_train_histone)){ # data for training histone marks\n",
        "      gr <- sample(gr, min(length(gr), ARGS$max_peaks_train_histone))\n",
        "      counts <- count_reads(gr, ARGS$local_bam_files[i], genome = get_bsgenome(ARGS$genome_versions[i]),  fragment_size_range = c(0, ARGS$fragment_size_range_max))\n",
        "      valid <- counts >= ARGS$min_reads_train_histone\n",
        "    }else{ # data for testing histone marks\n",
        "      gr <- sample(gr, min(length(gr), ARGS$max_peaks_test_histone))\n",
        "      counts <- count_reads(gr, ARGS$local_bam_files[i], genome = get_bsgenome(ARGS$genome_versions[i]),  fragment_size_range = c(0, ARGS$fragment_size_range_max))\n",
        "      valid <- counts >= ARGS$min_reads_test_histone\n",
        "    }\n",
        "    gr <- gr[valid]\n",
        "  }\n",
        "\n",
        "  sprintf('total blocks for %s: %d', names(ARGS$local_bam_files)[i], length(gr)) %>% message()\n",
        "\n",
        "  x[[i]] <- read_vplot(\n",
        "    gr,\n",
        "    filenames = ARGS$local_bam_files[i], \n",
        "    genome = get_bsgenome(ARGS$genome_versions[i]),\n",
        "    bin_size = ARGS$bin_size,\n",
        "    fragment_size_range = c(0, ARGS$fragment_size_range_max),\n",
        "    fragment_size_interval = ARGS$fragment_size_interval \n",
        "  )\n",
        "}"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "EpI58cqzDOMH",
        "outputId": "9ec14a07-b8c0-4ff1-9ad8-6ee88b24608f"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "reading ./s3.msi.umn.edu/gongx030/datasets/dataset=Buenrostro_version=20170721a/GM12878_50k_cells_summits.bed\n",
            "\n",
            "read_bam | reading ./s3.msi.umn.edu/gongx030/datasets/dataset=Buenrostro_version=20170721a/GM12878_50k_cells.bam\n",
            "\n",
            "Warning message in NSBS(i, x, exact = exact, strict.upper.bound = !allow.append, :\n",
            "“subscript is an array, passing it thru as.vector() first”\n",
            "total blocks for GM12878: 49911\n",
            "\n",
            "read_bam | reading ./s3.msi.umn.edu/gongx030/datasets/dataset=Buenrostro_version=20170721a/GM12878_50k_cells.bam\n",
            "\n",
            "reading ./s3.msi.umn.edu/gongx030/datasets/dataset=K562_version=20230104a/K562_summits.bed\n",
            "\n",
            "read_bam | reading ./s3.msi.umn.edu/gongx030/datasets/dataset=K562_version=20230104a/K562.bam\n",
            "\n",
            "Warning message in NSBS(i, x, exact = exact, strict.upper.bound = !allow.append, :\n",
            "“subscript is an array, passing it thru as.vector() first”\n",
            "total blocks for K562: 19965\n",
            "\n",
            "read_bam | reading ./s3.msi.umn.edu/gongx030/datasets/dataset=K562_version=20230104a/K562.bam\n",
            "\n",
            "reading ./s3.msi.umn.edu/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/MEF_NoDox_summits.bed\n",
            "\n",
            "total blocks for MEF_NoDox: 32222\n",
            "\n",
            "read_bam | reading ./s3.msi.umn.edu/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/MEF_NoDox.bam\n",
            "\n",
            "reading ./s3.msi.umn.edu/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/MEF_Dox_D7_Flk1pos_summits.bed\n",
            "\n",
            "total blocks for MEF_Dox_D7_Flk1pos: 32222\n",
            "\n",
            "read_bam | reading ./s3.msi.umn.edu/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/MEF_Dox_D7_Flk1pos.bam\n",
            "\n",
            "reading ./s3.msi.umn.edu/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/EB_NoDox_D25_summits.bed\n",
            "\n",
            "total blocks for EB_NoDox_D25: 32222\n",
            "\n",
            "read_bam | reading ./s3.msi.umn.edu/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/EB_NoDox_D25.bam\n",
            "\n",
            "reading ./s3.msi.umn.edu/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/EB_Dox_D25_Flk1pos_summits.bed\n",
            "\n",
            "total blocks for EB_Dox_D25_Flk1pos: 32222\n",
            "\n",
            "read_bam | reading ./s3.msi.umn.edu/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/EB_Dox_D25_Flk1pos.bam\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "names(x) <- names(ARGS$local_bam_files)\n",
        "x <- VplotsList(x)"
      ],
      "metadata": {
        "id": "J7WfbpmTFdCd"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Get inputs for histone models"
      ],
      "metadata": {
        "id": "nvAmPknDDWVn"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "The inputs for predicting histone marks. For SeATC, the inputs are the latent representation. For MACS2 and NucleoATAC, the inputs are the PCA of the corresonding scaled signals."
      ],
      "metadata": {
        "id": "0Gex1Fk11xTk"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "### SeATAC"
      ],
      "metadata": {
        "id": "xgrfzixHJByf"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "x <- model %>% seatac::predict(x, batch_size = ARGS$batch_size, vplots = FALSE)"
      ],
      "metadata": {
        "id": "zthz7TRf5Y3m"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### MACS2"
      ],
      "metadata": {
        "id": "eIqRXks5JE6h"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "for (i in 1:length(ARGS$local_macs2_files)){\n",
        "  sprintf('reading %s', ARGS$local_macs2_files[i]) %>% message()\n",
        "  cvg <- rtracklayer::import(ARGS$local_macs2_files[i], which = rowRanges(x[[i]]))\n",
        "  seqlengths(cvg) <- seqlengths(cvg) + ARGS$block_size\n",
        "  cvg <- coverage(cvg, weight = 'score')\n",
        "  y <- cvg[rowRanges(x[[i]])] %>% as.matrix()\n",
        "  y <- (y - rowMins(y)) / (rowMaxs(y) - rowMins(y))\n",
        "  y[is.na(y) | is.infinite(y)] <- 0\n",
        "  rowData(x[[i]])$macs2 <- y\n",
        "}"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "fg7LUI_OKZHU",
        "outputId": "38f9e683-76ce-42fc-95b1-5706a0d5790f"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "reading ./s3.msi.umn.edu/gongx030/datasets/dataset=Buenrostro_version=20170721a/GM12878_50k_cells_treat_pileup.bw\n",
            "\n",
            "reading ./s3.msi.umn.edu/gongx030/datasets/dataset=K562_version=20230104a/K562_treat_pileup.bw\n",
            "\n",
            "reading ./s3.msi.umn.edu/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/MEF_NoDox_treat_pileup.bw\n",
            "\n",
            "reading ./s3.msi.umn.edu/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/MEF_Dox_D7_Flk1pos_treat_pileup.bw\n",
            "\n",
            "reading ./s3.msi.umn.edu/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/EB_NoDox_D25_treat_pileup.bw\n",
            "\n",
            "reading ./s3.msi.umn.edu/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/EB_Dox_D25_Flk1pos_treat_pileup.bw\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "u <- NULL\n",
        "gs <- NULL\n",
        "for (i in 1:length(ARGS$local_macs2_files)){\n",
        "  u <- rbind(u, rowData(x[[i]])$macs2)\n",
        "  gs <- c(gs, rep(i, nrow(rowData(x[[i]]))))\n",
        "}\n",
        "u <- prcomp_irlba(u, n = ARGS$latent_dim)$x\n",
        "h <- split(1:nrow(u), factor(gs))"
      ],
      "metadata": {
        "id": "EWXDdPSzuhv6"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "for (i in 1:length(ARGS$local_macs2_files)){\n",
        "  rowData(x[[i]])$macs2_pca <- u[h[[i]], ]\n",
        "}"
      ],
      "metadata": {
        "id": "qG7Nmz6qu0DJ"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### NucleoATAC"
      ],
      "metadata": {
        "id": "kpVkqenDJG_Y"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "for (i in 1:length(ARGS$local_nucleoatac_files)){\n",
        "  sprintf('reading %s', ARGS$local_nucleoatac_files[i]) %>% message()\n",
        "  cvg <- rtracklayer::import(ARGS$local_nucleoatac_files[i], which = rowRanges(x[[i]]))\n",
        "  seqlengths(cvg) <- seqlengths(cvg) + ARGS$block_size\n",
        "  cvg <- coverage(cvg, weight = 'score')\n",
        "  y <- cvg[rowRanges(x[[i]])] %>% as.matrix()\n",
        "  y <- (y - rowMins(y)) / (rowMaxs(y) - rowMins(y))\n",
        "  y[is.na(y) | is.infinite(y)] <- 0\n",
        "  rowData(x[[i]])$nucleoatac <- y\n",
        "}"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "6b4KTop9SFXf",
        "outputId": "2a391896-91cf-4219-af9f-9abfdf01e625"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "reading ./s3.msi.umn.edu/gongx030/datasets/dataset=Buenrostro_version=20170721a/GM12878_50k_cells.nucleoatac_signal.smooth.bw\n",
            "\n",
            "reading ./s3.msi.umn.edu/gongx030/datasets/dataset=K562_version=20230104a/K562.nucleoatac_signal.smooth.bw\n",
            "\n",
            "reading ./s3.msi.umn.edu/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/MEF_NoDox.nucleoatac_signal.smooth.bw\n",
            "\n",
            "reading ./s3.msi.umn.edu/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/MEF_Dox_D7_Flk1pos.nucleoatac_signal.smooth.bw\n",
            "\n",
            "reading ./s3.msi.umn.edu/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/EB_NoDox_D25.nucleoatac_signal.smooth.bw\n",
            "\n",
            "reading ./s3.msi.umn.edu/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/EB_Dox_D25_Flk1pos.nucleoatac_signal.smooth.bw\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "u <- NULL\n",
        "gs <- NULL\n",
        "for (i in 1:length(ARGS$local_nucleoatac_files)){\n",
        "  u <- rbind(u, rowData(x[[i]])$nucleoatac)\n",
        "  gs <- c(gs, rep(i, nrow(rowData(x[[i]]))))\n",
        "}\n",
        "u <- prcomp_irlba(u, n = ARGS$latent_dim)$x\n",
        "h <- split(1:nrow(u), factor(gs))"
      ],
      "metadata": {
        "id": "yOXRmNJNwJTO"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "for (i in 1:length(ARGS$local_nucleoatac_files)){\n",
        "  rowData(x[[i]])$nucleoatac_pca <- u[h[[i]], ]\n",
        "}"
      ],
      "metadata": {
        "id": "8CjhICdhwJTP"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Reading training histone data"
      ],
      "metadata": {
        "id": "kZrDPv0oHeQT"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "y_train <- NULL\n",
        "for (i in ARGS$dataset_train_histone){\n",
        "  y <- NULL\n",
        "  for (j in c('H3K27ac', 'H3K4me1', 'H3K4me3')){\n",
        "    sprintf('reading %s', ARGS$local_bw_files[[i]][j]) %>% message()\n",
        "    cvg <- rtracklayer::import(ARGS$local_bw_files[[i]][j], which = rowRanges(x[[i]]))\n",
        "    seqlengths(cvg) <- seqlengths(cvg) + ARGS$block_size\n",
        "    cvg <- coverage(cvg, weight = 'score')\n",
        "    yij <- cvg[rowRanges(x[[i]])] %>% as.matrix()\n",
        "    y <- c(y, yij %>% tf$cast(tf$float32) %>% tf$expand_dims(2L))\n",
        "  }\n",
        "  y <- tf$concat(y, axis = 2L)\n",
        "  y_train <- c(y_train, y)\n",
        "}"
      ],
      "metadata": {
        "id": "M02dTDROtJ2I",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "0b678c59-dcea-4845-fee9-e01c50ed0a5a"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "reading ./www.encodeproject.org/files/ENCFF440GZA/@@download/ENCFF440GZA.bigWig\n",
            "\n",
            "reading ./www.encodeproject.org/files/ENCFF372EXQ/@@download/ENCFF372EXQ.bigWig\n",
            "\n",
            "reading ./www.encodeproject.org/files/ENCFF154XCY/@@download/ENCFF154XCY.bigWig\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "y_train <- tf$concat(y_train, axis = 0L)\n",
        "y_train <- tfa$image$gaussian_filter2d(y_train, filter_shape = shape(1L, 20L), sigma = 1.0, padding = 'REFLECT')\n",
        "\n",
        "include_train_histone <- tf$reduce_sum(tf$cast(tf$reduce_sum(y_train, shape(1L)) > ARGS$min_histone_signal_train, tf$float32), axis = 1L) == 3L\n",
        "\n",
        "y_train <- (y_train - tf$reduce_min(y_train, axis = 1L, keepdims = TRUE)) / (tf$reduce_max(y_train, axis = 1L, keepdims = TRUE) - tf$reduce_min(y_train, axis = 1L, keepdims = TRUE))\n",
        "is_nan <- tf$math$is_nan(y_train)\n",
        "y_train <- tf$where(is_nan, tf$zeros_like(is_nan, dtype = tf$float32), y_train)\n",
        "y_train <- tf$boolean_mask(y_train, include_train_histone)"
      ],
      "metadata": {
        "id": "0KUmsfeg897c"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "table(as.numeric(include_train_histone))"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 71
        },
        "id": "ZFyZyaPRMYMd",
        "outputId": "947dd775-8e8a-47dd-ffe7-42f2a4afbbe1"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "\n",
              "    0     1 \n",
              "14546 35365 "
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Fitting histone model"
      ],
      "metadata": {
        "id": "Gtihesk6kgZw"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "df_losses <- NULL"
      ],
      "metadata": {
        "id": "nbDTZvG_Dc6X"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### MACS2"
      ],
      "metadata": {
        "id": "1gfxi3qWUaQQ"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "#### Training on GM12878"
      ],
      "metadata": {
        "id": "EPBOvWx17t4h"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "x_train <- NULL\n",
        "for (i in ARGS$dataset_train_histone){\n",
        "  x_train <- c(x_train, tf$cast(rowData(x[[i]])$macs2_pca, tf$float32))\n",
        "}\n",
        "x_train <- tf$concat(x_train, axis = 0L)\n",
        "x_train <- tf$boolean_mask(x_train, include_train_histone)"
      ],
      "metadata": {
        "id": "ESHDbpfF6C3a"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "tf$keras$backend$clear_session()\n",
        "\n",
        "model_histone <- keras_model_sequential(input_shape = c(ARGS$latent_dim)) %>%\n",
        "  layer_dense(128L, activation = \"relu\") %>%\n",
        "  layer_dropout(0.1) %>%\n",
        "  layer_dense(256L, activation = \"relu\") %>%\n",
        "  layer_dropout(0.1) %>%\n",
        "  layer_dense(512L, activation = \"relu\") %>%\n",
        "  layer_dropout(0.1) %>%\n",
        "  layer_dense(ARGS$block_size * 3L, activation = \"sigmoid\") %>%\n",
        "  layer_reshape(c(ARGS$block_size, 3L)) \n",
        "\n",
        "model_histone %>% compile(\n",
        "  optimizer = \"adam\",\n",
        "  loss = 'mse'\n",
        ")\n",
        "\n",
        "history <- model_histone %>% keras::fit(\n",
        "  x_train, \n",
        "  y_train,\n",
        "  epochs = 50L, \n",
        "  batch_size = ARGS$batch_size, \n",
        "  validation_split = 0.2,\n",
        "  callbacks = tf$keras$callbacks$EarlyStopping(monitor='val_loss', patience=5)\n",
        ")\n",
        "\n",
        "history"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 88
        },
        "id": "bbzffhnNVnI3",
        "outputId": "092debe5-2e44-42df-8716-6e67caf4ac04"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "\n",
              "Final epoch (plot to see history):\n",
              "    loss: 0.09515\n",
              "val_loss: 0.09617 "
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "#### Evaluation"
      ],
      "metadata": {
        "id": "xdLd-Z4l7nmy"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "for (i in which(!names(ARGS$remote_bam_files) %in% ARGS$dataset_train_histone)){\n",
        "\n",
        "  x_test <- rowData(x[[i]])$macs2_pca %>% tf$cast(tf$float32)\n",
        "\n",
        "  y_test <- NULL\n",
        "  for (j in 1:3){ # H3K27ac, H3K4me1, H3K4me3\n",
        "    if (!is.na(ARGS$local_bw_files[[i]][j])){\n",
        "      sprintf('reading %s', ARGS$local_bw_files[[i]][j]) %>% message()\n",
        "      cvg <- rtracklayer::import(ARGS$local_bw_files[[i]][j], which = rowRanges(x[[i]]))\n",
        "      seqlengths(cvg) <- seqlengths(cvg) + ARGS$block_size\n",
        "      cvg <- coverage(cvg, weight = 'score')\n",
        "      y_test <- cvg[rowRanges(x[[i]])] %>% as.matrix()\n",
        "      y_test <- tf$cast(y_test, tf$float32)\n",
        "      y_test <- tfa$image$gaussian_filter2d(y_test, filter_shape = shape(1L, 20L), sigma = 1.0, padding = 'REFLECT')\n",
        "\n",
        "      include <- tf$reduce_sum(tf$cast(y_test > ARGS$min_histone_signal_test, tf$float32), axis = c(1L)) > 0\n",
        "      sprintf('%d/%d', sum(as.numeric(include)), length(as.numeric(include))) %>% message()\n",
        "\n",
        "      y_test <- (y_test - tf$reduce_min(y_test, axis = 1L, keepdims = TRUE)) / (tf$reduce_max(y_test, axis = 1L, keepdims = TRUE) - tf$reduce_min(y_test, axis = 1L, keepdims = TRUE))\n",
        "      is_nan <- tf$math$is_nan(y_test)\n",
        "      y_test <- tf$where(is_nan, tf$zeros_like(is_nan, dtype = tf$float32), y_test)\n",
        "      y_pred <- model_histone(x_test)[, , j] \n",
        "      df2 <- data.frame(\n",
        "        method = 'macs2',\n",
        "        dataset = names(ARGS$remote_bam_files)[i],\n",
        "        type = names(ARGS$local_bw_files[[i]][j]),\n",
        "        loss = tf$keras$metrics$mse(tf$boolean_mask(y_test, include), tf$boolean_mask(y_pred, include)) %>% tf$reduce_mean() %>% as.numeric()\n",
        "      )\n",
        "      print(df2)\n",
        "      df_losses <- rbind(df_losses, df2)\n",
        "    }\n",
        "  }\n",
        "}"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "oiEi325Duc3m",
        "outputId": "44aa43b5-060f-4fcc-e8a4-15cf7392b036"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "reading ./www.encodeproject.org/files/ENCFF488FYZ/@@download/ENCFF488FYZ.bigWig\n",
            "\n",
            "2209/19965\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "  method dataset    type       loss\n",
            "1  macs2    K562 H3K27ac 0.08625094\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "reading ./www.encodeproject.org/files/ENCFF700ZWZ/@@download/ENCFF700ZWZ.bigWig\n",
            "\n",
            "1225/19965\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "  method dataset    type       loss\n",
            "1  macs2    K562 H3K4me1 0.07906256\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "reading ./www.encodeproject.org/files/ENCFF285EUA/@@download/ENCFF285EUA.bigWig\n",
            "\n",
            "2722/19965\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "  method dataset    type      loss\n",
            "1  macs2    K562 H3K4me3 0.1054259\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "reading ./s3.msi.umn.edu/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/MEF_NoDox_d0_H3K27ac_FE.bw\n",
            "\n",
            "5343/32222\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "  method   dataset    type       loss\n",
            "1  macs2 MEF_NoDox H3K27ac 0.08403791\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "reading ./s3.msi.umn.edu/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/MEF_Dox_d7_H3K27ac_FE.bw\n",
            "\n",
            "7858/32222\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "  method            dataset    type       loss\n",
            "1  macs2 MEF_Dox_D7_Flk1pos H3K27ac 0.08560304\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "reading ./s3.msi.umn.edu/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/EB_NoDox_12h_H3K27ac_FE.bw\n",
            "\n",
            "1201/32222\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "  method      dataset    type       loss\n",
            "1  macs2 EB_NoDox_D25 H3K27ac 0.08401173\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "reading ./s3.msi.umn.edu/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/EB_Dox_12h_H3K27ac_FE.bw\n",
            "\n",
            "3982/32222\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "  method            dataset    type      loss\n",
            "1  macs2 EB_Dox_D25_Flk1pos H3K27ac 0.1011173\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### NucleoATAC"
      ],
      "metadata": {
        "id": "Sfa2Oh1_V105"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "#### Training"
      ],
      "metadata": {
        "id": "TpmPMYEB8C2o"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "x_train <- NULL\n",
        "for (i in ARGS$dataset_train_histone){\n",
        "  x_train <- c(x_train, tf$cast(rowData(x[[i]])$nucleoatac_pca, tf$float32))\n",
        "}\n",
        "x_train <- tf$concat(x_train, axis = 0L)\n",
        "x_train <- tf$boolean_mask(x_train, include_train_histone)"
      ],
      "metadata": {
        "id": "QzlNxNOsV106"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "tf$keras$backend$clear_session()\n",
        "\n",
        "model_histone <- keras_model_sequential(input_shape = c(ARGS$latent_dim)) %>%\n",
        "  layer_dense(128L, activation = \"relu\") %>%\n",
        "  layer_dropout(0.1) %>%\n",
        "  layer_dense(256L, activation = \"relu\") %>%\n",
        "  layer_dropout(0.1) %>%\n",
        "  layer_dense(512L, activation = \"relu\") %>%\n",
        "  layer_dropout(0.1) %>%\n",
        "  layer_dense(ARGS$block_size * 3L, activation = \"sigmoid\") %>%\n",
        "  layer_reshape(c(ARGS$block_size, 3L)) \n",
        "\n",
        "model_histone %>% compile(\n",
        "  optimizer = \"adam\",\n",
        "  loss = 'mse'\n",
        ")\n",
        "\n",
        "history <- model_histone %>% keras::fit(\n",
        "  x_train, \n",
        "  y_train,\n",
        "  epochs = 50L, \n",
        "  batch_size = ARGS$batch_size, \n",
        "  validation_split = 0.2,\n",
        "  callbacks = tf$keras$callbacks$EarlyStopping(monitor='val_loss', patience=5)\n",
        ")\n",
        "\n",
        "history"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 88
        },
        "id": "CHnTbJM4iwpr",
        "outputId": "bca1dccf-fdaf-459d-ad57-96a18080c48e"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "\n",
              "Final epoch (plot to see history):\n",
              "    loss: 0.09688\n",
              "val_loss: 0.09798 "
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "#### Evaluation"
      ],
      "metadata": {
        "id": "0JTHCueOu1lo"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "for (i in which(!names(ARGS$remote_bam_files) %in% ARGS$dataset_train_histone)){\n",
        "\n",
        "  x_test <- rowData(x[[i]])$nucleoatac_pca %>% tf$cast(tf$float32)\n",
        "\n",
        "  y_test <- NULL\n",
        "  for (j in 1:3){ # H3K27ac, H3K4me1, H3K4me3\n",
        "    if (!is.na(ARGS$local_bw_files[[i]][j])){\n",
        "      sprintf('reading %s', ARGS$local_bw_files[[i]][j]) %>% message()\n",
        "      cvg <- rtracklayer::import(ARGS$local_bw_files[[i]][j], which = rowRanges(x[[i]]))\n",
        "      seqlengths(cvg) <- seqlengths(cvg) + ARGS$block_size\n",
        "      cvg <- coverage(cvg, weight = 'score')\n",
        "      y_test <- cvg[rowRanges(x[[i]])] %>% as.matrix()\n",
        "      y_test <- tf$cast(y_test, tf$float32)\n",
        "      y_test <- tfa$image$gaussian_filter2d(y_test, filter_shape = shape(1L, 20L), sigma = 1.0, padding = 'REFLECT')\n",
        "\n",
        "      include <- tf$reduce_sum(tf$cast(y_test > ARGS$min_histone_signal_test, tf$float32), axis = c(1L)) > 0\n",
        "      sprintf('%d/%d', sum(as.numeric(include)), length(as.numeric(include))) %>% message()\n",
        "\n",
        "      y_test <- (y_test - tf$reduce_min(y_test, axis = 1L, keepdims = TRUE)) / (tf$reduce_max(y_test, axis = 1L, keepdims = TRUE) - tf$reduce_min(y_test, axis = 1L, keepdims = TRUE))\n",
        "      is_nan <- tf$math$is_nan(y_test)\n",
        "      y_test <- tf$where(is_nan, tf$zeros_like(is_nan, dtype = tf$float32), y_test)\n",
        "      y_pred <- model_histone(x_test)[, , j] \n",
        "      include <- tf$reduce_sum(y_test, axis = c(1L)) > 0 # testing samples which are not all are zeros\n",
        "      df2 <- data.frame(\n",
        "        method = 'nucleoatac',\n",
        "        dataset = names(ARGS$remote_bam_files)[i],\n",
        "        type = names(ARGS$local_bw_files[[i]][j]),\n",
        "        loss = tf$keras$metrics$mse(tf$boolean_mask(y_test, include), tf$boolean_mask(y_pred, include)) %>% tf$reduce_mean() %>% as.numeric()\n",
        "      )\n",
        "      print(df2)\n",
        "      df_losses <- rbind(df_losses, df2)\n",
        "    }\n",
        "  }\n",
        "}"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "43d80663-a50f-436f-8927-f9e25f202a29",
        "id": "E8_ETsl9u1lo"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "reading ./www.encodeproject.org/files/ENCFF488FYZ/@@download/ENCFF488FYZ.bigWig\n",
            "\n",
            "2209/19965\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "      method dataset    type     loss\n",
            "1 nucleoatac    K562 H3K27ac 0.134843\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "reading ./www.encodeproject.org/files/ENCFF700ZWZ/@@download/ENCFF700ZWZ.bigWig\n",
            "\n",
            "1225/19965\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "      method dataset    type      loss\n",
            "1 nucleoatac    K562 H3K4me1 0.1157217\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "reading ./www.encodeproject.org/files/ENCFF285EUA/@@download/ENCFF285EUA.bigWig\n",
            "\n",
            "2722/19965\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "      method dataset    type      loss\n",
            "1 nucleoatac    K562 H3K4me3 0.0919894\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "reading ./s3.msi.umn.edu/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/MEF_NoDox_d0_H3K27ac_FE.bw\n",
            "\n",
            "5343/32222\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "      method   dataset    type       loss\n",
            "1 nucleoatac MEF_NoDox H3K27ac 0.08769004\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "reading ./s3.msi.umn.edu/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/MEF_Dox_d7_H3K27ac_FE.bw\n",
            "\n",
            "7858/32222\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "      method            dataset    type       loss\n",
            "1 nucleoatac MEF_Dox_D7_Flk1pos H3K27ac 0.09001515\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "reading ./s3.msi.umn.edu/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/EB_NoDox_12h_H3K27ac_FE.bw\n",
            "\n",
            "1201/32222\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "      method      dataset    type       loss\n",
            "1 nucleoatac EB_NoDox_D25 H3K27ac 0.08276749\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "reading ./s3.msi.umn.edu/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/EB_Dox_12h_H3K27ac_FE.bw\n",
            "\n",
            "3982/32222\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "      method            dataset    type      loss\n",
            "1 nucleoatac EB_Dox_D25_Flk1pos H3K27ac 0.1406597\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### SeATAC"
      ],
      "metadata": {
        "id": "Q1tjdiElHKwz"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "#### Training"
      ],
      "metadata": {
        "id": "9m7eWnLX6rSW"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "x_train <- NULL\n",
        "for (i in ARGS$dataset_train_histone){\n",
        "  x_train <- c(x_train, tf$cast(rowData(x[[i]])$vae_z_mean[, 1, ], tf$float32))\n",
        "}\n",
        "x_train <- tf$concat(x_train, axis = 0L)\n",
        "x_train <- tf$boolean_mask(x_train, include_train_histone)"
      ],
      "metadata": {
        "id": "DJcfPeT6HmKM"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "tf$keras$backend$clear_session()\n",
        "\n",
        "model_histone <- keras_model_sequential(input_shape = c(ARGS$latent_dim)) %>%\n",
        "  layer_dense(128L, activation = \"relu\") %>%\n",
        "  layer_dropout(0.1) %>%\n",
        "  layer_dense(256L, activation = \"relu\") %>%\n",
        "  layer_dropout(0.1) %>%\n",
        "  layer_dense(512L, activation = \"relu\") %>%\n",
        "  layer_dropout(0.1) %>%\n",
        "  layer_dense(ARGS$block_size * 3L, activation = \"sigmoid\") %>%\n",
        "  layer_reshape(c(ARGS$block_size, 3L)) \n",
        "\n",
        "model_histone %>% compile(\n",
        "  optimizer = \"adam\",\n",
        "  loss = 'mse'\n",
        ")\n",
        "\n",
        "history <- model_histone %>% keras::fit(\n",
        "  x_train, \n",
        "  y_train,\n",
        "  epochs = 50L, \n",
        "  batch_size = ARGS$batch_size, \n",
        "  validation_split = 0.2,\n",
        "  callbacks = tf$keras$callbacks$EarlyStopping(monitor='val_loss', patience=5)\n",
        ")\n",
        "\n",
        "history"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "5RZcPgSlgdfK",
        "outputId": "0eb0b886-4949-44f1-c4cd-2492a9b4607a"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "\n",
              "Final epoch (plot to see history):\n",
              "    loss: 0.09207\n",
              "val_loss: 0.09292 "
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "#### Evaluation"
      ],
      "metadata": {
        "id": "wuiwp62xsfIS"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "for (i in which(!names(ARGS$remote_bam_files) %in% ARGS$dataset_train_histone)){\n",
        "\n",
        "  x_test <- tf$cast(rowData(x[[i]])$vae_z_mean[, 1, ], tf$float32)\n",
        "\n",
        "  y_test <- NULL\n",
        "  for (j in 1:3){ # H3K27ac, H3K4me1, H3K4me3\n",
        "    if (!is.na(ARGS$local_bw_files[[i]][j])){\n",
        "      sprintf('reading %s', ARGS$local_bw_files[[i]][j]) %>% message()\n",
        "      cvg <- rtracklayer::import(ARGS$local_bw_files[[i]][j], which = rowRanges(x[[i]]))\n",
        "      seqlengths(cvg) <- seqlengths(cvg) + ARGS$block_size\n",
        "      cvg <- coverage(cvg, weight = 'score')\n",
        "      y_test <- cvg[rowRanges(x[[i]])] %>% as.matrix()\n",
        "      y_test <- tf$cast(y_test, tf$float32)\n",
        "      y_test <- tfa$image$gaussian_filter2d(y_test, filter_shape = shape(1L, 20L), sigma = 1.0, padding = 'REFLECT')\n",
        "\n",
        "      include <- tf$reduce_sum(tf$cast(y_test > ARGS$min_histone_signal_test, tf$float32), axis = c(1L)) > 0\n",
        "      sprintf('%d/%d', sum(as.numeric(include)), length(as.numeric(include))) %>% message()\n",
        "\n",
        "      y_test <- (y_test - tf$reduce_min(y_test, axis = 1L, keepdims = TRUE)) / (tf$reduce_max(y_test, axis = 1L, keepdims = TRUE) - tf$reduce_min(y_test, axis = 1L, keepdims = TRUE))\n",
        "      is_nan <- tf$math$is_nan(y_test)\n",
        "      y_test <- tf$where(is_nan, tf$zeros_like(is_nan, dtype = tf$float32), y_test)\n",
        "      y_pred <- model_histone(x_test)[, , j] \n",
        "      df2 <- data.frame(\n",
        "        method = 'seatac',\n",
        "        dataset = names(ARGS$remote_bam_files)[i],\n",
        "        type = names(ARGS$local_bw_files[[i]][j]),\n",
        "        loss = tf$keras$metrics$mse(tf$boolean_mask(y_test, include), tf$boolean_mask(y_pred, include)) %>% tf$reduce_mean() %>% as.numeric()\n",
        "      )\n",
        "      print(df2)\n",
        "      df_losses <- rbind(df_losses, df2)\n",
        "    }\n",
        "  }\n",
        "}"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "1hoT7lH9siCi",
        "outputId": "1c140da4-3b5f-4595-911f-87136eeef973"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "reading ./www.encodeproject.org/files/ENCFF488FYZ/@@download/ENCFF488FYZ.bigWig\n",
            "\n",
            "2209/19965\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "  method dataset    type       loss\n",
            "1 seatac    K562 H3K27ac 0.07571367\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "reading ./www.encodeproject.org/files/ENCFF700ZWZ/@@download/ENCFF700ZWZ.bigWig\n",
            "\n",
            "1225/19965\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "  method dataset    type       loss\n",
            "1 seatac    K562 H3K4me1 0.07753807\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "reading ./www.encodeproject.org/files/ENCFF285EUA/@@download/ENCFF285EUA.bigWig\n",
            "\n",
            "2722/19965\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "  method dataset    type       loss\n",
            "1 seatac    K562 H3K4me3 0.09398346\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "reading ./s3.msi.umn.edu/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/MEF_NoDox_d0_H3K27ac_FE.bw\n",
            "\n",
            "5343/32222\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "  method   dataset    type       loss\n",
            "1 seatac MEF_NoDox H3K27ac 0.06973742\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "reading ./s3.msi.umn.edu/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/MEF_Dox_d7_H3K27ac_FE.bw\n",
            "\n",
            "7858/32222\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "  method            dataset    type       loss\n",
            "1 seatac MEF_Dox_D7_Flk1pos H3K27ac 0.08203718\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "reading ./s3.msi.umn.edu/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/EB_NoDox_12h_H3K27ac_FE.bw\n",
            "\n",
            "1201/32222\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "  method      dataset    type       loss\n",
            "1 seatac EB_NoDox_D25 H3K27ac 0.06498452\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "reading ./s3.msi.umn.edu/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/EB_Dox_12h_H3K27ac_FE.bw\n",
            "\n",
            "3982/32222\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "  method            dataset    type      loss\n",
            "1 seatac EB_Dox_D25_Flk1pos H3K27ac 0.1002154\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Plot the testing losses"
      ],
      "metadata": {
        "id": "N7Ehi3PeH4ID"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "options(repr.plot.width = 12, repr.plot.height = 5)\n",
        "df_losses %>%\n",
        "  mutate(method = factor(method, c('macs2', 'nucleoatac', 'seatac'))) %>%\n",
        "  mutate(label = sprintf('%s_%s', dataset, type)) %>%\n",
        "\n",
        "  ggplot(aes(x = label, y = loss, fill = method)) +\n",
        "    geom_bar(stat = 'identity', position=position_dodge()) +\n",
        "    scale_fill_manual(values= c(macs2 = 'lightcoral', nucleoatac = 'purple', seatac = 'cyan3')) +\n",
        "    coord_flip()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 317
        },
        "id": "-JOlyUa4H7i4",
        "outputId": "40d42560-3a43-4143-9a48-93d407c306d1"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "plot without title"
            ],
            "image/png": "iVBORw0KGgoAAAANSUhEUgAABaAAAAJYCAIAAAAbpNZlAAAACXBIWXMAABJ0AAASdAHeZh94\nAAAgAElEQVR4nOzdd0BV5R/H8e+Fy95DVKYi4kBwmzsrZ+bWcubelSPJhYriNlPcSWXOMLVM\nc49SHL8y9yz3IBRFQRCBC/f+/lAR2RJ4Ofp+/cU55znP+Z5z/QX383ue56h0Op0AAAAAAAAo\nmYG+CwAAAAAAAPivCDgAAAAAAIDiEXAAAAAAAADFI+AAAAAAAACKR8ABAAAAAAAUj4ADAAAA\nAAAoHgEHAAAAAABQPAIOAAAAAACgeGp9FwAUUrGxscnJyfquQszMzIyNjePi4lJSUvRdCwqW\niYmJTqdLSkrSdyEoWGq12sLCIiEhITExUd+1oGAZGhqamJjEx8fruxAUOBsbm+Tk5EePHum7\nEBQ4Kyur2NhYfVeBPLKzs9N3CShwBBxA5rRabSHJFAwMDApPMShQOp2OD/q1Z2hoaGBgICJ8\n1q89lUolfNBvBgMDA5VKxWf9JjAwMOCDBgozpqgAAAAAAADFI+AAAAAAAACKR8ABAAAAAAAU\nj4ADAAAAAAAoHgEHAAAAAABQPAIOAAAAAACgeAQcAAAAAABA8Qg4AAAAAACA4hFwAAAAAAAA\nxVPruwAAOVAdO6bvEgDgdXPX3V3fJQAAgHzGCA4AAAAAAKB4BBwAAAAAAEDxCDgAAAAAAIDi\nEXAAAAAAAADFI+AAAAAAAACKR8ABAAAAAAAUj4ADAAAAAAAoHgEHAAAAAABQPAIOAAAAAACg\neAQcAAAAAABA8Qg4AAAAAACA4hFwAAAAAAAAxSPgAAAAAAAAikfAAQAAAAAAFI+AAwAAAAAA\nKB4BBwAAAAAAUDwCDgAAAAAAoHgEHAAAAAAAQPEIOAAAAAAAgOIRcAAAAAAAAMUj4AAAAAAA\nAIpHwAEAAAAAABRPre8CAOTg+3al9F0CALxutryi6ySIGL2iSxUmzY9o9F0CAOBNxAgOAAAA\nAACgeAQcAAAAAABA8Qg4AAAAAACA4hFwAAAAAAAAxSPgAAAAAAAAisdbVF7CuZCeozZHqUr3\nCJnd1il1b8T6oQNWXNE5tJ61rFeZ581ePNWp3VffdPfK5JBV06DVgypmczmrOv5LRtazerrv\n0nd9hv+v3ldLu3vlplQREZXaxMzSrri7t1+txs0b+znl/1ru50J6jjpUe9ayvmWe77uyvN/Q\nnZWe3trtsKVLNx2/HH4vNtHQysG1fL02XTvWcTHK5FxdzMnvJ0779XHN4YGf1imacHH3hs37\njp67FhGdaGLrXLpm886dGnlZiIhErh/eZ8WlF+tQVRkaGviuWb7fHwAAAABAAQg4XpKtU5Fr\n23dcadPNUyUiIrp/duy8VcTJJjLlhWY2bw+b2Mbj+baRnWvmhwytnbO7nLFx4sHloWff6utj\n/PKlvus/tX1JnTbpcUzk1TNhW1cFbNvZNmByDz+rnM/NVzpDu/INu37g7GBprIm6tG/t8plj\nHwaGDK6cLmvRhO+dE7jgpEO7oIldyluJSORfe8+IT5PurdyLmCdcP/TjivkB15ODg5oVFbFv\n9PncKompp0bumDF1j0u9GqQbAAAAAPCmIuB4SRa1G1XavXX7qc6DKhqKSPLJ7Xti637Q6Nzq\nsBeaqa2dPT09M+0hm0MZFWnYpkTY+m83vT+7vYsqw1Hdg5Prvl258+jVqCQTx1LVmnXr3cbX\nJrWZoYWTq+uTXMXT27fme3U9xwxbPmdVtaUDKxhlfW7c8XlDA0/6BgQPqW4pIol/fzd81O9u\nw4NH1bPLZc0ZFK/doX3qRhlf8+uHx506HS6VS6RpE3fhxylBP9yrOHjmsIYuT5OPkp2mzkxt\nULa8d/LVbl8fPBbbrJmVqG1cPG1SH8OlsOMRVvX71rMUEZEHx374Zv3B89fvPNSYObiVf6dz\nrw5VnQyftk249tvqlb8cPHszOtnU1sWr/sfDe1RN7QgAAAAAoFiswfGyjGs0ecdg3/Y/EkRE\n4g9vDzN6t2nVlx9ekVuWdXt0LH193bLfojMc0t3cMDlw3XXPjgFzF84Z3d710prAqZtvZ9mT\n2q1V+7qmUQfCzumyO9ey8oAv2pn9Pnfu3nsicUeWzNyU1PDzz/KebrxAm/jgyt7dJ+IcK/ql\nGbiijTy0cGTAhqT3xn/pn5puZKTRJOmMbWzMM+w/unX37aLvNXv2KSTEG5Rq1m/s9AWL5gT0\nqBa/dfLUDTe0IiKiC9888Yt5h1V1+06cu2j+VP+O1R1V2ny5MQAAAACAnjGC46UZ+DRt5PjZ\n9n3RtZvIb9v/KNY4uKzB/vSNojb7t9yc5px6Yzb618zkkFfPb75q45T+7BcUfb9vy20jVq46\nWeeTiiZp9utO/Lz+onPreQMbllCJuDl/Mvz2+eFrN5xuOdg3i57UJTxcZW/E7Tjxu5TNucbe\nXUd2Pzd8yczvI6x3/W7efkbfihlChRzuV0REXpgJo9k3ucOcIzqtTmfq0XjEtH6VnmdC0Tvm\nzEhxbjNzao8yJpIV7e3di9Zdcm/Tr7phuiNxB7fuf+jVqWnpZ0NXitf9qO3TH4s6dfo86lS3\nbWE3PuxSQrQnNvxw1q7FrNHtyxiKiBQr7lEubU9//fVXcHBw6qa/v3/58uVzuvMCZ2BgIJKY\nczsAAAoNW1tbfZfwqqnV6jfwrt9ABgYGfNBAYUbAkQdujZuWXbdj9+0KsuOcd9Ph7hKToUm6\nNThUlsUyP2TiYJ/j5QxLd+j19p6gb39qHtwpze6oGzfizX0qlHj2vd7Aq0I50w3Xrz8Q36yG\nW+h0OhGVqHI618Ct1Yg+xwcvXJtStueczt65GJ+Sfs2R8F8nz/ojbQN19f7BwV00cZEXD21Y\nNX+atf30j31Mnxyy9Kld4lrYrhVb6ga09cp0FQ3tnbC545ZcrzxsSufS6Qd4RO7eelRX8dOG\nz0eE6B6c3bJ+04Gzt+7FxCWl6DTxonGKFCkhty9fijPzq1wmfUTyTEJCQnh4eOqmRqMxNMyq\n7aujUmWcmwQAQKFWGH6BvmIqleoNvOs3Ex80UJgRcOSF47tNqixfs3CJLqJq53ccJJOAI7/W\n4HjCvFr3LpUGfPPtzkY9nu/U6XSS8btvdt+Gk69dCxcr7+IWOZ/78PLp64kGBroH4XcTxC3H\nARwZb8om3b8slXkRD48iIiW9fMob3Oq6KHR/26DGT9bMULu3GD/A78txC8cGPBoX2K1CujVQ\nk65vmznh2/DqI6YPqumYvmbdla3bLpjXGVP/+ToatzZMGv+zecv+vT8q5WRtbiyHgnuHJGtE\nnj4xyfoB1a1bd+/evambMTExUVFRWbZ+VSwtLfVdAgAAL6cw/AJ9lRwdHTUaTUxMxr8I8bqx\nt7e/f/++vqtAHjk6Ouq7BBQ41uDIE4s6Tes+OnnyUd2mdXL+8p8fHBr3but8ZvXKPx6n7nL0\n8DB/dPbMNd3Tbd2lM+cTrDzcsxq+kXxz44ZDCY5165dX5XCu7u6OOcGHbTtMmdTaaM+cefvz\n+T/iyckanU6jSU6zy8il8cgZ/rXifwoc/fVfD3TPD8SfD50wetm9t8dNH5wx3RDRHN+6O8Lx\nvWY1ng8yiT17/LJJvY7d6/t6uhR1tLN9cCf8WX/FvUpbxp869ndKhn4AAAAAAErHCI68Mao2\nKGRVLzG1ymL6RvLDf69cSTubwrKYp9N/yUJU7m17N9k57uff1PJ0UouqUtsOpT8Nnb246IAW\n5a2iT29cvDmy3Mdtny/AkfIo8tYtC9Fqnrwmdsum/REubQO6+ahFJJtzU67/NCvkvEevOZ18\nnHUju5z7fOGXW0pPbl48z1nYrV1Lt8d7+5UoamumjYv4J+zn0DMW1T6rlW72okGRukOmm1tM\nmDplZPzwoM/qFTWU2BMhAZO3Pnqr39C3raKuXIkSETGyc3Wze/bQ4w9s3Rdbon2zcmlqs3R2\nsXl0/uiF+AplzbX3T6wM2XlPxPvJJSq17Vj+wLJZ0617t3/L00aib5y+pK7VvEr+LKAKAAAA\nANAnAo48UhmbW2ezNkXMvjlD96XdUW/kJv86/+mKJn5dPq6+b86fSakluLYNmJDyzarQyUOi\nkkwdPat1Duzd4vlaHxK9d9agvaIyNDaztC/mUbpi16CxjSs6GWV/buKFlTNX/Vt1aHBzZwMR\n8Wjn3+f0kJCZoeVmdvbM8v0m2TO3Mb23e93itZExiQaWdk7ufh0CvmhRI5O1R1Q2VfpOmWQx\nadKXI2fEB/k3iT70+9WkFAlbPOb5O3iLtZu7tPvTuTB392z9K9mvf2OXFzrx7Taq+8KQmf33\nirGpZYk6XTrWnPZ96hNrOWG62YoVm5aMW/1QZ27nUrrBx//tQwEAAAAAFBIqnU6XcyvgzRMT\nE6PRaPRdhVhaWq4tx1tUAABK0vyI/n+BvkqswfHmYA0ORWMNjjcBa3AAAAAAAADFY4qKvkWs\nH9J/xdUMu0t2WRT8kase6smKUuoEAAAAALyRmKICZI4pKgAA5A1TVPC6YoqKojFF5U3AFBUA\nAAAAAKB4BBwAAAAAAEDxWIMDKOxa95ir7xIAAG8EzeDB+i4BAIC8YwQHAAAAAABQPAIOAAAA\nAACgeAQcAAAAAABA8Qg4AAAAAACA4hFwAAAAAAAAxSPgAAAAAAAAikfAAQAAAAAAFI+AAwAA\nAAAAKB4BBwAAAAAAUDwCDgAAAAAAoHgEHAAAAAAAQPEIOAAAAAAAgOIRcAAAAAAAAMUj4AAA\nAAAAAIpHwAEAAAAAABSPgAMAAAAAACgeAQcAAAAAAFA8Ag4AAAAAAKB4BBwAAAAAAEDxCDgA\nAAAAAIDiEXAAAAAAAADFI+AAAAAAAACKp9Z3AQByYDNhQnR0dHJysr4LQcEyNzfXarUJCQn6\nLgQFy9jY2NraOj4+Pj4+Xt+1oGCp1WozM7PY2Fh9FwIAwJuCERwAAAAAAEDxCDgAAAAAAIDi\nEXAAAAAAAADFI+AAAAAAAACKR8ABAAAAAAAUj4ADAAAAAAAoHq+JBQo71bFj+i4BAKBsd93d\n9V0CAAAFjhEcAAAAAABA8Qg4AAAAAACA4hFwAAAAAAAAxSPgAAAAAAAAikfAAQAAAAAAFI+A\nAwAAAAAAKB4BBwAAAAAAUDwCDgAAAAAAoHgEHAAAAAAAQPEIOAAAAAAAgOIRcAAAAAAAAMUj\n4AAAAAAAAIpHwAEAAAAAABSPgAMAAAAAACgeAQcAAAAAAFA8Ag4AAAAAAKB4BBwAAAAAAEDx\nCDgAAAAAAIDiEXAAAAAAAADFI+AAAAAAAACKR8ABAAAAAAAUj4ADAAAAAAAonlrfBQDIwfft\nSum7BACAsm3RdwHZa35Eo+8SAACvA0ZwAAAAAAAAxSPgAAAAAAAAikfAAQAAAAAAFI+AAwAA\nAAAAKB4BR2F3ZO6HLSftSRIRkbDpLVvPOqjnggAAAAAAKHwK9i0q50J6jtocpSrdI2R2W6fU\nvRHrhw5YcUXn0HrWsl5lnjd78VSndl99090rk0NWTYNWD6qYzeVERFRqEzNLu+Lu3n61Gjdv\n7OdklL/3lYtrRa4f3mfFpRfPUVUZGhr4rllOHT6TzZ2mk/TPlsUbDl29fPVGZJxxo8DQT6u8\n9P38B+dCeo46VHvWsr5lnu+7srzf0J2VntZ/O2zp0k3HL4ffi000tHJwLV+vTdeOdVyMMjlX\nF3Py+4nTfn1cc3jgp3WKJlzcvWHzvqPnrkVEJ5rYOpeu2bxzp0ZeFiJ5eLwAAAAAgNdbwb8m\n1tapyLXtO6606eapEhER3T87dt4q4mQTmfJCM5u3h01s4/F828jONfNDhtbO2V7uXf+p7Uvq\ntEmPYyKvngnbuipg2862AZN7+Fnlz+3k9lr2jT6fWyUxtWnkjhlT97jUq5H91++XutM0tAmJ\n6uKVmtSu/8+ahYfzdCsFSmdoV75h1w+cHSyNNVGX9q1dPnPsw8CQwZXTxU6a8L1zAhecdGgX\nNLFLeSsRifxr7xnxadK9lXsR84Trh35cMT/genJwULOieXy8AAAAAIDXV8EHHBa1G1XavXX7\nqc6DKhqKSPLJ7Xti637Q6NzqsBcLsXb29PTMvMasD2VkaOHk6vokG/H09q35Xl3PMcOWz1lV\nbenACkYiugcn1327cufRq1FJJo6lqjXr1ruNr41KJO74vKGBJ30DgodUtxSRxL+/Gz7qd7fh\nwaPq2eXxWmobF0+bZw11l8KOR1jV71vPMvviX+ZOU+4cnB8YfMa994TPm7j5tR3sJyIRD39K\n3yxsesvZ0sO/yKk1h69Ex2ptfRr3HNStmqNKJOunIRJ7bmPI8m3Hr917pDO1dfKo1WVE31r2\nuasrE8Vrd2ifulHG1/z64XGnTodL5RJp2sRd+HFK0A/3Kg6eOayhy9Pko2SnqTNTG5Qt7518\ntdvXB4/FNmtmJdk93gfHfvhm/cHz1+881Jg5uJV/p3OvDlWdDJ+2Tbj22+qVvxw8ezM62dTW\nxav+x8N7VE3tCAAAAACgWK9gDQ7jGk3eMdi3/Y8EEZH4w9vDjN5tWtW44K8rIiJqt1bt65pG\nHQg7pxPR3dwwOXDddc+OAXMXzhnd3vXSmsCpm2+LiFhWHvBFO7Pf587de08k7siSmZuSGn7+\nWbbpRg7XSktzdOvu20Xfa5Z/dx1/aWOg/7xrVb6YNaqJW069av/4cZNJu6lfr1i1fGpLw9+m\nTN0QrpPsnkbK0W8nrYoo22ticMg3C6aP6vGOh7kqf+rWJj64snf3iTjHin5pRqdoIw8tHBmw\nIem98V/6p6YbGWk0STpjGxvzDPvTPd6EeINSzfqNnb5g0ZyAHtXit06euuGGVkREdOGbJ34x\n77Cqbt+JcxfNn+rfsbqjSps/dwYAAAAA0K+CH8EhYuDTtJHjZ9v3RdduIr9t/6NY4+CyBvvT\nN4ra7N9yc5pz6o3Z6F8zk0NePb/5qo1T+rOzoS7h4Sp7I27Hid+ln9dfdG49b2DDEioRN+dP\nht8+P3zthtMtB/uKGHt3Hdn93PAlM7+PsN71u3n7GX0rZvgm/RLXqphmRkzcwa37H3p1alo6\nx5QgV3equ3dk6aQvD9h3mjK9tZdpbsqye69nZ18bAxFD9yYD2u7ptXrjyXaDK57I8mnERd6J\nt/KsWsmzmImIvb2Tm3eOl0j38YmIyAuTgjT7JneYc0Sn1elMPRqPmNav0vNYJnrHnBkpzm1m\nTu1RxiTLC2hv71607pJ7m37VDdMdyfB4i9f9qO3TH4s6dfo86lS3bWE3PuxSQrQnNvxw1q7F\nrNHtyxiKiBQr7lEubU8HDhwYP3586uasWbOqVHmlq5lkSqVSiSTm3A4AAMVycHDQdwn/lZGR\n0WtwF8iRSqXigwYKs1cRcIi4NW5adt2O3bcryI5z3k2Hu0tMhibp1uBQWRbL/JCJw8tOldDp\ndCIqUUnUjRvx5j4VSjz7ImzgVaGc6Ybr1x+Ir52IGLi1GtHn+OCFa1PK9pzT2TtPoy1Sr5VG\n5O6tR3UVP22YiwU1cnGnurPffXFQU/qz2V+865T+q35WSnh6po7UcShVyib+8vUocc3madT6\n4L2Ncyb0Pudb2a+Cb+W3alb1sM4hnEm/hEr4r5Nn/ZG2gbp6/+DgLpq4yIuHNqyaP83afvrH\nPk/TGUuf2iWuhe1asaVuQFuvTFfR0N4JmztuyfXKw6Z0Lp1+gEfGx6t7cHbL+k0Hzt66FxOX\nlKLTxIvGKVKkhNy+fCnOzK9ymayem1qttrJ6HssYGhpqtfof4GFgwKuOAACvucLwC/e/MDQ0\n1Ol0Sr8L5EYh+fsQeWNomNvvT1CuVxNwiOO7TaosX7NwiS6iaud3HCSTgCO/1uDIIPnatXCx\n8i5uITqdTlQZvqin2fPw8unriQYGugfhdxPE7eUHcKS5Virdla3bLpjXGVM/Nws95OJOVS5V\nqlv9fmDP/qu12+du/IZISnKaBV2Tk5OfFJbN07CtM2Rhhdanjx47eebkptlrvnP/eOrUtp7Z\n/fcgQ+U26f5lqcyLeHgUESnp5VPe4FbXRaH72wY1frJmhtq9xfgBfl+OWzg24NG4wG4V0i0H\nm3R928wJ34ZXHzF9UE3H9AVn8nhvbZg0/mfzlv17f1TKydrcWA4F9w5J1qTesmQd1dSsWfOX\nX35J3YyJiXnw4EE2N/1qWFrmsHILAABKVxh+4f4Xjo6OycnJMTEZ/8DF68be3l7p/1zfZI6O\njvouAQXuVf2fwxZ1mtZ9dPLko7pN6+QhOMi75JsbNxxKcKxbv7xKHD08zB+dPXPt2QoZuktn\nzidYebg/WWlDd3fHnODDth2mTGpttGfOvP33/9O1ntEc37o7wvG9ZjXybfkN28oDp054L/7H\ngHGrLzzK3SmXzp19NsVBd+PsuThzD3eHHJ6GqG08Kr/bpsdngfMnt7U7v33/9fyqXyQ5WaPT\naTTJaXYZuTQeOcO/VvxPgaO//utBmiVM4s+HThi97N7b46YPzphuZPp4Y88ev2xSr2P3+r6e\nLkUd7Wwf3Al/1l9xr9KW8aeO/Z2SoR8AAAAAgNK9ohEcIkbVBoWs6iWmVll8009++O+VK2mn\nH1gW83TKQxaS8ijy1i0L0WqevLp1y6b9ES5tA7r5qEWkUtsOpT8Nnb246IAW5a2iT29cvDmy\n3MdtfUVEUq7/NCvkvEevOZ18nHUju5z7fOGXW0pPbl482wAou2s9EX9g677YEu2blcvXIMnK\nt9eUSaaBgePHJgZM6uVnLZL04OatBxq5F50k2rjbV65cEZVlsZJPn58q+cjyuZuNO9dwfPz3\nr4s3/uvZbkwllaiyfhrhe7/bk+Rbo5xbEStVzKWwM1Gm7h4vs+xJOrd2Ld0e7+1XoqitmTYu\n4p+wn0PPWFT7rJbti60MitQdMt3cYsLUKSPjhwd9Vq+oocSeCAmYvPXRW/2Gvm0VdeVKlIiI\nkZ2rm92zf0OZPV5LZxebR+ePXoivUNZce//EypCd90SeLCJiUKltx/IHls2abt27/VueNhJ9\n4/Qlda3mVV5uLVkAAAAAQGH0ygIOURmbW2czjCFm35yh+9LuqDdyk3+dl79M9N5Zg/aKytDY\nzNK+mEfpil2Dxjau6PQkOVG5tg2YkPLNqtDJQ6KSTB09q3UO7N2imIgkXlg5c9W/VYcGN3c2\nEBGPdv59Tg8JmRlabmZnzyxf6pH9tURE5O6erX8l+/Vv7PLy95ED87KdJ00xnTxh0pjEUUED\nqj3YEjT0x9tPDoUvGXr4hVVaVdW69XE9FjJ2RXi8cTHfZqMHtXdRZfc0xMRCdXPHst9W3I1J\nMrQq4lm5c2Cvd/7DNAlzG9N7u9ctXhsZk2hgaefk7tch4IsWNTJZYERlU6XvlEkWkyZ9OXJG\nfJB/k+hDv19NSpGwxWOev1K4WLu5S7s/nQuT6eNV+XYb1X1hyMz+e8XY1LJEnS4da077/tkx\n15YTpputWLFpybjVD3Xmdi6lG3ych39jAAAAAIDCR6XT6XJuBcUKm95ytuHIjXnJit50MTEx\nGo1G31WIpaXl2nK8RQUA8DprfkT/v3D/C0dHR41GwxocbwJ7e/v7919+JjsKB9bgeBPwggYA\nAAAAAKB4r26KSr6JWD+k/4qrGXaX7LIo+CPXwn6tV1l8/lJu5QAAAACANwBTVIDMMUUFAIBX\ngykqUAqmqCgaU1TeBExRAQAAAAAAikfAAQAAAAAAFE+Ba3AAb5jWPebquwQAAERENIMH67sE\nAACyxAgOAAAAAACgeAQcAAAAAABA8Qg4AAAAAACA4hFwAAAAAAAAxSPgAAAAAAAAikfAAQAA\nAAAAFI+AAwAAAAAAKB4BBwAAAAAAUDwCDgAAAAAAoHgEHAAAAAAAQPEIOAAAAAAAgOIRcAAA\nAAAAAMUj4AAAAAAAAIpHwAEAAAAAABSPgAMAAAAAACgeAQcAAAAAAFA8Ag4AAAAAAKB4BBwA\nAAAAAEDxCDgAAAAAAIDiEXAAAAAAAADFI+AAAAAAAACKR8ABAAAAAAAUT63vAgDkwGbChOjo\n6OTkZH0XgoJlbm6u1WoTEhL0XQgKlrGxsbW1dXx8fHx8vL5rQcFSq9VmZmaxsbH6LgQAgDcF\nIzgAAAAAAIDiEXAAAAAAAADFI+AAAAAAAACKR8ABAAAAAAAUj4ADAAAAAAAoHgEHAAAAAABQ\nPF4TCxR2qmPH9F0CAADKcNfdXd8lAAD0hhEcAAAAAABA8Qg4AAAAAACA4hFwAAAAAAAAxSPg\nAAAAAAAAikfAAQAAAAAAFI+AAwAAAAAAKB4BBwAAAAAAUDwCDgAAAAAAoHgEHAAAAAAAQPEI\nOAAAAAAAgOIRcAAAAAAAAMUj4AAAAAAAAIpHwAEAAAAAABSPgAMAAAAAACgeAQcAAAAAAFA8\nAg4AAAAAAKB4BBwAAAAAAEDxCDgAAAAAAIDiEXAAAAAAAADFI+AAAAAAAACKR8ABAAAAAAAU\nj4ADAAAAAAAonlrfBQDIwfftSum7BAAA9Kz5EY2+SwAAFHaM4AAAAAAAAIpHwAEAAAAAABSP\ngAMAAAAAACgeAQcAAAAAAFA8Ag4AAAAAAKB4vEVFSc6F9Bx1qPasZX3LiIiILubk9xOn/fq4\n5vDAT+sUNXywZUz3r8+kaW7eNCh0UMVnW/FXdq1e/evhs7eiU0ztXbwa9w340MdIHl3cvWHz\nvqPnrkVEJ5rYOpeu2bxzp0ZeFrmuQURErizvN3RnpaDVgyqKyO2wpUs3Hb8cfi820dDKwbV8\nvTZdO9ZxMcq5/tTuEi58//non27q6o3Z6F8zHx7bU9kUBgAAAABQOgIOpdKE750TuOCkQ7ug\niV3KWz3ba1VvSFC7kk83DK2dn+1P+mf1mLGbUmp1+HhEOWcriY28+thEJSIS+dfeM+LTpHsr\n9yLmCdcP/bhifsD15OCgZkXzXJjO0K58w64fODtYGmuiLu1bu3zm2IeBIYMrp0sSMq9fRB7+\nuWjW4eKV3W4ey3MJ/6kwAAAAAIASEXAoUtyFH6cE/XCv4uCZwxq+MAbB0Lq4p1QQwzEAACAA\nSURBVKdnhvb/blry491qo5YMr/U0Sijj8/RIyU5TZ6Y2K1veO/lqt68PHott1swqQye5VLx2\nh/apG2V8za8fHnfqdLhULpGb+nWR22YvuNlw/AjTJX+mCTjCprecbTh4WqlTX286cjPezL3G\nh59+2tTi2MpFq/aci9Q5ln2vz9AeVR1UIiKScHXvyhU/Hzz3b5yhXXHvt9r07Pquh1luCwMA\nAAAAKBQBh+JoIw8tnPfVftP3x3/Zs7KN6sWDsftnfLzncbJlEVfvWq26tK/jZioiEnno0CUz\nv9oPVwUM+t+1R6aOHn6Nun78vreVKkPnGk2SztjJxjx/Kk18cO3g7hNxjpX9nNPszbp+zeXQ\n6WuSO8zs6GW4MX1numNrvlN/2GPch9YPDoXM/nrWtKOWSR4tPwnqZXB987y5s5aW+nZ0fQvR\n/bslaNQP2mZ9/HuWdTR4cGFLyMKxweqFo+rb5FwYAAAAAEDBCDiUJnrHnBkpzm1mTu1RxuTF\nI8ZudToPbOXj4WCSdPfCvnU/zPj8bJ+vprZ0VUnE7QhJuvbjz7W69gnoZfPw3MaQb0dPfBw8\nq73rC/mC9vbuResuubfpV91QshW12b/l5nT7Xhjxodk3ucOcIzqtTmfq0XjEtH6VjHNR/6OT\n38zY7thvboviKonMcE2dZf2+Q5p7G4iIe+cG28duTRi4vHs9WxEp0av1bzu/Of631K+iPfXT\nmrPleqzs0cBKRKRY0f6f3Dg6dOuB6PrNbXMqTOS3337z9/dP3Vy0aFGNGjWyfxKvSqK+CwAA\nQM8cHR31XUJ2jIyMCnmFyC980EBhRsChNJY+tUtcC9u1YkvdgLZeZmmPWPg17+j35Ecvb18/\n+0f9Zm7YcqFl/3Ki1Wol2aL+wGEtK6pFxHOY8a3eY3/Zdq59X5/Us7V3wuaOW3K98rApnUvn\ntCqFzdvDJrbxeL4d/uvkWX+kbaCu3j84uIsmLvLioQ2r5k+ztp/+sY9p9vU/ODBv9tGKQ+fW\ns8v8mipPr1JP3/mjcnCwF4dSpWyfHjK3tzd5/CA6UeT+xYux2qvzu7Scn/ZUu4gIEducChOx\nt7dPm2iYm5trNJocHkXBMzTMIW4CAOBNUBh+KWfFyMhIp9MlJyfruxAUOLVazQetXEZGLL73\n+iPgUBq1e4vxA/y+HLdwbMCjcYHdKmS1VoZF2XKucvBuZLKUU9vb24tYu7urU/vwcJatdyKT\nxOfJCIak69tmTvg2vPqI6YNqOmacuJKhBmvnF1f6sEn370hlXsTDo4hISS+f8ga3ui4K3d82\nqLFltvVfPX48+m7SpK67nmzqtKK7OK31QbcP5yzoXFJEVGp16hd9lahErX5+TZVKJTqdTkSr\n1YpZo8DQT6tkcRPZFCZSsWLFRYsWpbaNiYmJiYnJ8WEUNEtLy5wbAQDwuisMv5Sz4ujomJyc\nXJgrRH6xt7fng1YuRt+8CQg4FMjIpfHIGebzxs8OHB0/KqhfNbvMvs0/On/+pti+XUQtIm4+\nFWzkz5u3UsTOUEQk5eaNf8W2btEn6Ub8+dCgoJ8eNxo3vWdFm0x6+m+SkzU6nVaTNunOtP7y\nXWbPb6l91uT+ntkTNtn3Ce5Z1bb4S1ysmJeX5ePj/zuVVKWicU5tMykMAAAAAKBYBBzKZFCk\n7pDp5hYTpk4ZGT886LN6RQ3l5raFvz4uV7VUcVuTpLsX9q8PPagp2/P9ciIiBpVaty/z2/Il\n80sNaONr/fDsz0t2PvLu/H45EYk9ERIweeujt/oNfdsq6sqVKBERIztXN7scA4Is3Nq1dHu8\nt1+JorZm2riIf8J+Dj1jUe2zWrYvtspYv6m9m4d96nEzO2MRiyIeHi+3CKhh5fadfQ+FfBVk\n3atjvTKO6vh7ty78cSix7rA23rksDAAAAACgTAQciqWyqdJ3yiSLSZO+HDkjPsi/iZm18Z1d\naxf8GBWbbGzn5OL9gf/QtvXcngzuULm0GjdZ9d2yn6Z9HqK1LOZZpceUHh+4qETk2qHfryal\nSNjiMWGpPRdrN3dp94zvms0dcxvTe7vXLV4bGZNoYGnn5O7XIeCLFjXsMzZMX79bfkyJU7l8\nMH6WTejqjSunbIxKNLS0K1bCp2Yr+5cqDAAAAACgQCqdTqfvGoDCKCYmpjCsZ2Zpabm2HG9R\nAQC86Zof0f8v5aw4OjpqNBqWZngT2Nvb379/X99VII9Yg+NNYKDvAgAAAAAAAP4rpqggMxHr\nh/RfcTXD7pJdFgV/5KqHegAAAAAAyBYBBzJTvH3wpvb6LgIAAAAAgNxiigoAAAAAAFA8Ag4A\nAAAAAKB4TFEBCrvWPebquwQAAPRtob4LyNqTt6fkxwvvc6YZPPiVXAcAFIkRHAAAAAAAQPEI\nOAAAAAAAgOIRcAAAAAAAAMUj4AAAAAAAAIpHwAEAAAAAABSPgAMAAAAAACgeAQcAAAAAAFA8\nAg4AAAAAAKB4BBwAAAAAABRu6zuqVOquG/O/49D2KpVp11/zv2M9IOAAAAAAAKDQ+Gfj5MDA\n0FPJ+q5DeQg4AAAAAAAoNP7ZOHniRAKOPCDgAAAAAAAAikfAAQAAAADAy3u6Lkb0n4v6Nyjj\nZGlqWbR80xEbLieLJF36aVTrqm52ZqbWbtU7z/4j+sUTk8P3fDWgaSV3OzNjE6ti5Rp0n7bz\nxpMBG2cCK6haLE8U+aWbmeqJBgvuPT/z4bGQwQ3LF7M0MbV1r/bRlD13dC/2HH3iu89bVyvp\nYG5iYlW0TP0uEzZdSkh7XHdn/6zudUvZm5lYFPF+u9e8Q1HpOlA0tb4LAAAAAABAoXRnZzVt\netu2S7dRrU2ub1m8aPaHTQzWjftn0Ijbjft8MqFz3JFVc9eMaK7yvLa6jeWTM1Iufd2yzsAd\nmkod+ozq42WTcO33VYvGNN1xbPXRdZ1cSvZY9pvJuMZjdtQYs2NyI2MREdvSNs+udW5204Y3\nzT/s9nkL24dH1sxfHfBBlMPF3QNcnx5/dPCL+o1mnTau2mVQUG3nxIvbly2e1Gpb2MLDuwaV\nMRQRids/7O1GwRft3+47ZlhV++jjG2Y0blzM/VU/soJDwAEAAAAAQN5oT4T77T619D1rEZEh\nLex8fCd+2f7TRotOHxzgYSgi0rvMI7duofPXBLfp5ygicuvrfp9tS27+3bFNPUuoRERkwJBO\n5WpWCxw26bcOX79TonoD32IGIo7l6jdoYJruWsdvVki9lgysoy7ZbPm8pecGTCovIiJnZw+c\ndTq55pQ/fx9TwURE5JOBzfpVahLy+ZBlHbb3KSJy9qtP5v2tbrz4f1uf1jawR5UW5fqKmLyC\nB/UqMEUFAAAAAIA8qjZw5NPEQURV4d0GRURn0WZovycJgojYvvtuZdH+88+lJ5vhoSt+S3Lq\nMLRTscSEZ1JKd+tUTe7s3HnyJa4l5o2a1jOUy//8o32yffGnDafFst2o4RVS8wq7xoFDaknC\nrg1bHz5toHPo5N8ntTYD115jerj850dQaDCCAwAAAACAvDEoVapkmk17e3sRdcmSBul2HYmK\nerJ17vx5kYdfNzT7OkNX5pGRL3UtQ0dHW0mKiooVsRGRK1euiJTy8Xlh2Iezr6+9HL58+ZqI\n3+XLl0V8y5VLGwMY+PiUFTmQmztVAAIOAAAAAADyRqVWZ5gYoVZn+Kat0z1dzFOr1YqU7rdu\nRU/X9G0MipR9+Ws971mn04moVKoc+kjfQPcarTJKwAEAAAAAwKvh7e0tcua2pkLNmpZZNMkx\npMhcqVKlRC6dOZMgpZ4P4og4c+a+GFT3LPGswcVz55LFOTUJ0J4993deLlY4sQYHAAAAAACv\nRsmOH9c1Sto8bsSOBy8OnXj0b8TDJz9ZWVmJ3L9//yV7Lt2mra/EbZg593zSs10xuycFHxLT\nhu2aWz9pUEHuhc7+7ubTVTtEG758+vLwvN9MYcMIDgAAAAAAXpGSg5ct3F1/wNfvlzneqWer\nGl5FDO7fvHL28K8bT767KXJJAxGpWLOm6fxt8z8NMm7lZWuscqra9t3SxrnoucKIxf7rG84a\nXbvO2UGdajsnXtzx3eLN1+3eWRDcs8iTBsPnfbK6yYKBb713YnD7ynYxJzcsWh5dupwcu1KQ\nN/wKEXAAAAAAAPCqqL36/nLCJ2T6l9//+v209feTTR2KuZSq1HDkV739njSw6Ri85njyxNDZ\nQ3+JSUyRt+fffbe0Y256tqgzM+xQmUmBi35aPHZtnKFticptxn85aXQrr2df/K3eCd6/08V/\nzNIfgoYvt/So2sJ/x1qnr4p0fl0CDpXudVpRBMg/MTExGo1G31WIpaWlqalpdHR0cnKyvmtB\nwTI3N9dqtQkJCfouBAXL2NjY2to6Pj4+Pj5e37WgYKnVajMzs9jYWH0XggLn6Oio0WhiYmL0\nXQgKnL29/cvPGkBh4eiYq4gAisYaHAAAAAAAQPEIOAAAAAAAgOIRcAAAAAAAAMUj4AAAAAAA\nAIpHwAEAAAAAABSPgAMAAAAAACieOucmAPRKdeyYvksAALwid93d9V0CAABKxQgOAAAAAACg\neAQcAAAAAABA8ZiiAgAAAADAy4mNjS2Ibq2srAqi2zcEIzgAAAAAAIDiMYIDAAAAAIC80H71\nVT72ZjB8eD729gZiBAcAAAAAAFA8Ag4AAAAAAKB4BBwAAAAAAEDxCDgAAAAAAIDiEXAAAAAA\nAADFI+AAAAAAAACKR8ABAAAAAAAUT53F/uu/f//btVz1UOKdHg088q0eAAAAAABQSEQf+W7G\nvDXbwk5duhNvXqx09dafTJzQu5qtvsvKVFYBx5EFPXtuyFUP7dYRcAAAAAAA8Bq6vnXFPqnX\nb/owHw+bR2c2TBnV573TmhO7B5TUd2GZyCrgqD9h164BueqhqF/+VQMAAAAAADIV2l7VVb10\nX9U9n8z99XyMlU/LgG+/6WezPWBAwLID13WutXvM/X5GMxcDEYnYMXHYtHUHz1y9l2jlWq5u\nt4mzRzfzMHrazaNTK8eN/Wp92Pk7SRZFvat1mrZqRrMiEnXgqyGjFu86dTNaa+FUwrftpDXB\nbZ1FKk74/VBqAbXq1kg64fTJuu33Bwy018sjyFZWAYeTb8OGvq+0EgAAAAAAkB3tjvEjjANm\nbg5wjNgwtOsnH7Xbav/Yd0jI7tmGZ4J7d+/0WZVrGzrZijyKMagyYN4XFUvaG9w9FhowqGVr\nwxN/jfUxFNH9Pe/9OiOuvztx7vbWlRyTI07/fkyVLKLZNvyDgMt9f9i+onpRg+hb5w6fMlFl\ndv3EhASdWQkn61d937mSVcCRmeToKydPXY7UuNZ6r1zhnHADAAAAAMDrS2fXKXjZ4BqGIuIz\nscuSdxY/Wvzv9I5FRcTvy+Ervxm28w/p1ETE68NxXzw9o2SJCavD9zotXnt27CQ/Sdk1I3B/\nsc8Orx9d00hExNPLt46IyN1r12LsKzdrVNnTXMTZuUT5GplcPOXKdwOmHfX5fN4HLxMlvDq5\nfItK4j8/fFrL1alUtbcbv9/92wsiIvJvyPtORT9a9aAAqwMAAAAAAM+oKlWrYvj0RxcXZ3Gp\nUqXo00M2zs7msXfuxIuIaCPC5g9pW69K+ZKuxYsV8x3/p1y7dk1E5MrRow+s3mlc0yhdv0Xa\nftrDbFlTD9/GXYZO+WbL6Xva9FdOubq2R8PBZxov3zyxukmB3d9/kquA4+GOT9/rsuBv147j\nZ/WsmLrXuXVLvwebNmx9VGDFAQAAAACAVAbGxqmjJ1QqlRgbG0vaba1WKyIXZnzQaOIp70Ff\nrdkeduT4iSOT31YlJSWJiOh0OlGpMpl9UrT9srM3jq72b+z68MDcrpW96s88oXl+NOHMkrZ1\nex5p9EPYqk4lczlQ4tXLTWG3l0/99k6lwH3/WzHxk3ddn+8vUqtWqYSTJ/8usOIAAAAAAMDL\nub9/1zHzj8ZP79SgUpmSrsWdbl/5R/f0UKlq1ewe7t3xP00mpxkX8W388YgZ3207teeLYge/\n/uHM0/0xByc1ru9/s/OmsK9buxbadENytwbHiaNHtdUm9fRViyS/cMDd3V0iIiIKpjIAAAAA\nAPDSbL29i8Qc3HYopkFtm5R/d40Z8s1NkSeLahg2+mJ83R+/6NTBYfao1pWddHfO/P6XSZtP\nmjxcMeL7xw1a1i7v7qCK/Cv093CLChVKiIhE7Rr6XquF0a3mL+/iGH7iRLiIiEnRMuWKm+nv\n/rKSm4AjJSVFTEwymWMTGRkpRi7pp+4AAAAAAAB9MWgwZcO0/kM+Kl1czCzs/TpMGt+63cin\nx1Rlh24Nsxozeu4njSbc01oV867RdXoHEQsb1fml/itH37j72MjBo1LjwO1fdrMTETm9YdXJ\nx8kSOrBBaGr/pUYevzS9kh5uLAcqnU6XU5uLQRW955Rff+OHdpYJqz4wG1X28K0va4roLs2s\nWX6k4fTwQ8OdX0WpwCsVExOj0WQ2buvVsrS0XFsuUd9VAAAA5LPmR/T/h9bLsre3v3//vr6r\nQB45Ojrmb4exsbEiov3qq3zs02D4cCsrq3zs8E2Tm+kzpTt3rxX744g+y848X0807tzqgR0m\n/WnQqEdH0g0AAAAAAKBfuXp5banPvp+3u/7gXr6bRhVRy8NjH3t9d/3ygySDYi2WLulDvgEA\nAAAAAPQsdwugqr0Hbj6xb+GQxiVtza0NHkRE21R4f9DcPcd+7u1ZmFdQBQAAAAAAb4ZcjeAQ\nETEsVm/Q3HqD5hZkMQAAAAAAAHnBAAwAAAAAAKB4uR7BIbrYi3t/+uX3E1fuPNJZFC1VqUHr\ntu96WakKsDakdy6k56hDtWct61tGRER0MSe/nzjt18c1hwd+Wqeo4YMtY7p/fSZNc/OmQaGD\nKj7bir+ya/XqXw+fvRWdYmrv4tW4b8CHPkby6OLuDZv3HT13LSI60cTWuXTN5p07NfKyyHUN\nIiJyZXm/oTsrBa0eVFFEboctXbrp+OXwe7GJhlYOruXrtenasc6TlwlnX39qdwkXvv989E83\ndfXGbPSvmQ+P7amkf7Ys3nDo6uWrNyLjjBsFhn5aJf/6BgAAAADoWe4CDt2dbSPbdPvycFSa\nV8pOHjmizherfprWxKmASkO2NOF75wQuOOnQLmhil/KpLxKyqjckqF3JpxuG1qkrwCb9s3rM\n2E0ptTp8PKKcs5XERl59bKISEYn8a+8Z8WnSvZV7EfOE64d+XDE/4HpycFCzonkuTGdoV75h\n1w+cHSyNNVGX9q1dPnPsw8CQwZWNclO/iDz8c9Gsw8Uru908lucSMqdNSFQXr9Skdv1/1iw8\nnM99AwAAAAD0LFcBxz9z2rSZ9adD/UELRnSpV87NTh7cPLd/1aypS6e3blv01IGhpQu6SqQT\nd+HHKUE/3Ks4eOawhi5pkwND6+Kenp4Z2v+7acmPd6uNWjK81tMooYzP0yMlO02dmdqsbHnv\n5Kvdvj54LLZZszy/fLl47Q7tUzfK+JpfPzzu1OlwqVwiN/XrIrfNXnCz4fgRpkv+TBNwhE1v\nOdtw8LRSp77edORmvJl7jQ8//bSpxbGVi1btORepcyz7Xp+hPao6PBlOlHB178oVPx8892+c\noV1x77fa9Oz6roeZiIipX9vBfiIS8fCnvN4bAAAAAKRhMHy4vkvAc7kJOE58t+iwQZ1Zv+0Z\n4f20uZubl1/tZg093q4yauF3J4dOq5h9B8hP2shDC+d9td/0/fFf9qxsk26OUOz+GR/veZxs\nWcTVu1arLu3ruJmKiEQeOnTJzK/2w1UBg/537ZGpo4dfo64fv++dyQQjjSZJZ+xkY54/lSY+\nuHZw94k4x8p+aV4mnE39msuh09ckd5jZ0ctwY/rOdMfWfKf+sMe4D60fHAqZ/fWsaUctkzxa\nfhLUy+D65nlzZy0t9e3o+hai+3dL0KgftM36+Pcs62jw4MKWkIVjg9ULR9W3yUW9jx8/vn//\nfuqmiYmJoaFhNu1fDZWKeWAAAOA1VBj+0MoDhZYNvCFyE3DcvHlTagzt6p2urVHZjztVHzXi\nxg0RAo5XJ3rHnBkpzm1mTu1RxuTFI8ZudToPbOXj4WCSdPfCvnU/zPj8bJ+vprZ0VUnE7QhJ\nuvbjz7W69gnoZfPw3MaQb0dPfBw8q73rC9+dtbd3L1p3yb1Nv+o5/Gc7arN/y83p9r0w4kOz\nb3KHOUd0Wp3O1KPxiGn9Khnnov5HJ7+Zsd2x39wWxVUSmeGaOsv6fYc09zYQEffODbaP3Zow\ncHn3erYiUqJX6992fnP8b6lfRXvqpzVny/VY2aOBlYhIsaL9P7lxdOjWA9H1m9tmf0siIv/7\n3//8/f1TNxctWlSjRo2cT3sVEvRdAAAAQD6zs7PTdwl5odCyUXB+8tPmY29tT/EakP8kNwGH\nq6ur/B0Xl/FAbGysuLm55XtRyIalT+0S18J2rdhSN6Ctl1naIxZ+zTv6PfnRy9vXz/5Rv5kb\ntlxo2b+caLVaSbaoP3BYy4pqEfEcZnyr99hftp1r39cn9WztnbC545ZcrzxsSufS6ZbLyMDm\n7WET23g83w7/dfKsP9I2UFfvHxzcRRMXefHQhlXzp1nbT//YxzT7+h8cmDf7aMWhc+tl8StD\n5elV6un/2FUODvbiUKrUs8zC3N7e5PGD6ESR+xcvxmqvzu/Scn7aU+0iIkRyEXA4OTk1bNgw\nddPa2joxMTHn0wqYWp37lYABAAAUozD8ofWyjI2Nk5KS9F0F8sjExCTnRlC43Hx3qtxjQI35\nCydv6f198yJpdkdunrz0wluDV1cqqNqQGbV7i/ED/L4ct3BswKNxgd0qZLVWhkXZcq5y8G5k\nspRT29vbi1i7u6tT+/Bwlq13IpPE58nQiqTr22ZO+Da8+ojpg2o65jwjQm3t/OJKHzbp/h2p\nzIt4eBQRKenlU97gVtdFofvbBjW2zLb+q8ePR99NmtR115NNnVZ0F6e1Puj24ZwFnUuKiEqt\nTh1XohLVC9/7VSqV6HQ6Ea1WK2aNAkM/rZKXaR0+Pj7Tp09P3YyJiYmNjc1DP/nL0tJS3yUA\nAADkv8Lwh9bLsre3V2LZeIKA402QVcDx75GNf4anbpXq2cN+ZKtyJzv07ly/vJudPLhxbt+a\nb9Zd9eg/o1TcvyLOWfSCAmHk0njkDPN542cHjo4fFdSvml1m3+YfnT9/U2zfLqIWETefCjby\n581bKWJnKCKScvPGv2Jbt+iTdCP+fGhQ0E+PG42b3rNibpaqeDnJyRqdTqtJzqn+8l1mz2+Z\nOrrr/p7ZEzbZ9wnuWdW2+EtcrJiXl+Xj4/87lVSlonHOrQEAAAAAr4+sAo5DM9p02JB+54nQ\nmSdC0+64v3hgu0hH3fr26VuigBkUqTtkurnFhKlTRsYPD/qsXlFDublt4a+Py1UtVdzWJOnu\nhf3rQw9qyvZ8v5yIiEGl1u3L/LZ8yfxSA9r4Wj88+/OSnY+8O79fTkRiT4QETN766K1+Q9+2\nirpyJUpExMjO1c0urwHBrV1Lt8d7+5UoamumjYv4J+zn0DMW1T6rlW6GSMb6Te3dPOxTj5vZ\nGYtYFPHweLnszLBy+86+h0K+CrLu1bFeGUd1/L1bF/44lFh3WBtvEUl6cPPWA43ci04Sbdzt\nK1euiMqyWEmn/FlRFQAAAACgV1kFHLVGrFvXMVc9uNbKv2rwElQ2VfpOmWQxadKXI2fEB/k3\nMbM2vrNr7YIfo2KTje2cXLw/8B/atp7bk8EdKpdW4yarvlv207TPQ7SWxTyr9JjS4wMXlYhc\nO/T71aQUCVs8Jiy152Lt5i7tnvFds7ljbmN6b/e6xWsjYxINLO2c3P06BHzRooZ9xobp63fL\naeWP3FC5fDB+lk3o6o0rp2yMSjS0tCtWwqdmq6dXv7UlaOiPt5/8HL5k6GERg3pjNvrXzIfr\nAgAAAAD0TKXT6fRdA1AYxcTEaDQafVchlpaWa8spbwkuAACA7DU/ov8/tF6Wvb39/fv39V0F\n8sjR0TF/O3yyIEu+v0XFyiqrZRaRM15CAwAAAAAAFC/Xb6DUPbyw+5ftf1wIvx+v0aYd9FGp\n59weFQugMuhTxPoh/VdczbC7ZJdFwR+56qEeAAAAAEC++rWHZYt7Cx7/2sNU35Xkl9wFHLGH\nJjZrOfFgVGazWdrVJeB4/RRvH7yJpWMBAAAAAHmxsatpm7hvNBu75npURT7IzRSVlIMTugQe\nNmk69Zdj5+Y1kuKDfr16Yt+asY3dHKt/sf3aspYFXiQAAAAAAEB2chNwnNj48zW7D+etHd2y\nckk7YzEwcyhRsX6nyVu3D9UFD15w9lXmMQAAAAAAvJlC26vUH83fGdiuorudta1zhQ/G7vz3\nySqn9xY0ULkOPZDaMmFVa5Vp11+fbj06tXJ4i8rutqYm5g7ulZqM3HY3Y99xJ5cPed/PxdrU\nzL6EX7MhK87EPjuijdg9pfNbJe1MjU3tPWt3m/n73WeTOyJ2TOzYoIKbo4WZVbHSNdpP2nb9\nyerB2/vYtlmdKL90M1KpVCpVwyX3smmc6wpzlpt0Ijw8XCrVq2clIgYGBvLszRKG5bt2rBww\nf/2xWTWr5eXSAHKldY+5+i4BAFAoaAYP1ncJyISjo6NGo4mJidF3IQDeCCkbp87x/WHbP+uL\nJhyb0rzeh59WuLahk212Z+j+nvd+nRHX3504d3vrSo7JEad/P6ZKTt/m4oIW9SamDJz7w6xa\nboYRhxcO7f9OT6Nz6zsVEd35ma2aTX3ce/4vy+o5RP42e+CQpm2Mzx4YWkpEHsUYVBkw74uK\nJe0N7h4LDRjUsrXhib/G+hg2/Sb654T0U1Syapy7CnMlNwGHhYXFs1TD2MbGNOrff5NEjEXE\n2tpabt++nZfrAgAAAACAl+Tdf05AA2cRMa36xeCmgQN2/iGdmmTTPmXXjMD9xT47vH50TSMR\nEU8v3zoZ2uydGRhWZ8adGV0cREQ8S85fenZblUU/3uk02GnXrOlHvD8/xGQ/FgAAIABJREFU\nubiXn0qkfOmQlVcOVp8887ehX78j4vXhuC+edlGyxITV4XudFq89O3aS3//bu/OAqKq/j+Nn\nAFmGHVmUXQQUEVwzFVErl8zSRC33Jdc0cws1RUVRXHMrl7KyTFNLzTS13EoJ7feYu7giihuK\nCwybwAxznz9AwIUBDRmuvF9/Mfeee+73XhwGPp5z7lMLKbJxSSosmZJMUfHy8hIXz5+XhBAi\nsE6dnD/X/3RLEkITv3ZDtHB15aEaAAAAAACUAYOafr4PvzZ1dLRKvX07Q+cBcUeOJFm+1iY3\nOyjC5cOH7+X8PtBeka9+5CURGxsrxPWYGJV1cPNARV5Tw4YtmpnfO306QQghtAlRn48MCa5f\nq5pr1SpVAqb8n7hy5UpRJymycUkqLJmSBBzV2rTxub1p3T61EMKt+8C2mVv6+XnXq+cT+PHe\nzPr9+/AIFQAAAAAAyoDC0PDRP+O1Wq3IXU6isJycnLyvJEkSCoVC6JKTkyMsB+zUSo9IWBhU\n1OG5W87Nebv1tJO+wxb8+HvU4WPHD89oocjOzi7iHEU3LkmFJVOSgEPUn7D31N7xgVohhHAZ\nuHbn/J4BpskpFgHvhW/fNqFWaZQBAAAAAACej42jo/H9u3cfrv4pnT17Pu/L6g0b2qbs++Mf\ndVGH5rVJ3bVl74Mnd7nVrm2dHLX/5MOetf/u/zu9cu3aVYS4f2D3UeX7U2Z3b1m3RjXXqo63\n4i5I+ccZGxsLjaZgJQ0djUtSYcmUKOBQWLvVru3tYJL7qnLw2O8PxFy+dOrv9VPbOhv+5xIA\nAAAAAMDzM2jW5g2xfcWKCw+E0NyOihj/9ZW8PYatx01plrCke9dZm/85eznuzD9bl33xR8Kj\nRxu1mTCtZfLKPh3Cfow6denyhWNRvy4b22f+/4QQitbjPn3l/Ge9PlwVdfZizF9fDey9OD4o\nLLSlEMLG19dBFb3zoEoSQnNz94SRX18r6LJ69eri5N5fz968c/duSqakq3FJKizhXXiegwAA\nAAAAQLlRpc/ydQOkJS1c7at4v7lQO3p8u4d7FDVH7Yha/nbmjx+1rlMjIPi9qdtuPvG4EUWN\nEdsPre1juWtSx0Z+fg3b9pm6+Y6Xn3Pu4aG/7pjof2xah7q1GoTMPttg2u+/jKwuhBAGLWdu\nmuW9532fqm4eNdotMRg/5d2CERA1hswP8z82qrGXk4NDyHf3dDYuSYUlopAk6Wnb4//67s8r\nJerB87V+LT2e59RAuaZSqR4+ElmfLCwssubM0XcVAIBygcfElk88JrbisLOzu3//vr6rwHOy\nt7cv3Q5TU1OFEJtzl3IoJSEnDSwtLUuxw4qmqFTk8Bf9+28qUQ+dfybgAAAAAAAAelVUwNF8\n6u7dQ0vUg9PTH3ELAAAAAABQVooKOBwDWrUKKNNKAAAAAAAAnhOLjAIAAAAAANkj4AAAAAAA\nALJHwAEAAAAAAGSPgAMAAAAAAMheUYuMAgAAAAAAXUJOMmigHOGbAQAAAAAAZI8RHAAAAAAA\nPA+rixdLsbcUH59S7K0CYgQHAAAAAACQPQIOAAAAAAAgewQcAAAAAABA9gg4AAAAAACA7LHI\nKFDeWU+dmpycrNFo9F0IXiylUqnVajMzM/VdCF4sY2NjKyurjIyMjIwMfdeCF8vIyMjMzCw1\nNVXfhQAAUFEwggMAAAAAAMgeAQcAAAAAAJA9Ag4AAAAAACqcLb1MFe+ueZlmwhNwAAAAAAAA\n2SPgAAAAAABABu79vaBXMx8nK1MTi8putVuO3Hzz4Z60E9+PfCvQxcrUzM4zsN3I1acfLnKd\n8Me0bi1ru9mbm1lW8WnUZfrOeLUQQojfB9p0Wpslfu1dSaFQKBStVtzV0VgIIUT6yR/GvFPP\n3cbURFnZvW7b8TvvlOGVlwxPUQEAAAAAoNxT7xzzdtilQet+X/2Kk0Hy9TOHTpoohBBCSBe/\neCd4Ws6Hi9bNa+JmmHBo6aghr/WvdGZjdwch0lUG9YcuGVenmp3BnaPrw4Z1eNfw+L+T/A3f\n/Dr5l0zTTmlfq7f0ys8FimoshHR+yVtBn8S/Pm3R7+/WtdcknPrrqKL8TW5RSJKk7xqA8kil\nUqnV6uLbvWAWFhampqY8JrYi4DGxFQSPia04eExsxWFvb69Wq1Uqlb4LwQtnZ2d3//59fVeB\n52Rvb1+6Heb+hLe6eLEU+0zx8bG0tCxy953lLRxn+u+9sOx15SPbc/YOcmp7fc7tnQMq5205\nNsm7/oHRt6KGOz3exectHJe3OHFmeqAQYkuvxwOOIhvn7PrAoW3UB4fOzG9c6bkv78VjBAdQ\n3imOHtV3CQCAUnDH3V3fJQAA5MwhZES/Bb3f9Ihq2eaNFq+17vBuuwB7AyHE5cOH7+WcGGiv\nGFi4dZXYWCGchDYhaunshT9FnbuemPRAI2WpRJbHFSECn3qGIhvHHTmSZPlam3KdbgjW4AAA\nAAAAQA6cuqyKuXpkbWgb15S/F/Wq59187nG1ECInJ0dYDtiplR6RsDBICHFuztutp530Hbbg\nx9+jDh87fnhGC0V2dnYR/RfdWJIkoVAoyuxKnxMBBwAAAAAAsmDsENCmzydzvt15cu+4KtFf\nrjsthKjesKFt6q4tex882fz+gd1Hle9Pmd29Zd0a1VyrOt6Ku1CwSIWxsbEoPBNeR+PqDRva\npuz74x/9z+HXiYADAAAAAIBy7/zqTz798rdDp+Ju3Lx87Lf1f90wr13bUwhh1GbCtJbJK/t0\nCPsx6tSlyxeORf26bGyf+f8TQtj4+jqoonceVElCaG7unjDy62sF/VWvXl2c3Pvr2Zt37t5N\nyZR0NTZsPW5Ks4Ql3bvO2vzP2ctxZ/7ZuuyLPxLK/A4Uh4ADAAAAAIByz9xacfaH0K4tAryq\n12s/7v98w3//pretEEIoaozYfmhtH8tdkzo28vNr2LbP1M13vPychRAGLWdumuW9532fqm4e\nNdotMRg/5V3D/P5qDJkf5n9sVGMvJweHkO/u6WysqDlqR9TytzN//Kh1nRoBwe9N3XazHC7p\nyVNUgKcrP09RMTtzRt9VAABKAYuMvqx4ikrFwVNUZO1leIoKisMIDgAAAAAAIHsEHAAAAAAA\nQPYIOAAAAAAAgOwRcAAAAAAAANkj4AAAAAAAALJHwAEAAAAAAGSPgAMAAAAAAMgeAQcAAAAA\nAJA9I30XAAAAAACALKX4+Oi7BBQg4AAAAAAA4NlYWlrquwQ8jikqAAAAAABA9gg4AAAAAACA\n7BFwAAAAAAAA2SPgAAAAAAAAskfAAQAAAAAAZI+AAwAAAAAAyJ5CkiR91wCURyqVSq1W67sK\nYWFhscEvS99VAAAA4GXQ/rD+f7/VF3t7e32XgBeOERwAAAAAAED2CDgAAAAAAIDsEXAAAAAA\nAADZI+AAAAAAAACyR8ABAAAAAABkj4ADAAAAAADInpG+C8B/cmZl/wkHm85bNaiGEEIISXXi\nu2mzfnvQeEz4iCAnw6TtE/t+ebpQc+WbEeuH1Xn4KiNu99q1vx2KuZ6cY2rn4t1mUNh7/pVE\n+sU9m7btP3LmSkJylomNs0/j9j26t/Y2L3ENQggh4r4fPGpX3Yi1w+oIIW5FffXV1mOXbtxN\nzTK0rOxaK7hTr25BLpWKrT+z6EoSN44ZuDr20ToU9UetD3/d7L/cTgAAAACAXBFwvDzUN/Yt\nDP/iROXOEdN61rJ8uNUyeGRE52p5LwytnB9uz76wduKkrTlNuvb5xM/ZUqQmXn5gohBCiMR/\n950W/m37dnR3UGbGH/xp9edh8ZrFEe2cnrswydC2VqtebztXtjBW34vdv+H7uZNSwlcOr1ep\nmPp1VGLXeuyi+ln5hyb+MSdyr0twI9INAAAAAKioCDheEmnnfpoZse5uneFzR7dyKZwcGFpV\n9fLyeqL9za0rfrrTcMKKMU3yopAa/nl7qnWPnJvfrGYtX83l3l9GH01t187yiU5KqGrTrl3y\nX9QIUMYfmnzy1A1Rz7OY+nVUYmTt4mX9cJcUG3UswbL5oGALIYQQSUfXfb0x+mz87RS1WWW3\nWq/1+KBrA0fDvLaZV/5c+8Ov0THXkjWmNi7ezfuM6dcgvyMAAAAAgGwRcLwEtIkHly5ZcMD0\nrSnz+9ezVjy6M/XAnD57H2gsHFx9m3Ts2SXIzVQIIRIPHow1C2yasiZs2D9X0k3tPQJb9+rz\nlq+l4onO1epsydjRWlk6lWYlXYneczzNvl6gc6GtuuovvhL1kR17bjm9EdrAOPd1ZoZB9XaD\nQzydLA1Ulw6sWT4j0mDxgvfcDYSQbmybNu7bO4E9B017tZqVNunKqUsKbUFHqamp169fz39Z\nuXJlY2PjUrjq/8bAgIVyAAAAUDqMjPgDEC8z/n3LX/IfC+fkOHeaG9mvhsmje4zdgnp82NHf\no7JJ9p1z+39eN2dszMAFkR1cFSLhVoLIvvLTL016DQz7wDrlzJaV33w67cHieV1cH8kXtLf2\nLPs51r3T4FcMhU73toV22PbYtkdGfKj3z+i68LCklSRTjzafzBpctyA5KLr+ElWSFr3jQIp3\n9zd9HlZetdn7IXlfOjl2H3vvZO+dUVff6+kptMc3rYuxfWfep11qGAohRJWqHn6Fe/r3339D\nQ0PzXy5btqxRo0a6r7usPNB3AQAAAHgZ2NjY6LsE4AUi4JA/C/+mnleidq/e3iwsxPuRVSjM\nA9t3C8z90ts3INAuffDcTdvPdRjiJ7RardCYN/9wdIc6RkIIr9HG1wdM+nXnmS6D/POP1t6O\nWjR5RXy90TN7+Dy2XMYTrFuMntbJo+D1jd9mzPtf4QZGrwxZvLinOi3x4sFNaz6fZWU3u4+/\naXH1l6SSxD07jkh1RrQqGBEiJcVs37j175jrd1Vp2TmSOkOoHROF8BS3LsWmmQXWq1FUWOPh\n4dG3b9/8l5UrV37wQP/JQqVKxd18AAAAoGTKw++3+mJmxop9Lz8CDvkzcn9nytDA+ZOXTgpL\nnxzeu3ZRa2WY1/RzFdF3EjXCz8jOzk4IK3d3o/w+PJzFjtuJ2cI/d2hFdvzOuVO/ufHKJ7OH\nNbYvctZIQQ1Wzo+u9GH92L8shdLBw8NBiGre/rUMrvdatv5ASEQbi5LUr7MSKW7HznPKoInN\nC9bRuL5p+pRflB2GDHi/uqOV0lgcXDxgpUYthBCSJAmFKPpivLy8RowYkf9SpVKlp6cXe+kv\nmoWFhb5LAAAAwEuiPPx+qy8EHBUB0/tfCpVc2oyfE9okY3P4p1/+myQ9vVH62bPXhI2Dg5EQ\nws2/trW4de16Tt6+nGtXbwobJ6fcdCPj7Pqpn66622Ly7OElSTeekUajliS1WlOS+oupRH1s\nx54E+zfaNSqY8JIac+ySSXC3vs0DvFyc7G1tkm7feNhfVW8fi4yTR8/nPNEPAAAAAEDuGMHx\nsjBwaDZyttJ8auTM8RljIj4OdjIU13Yu/e2BX4PqVW1Msu+cO7BxfbS6Zv+3/IQQwqDuu11q\n/Pn9is+rD+0UYJUS88uKXem+Pd7yE0KkHl8ZNmNH+quDR7WwvBcXd08IISrZurrZPu96m9d3\nf/V7hm+gp5ONmTYt4ULUL+tPmzf8uMljs/+eUn+xlWT8vWN/qmeXdn6FcjoLZxfr9LNHzmXU\nrqnU3j/+w8pdd4XwzT1F3ZButf5eNW+21YAur3pZi+Srp2KNmrSvb/ucFwYAAAAAKD8IOF4i\nCuv6g2ZON58+ff74ORkRoW3NrIxv797wxU/3UjXGto4uvm+HjgoJdssdB6Fw6Th5huLbVZtn\njV2ptajiVb/fzH5vuyiEEFcO/nU5O0dELZ8Yld9zlc6Lvur75LNmS0ZpbXp3z8/LNySqsgws\nbB3dA7uGjXunkV3x9ScXU8mdvTv+1QQOaePySCcBvSf0Xbpy7pB9wtjUwjOoZ7fGs757uM+1\nw9TZZqtXb10xeW2KpLR18WnZJ+g5rwoAAAAAUK4oJKmICQ1AxaZSqdRqtb6rEBYWFhv8svRd\nBQAAAF4G7Q/r//dbfbG3t9d3CXjhWIMDAAAAAADIHlNUUDIJG0cOWX35ic3Vei5b/L6rHuoB\nAAAAAKAQAg6UTNUui7d20XcRAAAAAAA8HVNUAAAAAACA7BFwAAAAAAAA2WOKClDevdtvkb5L\nAADgcerhw/VdQnlhb2+vVqtVKpW+C8ELZ2dnd//+fX1XAaBIjOAAAAAAAACyR8ABAAAAAABk\nj4ADAAAAAADIHgEHAAAAAACQPQIOAAAAAAAgewQcAAAAAABA9gg4AAAAAACA7BFwAAAAAAAA\n2SPgAAAAAAAAskfAAQAAAAAAZI+AAwAAAAAAyB4BBwAAAAAAkD0CDgAAAAAAIHsEHAAAAAAA\nQPYIOAAAAAAAgOwRcAAAAAAAANkj4AAAAAAAALJHwAEAAAAAAGSPgAMAAAAAAMgeAQcAAAAA\nAJA9Ag4AAAAAACB7RvouAEAxrKdOTU5O1mg0+i4EL5ZSqdRqtZmZmfouBC+WsbGxlZVVRkZG\nRkaGvmvBi2VkZGRmZpaamqrvQgAAqCgYwQEAAAAAAGSPgAMAAAAAAMgeAQcAAAAAAJA9Ag4A\nAAAAACB7BBwAAAAAAED2CDgAAAAAAIDsEXAAAAAAAADZM9J3AQCKoTh6VN8lAIAe3HF313cJ\nAABAThjBAQAAAAAAZI+AAwAAAAAAyB4BBwAAAAAAkD0CDgAAAAAAIHsEHAAAAAAAQPYIOAAA\nAAAAgOwRcAAAAAAAANkj4AAAAAAAALJHwAEAAAAAAGSPgAMAAAAAAMgeAQcAAAAAAJA9Ag4A\nAAAAACB7BBwAAAAAAED2CDgAAAAAAIDsEXAAAAAAAADZI+AAAAAAAACyR8ABAAAAAABkj4AD\nAAAAAADIHgEHAAAAAACQPQIOAAAAAAAgewQcAAAAAABA9oz0XQCAYnzXubq+SwAAPdiu7wJK\nQ6YQlfRdw1O0P6zWdwkAAJQ+RnAAAAAAAADZI+AAAAAAAACyR8ABAAAAAABkj4ADAAAAAADI\nHgEHAAAAAACQPQIOAAAAAAAgey//Y2LPrOw/Ydu9R7c5dl7wdV/vR3YZmljaOnk1bNezT/ua\nFsX1ZhkUumJ8sGXetthvB475J3jBV329S1qJwsjEzMK2qrtvYJM27dsEOpb+A+SKO1f6xT2b\ntu0/cuZKQnKWiY2zT+P2Pbq39jYXQgiRtH1i3y9PF+pM+WbE+mF1dJ7rYNN5qwbVKNgW9/3g\nUbvqRqwdVkcIcSvqq6+2Hrt0425qlqFlZddawZ16dQtyqfSUYyXVie+mzfrtQeMx4SOCnDKL\nLjJx45iBq2MfrUNRf9T68NfN/tuNAwAAAADI08sfcAghhHWL0dM6eRS8rmTr+vBLm9dDI7tU\nE1JOZnL8wfVffhmebvNtaJBSV2/GxlnR36+PeXWQv/EzV5J7Okmb/UCVePl01I41YTt3hYTN\n6BdoWfyxpXmuxH/3nRb+bft2dHdQZsYf/Gn152HxmsUR7ZzyjrUMHhnRuVreC0Mr5/9UiGRo\nW6tVr7edK1sYq+/F7t/w/dxJKeErh9d7LNdR39i3MPyLE5U7R0zrWctS6CzSrvXYRfWz8g9N\n/GNO5F6X4EakGwAAAABQUVWMgMPIytnLy+upuwzNHV1dXYUQws3DW5z6Y9KJ2JsiSOdYDIdW\nnTyjNn6z9a3PurgontgrJZ34+Zsfdh25fC/bxL56w3a9B3QKsM5vVnA64eUb0PiNZl4TR3+/\ncE3Drz6sXanoY9OOLRkVfiIgbPHIVyyEEFnnvx0z4S+3MYsnBNvqqFPXuap1j5yb37BmLV/N\n5d5fRh9NbdcuL2gxtKpa1B17dlWbdu2S/6JGgDL+0OSTp26Iep6F2qSd+2lmxLq7dYbPHd3K\nJS/50FGkkbWLl/XDXVJs1LEEy+aDgnOH3iQdXff1xuiz8bdT1GaV3Wq91uODrg0cDfPaZl75\nc+0Pv0bHXEvWmNq4eDfvM6Zfg/yOAAAAAACyVTECjhKQctKuH/rzRJpVLT+X4tpaNOvX7cSw\nNav+bBX2us1j3VzbNCP85+zWQ8JG1rJMPrVl+crwSKOlczpUeXpPRm4duzTbMOPvqDNDa9cR\nRR5rUW/ouM5jxi9aFLA47HXTwyvmbs1uNe1jnelGMed6LJhRq7MlY0frgoErqQfm9Nn7QGPh\n4OrbpGPPLkFups90rqJos5KuRO85nmZfL7DQoBBt4sGlSxYcMH1ryvz+9ayfzIyKKvLh9iM7\n9txyeiO0Qd6AmswMg+rtBod4OlkaqC4dWLN8RqTB4gXvuRsIId3YNm3ct3cCew6a9mo1K23S\nlVOXFNqCjhITE0+ePJn/0s/Pz8rKqhSu+r8xNDQsvhEAAM/CxMRE3yW8bAwMDLirFYFCoeAb\nDZRnFSPguLcttMO2gpcGwRO3hDbO27X9086/K4Sk1WhyDGzr9A4bWpJ5Dk5vDeqw85Mf1pwI\n+qhO4R9x0vFfNl50fnfJh608FUK4OX805tbZMRs2neowPKCInow8PVzFvoRbaSIwVsexxr69\nxvc9M2bF3O8SrHb/pewyZ1AdndNoijlXncIzYrS39iz7Oda90+BXcv+SNnYL6vFhR3+PyibZ\nd87t/3ndnLExAxdEdnAtMnkQT9xhIYQQj8y6Ue+f0XXhYUkrSaYebT6ZNbhuwfSe5D8Wzslx\n7jQ3sl+Noj8uHi+yQFr0jgMp3t3f9HlYX9Vm74fkfenk2H3svZO9d0Zdfa+np9Ae37Quxvad\neZ92qWEohBBVqnr4Fe4pJiZmwoQJ+S+XLVvm4lJs2gUAgPxYWr6AubEVm6GhIXe1guAbDZRn\nFSPgeGwNDoVFwYAKm5ajI7tUE0KbnXbr/K4fvo2INJ8zrZ3OP+aFEMLQp+sHLfZGfLO5/eLu\nhTbfu3o1Q+lf2/Ph4Qbetf1MN8XHJ4mAooZbSJIkhEIoijvWwK3jJwOPDV+6Iadm/4U9fJ99\n+Y/C5yqgvR21aPKK+HqjZ/bwyZsZYh7Yvltg7pfevgGBdumD527afq7DEL8n+ivw+ConN36b\nMe9/hRsYvTJk8eKe6rTEiwc3rfl8lpXd7D7+eaNCLPybel6J2r16e7OwEO+npktPKTJf4p4d\nR6Q6I1oVjAiRkmK2b9z6d8z1u6q07BxJnSHUjolCeIpbl2LTzALr1ShqRET16tVHjBiR/9LB\nwSE9PV3HRZcN/pcAAFDqysMH3MvE3Nw8JycnMzNT34XghVMqlRkZGfquAs/J3Nxc3yXghasY\nAUeJ1uAQ7l41TC4fnrx225l2H/oX26eyYd+edYd+/c2u1v0KNkqSJBRPhCNPbimguXLlhrD0\nrWpe/LEpl07FZxkYSEk37mQKt2cfwFHoXHmy43fOnfrNjVc+mT2ssX0RRZrX9HMV0XcSNcJP\nxz+WJ+6w9WONFUoHDw8HIap5+9cyuN5r2foDIRFtctfMMHJ/Z8rQwPmTl04KS58c3rv2Y6G4\nziKluB07zymDJjYvWEfj+qbpU35Rdhgy4P3qjlZKY3Fw8YCVGrUQed8dUfQ3w93dvW/fvvkv\nVSrVgwcPir7mMsIUFQBAqSsPH3AvE3Nzc61Wy12tCMzMzPhGyxcBR0VgoO8CyhtJZGdlSSVq\nWrnNgBDn02t/+F/BDzl7Dw9leszpKw87kGJPn8209HAvaviG5tqWTQcz7Zs1r6Uo5ljpzh8L\nFx+y6Tpz+ruV9i5ccuD+M19Z4XMJIUTG2fVTP111t8Xk2cOLTDeEEOlnz14TNg4OpReFaTRq\nSVKrNYU2VXJpM35OaJOMzeGffvlvUqHbX0yR6mM79iTYv9GuUcGAltSYY5dMgrv1bR7g5eJk\nb2uTdPvGw/6qevtYZJw8ej6n1C4FAAAAAFBeVIwRHJqUm3Fxhec2WFTxcswdApGTnnj9urkQ\n2uy02+d3rd6XYtesac3iJqjkUbiHDGi7a/IvfxoJu7wtdUO6+oxY/9lyp6Hv5C4Uui3Rr09I\nwQIceafTqnMf3bp964EEl5Cw3v5GQggdx+bEb5638qzHBwu7+ztL43ueGbt0/nafGe2r6syn\ndJ0r9fjKsBk70l8dPKqF5b24uHtCCFHJ1tXN1lhc27n0twd+DapXtTHJvnPuwMb10eqa/d/S\nNT+lGNd3f/V7hm+gp5ONmTYt4ULUL+tPmzf8uMljy7MaODQbOVtpPjVy5viMMREfBzsZ6ioy\nV8bfO/anenZp51foPlg4u1innz1yLqN2TaX2/vEfVu66K4Rv7inqhnSr9feqebOtBnR51cta\nJF89FWvUpH39Z1usFQAAAABQHlWMgEO1f+Go/YU3BI/fGhokhBAied+8YfuEEIpKSht711od\nQnu836jksz9MAnv2eWX/wv/LfrhB4RoSNjXn6zXrZ4y8l21q79WwR/iAdwo9QiX3dApDYzML\nuyoePnV6RUxqU8exku5js879MHfNzQajFrd3NhBCeHQOHXhq5Mq56/3m9vB6fEkKUbJzXTn4\n1+XsHBG1fGJUfvsqnRd91ddLmFkZ39694Yuf7qVqjG0dXXzfDh0VEuxWwsznaZTWpnf3/Lx8\nQ6Iqy8DC1tE9sGvYuHca2T3ZUGFdf9DM6ebTp88fPycjIrRtctFFCiGEuLN3x7+awCFtHlkI\nVBHQe0LfpSvnDtknjE0tPIN6dms867uH+1w7TJ1ttnr11hWT16ZISlsXn5Z9gp7/wgAAAAAA\n5YdCkko2HwOoYFQqlVqt1ncVwsLCYoNflr6rAAC8VNof1v8H3MvE3t5erVarVCp9F4IXzs7O\n7v79Z58qjvLB3t5e3yXghWMNDgAAAAAAIHsVY4rKs0rYOHLI6stPbK7Wc9ni9131UE9RyrJO\nudwTAAAAAECFxBQV4OmYogIAeFkxRaV0MUWl4mCKiqwxRaUiYIoKAAAAAACQPQIOAAAAAAAg\ne6zBAZR37/ZbpO8SAAAvl6VP36wePrxs6wAAoDQxggMAAAAAAMj9Bw5lAAAYZUlEQVQeAQcA\nAAAAAJA9Ag4AAAAAACB7BBwAAAAAAED2CDgAAAAAAIDsEXAAAAAAAADZI+AAAAAAAACyR8AB\nAAAAAABkj4ADAAAAAADIHgEHAAAAAACQPQIOAAAAAAAgewQcAAAAAABA9gg4AAAAAACA7BFw\nAAAAAAAA2SPgAAAAAAAAskfAAQAAAAAAZI+AAwAAAAAAyB4BBwAAAAAAkD0CDgAAAAAAIHsE\nHAAAAAAAQPYIOAAAAAAAgOwZ6bsAAMWwnjo1OTlZo9HouxC8WEqlUqvVZmZm6rsQvFjGxsZW\nVlYZGRkZGRn6rgUvlpGRkZmZWWpqqr4LAQCgomAEBwAAAAAAkD0CDgAAAAAAIHsEHAAAAAAA\nQPYIOAAAAAAAgOwRcAAAAAAAANkj4AAAAAAAALJHwAEAAAAAAGTPSN8FACiG4uhRfZcAAACK\ndvWqEOKOu7u+6wCAio4RHAAAAAAAQPYIOAAAAAAAgOwRcAAAAAAAANkj4AAAAAAAALJHwAEA\nAAAAAGSPgAMAAAAAAMgeAQcAAAAAAJA9Ag4AAAAAACB7BBwAAAAAAED2CDgAAAAAAIDsEXAA\nAAAAAADZI+AAAAAAAACyR8ABAAAAAABkj4ADAAAAAADIHgEHAAAAAACQPQIOAAAAAAAgewQc\nAAAAAABA9gg4AAAAAACA7BFwAAAAAAAA2SPgAAAAAAAAskfAAQAAAAAAZE8hSZK+awDKI5VK\npVar9V2FsLCw2OCXpe8qAAAAgOfR/rD+f6POZW9vr+8S8MIxggMAAAAAAMgeAQcAAAAAAJA9\nAg4AAAAAACB7BBwAAAAAAED2CDjKu8OL3uswfW+2EEKIqNkd3p0XreeCAAAAAAAof4zK4Bxn\nVvafsO3eo9scOy/4uq/3I7sMTSxtnbwatuvZp31Ni5L0pjAyMbOwreruG9ikTfs2gY6VXmDl\nTz9X+sU9m7btP3LmSkJylomNs0/j9j26t/Y2F0IIkbR9Yt8vTxfqTPlmxPphdUpwrocs34xY\nq+OAQrIvbF++6eDlS5evJqYZtw5fP6L+s13nf3NmZf8JB5vOWzWoRsG2uO8Hj9pVN6/+W1Ff\nfbX12KUbd1OzDC0ru9YK7tSrW5BLpaccK6lOfDdt1m8PGo8JHxHklFn07U3cOGbg6thH61DU\nH7U+/HWzsrloAAAAAED5UhYBhxBCWLcYPa2TR8HrSrauD7+0eT00sks1IeVkJscfXP/ll+Hp\nNt+GBil19ZZ7iKTNfqBKvHw6aseasJ27QsJm9Au0LP3KdZ0r8d99p4V/274d3R2UmfEHf1r9\neVi8ZnFEO6e8Yy2DR0Z0rpb3wtDKubhzPXqXSnBAHm1mllHVum2bNr/w49JDz3h9ZUAytK3V\nqtfbzpUtjNX3Yvdv+H7upJTwlcPrPZZIqW/sWxj+xYnKnSOm9axlKXTeXrvWYxfVL3h2auIf\ncyL3ugQ3It0AAAAAgIqqrAIOIytnLy+vp+4yNHd0dXUVQgg3D29x6o9JJ2JviiBvXb0VHCK8\nfAMav9HMa+Lo7xeuafjVh7UrCSElnfj5mx92Hbl8L9vEvnrDdr0HdAqwVgiRdmzJqPATAWGL\nR75iIYTIOv/tmAl/uY1ZPCHY9jnPVa175Nz8hjVr+Wou9/4y+mhqu3Z5QYuhVdWirvqpdNyl\nJ+Tcjv48fPFp9wFTx7Z1CwwZHiiESEjZ/HizqNkdPhP9Qh1O/ngoLjlVa+Pfpv+w3g3tFUIU\nfaOESD2zZeX3O49duZsumdo4ejTp+cmgJnYlv47HVG3atUv+ixoByvhDk0+euiHqeRZqk3bu\np5kR6+7WGT53dCuXvORDx+01snbxsn64S4qNOpZg2XxQcO7An6Sj677eGH02/naK2qyyW63X\nenzQtYGjYV7bzCt/rv3h1+iYa8kaUxsX7+Z9xvRrkN8RAAAAAEC2yirgKAEpJ+36oT9PpFnV\n8nN5tiON3Dp2abZhxt9RZ4bWriOubZoR/nN26yFhI2tZJp/asnxleKTR0jkdqgiLekPHdR4z\nftGigMVhr5seXjF3a3araR/rTDeKOZfi0X1qdbZk7GhdMPgk9cCcPnsfaCwcXH2bdOzZJcjN\n9NkurEgZsVtmTV+XGjxu3sCGtopiGmv/99PWzpMivwywUl/d9dnUmZHKLz7r4qKQirxROUe+\nmb4moW3otNE+Ngbpd6+eu1KpuHOUkDYr6Ur0nuNp9vUCC41O0SYeXLpkwQHTt6bM71/PushT\nPXF7H24/smPPLac3QhsY577OzDCo3m5wiKeTpYHq0oE1y2dEGixe8J67gRDSjW3Txn17J7Dn\noGmvVrPSJl05dUmhLZ0rAwAAAADoV1kFHPe2hXbYVvDSIHjiltDGebu2f9r5d4WQtBpNjoFt\nnd5hQ599poGRp4er2JdwK00Exv6y8aLzu0s+bOWpEMLN+aMxt86O2bDpVIfhAUIY+/Ya3/fM\nmBVzv0uw2v2XssucQXV0ToUp5lx1Cs+I0d7as+znWPdOg1/JHSxg7BbU48OO/h6VTbLvnNv/\n87o5Y2MGLojs4KorKnjkLnn3/3pBJ8cnG0l3D381ff7fdt1nzn7Xu0SJie0b/XsEWBsIYeje\ndmjI3g/WbjnReXid40XeqLTE2xmWXg3qelUxEcLOztHNt9hTPPb9FUII8ch8IfX+GV0XHpa0\nkmTq0eaTWYPrGufvSv5j4Zwc505zI/vVMCnyBI/f3gJp0TsOpHh3f9Pn4Z2t2uz9kLwvnRy7\nj713svfOqKvv9fQU2uOb1sXYvjPv0y41DIUQokpVD7/CPd28efOff/7Jf9mwYUM7u+cft1Ja\nDA0fv2IAAABALkxNS+s/eYHi6WkNDoVFlfyvbVqOjuxSTQhtdtqt87t++DYi0nzOtHY6g4An\nSZIkhEIoxL2rVzOU/rU9Hx5u4F3bz3RTfHySCLAVQhi4dfxk4LHhSzfk1Oy/sIevsa4+iz1X\nAe3tqEWTV8TXGz2zh0/e/ArzwPbdAnO/9PYNCLRLHzx30/ZzHYb4PdFfgUfukknlp/1xLcV8\nOy5a7fPxZ+NedyzpH76eXl75T8upXL26dcal+HvCVceNavL2G1sWTh1wJqBeYO2Aeq82buBh\nVcy34/E1Vm78NmPe/wo3MHplyOLFPdVpiRcPblrz+Swru9l9/PN+1Fn4N/W8ErV79fZmYSHe\nT822nnJ78yXu2XFEqjOiVcGIECkpZvvGrX/HXL+rSsvOkdQZQu2YKISnuHUpNs0ssF6Nou7b\n+fPnIyMj818uW7bM3d1d93UDAAAA0MHCQscDJIBSVq7W4BDuXjVMLh+evHbbmXYf+j9L95or\nV24IS9+q5kKSJKF44q/xQltSLp2KzzIwkJJu3MkUbs8+gKPQufJkx++cO/WbG698MntYY/si\nggDzmn6uIvpOokb46bjlJViDQ+FS/xXLv/7ee+By0y4lG78hRI4mp1D5Go0QQui8UTZBI5fW\nfvfUkaMnTp/Y+tmP37r3iYwM8dKVpzxRufVjl6lQOnh4OAhRzdu/lsH1XsvWHwiJaJP7s87I\n/Z0pQwPnT146KSx9cnjv2o+tFKvz9kpxO3aeUwZNbF6wjsb1TdOn/KLsMGTA+9UdrZTG4uDi\nASs16vxLFkVHNTVq1Jg4cWL+yypVqqSlpem46LJhYlL0uBYAAACgfCsPv1HnImqpCMrRGhwP\nSSI7K0sSOv4QfZzm2pZNBzPt32heSyEMPTyU6TGnr0h1cscmSLGnz2ZaBrjnrrQh3flj4eJD\nNl1njsleOnXhksDFE5o/4xyEwucSQoiMs+sjIjY/aD15dv86OharTD979pqwaeHw3++3Tb0P\nJ7QwmRIRNjkrPLxnTfPijxCxZ2Ky2jY0EUII6WrMmTRlLffKwl7njRJG1h71Xveo93qnvu1X\nDxv9+4H4kGdZLVUnjUYtSVq1ptCmSi5txs9RLpnyWfinGRMiBhesK1LM7VUf27Enwf6NUY0K\nhuKkxhy7ZBI8s2/zACGEENLF2zckkTvHpqq3j0XGyaPnc+o+fRCHs7NzSEhI/kuVSpWZmfnf\nrrUUGBmVwzcpAAAAUCLl4TfqXAQcFUFZ/e2kSbkZF1d4doFFFS/H3OETOemJ16+bC6HNTrt9\nftfqfSl2zZrW1J1u5B2iVec+unX71gMJLiFhvf2NhBB1Q7r6jFj/2XKnoe/krp25LdGvT0iA\nEELkxG+et/KsxwcLu/s7S+N7nhm7dP52nxntqxo877lSj68Mm7Ej/dXBo1pY3ouLuyeEEJVs\nXd1sjcW1nUt/e+DXoHpVG5PsO+cObFwfra7Z/y1d81NKzjLgg5nTTcPDp0zKCpv+QaCVENlJ\n164nqcXd5GyhTbsVFxcnFBZVquXdYIXm8PeLthn3aGT/4Pxvy7fc9Oo8sa5CKIq+UTf2fbs3\nO6CRn5uDpUIVG3X6nqm7x1MWAymp67u/+j3DN9DTycZMm5ZwIeqX9afNG37cxObRVgYOzUbO\nVppPjZw5PmNMxMfBToa6bm+ujL937E/17NLOr9B30MLZxTr97JFzGbVrKrX3j/+wctddkRdw\nGNQN6Vbr71XzZlsN6PKql7VIvnoq1qhJ+/rPtswsAAAAAKA8KquAQ7V/4aj9hTcEj98aGiSE\nECJ537xh+4QQikpKG3vXWh1Ce7zfqJiZI7mHKAyNzSzsqnj41OkVMalNHcfc/EThGhI2Nefr\nNetnjLyXbWrv1bBH+IB3qgghss79MHfNzQajFrd3NhBCeHQOHXhq5Mq56/3m9vB6fGGHEp7r\nysG/LmfniKjlE6Py21fpvOirvl7CzMr49u4NX/x0L1VjbOvo4vt26KiQYLdSehiJEMqaPabP\nNJ0xdfrErAkRQxsmbY8Y9dOt3F03Vow69MgyroqGvQe6Hl05afWNDOMqAe0+HdbFRaHrRgkT\nc8W1P1b9ufqOKtvQ0sGrXo/wD177D2mn0tr07p6fl29IVGUZWNg6ugd2DRv3TqOnjJxRWNcf\nNHO6+fTp88fPyYgIbZtc9O0VQghxZ++OfzWBQ9o88tAdRUDvCX2Xrpw7ZJ8wNrXwDOrZrfGs\n7x7uc+0wdbbZ6tVbV0xemyIpbV18WvYJev4LAwAAAACUHwpJkvRdA16gqNkdPjMcvyWUv+Of\nmUqlUqvV+q5CWFhYbPDL0ncVAAAAwPNof1j/v1Hnsre313cJeOF0zs4AAAAAAACQg/K6fmHC\nxpFDVl9+YnO1nssWv+/KufRAvpUDAAAAACoApqgAT8cUFQAAAOA/YooKyhJTVAAAAAAAgOwR\ncAAAAAAAANkj4AAAAAAAALJXXhcZBfDQu/0W6bsEAADwzNTDh+u7BJQyOzu7+/fv67sKAEVi\nBAcAAAAAAJA9Ag4AAAAAACB7BBwAAAAAAED2CDgAAAAAAIDsEXAAAAAAAADZI+AAAAAAAACy\nR8ABAAAAAABkj4ADAAAAAADIHgEHAAAAAACQPQIOAAAAAAAgewQcAAAAAABA9gg4AAAAAACA\n7BFwAAAAAAAA2SPgAAAAAAAAskfAAQAAAAAAZI+AAwAAAAAAyB4BBwAAAAAAkD0CDgAAAAAA\nIHsEHAAAAAAAQPYIOAAAAAAAgOwRcAAAAAAAANkz0ncBAIphPXVqcnKyRqPRdyF4sZRKpVar\nzczM1HcheLGMjY2trKwyMjIyMjL0XQteLCMjIzMzs9TUVH0XghfO3t5erVarVCp9FwIAFR0j\nOAAAAAAAgOwRcAAAAAAAANkj4AAAAAAAALJHwAEAAAAAAGSPgAMAAAAAAMgeAQcAAAAAAJA9\nAg4AAAAAACB7BBwAAAAAAED2CDgAAAAAAIDsEXAAAAAAAADZU0iSpO8aABRpx44dx48fHzhw\noKOjo75rAVAKLly4sHHjxpYtWzZt2lTftQAoBVqtdvbs2a6urn369NF3LQBQ0TGCAyjXjh8/\nvnnzZpVKpe9CAJSOW7dubd68+dy5c/ouBECp2bx58/79+/VdBQCAgAMAAAAAAMgfAQcAAAAA\nAJA9Ag4AAAAAACB7LDIKAAAAAABkjxEcAAAAAABA9gg4AAAAAACA7BFwAAAAAAAA2TPSdwFA\nxaS98++6lT/sOX49RWHtVrdV78HdG1RWPGvLkncCoAyUwvs6afvEvl+eLtRS+WbE+mF1yqJ6\nAE8o6Zs6+8L25ZsOXr50+WpimnHr8PUj6j9HJwCA/84wPDxc3zUAFY76/OpxU37NajoodHi3\nFm7Jf61e9Ze6QdvAyk8OqdLRsuSdACgDpfK+zry4d8uFGiPnjun+Zq429avZKPnPCEAfnuFN\nfeP0iTSnV1q8annl8HXHll1erfocnQAA/jt+ugJlL/PQ5t9uuHQaM+S1mh4etV4fOrqTc/zW\nzf+X9UwtS94JgDJQKu9rIYQQhlZVvfJ52JuW5VUAyPcMn7OmgSHD+3Vt1yKwyuNvWD6sAaBM\nEXAAZS/u7Nks+/oN3B++rla/nm3mmZi4Z2pZ8k4AlIFSeV8LIYRIPTCnz3vv9fhg+LjZa6Kv\nZb7owgE8Xal8zvJhDQBlimGvQJmTku4nC1tbm4Ittra2Ivn+fUkIRUlblrwTAGWgVN7XQmHs\nFtTjw47+HpVNsu+c2//zujljYwYuiOzgyrsaKGul8jnLhzUAlC0CDqDsSU/d+rTfdHS0LHkn\nAMpAqbyvhXlg+26BuRu8fQMC7dIHz920/VyHIX6lUSKAZ1Eqn7N8WANAmWKKClDmFHZ2NiIp\nKblgS1JSkrCxtXviFx4dLUveCYAyUCrv68eZ1/RzFUl3EjUvpGQAupTK5ywf1gBQtgg4gLLn\n5edncvfokWsPX18+eizJtJa/1zO1LHknAMpAqbyvH5N+9uw1YePgwGhLQA9K5XOWD2sAKFM8\nJhYoe0bOjg+iN//6fyoHTweDxGPrv1h93LrjxwPqVzYQQhW9fOrSo1ZBDV0r6W6pqxMAZa5U\n3tfXdi798VSalJ2Znnzz3N8/f7Hqr+Tq3T/q7mfN//YCZa7kb2ohspOuxSfcS7p1JupAnFnN\nBi6GSclqE1vzSnxYA0DZUkjS0ycHAniRtImHf1y5Zs+JaynC2q1eq96DejS0VwghxN2t4z/4\n2vDDNZHtrIppqXMXgLL339/Xd6NXfrHp37ib91I1xraOLr5NO3QPCfZQ6vGagAqtxG/quDWD\nR/10q/ChBsETt4Q21tkJAKDUEXAAAAAAAADZY3wcAAAAAACQPQIOAAAAAAAgewQcAAAAAABA\n9gg4AAAAAACA7BFwAAAAAAAA2SPgAAAAAAAAskfAAQAAXiJ7htooFG9/l6bvOgAAQFkj4AAA\nAAAAALJHwAEAAAAAAGSPgAMAAAAAAMgeAQcAAHipJR//duy7DatVVpqYWDrVaN5z6tbYzIK9\nmmu75g58I8DT0dLE1NLe1afhW/0WHEwt2V4AAFCeGOm7AAAAgBcmPXpc89bzThk36Dksoqlz\n1sXfVy2f3nFn1NJDu4fVMBRCHf1pq7fmJwb2GD7xYx87g5Rb8eeifztwImVMU0tR3F4AAFC+\nEHAAAICXVsxnH847pWk88//+mljbRAghPvqw3eC6bVeOHbmq6+8DHcSpXzddyHlt6Z61w+zy\nj5me/5XuvQAAoHxhigoAAHhZXdy86ZSw6DxhTG66IYQQtm3CRzYRmbs37UgRQljb2Ahx9vd1\nhxPVTzlc914AAFC+EHAAAICXVVxcnBDV/f1NC290DgiwE9pLl64IIaoPmz+pkXr7R42qVvZo\n+GbPjyNX7buckd9S914AAFC+EHAAAICXlSRJQigUiqJb2Lw+49Dl2AM/LRrZvnrW8XXhH7xR\no+b7G26UaC8AAChXCDgAAMDLqnr16kLEnj6dWXhjwunT94WBl5dn3msDy2rBXUdELNvwZ8zN\nuHXd7K79NHHJ4fzWuvcCAIDyg4ADAAC8rHw6hQSItE1zF53NfrhJtWf64oPCtFXn9lZCiJT7\n9zWF2ldybdzAVYj79+8LUexeAABQvvAUFQAA8NKq/cny0I2t5n3aNChmWPemzlkX//h2+bZ4\n29e+WNzfQQghdg1xHhH/due2DXw9XaxzEmO2f730iKHX6N7NRfF7AQBA+aKQJEnfNQAAAJSS\nPUNtWn/ZbFXqb/0s8rYkH/tmeviyzVFnbqYZ2njWa9X9k+mfdvTOXXc0dvOkOd//FX30wvU7\nKo3S0dXnlXcGjhs3oImTQfF7AQBA+ULAAQAAAAAAZI//gQAAAAAAALJHwAEAAAAAAGSPgAMA\nAAAAAMgeAQcAAAAAAJA9Ag4AAAAAACB7BBwAAAAAAED2CDgAAAAAAIDsEXAAAAAAAADZI+AA\nAAAAAACyR8ABAAAAAABkj4ADAAAAAADI3v8DqMIT795iFbkAAAAASUVORK5CYII="
          },
          "metadata": {
            "image/png": {
              "width": 720,
              "height": 300
            }
          }
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "df_losses %>%\n",
        "  mutate(method = factor(method, c('macs2', 'nucleoatac', 'seatac'))) %>%\n",
        "  mutate(label = sprintf('%s_%s', dataset, type)) %>%\n",
        "  group_by(method) %>%\n",
        "  summarize(loss_mean = mean(loss), loss_std = sd(loss))"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 192
        },
        "id": "ADhVQCHqFtR8",
        "outputId": "76ee8333-e5de-4c1b-c973-6b1a155e3aa1"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/html": [
              "<table class=\"dataframe\">\n",
              "<caption>A tibble: 3 × 3</caption>\n",
              "<thead>\n",
              "\t<tr><th scope=col>method</th><th scope=col>loss_mean</th><th scope=col>loss_std</th></tr>\n",
              "\t<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
              "</thead>\n",
              "<tbody>\n",
              "\t<tr><td>macs2     </td><td>0.08935848</td><td>0.009858105</td></tr>\n",
              "\t<tr><td>nucleoatac</td><td>0.10624094</td><td>0.023993544</td></tr>\n",
              "\t<tr><td>seatac    </td><td>0.08060138</td><td>0.012655928</td></tr>\n",
              "</tbody>\n",
              "</table>\n"
            ],
            "text/markdown": "\nA tibble: 3 × 3\n\n| method &lt;fct&gt; | loss_mean &lt;dbl&gt; | loss_std &lt;dbl&gt; |\n|---|---|---|\n| macs2      | 0.08935848 | 0.009858105 |\n| nucleoatac | 0.10624094 | 0.023993544 |\n| seatac     | 0.08060138 | 0.012655928 |\n\n",
            "text/latex": "A tibble: 3 × 3\n\\begin{tabular}{lll}\n method & loss\\_mean & loss\\_std\\\\\n <fct> & <dbl> & <dbl>\\\\\n\\hline\n\t macs2      & 0.08935848 & 0.009858105\\\\\n\t nucleoatac & 0.10624094 & 0.023993544\\\\\n\t seatac     & 0.08060138 & 0.012655928\\\\\n\\end{tabular}\n",
            "text/plain": [
              "  method     loss_mean  loss_std   \n",
              "1 macs2      0.08935848 0.009858105\n",
              "2 nucleoatac 0.10624094 0.023993544\n",
              "3 seatac     0.08060138 0.012655928"
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "wilcox.test(df_losses$loss[df_losses$method == 'seatac'], df_losses$loss[df_losses$method == 'macs2'], alternative = 'less')"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 213
        },
        "id": "XKpi8MvHGICu",
        "outputId": "717e1809-5171-426c-d371-d05179c1d055"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "\n",
              "\tWelch Two Sample t-test\n",
              "\n",
              "data:  df_losses$loss[df_losses$method == \"seatac\"] and df_losses$loss[df_losses$method == \"macs2\"]\n",
              "t = -1.4443, df = 11.322, p-value = 0.08788\n",
              "alternative hypothesis: true difference in means is less than 0\n",
              "95 percent confidence interval:\n",
              "        -Inf 0.002103849\n",
              "sample estimates:\n",
              " mean of x  mean of y \n",
              "0.08060138 0.08935848 \n"
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "wilcox.test(df_losses$loss[df_losses$method == 'seatac'], df_losses$loss[df_losses$method == 'nucleoatac'], alternative = 'less')"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 124
        },
        "id": "mtDi4Nb4G57y",
        "outputId": "afd13e6e-a5c7-4cf9-979f-1639aedde6cf"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "\n",
              "\tWilcoxon rank sum exact test\n",
              "\n",
              "data:  df_losses$loss[df_losses$method == \"seatac\"] and df_losses$loss[df_losses$method == \"nucleoatac\"]\n",
              "W = 8, p-value = 0.01894\n",
              "alternative hypothesis: true location shift is less than 0\n"
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Read the raw H3K37ac signal for Etv2 data"
      ],
      "metadata": {
        "id": "KC-1IM7YHsEt"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "etv2_samples <- c('MEF_NoDox','MEF_Dox_D7_Flk1pos','EB_NoDox_D25','EB_Dox_D25_Flk1pos')"
      ],
      "metadata": {
        "id": "vgRdzrGAMXAq"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "y <- NULL\n",
        "for (i in which(names(ARGS$remote_bam_files) %in% etv2_samples)){\n",
        "  j <- 'H3K27ac'\n",
        "  sprintf('reading %s', ARGS$local_bw_files[[i]][j]) %>% message()\n",
        "  cvg <- rtracklayer::import(ARGS$local_bw_files[[i]][j], which = rowRanges(x[[i]]))\n",
        "  seqlengths(cvg) <- seqlengths(cvg) + ARGS$block_size\n",
        "  cvg <- coverage(cvg, weight = 'score')\n",
        "  yi <- cvg[rowRanges(x[[i]])] %>% as.matrix() %>% tf$cast(tf$float32)\n",
        "  yi <- tf$expand_dims(yi, axis = 1L)\n",
        "  y <- c(y, yi)\n",
        "}\n",
        "y <- tf$concat(y, axis = 1L)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "emsp2EO9Hxuz",
        "outputId": "927714c3-b4b7-40fd-af18-11bcf064d4e9"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "reading ./s3.msi.umn.edu/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/MEF_NoDox_d0_H3K27ac_FE.bw\n",
            "\n",
            "reading ./s3.msi.umn.edu/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/MEF_Dox_d7_H3K27ac_FE.bw\n",
            "\n",
            "reading ./s3.msi.umn.edu/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/EB_NoDox_12h_H3K27ac_FE.bw\n",
            "\n",
            "reading ./s3.msi.umn.edu/gongx030/datasets/dataset=Etv2PioneerChIPseq_version=20191203a/EB_Dox_12h_H3K27ac_FE.bw\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "y <- y %>% tf$expand_dims(3L)"
      ],
      "metadata": {
        "id": "f4L6F9QxODPd"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Predicting histone signal for Etv2 data"
      ],
      "metadata": {
        "id": "yd8-rqFu_gra"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "histones <- c('H3K27ac', 'H3K4me1', 'H3K4me3')"
      ],
      "metadata": {
        "id": "L-D2CRzvLk7D"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "y_pred <- NULL\n",
        "for (i in which(names(ARGS$remote_bam_files) %in% etv2_samples)){\n",
        "  x_test <- tf$cast(rowData(x[[i]])$vae_z_mean[, 1, ], tf$float32)\n",
        "  yi <- model_histone(x_test) \n",
        "  yi <- tf$expand_dims(yi, axis = 1L)\n",
        "  y_pred <- c(y_pred, yi)\n",
        "}\n",
        "y_pred <- tf$concat(y_pred, axis = 1L)"
      ],
      "metadata": {
        "id": "g0OuK4FLlf30"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Read Etv2 data again"
      ],
      "metadata": {
        "id": "PPVeH0AQMPyS"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "x2 <- read_vplot(\n",
        "  gr_etv2,\n",
        "  filenames = ARGS$local_bam_files[etv2_samples],\n",
        "  genome = get_bsgenome('mm10'),\n",
        "  bin_size = ARGS$bin_size,\n",
        "  fragment_size_range = c(0, ARGS$fragment_size_range_max),\n",
        "  fragment_size_interval = ARGS$fragment_size_interval \n",
        ")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "5hN_3jZ-Lyma",
        "outputId": "6be0ab4a-8c1c-4e57-c1a3-602883310e90"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "read_bam | reading ./s3.msi.umn.edu/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/MEF_NoDox.bam\n",
            "\n",
            "read_bam | reading ./s3.msi.umn.edu/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/MEF_Dox_D7_Flk1pos.bam\n",
            "\n",
            "read_bam | reading ./s3.msi.umn.edu/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/EB_NoDox_D25.bam\n",
            "\n",
            "read_bam | reading ./s3.msi.umn.edu/gongx030/datasets/dataset=Etv2ATAC_version=20190228b/EB_Dox_D25_Flk1pos.bam\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "y <- as.array(y)\n",
        "dimnames(y)[[4]] <- 'H3K27ac'"
      ],
      "metadata": {
        "id": "4-siFFfSN3W8"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "y_pred <- as.array(y_pred)\n",
        "dimnames(y_pred)[[4]] <- histones"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 97
        },
        "id": "MPqDiTtjOISd",
        "outputId": "77f3a2c1-2920-4ff1-882c-87181de45641"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "error",
          "ename": "ERROR",
          "evalue": "ignored",
          "traceback": [
            "Error in dimnames(y)[[4]] <- histones: length of 'dimnames' [4] not equal to array extent\nTraceback:\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "rowData(x2)$histone <- y\n",
        "rowData(x2)$histone_pred <- y_pred"
      ],
      "metadata": {
        "id": "D7bbtt9wNtt6"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "x2"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 249
        },
        "id": "m7GYNc5CyrtB",
        "outputId": "b292efa4-1916-454d-ecac-6adbc1c06f51"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "class: Vplots\n",
              "dim: 32222 4 64 128\n",
              "assays(1): counts\n",
              "rowData(3): score histone histone_pred\n",
              "dimdata:\n",
              "grange: names(1): id\n",
              "sample: names(3): id filename name\n",
              "interval: names(2): id center\n",
              "bin: names(2): id position\n",
              "## fragment_size_range:0,640\n",
              "## fragment_size_interval:10\n",
              "## bin_size:5\n",
              "## window_size:640"
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Save representations"
      ],
      "metadata": {
        "id": "QyuDw9nhwrpx"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "dir <- gsub('(.+)/.+', '\\\\1', ARGS$local_representation_file)\n",
        "if (!file.exists(dir)) dir.create(dir, recursive = TRUE)\n",
        "sprintf('writing %s', ARGS$local_representation_file) %>% message()\n",
        "saveRDS(x2, file = ARGS$local_representation_file)\n",
        "command <- sprintf('s3cmd sync %s %s',  ARGS$local_representation_file,  ARGS$s3_representation_file)\n",
        "message(command)\n",
        "system(command)\n",
        "command <- sprintf('s3cmd setacl -P %s',  ARGS$s3_representation_file)\n",
        "message(command)\n",
        "system(command)\n",
        "message(ARGS$remote_representation_file)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "DXLqGYK_xasM",
        "outputId": "d321445e-5601-46fd-a8a3-1d360cf754db"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "writing ./s3.msi.umn.edu/gongx030/projects/seatac_project/etv2_20230108a.RData\n",
            "\n",
            "s3cmd sync ./s3.msi.umn.edu/gongx030/projects/seatac_project/etv2_20230108a.RData s3://gongx030/projects/seatac_project/etv2_20230108a.RData\n",
            "\n",
            "s3cmd setacl -P s3://gongx030/projects/seatac_project/etv2_20230108a.RData\n",
            "\n",
            "https://s3.msi.umn.edu/gongx030/projects/seatac_project/etv2_20230108a.RData\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## SessionInfo"
      ],
      "metadata": {
        "id": "TRUcghfUpo3a"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "sessionInfo()"
      ],
      "metadata": {
        "id": "xID7mUnopoIS",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "outputId": "0d471557-7d42-42a3-ec02-de81bc83634b"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "R version 4.2.1 (2022-06-23)\n",
              "Platform: x86_64-conda-linux-gnu (64-bit)\n",
              "Running under: CentOS Linux 7 (Core)\n",
              "\n",
              "Matrix products: default\n",
              "BLAS/LAPACK: /panfs/jay/groups/18/garrydj/gongx030/.conda/envs/r421_py37_tf26/lib/libopenblasp-r0.3.21.so\n",
              "\n",
              "locale:\n",
              " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
              " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
              " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
              " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
              " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
              "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
              "\n",
              "attached base packages:\n",
              "[1] stats4    stats     graphics  grDevices utils     datasets  methods  \n",
              "[8] base     \n",
              "\n",
              "other attached packages:\n",
              " [1] BSgenome.Hsapiens.UCSC.hg19_1.4.3  BSgenome.Mmusculus.UCSC.mm10_1.4.3\n",
              " [3] BSgenome_1.64.0                    rtracklayer_1.56.1                \n",
              " [5] Biostrings_2.64.1                  XVector_0.36.0                    \n",
              " [7] bpt_0.1.005                        seatac_0.4.2-002                  \n",
              " [9] SummarizedExperiment_1.26.1        Biobase_2.56.0                    \n",
              "[11] MatrixGenerics_1.8.1               matrixStats_0.63.0                \n",
              "[13] GenomicRanges_1.48.0               GenomeInfoDb_1.32.4               \n",
              "[15] IRanges_2.30.1                     S4Vectors_0.34.0                  \n",
              "[17] BiocGenerics_0.42.0                tfdatasets_2.9.0                  \n",
              "[19] tfprobability_0.15.1               tensorflow_2.11.0                 \n",
              "[21] motifmatchr_1.18.0                 chromVARmotifs_0.2.0              \n",
              "[23] chromVAR_1.18.0                    irlba_2.3.5.1                     \n",
              "[25] pROC_1.18.0                        stringr_1.5.0                     \n",
              "[27] tidyr_1.2.1                        dplyr_1.0.10                      \n",
              "[29] Matrix_1.5-3                       devtools_2.4.5                    \n",
              "[31] usethis_2.1.6                      BiocManager_1.30.18               \n",
              "\n",
              "loaded via a namespace (and not attached):\n",
              "  [1] backports_1.4.1             uuid_1.1-0                 \n",
              "  [3] plyr_1.8.7                  repr_1.1.4                 \n",
              "  [5] lazyeval_0.2.2              BiocParallel_1.30.4        \n",
              "  [7] tfruns_1.5.1                ggplot2_3.4.0              \n",
              "  [9] TFBSTools_1.34.0            digest_0.6.31              \n",
              " [11] htmltools_0.5.4             GO.db_3.15.0               \n",
              " [13] fansi_1.0.3                 magrittr_2.0.3             \n",
              " [15] memoise_2.0.1               tzdb_0.3.0                 \n",
              " [17] remotes_2.4.2               readr_2.1.3                \n",
              " [19] annotate_1.74.0             R.utils_2.12.0             \n",
              " [21] prettyunits_1.1.1           colorspace_2.0-3           \n",
              " [23] rappdirs_0.3.3              blob_1.2.3                 \n",
              " [25] callr_3.7.3                 crayon_1.5.2               \n",
              " [27] RCurl_1.98-1.9              jsonlite_1.8.4             \n",
              " [29] TFMPvalue_0.0.9             zeallot_0.1.0              \n",
              " [31] glue_1.6.2                  gtable_0.3.1               \n",
              " [33] zlibbioc_1.42.0             DelayedArray_0.22.0        \n",
              " [35] pkgbuild_1.4.0              abind_1.4-5                \n",
              " [37] scales_1.2.1                DBI_1.1.3                  \n",
              " [39] miniUI_0.1.1.1              Rcpp_1.0.9                 \n",
              " [41] viridisLite_0.4.1           xtable_1.8-4               \n",
              " [43] reticulate_1.26             bit_4.0.5                  \n",
              " [45] DT_0.26                     profvis_0.3.7              \n",
              " [47] htmlwidgets_1.6.0           httr_1.4.4                 \n",
              " [49] gplots_3.1.3                ellipsis_0.3.2             \n",
              " [51] farver_2.1.1                urlchecker_1.0.1           \n",
              " [53] pkgconfig_2.0.3             XML_3.99-0.11              \n",
              " [55] R.methodsS3_1.8.2           here_1.0.1                 \n",
              " [57] utf8_1.2.2                  labeling_0.4.2             \n",
              " [59] tidyselect_1.2.0            rlang_1.0.6                \n",
              " [61] reshape2_1.4.4              later_1.3.0                \n",
              " [63] AnnotationDbi_1.58.0        munsell_0.5.0              \n",
              " [65] tools_4.2.1                 cachem_1.0.6               \n",
              " [67] cli_3.5.0                   DirichletMultinomial_1.38.0\n",
              " [69] generics_0.1.3              RSQLite_2.2.18             \n",
              " [71] evaluate_0.19               fastmap_1.1.0              \n",
              " [73] yaml_2.3.6                  processx_3.8.0             \n",
              " [75] bit64_4.0.5                 fs_1.5.2                   \n",
              " [77] caTools_1.18.2              purrr_1.0.0                \n",
              " [79] KEGGREST_1.36.3             whisker_0.4.1              \n",
              " [81] mime_0.12                   R.oo_1.25.0                \n",
              " [83] poweRlaw_0.70.6             pracma_2.4.2               \n",
              " [85] compiler_4.2.1              rstudioapi_0.14            \n",
              " [87] plotly_4.10.0               png_0.1-8                  \n",
              " [89] tibble_3.1.8                stringi_1.7.8              \n",
              " [91] ps_1.7.2                    desc_1.4.2                 \n",
              " [93] lattice_0.20-45             IRdisplay_1.1              \n",
              " [95] CNEr_1.32.0                 keras_2.11.0               \n",
              " [97] vctrs_0.5.1                 pillar_1.8.1               \n",
              " [99] lifecycle_1.0.3             tfautograph_0.3.2          \n",
              "[101] data.table_1.14.6           bitops_1.0-7               \n",
              "[103] httpuv_1.6.7                R6_2.5.1                   \n",
              "[105] BiocIO_1.6.0                promises_1.2.0.1           \n",
              "[107] KernSmooth_2.23-20          sessioninfo_1.2.2          \n",
              "[109] codetools_0.2-18            gtools_3.9.3               \n",
              "[111] assertthat_0.2.1            seqLogo_1.62.0             \n",
              "[113] pkgload_1.3.2               rprojroot_2.0.3            \n",
              "[115] rjson_0.2.21                withr_2.5.0                \n",
              "[117] GenomicAlignments_1.32.1    Rsamtools_2.12.0           \n",
              "[119] GenomeInfoDbData_1.2.8      parallel_4.2.1             \n",
              "[121] hms_1.1.2                   grid_4.2.1                 \n",
              "[123] IRkernel_1.3                pbdZMQ_0.3-7               \n",
              "[125] shiny_1.7.4                 base64enc_0.1-3            \n",
              "[127] restfulr_0.0.15            "
            ]
          },
          "metadata": {}
        }
      ]
    }
  ]
}